<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="Sequence properties" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="Sequence properties">
    
    <title>Sequence properties | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    rnaseq_normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    rnaseq_normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="Sequence-Properties" class="anchor" href="#Sequence-Properties">Sequence Properties</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BioFSharp/gh-pages?filepath=01_04_sequence_properties.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-binder.svg" alt="Binder" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.fsx"><img src="https://csbiology.github.io/BioFSharp/img/badge-script.svg" alt="Script" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-notebook.svg" alt="Notebook" /></a></p>
<p><em>Summary:</em> This example shows how to calculate properties of amino acid sequences in BioFSharp</p>
<h2><a name="General" class="anchor" href="#General">General</a></h2>
<p>BioFSharp comes equipped with a range of numerical values for important amino acid properties. To access them in an easy fashion, you can use the <code>initGetAminoProperty</code> function in the following way. The result is a mapping function, which assigns a value to each compatible amino acid.<br />
In this tutorial our aim is to find out the hydrophobicity of a peptide. We start by calling the aforementioned function.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="m">AminoProperties</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="fn">getHydrophobicityIndex</span>  <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="fn">initGetAminoProperty</span> <span onmouseout="hideTip(event, 'fs5', 5)" onmouseover="showTip(event, 'fs5', 5)" class="rt">AminoProperty</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 6)" onmouseover="showTip(event, 'fs6', 6)" class="id">HydrophobicityIndex</span>

<span onmouseout="hideTip(event, 'fs3', 7)" onmouseover="showTip(event, 'fs3', 7)" class="fn">getHydrophobicityIndex</span> <span onmouseout="hideTip(event, 'fs7', 8)" onmouseover="showTip(event, 'fs7', 8)" class="m">AminoAcidSymbols</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 9)" onmouseover="showTip(event, 'fs8', 9)" class="vt">AminoAcidSymbol</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 10)" onmouseover="showTip(event, 'fs9', 10)" class="id">Ala</span> 
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">0.61
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 11)" onmouseover="showTip(event, 'fs10', 11)" class="fn">getHydrophobicityIndexZ</span>  <span class="o">=</span> <span onmouseout="hideTip(event, 'fs11', 12)" onmouseover="showTip(event, 'fs11', 12)" class="fn">initGetAminoPropertyZnorm</span> <span onmouseout="hideTip(event, 'fs5', 13)" onmouseover="showTip(event, 'fs5', 13)" class="rt">AminoProperty</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 14)" onmouseover="showTip(event, 'fs6', 14)" class="id">HydrophobicityIndex</span>

<span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="fn">getHydrophobicityIndexZ</span> <span onmouseout="hideTip(event, 'fs7', 16)" onmouseover="showTip(event, 'fs7', 16)" class="m">AminoAcidSymbols</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 17)" onmouseover="showTip(event, 'fs8', 17)" class="vt">AminoAcidSymbol</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 18)" onmouseover="showTip(event, 'fs9', 18)" class="id">Ala</span> 
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">-0.4813502112
</code></pre></td></tr></table>
<p>With this function you might easily estimate the hydrophobictiy of our peptide by calling it on every element with a map. Usually close amino acids in a peptide influence each other. To cover this you can use the <code>ofWindowedBioArray</code> function. It also takes a window size and calculates the value of the property of every amino acid in the chain with regards to the effect of the adjacent amino acids in this window.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 19)" onmouseover="showTip(event, 'fs12', 19)" class="id">peptide</span> <span class="o">=</span> 
    <span class="s">&quot;REYAHMIGMEYDTVQK&quot;</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs13', 20)" onmouseover="showTip(event, 'fs13', 20)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs14', 21)" onmouseover="showTip(event, 'fs14', 21)" class="id">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs15', 22)" onmouseover="showTip(event, 'fs15', 22)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 23)" onmouseover="showTip(event, 'fs16', 23)" class="id">map</span> <span onmouseout="hideTip(event, 'fs7', 24)" onmouseover="showTip(event, 'fs7', 24)" class="m">AminoAcidSymbols</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 25)" onmouseover="showTip(event, 'fs17', 25)" class="id">aminoAcidSymbol</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 26)" onmouseover="showTip(event, 'fs18', 26)" class="id">peptidehydrophobicites</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs12', 27)" onmouseover="showTip(event, 'fs12', 27)" class="id">peptide</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs15', 28)" onmouseover="showTip(event, 'fs15', 28)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 29)" onmouseover="showTip(event, 'fs16', 29)" class="id">map</span> <span onmouseout="hideTip(event, 'fs3', 30)" onmouseover="showTip(event, 'fs3', 30)" class="fn">getHydrophobicityIndex</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">[|0.6; 0.47; 1.88; 0.61; 0.61; 1.18; 2.22; 0.07; 1.18; 0.47; 1.88; 0.46;
0.05; 1.32; 0.0; 1.15|]
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 31)" onmouseover="showTip(event, 'fs19', 31)" class="id">peptidehydrophobicites&#39;</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs12', 32)" onmouseover="showTip(event, 'fs12', 32)" class="id">peptide</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs2', 33)" onmouseover="showTip(event, 'fs2', 33)" class="m">AminoProperties</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 34)" onmouseover="showTip(event, 'fs20', 34)" class="id">ofWindowedBioArray</span> <span class="n">3</span> <span onmouseout="hideTip(event, 'fs3', 35)" onmouseover="showTip(event, 'fs3', 35)" class="fn">getHydrophobicityIndex</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">
</code></pre></td></tr></table>
<p>In the last step you can then just sum or average over the values to get a summary value of the hydrophobicity, depending on wether you want a length dependent or independent value.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs15', 36)" onmouseover="showTip(event, 'fs15', 36)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs21', 37)" onmouseover="showTip(event, 'fs21', 37)" class="id">sum</span> <span onmouseout="hideTip(event, 'fs18', 38)" onmouseover="showTip(event, 'fs18', 38)" class="id">peptidehydrophobicites</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">[|0.89; 0.834; 0.8916666667; 1.081428571; 1.005714286; 1.107142857;
0.9057142857; 1.087142857; 1.065714286; 0.9042857143; 0.7757142857;
0.7657142857; 0.7614285714; 0.81; 0.596; 0.63|]
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs15', 39)" onmouseover="showTip(event, 'fs15', 39)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs21', 40)" onmouseover="showTip(event, 'fs21', 40)" class="id">sum</span> <span onmouseout="hideTip(event, 'fs19', 41)" onmouseover="showTip(event, 'fs19', 41)" class="id">peptidehydrophobicites&#39;</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">14.11166667
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs15', 42)" onmouseover="showTip(event, 'fs15', 42)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 43)" onmouseover="showTip(event, 'fs22', 43)" class="id">average</span> <span onmouseout="hideTip(event, 'fs18', 44)" onmouseover="showTip(event, 'fs18', 44)" class="id">peptidehydrophobicites</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">14.11166667
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs15', 45)" onmouseover="showTip(event, 'fs15', 45)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 46)" onmouseover="showTip(event, 'fs22', 46)" class="id">average</span> <span onmouseout="hideTip(event, 'fs19', 47)" onmouseover="showTip(event, 'fs19', 47)" class="id">peptidehydrophobicites&#39;</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">0.8819791667
</code></pre></td></tr></table>
<h2><a name="Isoelectric-Point" class="anchor" href="#Isoelectric-Point">Isoelectric Point</a></h2>
<p>The isoelectric point (pI) of a protein is the point at which it carries as many positive as negative charges.
Therefore the overall charge is zero. Knowing this value can e.g. be useful for isolation of single proteins in a voltage gradient.<br />
The implementation is based on: <a href="http://fields.scripps.edu/DTASelect/20010710-pI-Algorithm.pdf">this document</a>.
In principle, the distinct amino acids in the protein are counted.
By using the <a href="https://en.wikipedia.org/wiki/Henderson-Hasselbalch_equation">Henderson-Hasselbalch equation</a> and the pKr values, the theoretic charge states of the amino acids for a specific pH can be calculated.
Multiplying those charge states with the count of the associated amino acids and adding those products together then gives the overall charge of the protein. This is only done with the amino acids, which might be charged (basic, acidic).
The isoelectric point is the pH value for which this function returns zero. It is found by <a href="https://en.wikipedia.org/wiki/Bisection_method">bisection</a> (also called Binary Search).<br />
Disclaimer: Keep in mind, that this algorithm ignores post-translational modifications and interactions of the amino acids with each other. Therefore it is only intented to be a rough approximation and should be used as such.</p>
</br>
<p>The function for finding the isoelectric point is found in the <code>IsoelectricPoint</code> module.</p>
<ul>
<li>Besides the peptide sequence in form of a <code>AminoAcidSymbol</code> Seq, it takes</li>
<li>a mapping function, which maps an <code>AminoAcidSymbol</code> to a float representing the pKr and</li>
<li>an accuracy value. The found pI has to be at least as close to zero as this accuracy value</li>
</ul>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//AA sequence</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 48)" onmouseover="showTip(event, 'fs23', 48)" class="id">myProteinForPI</span> <span class="o">=</span> 
    <span class="s">&quot;ATPIIEMNYPWTMNIKLSSDACMTNWWPNCMTLKIIA&quot;</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs24', 49)" onmouseover="showTip(event, 'fs24', 49)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs25', 50)" onmouseover="showTip(event, 'fs25', 50)" class="id">map</span> <span onmouseout="hideTip(event, 'fs7', 51)" onmouseover="showTip(event, 'fs7', 51)" class="m">AminoAcidSymbols</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 52)" onmouseover="showTip(event, 'fs17', 52)" class="id">aminoAcidSymbol</span>

<span class="c">//accuracy in z</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs26', 53)" onmouseover="showTip(event, 'fs26', 53)" class="id">acc</span> <span class="o">=</span> <span class="n">0.5</span>

<span onmouseout="hideTip(event, 'fs27', 54)" onmouseover="showTip(event, 'fs27', 54)" class="m">IsoelectricPoint</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 55)" onmouseover="showTip(event, 'fs28', 55)" class="id">tryFind</span> <span onmouseout="hideTip(event, 'fs27', 56)" onmouseover="showTip(event, 'fs27', 56)" class="m">IsoelectricPoint</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 57)" onmouseover="showTip(event, 'fs29', 57)" class="id">getpKr</span> <span onmouseout="hideTip(event, 'fs26', 58)" onmouseover="showTip(event, 'fs26', 58)" class="id">acc</span> <span onmouseout="hideTip(event, 'fs23', 59)" onmouseover="showTip(event, 'fs23', 59)" class="id">myProteinForPI</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">Some 7.0
</code></pre></td></tr></table>

        <div class="fsdocs-tip" id="fs1">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs2">module AminoProperties

from BioFSharp<br /><em>&lt;summary&gt;
Contains functionalities for obtaining included literary data on key amino acid properties
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs3">val getHydrophobicityIndex: (AminoAcidSymbols.AminoAcidSymbol -&gt; float)</div>
<div class="fsdocs-tip" id="fs4">val initGetAminoProperty: property: AminoProperty -&gt; (AminoAcidSymbols.AminoAcidSymbol -&gt; float)<br /><em>&lt;summary&gt;
Returns a simple mapping function for the given amino acid property
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs5">type AminoProperty =
  | GravyScore
  | HydrophobicityIndex
  | HydrophobicityFasman
  | MeltingPointFasman
  | OpticalRotationFasman
  | PK_NFasman
  | PK_CFasman
  | NormalizedBetaSheet
  | NormalizedHelix
  | NormalizedTurn
  ...
  static member toString: (AminoProperty -&gt; string)<br /><em>&lt;summary&gt;
Union case of amino acid properties, referencing the according included information in this library. Use &quot;initGetAminoProperty&quot; function to obtain a simple mapping function
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs6">union case AminoProperty.HydrophobicityIndex: AminoProperty<br /><em>&lt;summary&gt;
 Hydrophobicity index (Argos et al., 1982)
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs7">module AminoAcidSymbols

from BioFSharp<br /><em>&lt;summary&gt;
Contains the AminoAcidSymbol type and its according functions. The AminoAcidSymbol type is a lightweight, efficient presentation of amino acids
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs8">[&lt;Struct&gt;]
type AminoAcidSymbol =
  interface IBioItem
  interface IComparable
  override Equals: other: obj -&gt; bool
  override GetHashCode: unit -&gt; int
  override ToString: unit -&gt; string
  static member op_Explicit: value: int -&gt; AminoAcidSymbol + 2 overloads
  static member Ala: AminoAcidSymbol
  static member Arg: AminoAcidSymbol
  static member Asn: AminoAcidSymbol
  static member Asp: AminoAcidSymbol
  ...</div>
<div class="fsdocs-tip" id="fs9">property AminoAcidSymbols.AminoAcidSymbol.Ala: AminoAcidSymbols.AminoAcidSymbol with get<br /><em>&lt;summary&gt;
 &#39;A&#39;  *Alanin
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs10">val getHydrophobicityIndexZ: (AminoAcidSymbols.AminoAcidSymbol -&gt; float)</div>
<div class="fsdocs-tip" id="fs11">val initGetAminoPropertyZnorm: property: AminoProperty -&gt; (AminoAcidSymbols.AminoAcidSymbol -&gt; float)<br /><em>&lt;summary&gt;
Returns a simple mapping function for the given amino acid property. Normalizes the values to the Z-Norm scale
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs12">val peptide: AminoAcidSymbols.AminoAcidSymbol array</div>
<div class="fsdocs-tip" id="fs13">Multiple items<br />module BioArray

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioArray

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioArray type and its according functions. The BioArray type is an array of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs14">val ofAminoAcidString: s: #(char seq) -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 Generates amino acid sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs15">module Array

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs16">val map: mapping: (&#39;T -&gt; &#39;U) -&gt; array: &#39;T array -&gt; &#39;U array</div>
<div class="fsdocs-tip" id="fs17">val aminoAcidSymbol: a: &#39;a -&gt; AminoAcidSymbols.AminoAcidSymbol (requires member op_Explicit)<br /><em>&lt;summary&gt;
 Maps input to AminoAcidSymbol if possible
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs18">val peptidehydrophobicites: float array</div>
<div class="fsdocs-tip" id="fs19">val peptidehydrophobicites&#39;: float array</div>
<div class="fsdocs-tip" id="fs20">val ofWindowedBioArray: n: int -&gt; pf: (&#39;a -&gt; float) -&gt; source: BioArray.BioArray&lt;&#39;a&gt; -&gt; float array (requires &#39;a :&gt; IBioItem)<br /><em>&lt;summary&gt;
 Returns an array of sliding windows based property averages.
 Each window contains the n elements surrounding the current element
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs21">val sum: array: &#39;T array -&gt; &#39;T (requires member (+) and member Zero)</div>
<div class="fsdocs-tip" id="fs22">val average: array: &#39;T array -&gt; &#39;T (requires member (+) and member DivideByInt and member Zero)</div>
<div class="fsdocs-tip" id="fs23">val myProteinForPI: AminoAcidSymbols.AminoAcidSymbol seq</div>
<div class="fsdocs-tip" id="fs24">module Seq

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs25">val map: mapping: (&#39;T -&gt; &#39;U) -&gt; source: &#39;T seq -&gt; &#39;U seq</div>
<div class="fsdocs-tip" id="fs26">val acc: float</div>
<div class="fsdocs-tip" id="fs27">module IsoelectricPoint

from BioFSharp<br /><em>&lt;summary&gt;
Finding the isoelectric point of peptides
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs28">val tryFind: pKrFunc: (AminoAcidSymbols.AminoAcidSymbol -&gt; float) -&gt; accuracy: float -&gt; aaSeq: AminoAcidSymbols.AminoAcidSymbol seq -&gt; float option<br /><em>&lt;summary&gt;
Finds the pH for which the global charge of the aaSeq is closer to 0 than the given accuracy.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs29">val getpKr: (AminoAcidSymbols.AminoAcidSymbol -&gt; float)<br /><em>&lt;summary&gt;
Maps AminoAcidSymbol to default pK value of it&#39;s sidechain. Returns 0.0 if sidechain is neither acidic nor basic
&lt;/summary&gt;</em></div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#Sequence-Properties">
      Sequence Properties
    </a>
  </li>
  <li class="level-2">
    <a href="#General">
      General
    </a>
  </li>
  <li class="level-2">
    <a href="#Isoelectric-Point">
      Isoelectric Point
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>