<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="BioCollections" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="BioCollections">
    
    <title>BioCollections | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.1/es5/tex-mml-chtml.js"></script>
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="BioCollections" class="anchor" href="#BioCollections">BioCollections</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BioFSharp/gh-pages?filepath=01_03_biocollections.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-binder.svg" alt="Binder" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/01_03_biocollections.fsx"><img src="https://csbiology.github.io/BioFSharp/img/badge-script.svg" alt="Script" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/01_03_biocollections.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-notebook.svg" alt="Notebook" /></a></p>
<p><em>Summary:</em> This example shows how to use collections of biological items in BioFSharp</p>
<p>Analogous to the build-in collections BioFSharp provides BioSeq, BioList and BioArray for individual collection specific optimized operations.
The easiest way to create them are the <code>ofBioItemString</code> -functions</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">s1</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 4)" onmouseover="showTip(event, 'fs4', 4)" class="id">ofAminoAcidString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 5)" onmouseover="showTip(event, 'fs5', 5)" class="id">s2</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 6)" onmouseover="showTip(event, 'fs6', 6)" class="m">BioList</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 7)" onmouseover="showTip(event, 'fs7', 7)" class="id">ofAminoAcidString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs8', 8)" onmouseover="showTip(event, 'fs8', 8)" class="id">s3</span> <span class="o">=</span> <span class="s">&quot;TAGCAT&quot;</span>  <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 9)" onmouseover="showTip(event, 'fs9', 9)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 10)" onmouseover="showTip(event, 'fs10', 10)" class="id">ofNucleotideString</span> 

<span class="c">//Peptide represented as a Bioseq</span>
<span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs3', 11)" onmouseover="showTip(event, 'fs3', 11)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 12)" onmouseover="showTip(event, 'fs4', 12)" class="id">ofAminoAcidString</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//Peptide represented as a BioList</span>
<span class="s">&quot;PEPTIDE&quot;</span><span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 13)" onmouseover="showTip(event, 'fs6', 13)" class="m">BioList</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 14)" onmouseover="showTip(event, 'fs7', 14)" class="id">ofAminoAcidString</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//Nucleotide sequence represented as a BioArray</span>
<span class="s">&quot;TAGCAT&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 15)" onmouseover="showTip(event, 'fs9', 15)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 16)" onmouseover="showTip(event, 'fs10', 16)" class="id">ofNucleotideString</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<h2><a name="Nucleotides" class="anchor" href="#Nucleotides">Nucleotides</a></h2>
<p><img src="img/Nucleotides.svg" alt="Nucleotides1" /></p>
<p><strong>Figure 1: Selection of covered nucleotide operations</strong> (A) Biological principle. (B) Workflow with <code>BioSeq</code>. (C) Other covered functionalities.</p>
<p>Let's imagine you have a given gene sequence and want to find out what the according protein might look like.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 17)" onmouseover="showTip(event, 'fs11', 17)" class="id">myGene</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs9', 18)" onmouseover="showTip(event, 'fs9', 18)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 19)" onmouseover="showTip(event, 'fs10', 19)" class="id">ofNucleotideString</span> <span class="s">&quot;ATGGCTAGATCGATCGATCGGCTAACGTAA&quot;</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<p>Yikes! Unfortunately we got the 5'-3' coding strand. For proper transcription we should get the complementary strand first:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 20)" onmouseover="showTip(event, 'fs12', 20)" class="id">myProperGene</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs9', 21)" onmouseover="showTip(event, 'fs9', 21)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 22)" onmouseover="showTip(event, 'fs13', 22)" class="id">complement</span> <span onmouseout="hideTip(event, 'fs11', 23)" onmouseover="showTip(event, 'fs11', 23)" class="id">myGene</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<p>Now let's transcribe and translate it:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 24)" onmouseover="showTip(event, 'fs14', 24)" class="id">myTranslatedGene</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs12', 25)" onmouseover="showTip(event, 'fs12', 25)" class="id">myProperGene</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 26)" onmouseover="showTip(event, 'fs9', 26)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs15', 27)" onmouseover="showTip(event, 'fs15', 27)" class="id">transcribeTemplateStrand</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 28)" onmouseover="showTip(event, 'fs9', 28)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 29)" onmouseover="showTip(event, 'fs16', 29)" class="id">translate</span> <span class="n">0</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<p>Of course, if your input sequence originates from the coding strand, you can directly transcribe it to mRNA since the
only difference between the coding strand and the mRNA is the replacement of 'T' by 'U' (Figure 1B)</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 30)" onmouseover="showTip(event, 'fs17', 30)" class="id">myTranslatedGeneFromCodingStrand</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs11', 31)" onmouseover="showTip(event, 'fs11', 31)" class="id">myGene</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 32)" onmouseover="showTip(event, 'fs9', 32)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 33)" onmouseover="showTip(event, 'fs18', 33)" class="id">transcribeCodingStrand</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 34)" onmouseover="showTip(event, 'fs9', 34)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 35)" onmouseover="showTip(event, 'fs16', 35)" class="id">translate</span> <span class="n">0</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<p>Other Nucleotide conversion operations are also covered:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 36)" onmouseover="showTip(event, 'fs19', 36)" class="id">mySmallGene</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 37)" onmouseover="showTip(event, 'fs3', 37)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 38)" onmouseover="showTip(event, 'fs20', 38)" class="id">ofNucleotideString</span>  <span class="s">&quot;ATGTTCCGAT&quot;</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 39)" onmouseover="showTip(event, 'fs21', 39)" class="id">smallGeneRev</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 40)" onmouseover="showTip(event, 'fs3', 40)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 41)" onmouseover="showTip(event, 'fs22', 41)" class="id">reverse</span> <span onmouseout="hideTip(event, 'fs19', 42)" onmouseover="showTip(event, 'fs19', 42)" class="id">mySmallGene</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 43)" onmouseover="showTip(event, 'fs23', 43)" class="id">smallGeneComp</span>    <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 44)" onmouseover="showTip(event, 'fs3', 44)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 45)" onmouseover="showTip(event, 'fs24', 45)" class="id">complement</span> <span onmouseout="hideTip(event, 'fs19', 46)" onmouseover="showTip(event, 'fs19', 46)" class="id">mySmallGene</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 47)" onmouseover="showTip(event, 'fs25', 47)" class="id">smallGeneRevComp</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 48)" onmouseover="showTip(event, 'fs3', 48)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 49)" onmouseover="showTip(event, 'fs26', 49)" class="id">reverseComplement</span> <span onmouseout="hideTip(event, 'fs19', 50)" onmouseover="showTip(event, 'fs19', 50)" class="id">mySmallGene</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<h2><a name="AminoAcids" class="anchor" href="#AminoAcids">AminoAcids</a></h2>
<h3><a name="Basics" class="anchor" href="#Basics">Basics</a></h3>
<p>Some functions which might be needed regularly are defined to work with nucleotides and amino acids:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 51)" onmouseover="showTip(event, 'fs27', 51)" class="id">myPeptide</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs3', 52)" onmouseover="showTip(event, 'fs3', 52)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 53)" onmouseover="showTip(event, 'fs4', 53)" class="id">ofAminoAcidString</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 54)" onmouseover="showTip(event, 'fs28', 54)" class="id">myPeptideFormula</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 55)" onmouseover="showTip(event, 'fs3', 55)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 56)" onmouseover="showTip(event, 'fs29', 56)" class="id">toFormula</span> <span onmouseout="hideTip(event, 'fs27', 57)" onmouseover="showTip(event, 'fs27', 57)" class="id">myPeptide</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs30', 58)" onmouseover="showTip(event, 'fs30', 58)" class="m">Formula</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 59)" onmouseover="showTip(event, 'fs31', 59)" class="id">toString</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs32', 60)" onmouseover="showTip(event, 'fs32', 60)" class="id">myPeptideMass</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs3', 61)" onmouseover="showTip(event, 'fs3', 61)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 62)" onmouseover="showTip(event, 'fs33', 62)" class="id">toAverageMass</span> <span onmouseout="hideTip(event, 'fs27', 63)" onmouseover="showTip(event, 'fs27', 63)" class="id">myPeptide</span> 
</code></pre>
Warning: Output, it-value and value references require --eval
<h3><a name="Digestion" class="anchor" href="#Digestion">Digestion</a></h3>
<p>BioFSharp also comes equipped with a set of tools aimed at cutting apart amino acid sequences. To demonstrate the usage, we'll throw some <code>trypsin</code> at the small RuBisCO subunit of <em>Arabidopos thaliana</em>:<br />
In the first step, we define our input sequence and the protease we want to use.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs34', 64)" onmouseover="showTip(event, 'fs34', 64)" class="id">RBCS</span> <span class="o">=</span> 
    <span class="s">&quot;&quot;&quot;MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT&quot;&quot;&quot;</span> 
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 65)" onmouseover="showTip(event, 'fs9', 65)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs35', 66)" onmouseover="showTip(event, 'fs35', 66)" class="id">ofAminoAcidString</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs36', 67)" onmouseover="showTip(event, 'fs36', 67)" class="id">trypsin</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 68)" onmouseover="showTip(event, 'fs37', 68)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 69)" onmouseover="showTip(event, 'fs38', 69)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 70)" onmouseover="showTip(event, 'fs39', 70)" class="id">getProteaseBy</span> <span class="s">&quot;Trypsin&quot;</span>
</code></pre>
<p>With these two things done, digesting the protein is a piece of cake. For doing this, just use the <code>digest</code> function.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs40', 71)" onmouseover="showTip(event, 'fs40', 71)" class="id">digestedRBCS</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 72)" onmouseover="showTip(event, 'fs37', 72)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 73)" onmouseover="showTip(event, 'fs41', 73)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs42', 74)" onmouseover="showTip(event, 'fs42', 74)" class="id">digest</span> <span onmouseout="hideTip(event, 'fs36', 75)" onmouseover="showTip(event, 'fs36', 75)" class="id">trypsin</span> <span class="n">0</span> <span onmouseout="hideTip(event, 'fs34', 76)" onmouseover="showTip(event, 'fs34', 76)" class="id">RBCS</span> 

<span onmouseout="hideTip(event, 'fs40', 77)" onmouseover="showTip(event, 'fs40', 77)" class="id">digestedRBCS</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs43', 78)" onmouseover="showTip(event, 'fs43', 78)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs44', 79)" onmouseover="showTip(event, 'fs44', 79)" class="id">head</span>
</code></pre>
Warning: Output, it-value and value references require --eval
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">(*</span>
<span class="c">In reality, proteases don&#39;t always completely cut the protein down. Instead, some sites stay intact and should be considered for in silico analysis. </span>
<span class="c">This can easily be done with the `concernMissCleavages` function. It takes the minimum and maximum amount of misscleavages you want to have and also the digested protein. As a result you get all possible combinations arising from this information.</span>
<span class="c">*)</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs45', 80)" onmouseover="showTip(event, 'fs45', 80)" class="id">digestedRBCS&#39;</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 81)" onmouseover="showTip(event, 'fs37', 81)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 82)" onmouseover="showTip(event, 'fs41', 82)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 83)" onmouseover="showTip(event, 'fs46', 83)" class="id">concernMissCleavages</span> <span class="n">0</span> <span class="n">2</span> <span onmouseout="hideTip(event, 'fs40', 84)" onmouseover="showTip(event, 'fs40', 84)" class="id">digestedRBCS</span>

<span onmouseout="hideTip(event, 'fs40', 85)" onmouseover="showTip(event, 'fs40', 85)" class="id">digestedRBCS</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs43', 86)" onmouseover="showTip(event, 'fs43', 86)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs47', 87)" onmouseover="showTip(event, 'fs47', 87)" class="id">item</span> <span class="n">1</span>
</code></pre>
Warning: Output, it-value and value references require --eval

        <div class="fsdocs-tip" id="fs1">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs2">val s1: BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs3">Multiple items<br />module BioSeq

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioSeq

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioSeq type and its according functions. The BioSeq type is a sequence of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs4">val ofAminoAcidString: s: #(char seq) -&gt; BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 Generates AminoAcid sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs5">val s2: BioList.BioList&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs6">Multiple items<br />module BioList

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioList

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioList type and its according functions. The BioList type is a List of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs7">val ofAminoAcidString: s: #(char seq) -&gt; BioList.BioList&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 Generates amino acid sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs8">val s3: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs9">Multiple items<br />module BioArray

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioArray

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioArray type and its according functions. The BioArray type is an array of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs10">val ofNucleotideString: s: #(char seq) -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Generates nucleotide sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs11">val myGene: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs12">val myProperGene: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs13">val complement: nucs: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Create the complement DNA or cDNA (from RNA) strand. For example, the sequence &quot;ATGC&quot; is converted to &quot;TACG&quot;
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs14">val myTranslatedGene: BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs15">val transcribeTemplateStrand: nucs: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Transcribe a given DNA template strand (3&#39;-----5&#39;)
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs16">val translate: nucleotideOffset: int -&gt; rnaSeq: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 translates nucleotide sequence to aminoacid sequence    
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs17">val myTranslatedGeneFromCodingStrand: BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs18">val transcribeCodingStrand: nucs: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Transcribe a given DNA coding strand (5&#39;-----3&#39;)
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs19">val mySmallGene: BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs20">val ofNucleotideString: s: #(char seq) -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Generates nucleotide sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs21">val smallGeneRev: BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs22">val reverse: nucs: Nucleotides.Nucleotide seq -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Create the reverse DNA or RNA strand. For example, the sequence &quot;ATGC&quot; is converted to &quot;CGTA&quot;
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs23">val smallGeneComp: BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs24">val complement: nucs: Nucleotides.Nucleotide seq -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Create the complement DNA or cDNA (from RNA) strand. For example, the sequence &quot;ATGC&quot; is converted to &quot;TACG&quot;
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs25">val smallGeneRevComp: BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs26">val reverseComplement: nucs: Nucleotides.Nucleotide seq -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Create the reverse complement strand meaning antiparallel DNA strand or the cDNA (from RNA) respectivly. For example, the sequence &quot;ATGC&quot; is converted to &quot;GCAT&quot;. &quot;Antiparallel&quot; combines the two functions &quot;Complement&quot; and &quot;Inverse&quot;.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs27">val myPeptide: BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs28">val myPeptideFormula: string</div>
<div class="fsdocs-tip" id="fs29">val toFormula: bs: #IBioItem seq -&gt; Formula.Formula<br /><em>&lt;summary&gt;
 Returns formula
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs30">module Formula

from BioFSharp<br /><em>&lt;summary&gt;
Contains functionality for working with molecules as a formula of their elements and formulas of biologically relevant molecules
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs31">val toString: f: Formula.Formula -&gt; string<br /><em>&lt;summary&gt;
 Returns Formula as string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs32">val myPeptideMass: float</div>
<div class="fsdocs-tip" id="fs33">val toAverageMass: bs: #IBioItem seq -&gt; float<br /><em>&lt;summary&gt;
 Returns average mass of the given sequence
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs34">val RBCS: BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="fsdocs-tip" id="fs35">val ofAminoAcidString: s: #(char seq) -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 Generates amino acid sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs36">val trypsin: Digestion.Protease</div>
<div class="fsdocs-tip" id="fs37">module Digestion

from BioFSharp<br /><em>&lt;summary&gt;
Contains types and functions needed to digest amino acid sequences with proteases
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs38">module Table

from BioFSharp.Digestion<br /><em>&lt;summary&gt;
Contains frequently needed proteases
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs39">val getProteaseBy: name: string -&gt; Digestion.Protease</div>
<div class="fsdocs-tip" id="fs40">val digestedRBCS: Digestion.DigestedPeptide&lt;int&gt; array</div>
<div class="fsdocs-tip" id="fs41">module BioArray

from BioFSharp.Digestion</div>
<div class="fsdocs-tip" id="fs42">val digest: protease: Digestion.Protease -&gt; proteinID: &#39;a -&gt; aas: AminoAcids.AminoAcid array -&gt; Digestion.DigestedPeptide&lt;&#39;a&gt; array<br /><em>&lt;summary&gt;
 Takes Proteinsequence as input and returns Array of resulting DigestedPeptides
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs43">module Seq

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs44">val head: source: &#39;T seq -&gt; &#39;T</div>
<div class="fsdocs-tip" id="fs45">val digestedRBCS&#39;: Digestion.DigestedPeptide&lt;int&gt; array</div>
<div class="fsdocs-tip" id="fs46">val concernMissCleavages: minMissCleavages: int -&gt; maxMisscleavages: int -&gt; digestedPeptides: Digestion.DigestedPeptide&lt;&#39;a&gt; array -&gt; Digestion.DigestedPeptide&lt;&#39;a&gt; array (requires equality)<br /><em>&lt;summary&gt;
 Returns Sequence of DigestedPeptides including those resulting of one or more Misscleavage events.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs47">val item: index: int -&gt; source: &#39;T seq -&gt; &#39;T</div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#BioCollections">
      BioCollections
    </a>
  </li>
  <li class="level-2">
    <a href="#Nucleotides">
      Nucleotides
    </a>
  </li>
  <li class="level-2">
    <a href="#AminoAcids">
      AminoAcids
    </a>
  </li>
  <li class="level-3">
    <a href="#Basics">
      Basics
    </a>
  </li>
  <li class="level-3">
    <a href="#Digestion">
      Digestion
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>