<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="Blast" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="Blast">
    
    <title>Blast | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.1/es5/tex-mml-chtml.js"></script>
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header active">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    rnaseq_normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header active">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    rnaseq_normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="Blast" class="anchor" href="#Blast">Blast</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BioFSharp/gh-pages?filepath=02_02_blast.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-binder.svg" alt="Binder" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/02_02_blast.fsx"><img src="https://csbiology.github.io/BioFSharp/img/badge-script.svg" alt="Script" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/02_02_blast.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-notebook.svg" alt="Notebook" /></a></p>
<p><em>Summary:</em> This example shows how to perform blast with BioFSharp</p>
<p>BioFSharp's BlastWrapper is a tool for performing different tasks in NCBI BLAST console applications (version 2.2.31+).
It is able to create BLAST databases and perform <strong>blastN</strong> or <strong>blastP</strong> queries, while providing a way to set
output parameter for creating a custom output format.
Official documentation for all BLAST applications can be found <a href="http://www.ncbi.nlm.nih.gov/books/NBK279690">here</a>.</p>
<p>For the purpose of this tutorial, we will build a protein database using a <code>.fastA</code> file containing chloroplast proteins
of Chlamydomonas reinhardtii included in BioFSharp/docs/content/data.</p>
<p>Our query protein for the subsequent BLAST search will be the <a href="http://www.ncbi.nlm.nih.gov/protein/7525013?report=fasta">photosystem II protein D1</a> from Arabidopsis thaliana chloroplast.</p>
<h2><a name="Creation-of-a-BLAST-database" class="anchor" href="#Creation-of-a-BLAST-database">Creation of a BLAST database</a></h2>
<p>We will use the minimal amount of parameters needed to create a BLAST database from an input file.
The created database files will have the same name as the input file and will be located in the same folder.
However, there are many parameters you can use to specify your database. Please refer to the <a href="http://www.ncbi.nlm.nih.gov/books/NBK279675/">NCBI user manual</a> for more information.</p>
<p>First, lets specify the path of our input and the type of our database. Use a string for the input path and the provided <code>MakeDbParams</code> type
for every other parameter.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="id">IO</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="m">BlastNCBI</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="m">Parameters</span>

<span class="c">///path and name of the input file/output database. </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="id">inputFile</span> <span class="o">=</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Chlamy_Cp.fastA&quot;</span><span class="pn">)</span>

<span class="c">///defines database type (in this case: a protein database)</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="id">typeOfDatabase</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 8)" onmouseover="showTip(event, 'fs4', 8)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="rt">MakeDbParams</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="id">DbType</span> <span onmouseout="hideTip(event, 'fs4', 11)" onmouseover="showTip(event, 'fs4', 11)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 12)" onmouseover="showTip(event, 'fs9', 12)" class="id">Protein</span>
</code></pre>
<p>The wrapper needs to know the path of the ncbi applications.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///the path of the /bin folder where the BLAST applications are located</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 13)" onmouseover="showTip(event, 'fs10', 13)" class="id">ncbiPath</span> <span class="o">=</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/../../lib/ncbi-blast/bin&quot;</span><span class="pn">)</span>
</code></pre>
<p>We now provide the wrapper our ncbi path, the input path and a sequence of parameters (containing just one parameter in this case, the database type).</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs11', 14)" onmouseover="showTip(event, 'fs11', 14)" class="fn">BlastWrapper</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="id">ncbiPath</span><span class="pn">)</span><span class="pn">.</span><span class="id">makeblastdb</span> <span onmouseout="hideTip(event, 'fs5', 16)" onmouseover="showTip(event, 'fs5', 16)" class="id">inputFile</span> <span class="pn">(</span><span class="pn">[</span><span onmouseout="hideTip(event, 'fs6', 17)" onmouseover="showTip(event, 'fs6', 17)" class="id">typeOfDatabase</span><span class="pn">;</span><span class="pn">]</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 18)" onmouseover="showTip(event, 'fs12', 18)" class="fn">seq</span><span class="pn">&lt;</span><span onmouseout="hideTip(event, 'fs4', 19)" onmouseover="showTip(event, 'fs4', 19)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 20)" onmouseover="showTip(event, 'fs7', 20)" class="id">MakeDbParams</span><span class="pn">&gt;</span><span class="pn">)</span>
</code></pre>
<p>Your console output will look like this:</p>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="txt">Starting Makeblastdb...


Building a new DB, current time: 12/05/2018 09:02:20
New DB name:   C:\Users\Kevin\source\repos\CSBiology\BioFSharp\docsrc\content/data/Chlamy_Cp.fastA
New DB title:  C:\Users\Kevin\source\repos\CSBiology\BioFSharp\docsrc\content/data/Chlamy_Cp.fastA
Sequence type: Protein
Deleted existing Protein BLAST database named C:\Users\Kevin\source\repos\CSBiology\BioFSharp\docsrc\content/data/Chlamy_Cp.fastA
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 74 sequences in 0.0107442 seconds.
Makeblastdb done.
</code></pre></td></tr></table>
<p>This creates 3 new files in our directory:
<code>Chlamy_Cp.fastA.phr</code>,<code>Chlamy_Cp.fastA.pin</code> and <code>Chlamy_Cp.fastA.psq</code>.</p>
<p>We have sucesssfully created our search database.</p>
<h2><a name="Creating-a-fasta-file-from-an-aminoacid-string" class="anchor" href="#Creating-a-fasta-file-from-an-aminoacid-string">Creating a fasta file from an aminoacid string</a></h2>
<p><em>Note: this step is not necessary if you want to use an already existing file as query. If this is the case, skip to step 3.</em></p>
<p>First, lets specify a string with our aminoacid sequence and convert it to a <code>BioSeq</code>.
For more information about <code>BioSeq</code>, please refer to this <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-bioseq.html">documentation</a></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Raw string of the aminoacid sequence of our query protein</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 21)" onmouseover="showTip(event, 'fs13', 21)" class="id">aminoacidString</span> <span class="o">=</span> <span class="s">&quot;MTAILERRESESLWGRFCNWITSTENRLYIGWFGVLMIPTLLTATSVFIIAFIAAPPVDIDGIREPVSGS</span>
<span class="s">LLYGNNIISGAIIPTSAAIGLHFYPIWEAASVDEWLYNGGPYELIVLHFLLGVACYMGREWELSFRLGMR</span>
<span class="s">PWIAVAYSAPVAAATAVFLIYPIGQGSFSDGMPLGISGTFNFMIVFQAEHNILMHPFHMLGVAGVFGGSL</span>
<span class="s">FSAMHGSLVTSSLIRETTENESANEGYRFGQEEETYNIVAAHGYFGRLIFQYASFNNSRSLHFFLAAWPV</span>
<span class="s">VGIWFTALGISTMAFNLNGFNFNQSVVDSQGRVINTWADIINRANLGMEVMHERNAHNFPLDLAAVEAPS</span>
<span class="s">TNG&quot;</span>
<span class="c">///header for the .fastA file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 22)" onmouseover="showTip(event, 'fs14', 22)" class="id">header</span> <span class="o">=</span> <span class="s">&quot;&gt;gi|7525013|ref|NP_051039.1| photosystem II protein D1 (chloroplast) [Arabidopsis thaliana]&quot;</span>

<span class="c">///Query sequency represented as a sequence of `AminoAcid` one of BioFSharp&#39;s `BioItems`</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 23)" onmouseover="showTip(event, 'fs15', 23)" class="id">querySequence</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs16', 24)" onmouseover="showTip(event, 'fs16', 24)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 25)" onmouseover="showTip(event, 'fs17', 25)" class="id">ofAminoAcidString</span> <span onmouseout="hideTip(event, 'fs13', 26)" onmouseover="showTip(event, 'fs13', 26)" class="id">aminoacidString</span>
</code></pre>
<p>We will now use BioFSharp's <code>FastA</code> library to create a <code>FASTA</code> item and write it to a file.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///path and name of the query file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 27)" onmouseover="showTip(event, 'fs18', 27)" class="id">queryFastaPath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/testQuery.fastA&quot;</span>

<span class="c">///FastaItem containing header string and query sequence</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 28)" onmouseover="showTip(event, 'fs19', 28)" class="id">queryFastaItem</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs20', 29)" onmouseover="showTip(event, 'fs20', 29)" class="m">FastA</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs21', 30)" onmouseover="showTip(event, 'fs21', 30)" class="id">createFastaItem</span> <span onmouseout="hideTip(event, 'fs14', 31)" onmouseover="showTip(event, 'fs14', 31)" class="id">header</span> <span onmouseout="hideTip(event, 'fs15', 32)" onmouseover="showTip(event, 'fs15', 32)" class="id">querySequence</span>
</code></pre>
<p>To create our <code>.fastA</code> file, we need to use the <code>BioItem.symbol</code> converter, which will convert the 3 letter code of the aminoacids in our biosequence
to the one letter symbol (eg. Met -&gt; M)</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs20', 33)" onmouseover="showTip(event, 'fs20', 33)" class="m">FastA</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 34)" onmouseover="showTip(event, 'fs22', 34)" class="id">write</span> <span onmouseout="hideTip(event, 'fs23', 35)" onmouseover="showTip(event, 'fs23', 35)" class="m">BioItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 36)" onmouseover="showTip(event, 'fs24', 36)" class="id">symbol</span> <span onmouseout="hideTip(event, 'fs18', 37)" onmouseover="showTip(event, 'fs18', 37)" class="id">queryFastaPath</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs19', 38)" onmouseover="showTip(event, 'fs19', 38)" class="id">queryFastaItem</span><span class="pn">;</span><span class="pn">]</span> 
</code></pre>
<h2><a name="Performing-the-BLAST-search" class="anchor" href="#Performing-the-BLAST-search">Performing the BLAST search</a></h2>
<p>We have created our search database and the query we want to find. Before we can perform the actual search, we need to define the BLAST prameters.</p>
<p><em>Note: custom output formats can only be specified for output types <code>CSV</code>, <code>tabular</code> and <code>tabular with comments</code>. For more information, check
the <a href="http://www.ncbi.nlm.nih.gov/books/NBK279675/">options for the command-line applicaions</a></em></p>
<p>First, lets specify the overall output type. This will define the outline of our output. We want our output to be in tabular form, with added information
in the form of comments.</p>
<p><em>Note: when not specified otherwise, the output type will be <code>pairwise</code></em></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///overall outline of the output </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 39)" onmouseover="showTip(event, 'fs25', 39)" class="id">outputType</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs26', 40)" onmouseover="showTip(event, 'fs26', 40)" class="en">OutputType</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs27', 41)" onmouseover="showTip(event, 'fs27', 41)" class="id">TabularWithComments</span>
</code></pre>
<p>We have a large selection of parameters that we can include in the output.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///a sequence of custom output format parameters</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 42)" onmouseover="showTip(event, 'fs28', 42)" class="id">outputFormat</span><span class="o">=</span> 
    
    <span class="pn">[</span>   
        <span onmouseout="hideTip(event, 'fs29', 43)" onmouseover="showTip(event, 'fs29', 43)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs30', 44)" onmouseover="showTip(event, 'fs30', 44)" class="id">Query_SeqId</span><span class="pn">;</span> 
        <span onmouseout="hideTip(event, 'fs29', 45)" onmouseover="showTip(event, 'fs29', 45)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 46)" onmouseover="showTip(event, 'fs31', 46)" class="id">Subject_SeqId</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 47)" onmouseover="showTip(event, 'fs29', 47)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs32', 48)" onmouseover="showTip(event, 'fs32', 48)" class="id">Query_Length</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 49)" onmouseover="showTip(event, 'fs29', 49)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 50)" onmouseover="showTip(event, 'fs33', 50)" class="id">Subject_Length</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 51)" onmouseover="showTip(event, 'fs29', 51)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs34', 52)" onmouseover="showTip(event, 'fs34', 52)" class="id">AlignmentLength</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 53)" onmouseover="showTip(event, 'fs29', 53)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs35', 54)" onmouseover="showTip(event, 'fs35', 54)" class="id">MismatchCount</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 55)" onmouseover="showTip(event, 'fs29', 55)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs36', 56)" onmouseover="showTip(event, 'fs36', 56)" class="id">IdentityCount</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 57)" onmouseover="showTip(event, 'fs29', 57)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs37', 58)" onmouseover="showTip(event, 'fs37', 58)" class="id">PositiveScoringMatchCount</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 59)" onmouseover="showTip(event, 'fs29', 59)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 60)" onmouseover="showTip(event, 'fs38', 60)" class="id">Evalue</span><span class="pn">;</span>
        <span onmouseout="hideTip(event, 'fs29', 61)" onmouseover="showTip(event, 'fs29', 61)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 62)" onmouseover="showTip(event, 'fs39', 62)" class="id">Bitscore</span><span class="pn">;</span>
    <span class="pn">]</span> 
        <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs40', 63)" onmouseover="showTip(event, 'fs40', 63)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 64)" onmouseover="showTip(event, 'fs41', 64)" class="id">toSeq</span>
</code></pre>
<p>Finally, we create a <code>BlastParam</code> of the type <code>OutputTypeCustom</code> from a touple of <code>outputType</code> and <code>outputFormat</code>.</p>
<p><em>Note: No touple required if you want to use the default output format. If this is the case,
just create a <code>BlastParam</code> of type <code>OutputType</code>.</em></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///The final output format</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs42', 65)" onmouseover="showTip(event, 'fs42', 65)" class="id">customOutputFormat</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs43', 66)" onmouseover="showTip(event, 'fs43', 66)" class="uc">OutputTypeCustom</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs25', 67)" onmouseover="showTip(event, 'fs25', 67)" class="id">outputType</span> <span class="pn">,</span> <span onmouseout="hideTip(event, 'fs28', 68)" onmouseover="showTip(event, 'fs28', 68)" class="id">outputFormat</span><span class="pn">)</span>
</code></pre>
<p>We now have everything set up to perform the BLAST search. As we are talking about proteins, we will use blastP. The parameters needed for the Wrapper function are:</p>
<ul>
<li>path of the ncbi/bin folder</li>
<li>path and name of the search database</li>
<li>path and name of the query</li>
<li>path and name of our output file</li>
<li>a sequence of BLAST parameters, containing any parameters additional to the above (like our customOutputFormat)</li>
</ul>
<p><em>Note: in this case we can use the string <code>inputFile</code> that we used above for creating our database, as we did not specify another path or name for our database. Adjust accordingly if
done otherwise</em></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///output file of the BLAST search</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 69)" onmouseover="showTip(event, 'fs44', 69)" class="id">outputPath</span> <span class="o">=</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Output.txt&quot;</span><span class="pn">)</span> 
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs11', 70)" onmouseover="showTip(event, 'fs11', 70)" class="fn">BlastWrapper</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs10', 71)" onmouseover="showTip(event, 'fs10', 71)" class="id">ncbiPath</span><span class="pn">)</span><span class="pn">.</span><span class="id">blastP</span> <span onmouseout="hideTip(event, 'fs5', 72)" onmouseover="showTip(event, 'fs5', 72)" class="id">inputFile</span> <span onmouseout="hideTip(event, 'fs18', 73)" onmouseover="showTip(event, 'fs18', 73)" class="id">queryFastaPath</span> <span onmouseout="hideTip(event, 'fs44', 74)" onmouseover="showTip(event, 'fs44', 74)" class="id">outputPath</span> <span class="pn">(</span><span class="pn">[</span><span onmouseout="hideTip(event, 'fs42', 75)" onmouseover="showTip(event, 'fs42', 75)" class="id">customOutputFormat</span><span class="pn">;</span><span class="pn">]</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 76)" onmouseover="showTip(event, 'fs12', 76)" class="fn">seq</span><span class="pn">&lt;</span><span onmouseout="hideTip(event, 'fs45', 77)" onmouseover="showTip(event, 'fs45', 77)" class="rt">BlastParams</span><span class="pn">&gt;</span><span class="pn">)</span>
</code></pre>
<p>As you can see in the result file, the format is tab separated and contains the fields we specified in our our <code>customOutputFormat</code>.</p>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="txt"># BLASTP 2.2.31+
# Query: &gt;gi|7525013|ref|NP_051039.1| photosystem II protein D1 (chloroplast) [Arabidopsis thaliana]
# Database: C:\Users\Kevin\source\repos\CSBiology\BioFSharp\docsrc\content/data/Chlamy_Cp.fastA
# Fields: query id, subject id, query length, subject length, alignment length, mismatches, identical, positives, evalue, bit score
# 8 hits found
&gt;gi|7525013|ref|NP_051039.1|    sp|P19547|    353    353    346     20      326     338     0.0     645
&gt;gi|7525013|ref|NP_051039.1|    sp|P19546|    353    353    346     20      326     338     0.0     645
&gt;gi|7525013|ref|NP_051039.1|    sp|P19587|    353    353    303     199     93      162     9e-045  152
&gt;gi|7525013|ref|NP_051039.1|    sp|P19592|    353    490    49      32      14      24      2.2     21.6
&gt;gi|7525013|ref|NP_051039.1|    sp|P19592|    353    490    15      10      5       12      8.7     19.6
&gt;gi|7525013|ref|NP_051039.1|    sp|P19571|    353    3121   25      13      11      15      5.4     20.4
&gt;gi|7525013|ref|NP_051039.1|    sp|P19580|    353    470    30      20      10      12      6.5     20.0
&gt;gi|7525013|ref|NP_051039.1|    sp|P19565|    353    627    51      26      16      23      8.6     20.0
# BLAST processed 1 queries
</code></pre></td></tr></table>

        <div class="fsdocs-tip" id="fs1">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs2">namespace BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs3">module BlastNCBI

from BioFSharp.IO<br /><em>&lt;summary&gt;
Blast Wrapper
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs4">module Parameters

from BioFSharp.IO.BlastNCBI</div>
<div class="fsdocs-tip" id="fs5">val inputFile: string<br /><em>path and name of the input file/output database. </em></div>
<div class="fsdocs-tip" id="fs6">val typeOfDatabase: MakeDbParams<br /><em>defines database type (in this case: a protein database)</em></div>
<div class="fsdocs-tip" id="fs7">type MakeDbParams =
  | Input of string
  | Output of String
  | DbType of DbType
  | MaskData of string
  | ParseSeqIds</div>
<div class="fsdocs-tip" id="fs8">union case MakeDbParams.DbType: DbType -&gt; MakeDbParams</div>
<div class="fsdocs-tip" id="fs9">union case DbType.Protein: DbType</div>
<div class="fsdocs-tip" id="fs10">val ncbiPath: string<br /><em>the path of the /bin folder where the BLAST applications are located</em></div>
<div class="fsdocs-tip" id="fs11">Multiple items<br />type BlastWrapper =
  new: rootPath: string -&gt; BlastWrapper
  member blastN: searchDB: string -&gt; query: string -&gt; output: string -&gt; ps: BlastParams seq -&gt; unit
  member blastP: searchDB: string -&gt; query: string -&gt; output: string -&gt; ps: BlastParams seq -&gt; unit
  member makeblastdb: searchDB: string -&gt; ps: MakeDbParams seq -&gt; unit<br /><em>&lt;summary&gt;
A Wrapper to perform different BLAST tasks
&lt;/summary&gt;</em><br /><br />--------------------<br />new: rootPath: string -&gt; BlastWrapper</div>
<div class="fsdocs-tip" id="fs12">Multiple items<br />val seq: sequence: &#39;T seq -&gt; &#39;T seq<br /><br />--------------------<br />type &#39;T seq = System.Collections.Generic.IEnumerable&lt;&#39;T&gt;</div>
<div class="fsdocs-tip" id="fs13">val aminoacidString: string<br /><em>Raw string of the aminoacid sequence of our query protein</em></div>
<div class="fsdocs-tip" id="fs14">val header: string<br /><em>header for the .fastA file</em></div>
<div class="fsdocs-tip" id="fs15">val querySequence: BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;<br /><em>Query sequency represented as a sequence of `AminoAcid` one of BioFSharp&#39;s `BioItems`</em></div>
<div class="fsdocs-tip" id="fs16">Multiple items<br />module BioSeq

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioSeq

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioSeq type and its according functions. The BioSeq type is a sequence of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs17">val ofAminoAcidString: s: #(char seq) -&gt; BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;<br /><em>&lt;summary&gt;
 Generates AminoAcid sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs18">val queryFastaPath: string<br /><em>path and name of the query file</em></div>
<div class="fsdocs-tip" id="fs19">val queryFastaItem: FastA.FastaItem&lt;BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;&gt;<br /><em>FastaItem containing header string and query sequence</em></div>
<div class="fsdocs-tip" id="fs20">module FastA

from BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs21">val createFastaItem: header: string -&gt; sequence: &#39;a -&gt; FastA.FastaItem&lt;&#39;a&gt;<br /><em>&lt;summary&gt;
 Creates with header line and sequence.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs22">val write: toString: (&#39;T -&gt; char) -&gt; filePath: string -&gt; data: FastA.FastaItem&lt;#(&#39;T seq)&gt; seq -&gt; unit<br /><em>&lt;summary&gt;
 Writes FastaItem to file. Converter determines type of sequence by converting type -&amp;gt; char. If file already exists the data is overwritten.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs23">module BioItem

from BioFSharp<br /><em>&lt;summary&gt;
 Basic functions on IBioItems interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs24">val symbol: bItem: #IBioItem -&gt; char<br /><em>&lt;summary&gt;
 Returns the symbol of the bio item
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs25">val outputType: OutputType<br /><em>overall outline of the output </em></div>
<div class="fsdocs-tip" id="fs26">Multiple items<br />union case BlastParams.OutputType: OutputType -&gt; BlastParams<br /><br />--------------------<br />[&lt;Struct&gt;]
type OutputType =
  | Pairwise = 0
  | Query_anchored = 1
  | Query_anchored_NoIdentities = 2
  | Query_anchored_Flat = 3
  | Query_anchored_Flat_NoIdentities = 4
  | XML = 5
  | Tabular = 6
  | TabularWithComments = 7
  | TextASN1 = 8
  | BinaryASN1 = 9
  | CSV = 10
  | BLAST_ArchiveFormat = 11
  | JSON_Seqalign = 12
  | JSON_Blast = 13
  | XML2_Blast = 14</div>
<div class="fsdocs-tip" id="fs27">OutputType.TabularWithComments: OutputType = 7</div>
<div class="fsdocs-tip" id="fs28">val outputFormat: OutputCustom seq<br /><em>a sequence of custom output format parameters</em></div>
<div class="fsdocs-tip" id="fs29">type OutputCustom =
  | Query_SeqId
  | Query_GI
  | Query_Accesion
  | Query_Accesion_Version
  | Query_Length
  | Subject_SeqId
  | Subject_All_SeqIds
  | Subject_GI
  | Subject_All_GIs
  | Subject_Accession
  ...</div>
<div class="fsdocs-tip" id="fs30">union case OutputCustom.Query_SeqId: OutputCustom</div>
<div class="fsdocs-tip" id="fs31">union case OutputCustom.Subject_SeqId: OutputCustom</div>
<div class="fsdocs-tip" id="fs32">union case OutputCustom.Query_Length: OutputCustom</div>
<div class="fsdocs-tip" id="fs33">union case OutputCustom.Subject_Length: OutputCustom</div>
<div class="fsdocs-tip" id="fs34">union case OutputCustom.AlignmentLength: OutputCustom</div>
<div class="fsdocs-tip" id="fs35">union case OutputCustom.MismatchCount: OutputCustom</div>
<div class="fsdocs-tip" id="fs36">union case OutputCustom.IdentityCount: OutputCustom</div>
<div class="fsdocs-tip" id="fs37">union case OutputCustom.PositiveScoringMatchCount: OutputCustom</div>
<div class="fsdocs-tip" id="fs38">union case OutputCustom.Evalue: OutputCustom</div>
<div class="fsdocs-tip" id="fs39">union case OutputCustom.Bitscore: OutputCustom</div>
<div class="fsdocs-tip" id="fs40">Multiple items<br />module List

from Microsoft.FSharp.Collections<br /><br />--------------------<br />type List&lt;&#39;T&gt; =
  | op_Nil
  | op_ColonColon of Head: &#39;T * Tail: &#39;T list
  interface IReadOnlyList&lt;&#39;T&gt;
  interface IReadOnlyCollection&lt;&#39;T&gt;
  interface IEnumerable
  interface IEnumerable&lt;&#39;T&gt;
  member GetReverseIndex: rank: int * offset: int -&gt; int
  member GetSlice: startIndex: int option * endIndex: int option -&gt; &#39;T list
  static member Cons: head: &#39;T * tail: &#39;T list -&gt; &#39;T list
  member Head: &#39;T
  member IsEmpty: bool
  member Item: index: int -&gt; &#39;T with get
  ...</div>
<div class="fsdocs-tip" id="fs41">val toSeq: list: &#39;T list -&gt; &#39;T seq</div>
<div class="fsdocs-tip" id="fs42">val customOutputFormat: BlastParams<br /><em>The final output format</em></div>
<div class="fsdocs-tip" id="fs43">union case BlastParams.OutputTypeCustom: OutputType * OutputCustom seq -&gt; BlastParams</div>
<div class="fsdocs-tip" id="fs44">val outputPath: string<br /><em>output file of the BLAST search</em></div>
<div class="fsdocs-tip" id="fs45">type BlastParams =
  | SearchDB of string
  | Query of string
  | Output of string
  | OutputType of OutputType
  | OutputTypeCustom of OutputType * OutputCustom seq
  | Num_threads of int
  | Max_Hits of int</div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#Blast">
      Blast
    </a>
  </li>
  <li class="level-2">
    <a href="#Creation-of-a-BLAST-database">
      Creation of a BLAST database
    </a>
  </li>
  <li class="level-2">
    <a href="#Creating-a-fasta-file-from-an-aminoacid-string">
      Creating a fasta file from an aminoacid string
    </a>
  </li>
  <li class="level-2">
    <a href="#Performing-the-BLAST-search">
      Performing the BLAST search
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>