<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="Alignment" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="Alignment">
    
    <title>Alignment | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.1/es5/tex-mml-chtml.js"></script>
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header active">
  Algorithms
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_03_biocollections.html">
    BioCollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header active">
  Algorithms
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="Pairwise-Alignment" class="anchor" href="#Pairwise-Alignment">Pairwise Alignment</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BioFSharp/gh-pages?filepath=02_01_alignment.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-binder.svg" alt="Binder" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/02_01_alignment.fsx"><img src="https://csbiology.github.io/BioFSharp/img/badge-script.svg" alt="Script" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/02_01_alignment.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-notebook.svg" alt="Notebook" /></a></p>
<p><em>Summary:</em> This example shows how to perform sequence alignments in BioFSharp</p>
<p>In this short tutorial, the usage of the pairwise alignment implementation is demonstrated.
For global alignments, the <strong>NeedlemanWunsch</strong> algorithm is used. For local alignments, the <strong>SmithWaterman</strong> algorithm is used.</p>
<p>For both implementations, the gapvalues are evaluated using the <strong>affine</strong> gapscoremodel.</p>
<h2><a name="Aligning-aminoacid-and-nucleotide-sequences" class="anchor" href="#Aligning-aminoacid-and-nucleotide-sequences">Aligning aminoacid and nucleotide sequences</a></h2>
<p>For defining the scores of matching and missmatching characters, the <strong>scoring</strong> function is defined. In the case of amino acid or nucleotide sequence alignments, the integrated substitution-matrices can be used.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="id">Algorithm</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="m">PairwiseAlignment</span>

<span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="fn">aaScoring</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="m">ScoringMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="id">getScoringMatrixAminoAcid</span> <span onmouseout="hideTip(event, 'fs5', 8)" onmouseover="showTip(event, 'fs5', 8)" class="m">ScoringMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="rt">ScoringMatrixAminoAcid</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="id">BLOSUM62</span>

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 11)" onmouseover="showTip(event, 'fs9', 11)" class="fn">nucScoring</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 12)" onmouseover="showTip(event, 'fs5', 12)" class="m">ScoringMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 13)" onmouseover="showTip(event, 'fs10', 13)" class="id">getScoringMatrixNucleotide</span>  <span onmouseout="hideTip(event, 'fs5', 14)" onmouseover="showTip(event, 'fs5', 14)" class="m">ScoringMatrix</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs11', 15)" onmouseover="showTip(event, 'fs11', 15)" class="rt">ScoringMatrixNucleotide</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 16)" onmouseover="showTip(event, 'fs12', 16)" class="id">EDNA</span>
</code></pre>
<p>In order to align two sequences, not only the values for substitutions, but also the <strong>costs</strong> for gaps are needed. In this implementation, an affine gap-penalty is realized. An affine gap penalty weights continuations of gaps different than the opening of a gap.
The scoring function (in this case a substitution matrix) and the two gap penalty values are combined into the <code>Costs</code> type.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 17)" onmouseover="showTip(event, 'fs13', 17)" class="id">aaCosts</span> <span class="o">=</span> <span class="pn">{</span>
    <span class="prop">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">5</span>
    <span class="prop">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="fn">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 18)" onmouseover="showTip(event, 'fs4', 18)" class="fn">aaScoring</span> 
    <span class="pn">}</span>

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 19)" onmouseover="showTip(event, 'fs14', 19)" class="id">nucCosts</span> <span class="o">=</span> <span class="pn">{</span>
    <span class="prop">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="prop">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="fn">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs9', 20)" onmouseover="showTip(event, 'fs9', 20)" class="fn">nucScoring</span> 
    <span class="pn">}</span>
</code></pre>
<p>The alignment functions use <code>Arrays</code> as input. The Elements can be of any type, but require type equality. Also they need to have type equality with the input of the scoring function.
Both the global and local alignment algorithms take the same parameters (costs,firstSequence,secondSequence) and return the same format.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 21)" onmouseover="showTip(event, 'fs15', 21)" class="id">aaSeq1</span> <span class="o">=</span> <span class="s">&quot;ACDM&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 22)" onmouseover="showTip(event, 'fs16', 22)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 23)" onmouseover="showTip(event, 'fs17', 23)" class="id">ofAminoAcidSymbolString</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 24)" onmouseover="showTip(event, 'fs18', 24)" class="id">aaSeq2</span> <span class="o">=</span> <span class="s">&quot;MAACEDM&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 26)" onmouseover="showTip(event, 'fs17', 26)" class="id">ofAminoAcidSymbolString</span>

<span class="id">NeedlemanWunsch</span><span class="pn">.</span><span class="id">runAminoAcidSymbol</span> <span onmouseout="hideTip(event, 'fs13', 27)" onmouseover="showTip(event, 'fs13', 27)" class="id">aaCosts</span> <span onmouseout="hideTip(event, 'fs15', 28)" onmouseover="showTip(event, 'fs15', 28)" class="id">aaSeq1</span> <span onmouseout="hideTip(event, 'fs18', 29)" onmouseover="showTip(event, 'fs18', 29)" class="id">aaSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 12
  Sequences = [[-; -; A; C; -; D; M]; [M; A; A; C; E; D; M]] }
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">SmithWaterman</span><span class="pn">.</span><span class="id">runAminoAcidSymbol</span> <span onmouseout="hideTip(event, 'fs13', 30)" onmouseover="showTip(event, 'fs13', 30)" class="id">aaCosts</span> <span onmouseout="hideTip(event, 'fs15', 31)" onmouseover="showTip(event, 'fs15', 31)" class="id">aaSeq1</span> <span onmouseout="hideTip(event, 'fs18', 32)" onmouseover="showTip(event, 'fs18', 32)" class="id">aaSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 19
  Sequences = [[A; C; -; D; M]; [A; C; E; D; M]] }
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 33)" onmouseover="showTip(event, 'fs19', 33)" class="id">nucSeq1</span> <span class="o">=</span> <span class="s">&quot;ATGA&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 34)" onmouseover="showTip(event, 'fs16', 34)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 35)" onmouseover="showTip(event, 'fs20', 35)" class="id">ofNucleotideString</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 36)" onmouseover="showTip(event, 'fs21', 36)" class="id">nucSeq2</span> <span class="o">=</span> <span class="s">&quot;BATVAWG&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 37)" onmouseover="showTip(event, 'fs16', 37)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 38)" onmouseover="showTip(event, 'fs20', 38)" class="id">ofNucleotideString</span>

<span class="id">NeedlemanWunsch</span><span class="pn">.</span><span class="id">runNucleotide</span> <span onmouseout="hideTip(event, 'fs14', 39)" onmouseover="showTip(event, 'fs14', 39)" class="id">nucCosts</span> <span onmouseout="hideTip(event, 'fs19', 40)" onmouseover="showTip(event, 'fs19', 40)" class="id">nucSeq1</span> <span onmouseout="hideTip(event, 'fs21', 41)" onmouseover="showTip(event, 'fs21', 41)" class="id">nucSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 9
  Sequences = [[Gap; A; T; G; A; Gap; Gap]; [B; A; T; V; A; W; G]] }
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">SmithWaterman</span><span class="pn">.</span><span class="id">runNucleotide</span> <span onmouseout="hideTip(event, 'fs14', 42)" onmouseover="showTip(event, 'fs14', 42)" class="id">nucCosts</span> <span onmouseout="hideTip(event, 'fs19', 43)" onmouseover="showTip(event, 'fs19', 43)" class="id">nucSeq1</span> <span onmouseout="hideTip(event, 'fs21', 44)" onmouseover="showTip(event, 'fs21', 44)" class="id">nucSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 11
  Sequences = [[A; T; G; A]; [A; T; V; A]] }
</code></pre></td></tr></table>
<h2><a name="Aligning-anything-else" class="anchor" href="#Aligning-anything-else">Aligning anything else</a></h2>
<p>This implementation was aimed to be as generic as possible. To achieve this, the scoring function can be designed at will, the only constraints being the need for two input variables and the type equality.<br />
Also besides the alignment functions which only take BioItems as input and represent the gaps by the appropriate gaps of type BioItem. There is also a generic alignment function <code>runGeneric</code> which returns <code>lists of options</code>, where None represents gaps. Therefore any input type can be used, given it matches the cost function.</p>
<p>For example, one could use a simple <code>if .. then .. else</code> equality function to match nucleotides</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 45)" onmouseover="showTip(event, 'fs22', 45)" class="fn">scoring</span> <span onmouseout="hideTip(event, 'fs23', 46)" onmouseover="showTip(event, 'fs23', 46)" class="fn">a</span> <span onmouseout="hideTip(event, 'fs24', 47)" onmouseover="showTip(event, 'fs24', 47)" class="fn">b</span> <span class="o">=</span> 
    <span class="k">if</span> <span onmouseout="hideTip(event, 'fs23', 48)" onmouseover="showTip(event, 'fs23', 48)" class="fn">a</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs24', 49)" onmouseover="showTip(event, 'fs24', 49)" class="fn">b</span> <span class="k">then</span> <span class="n">2</span>
    <span class="k">else</span> <span class="o">-</span><span class="n">2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 50)" onmouseover="showTip(event, 'fs25', 50)" class="id">costs</span> <span class="o">=</span> <span class="pn">{</span>
    <span class="prop">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="prop">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="fn">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs22', 51)" onmouseover="showTip(event, 'fs22', 51)" class="fn">scoring</span>
    <span class="pn">}</span>

<span class="id">NeedlemanWunsch</span><span class="pn">.</span><span class="id">runGeneric</span> <span onmouseout="hideTip(event, 'fs25', 52)" onmouseover="showTip(event, 'fs25', 52)" class="id">costs</span> <span onmouseout="hideTip(event, 'fs19', 53)" onmouseover="showTip(event, 'fs19', 53)" class="id">nucSeq1</span> <span onmouseout="hideTip(event, 'fs21', 54)" onmouseover="showTip(event, 'fs21', 54)" class="id">nucSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = -1
  Sequences =
   [[None; Some A; Some T; Some G; Some A; None; None];
    [Some B; Some A; Some T; Some V; Some A; Some W; Some G]] }
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">SmithWaterman</span><span class="pn">.</span><span class="id">runGeneric</span> <span onmouseout="hideTip(event, 'fs25', 55)" onmouseover="showTip(event, 'fs25', 55)" class="id">costs</span> <span onmouseout="hideTip(event, 'fs19', 56)" onmouseover="showTip(event, 'fs19', 56)" class="id">nucSeq1</span> <span onmouseout="hideTip(event, 'fs21', 57)" onmouseover="showTip(event, 'fs21', 57)" class="id">nucSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 1
  Sequences =
   [[Some A; Some T; Some G; Some A]; [Some A; Some T; Some V; Some A]] }
</code></pre></td></tr></table>
<p>or also Integers:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs26', 58)" onmouseover="showTip(event, 'fs26', 58)" class="id">intSeq1</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">1</span><span class="pn">;</span><span class="n">2</span><span class="pn">;</span><span class="n">3</span><span class="pn">;</span><span class="n">4</span><span class="pn">;</span><span class="n">5</span><span class="pn">|]</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 59)" onmouseover="showTip(event, 'fs27', 59)" class="id">intSeq2</span> <span class="o">=</span> <span class="pn">[|</span><span class="n">1</span><span class="pn">;</span><span class="n">1</span><span class="pn">;</span><span class="n">2</span><span class="pn">;</span><span class="n">4</span><span class="pn">;</span><span class="n">6</span><span class="pn">;</span><span class="n">7</span><span class="pn">;</span><span class="pn">|]</span>

<span class="id">NeedlemanWunsch</span><span class="pn">.</span><span class="id">runGeneric</span> <span onmouseout="hideTip(event, 'fs25', 60)" onmouseover="showTip(event, 'fs25', 60)" class="id">costs</span> <span onmouseout="hideTip(event, 'fs26', 61)" onmouseover="showTip(event, 'fs26', 61)" class="id">intSeq1</span> <span onmouseout="hideTip(event, 'fs27', 62)" onmouseover="showTip(event, 'fs27', 62)" class="id">intSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = -2
  Sequences =
   [[Some 1; None; Some 2; Some 3; Some 4; Some 5; None];
    [Some 1; Some 1; Some 2; None; Some 4; Some 6; Some 7]] }
</code></pre></td></tr></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">SmithWaterman</span><span class="pn">.</span><span class="id">runGeneric</span> <span onmouseout="hideTip(event, 'fs25', 63)" onmouseover="showTip(event, 'fs25', 63)" class="id">costs</span> <span onmouseout="hideTip(event, 'fs26', 64)" onmouseover="showTip(event, 'fs26', 64)" class="id">intSeq1</span> <span onmouseout="hideTip(event, 'fs27', 65)" onmouseover="showTip(event, 'fs27', 65)" class="id">intSeq2</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = 0
  Sequences =
   [[Some 1; Some 2; Some 3; Some 4]; [Some 1; Some 2; None; Some 4]] }
</code></pre></td></tr></table>
<h1><a name="Multiple-sequence-alignment-with-ClustalOmega" class="anchor" href="#Multiple-sequence-alignment-with-ClustalOmega">Multiple sequence alignment with ClustalOmega</a></h1>
<p>Clustal Omega is a multiple sequence alignment (MSA) tool. This tutorial describes using it in F# via the ClustalOWrapper. For some more indepth information about which parameters to choose for your goal, also check out <a href="http://www.clustal.org/omega/README">the official tutorial</a>.</p>
<h2><a name="Aligning-sequences-from-files" class="anchor" href="#Aligning-sequences-from-files">Aligning sequences from files</a></h2>
<p>The first step is to create a wrapper-object. As optional input it takes a path to the clustalo executable you want to use. You have to fill this argument if you work with a precompiled verion or on linux.</p>
<p>you will have to <a href="http://www.clustal.org/omega/clustal-omega-1.2.2-win64.zip">install clustal omega</a> yourself to use this wrapper.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 66)" onmouseover="showTip(event, 'fs1', 66)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 67)" onmouseover="showTip(event, 'fs28', 67)" class="id">IO</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs29', 68)" onmouseover="showTip(event, 'fs29', 68)" class="m">ClustalOWrapper</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 69)" onmouseover="showTip(event, 'fs1', 69)" class="id">BioFSharp</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="id">cw</span> <span class="o">=</span> <span class="id">ClustalOWrapper</span><span class="pn">(</span><span class="s">&quot;path/where/you/extracted/clustal-omega/clustalo.exe&quot;</span><span class="pn">)</span>
</code></pre>
<p>The general structure of arguments the wrapper takes was kept the same as in the command line tool. In general, you need an <code>inputPath</code>, an <code>outputPath</code> and <code>parameters</code>. As there are several inputs possible, you have to choose what it is. As we want to align a normal sequence we just pick <code>SequenceFile</code>.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs32', 72)" onmouseover="showTip(event, 'fs32', 72)" class="id">inputPath</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs33', 73)" onmouseover="showTip(event, 'fs33', 73)" class="rt">Input</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs34', 74)" onmouseover="showTip(event, 'fs34', 74)" class="id">SequenceFile</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">@&quot;\data\Chlamy_Cp.fastA&quot;</span><span class="pn">)</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs35', 75)" onmouseover="showTip(event, 'fs35', 75)" class="id">outputPath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">@&quot;\data\Chlamy_Cp.aln&quot;</span>
</code></pre>
<p>As additional parameters go, we'll restrict input to FastA format and the output to Clustal format. Also we will use the <code>Force</code> parameter to force the overwrite of a possilby already existing file with the name <code>ChlamyCp.aln</code>.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//Input has to be in FastA format</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs36', 76)" onmouseover="showTip(event, 'fs36', 76)" class="id">inputModifier</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 77)" onmouseover="showTip(event, 'fs37', 77)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 78)" onmouseover="showTip(event, 'fs38', 78)" class="rt">ClustalParams</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 79)" onmouseover="showTip(event, 'fs39', 79)" class="id">Input</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs37', 80)" onmouseover="showTip(event, 'fs37', 80)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs40', 81)" onmouseover="showTip(event, 'fs40', 81)" class="rt">InputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 82)" onmouseover="showTip(event, 'fs41', 82)" class="id">Format</span> <span onmouseout="hideTip(event, 'fs37', 83)" onmouseover="showTip(event, 'fs37', 83)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs42', 84)" onmouseover="showTip(event, 'fs42', 84)" class="rt">FileFormat</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs43', 85)" onmouseover="showTip(event, 'fs43', 85)" class="id">FastA</span><span class="pn">]</span>
<span class="c">//Output has to be in Clustal format</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 86)" onmouseover="showTip(event, 'fs44', 86)" class="id">outputModifier</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 87)" onmouseover="showTip(event, 'fs37', 87)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 88)" onmouseover="showTip(event, 'fs38', 88)" class="rt">ClustalParams</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 89)" onmouseover="showTip(event, 'fs45', 89)" class="id">Output</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs37', 90)" onmouseover="showTip(event, 'fs37', 90)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 91)" onmouseover="showTip(event, 'fs46', 91)" class="rt">OutputCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs47', 92)" onmouseover="showTip(event, 'fs47', 92)" class="id">Format</span> <span onmouseout="hideTip(event, 'fs37', 93)" onmouseover="showTip(event, 'fs37', 93)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs42', 94)" onmouseover="showTip(event, 'fs42', 94)" class="rt">FileFormat</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs48', 95)" onmouseover="showTip(event, 'fs48', 95)" class="id">Clustal</span><span class="pn">]</span>
<span class="c">//Forces overwriting</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs49', 96)" onmouseover="showTip(event, 'fs49', 96)" class="id">forceModifier</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 97)" onmouseover="showTip(event, 'fs37', 97)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 98)" onmouseover="showTip(event, 'fs38', 98)" class="rt">ClustalParams</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs50', 99)" onmouseover="showTip(event, 'fs50', 99)" class="id">Miscallaneous</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs37', 100)" onmouseover="showTip(event, 'fs37', 100)" class="m">Parameters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs51', 101)" onmouseover="showTip(event, 'fs51', 101)" class="rt">MiscallaneousCustom</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs52', 102)" onmouseover="showTip(event, 'fs52', 102)" class="id">Force</span><span class="pn">]</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//Perform alignment</span>
<span onmouseout="hideTip(event, 'fs30', 103)" onmouseover="showTip(event, 'fs30', 103)" class="id">cw</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs53', 104)" onmouseover="showTip(event, 'fs53', 104)" class="id">AlignFromFile</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs32', 105)" onmouseover="showTip(event, 'fs32', 105)" class="id">inputPath</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs35', 106)" onmouseover="showTip(event, 'fs35', 106)" class="id">outputPath</span><span class="pn">,</span><span class="pn">[</span><span onmouseout="hideTip(event, 'fs36', 107)" onmouseover="showTip(event, 'fs36', 107)" class="id">inputModifier</span><span class="pn">;</span><span onmouseout="hideTip(event, 'fs44', 108)" onmouseover="showTip(event, 'fs44', 108)" class="id">outputModifier</span><span class="pn">;</span><span onmouseout="hideTip(event, 'fs49', 109)" onmouseover="showTip(event, 'fs49', 109)" class="id">forceModifier</span><span class="pn">]</span><span class="pn">)</span>
</code></pre>
<h2><a name="Aligning-sequences-directly-in-F-Interactive" class="anchor" href="#Aligning-sequences-directly-in-F-Interactive">Aligning sequences directly in F# Interactive</a></h2>
<p>With the <code>AlignSequences</code> method, one can also directly align sequences with the clustal tool and also directly receive the alignment directly in F#.</p>
<p>As input, it takes a collection of <code>TaggedSequence</code>s, and again a set of parameters. The default options can be used by not using any additional parameters.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs54', 110)" onmouseover="showTip(event, 'fs54', 110)" class="id">sequences</span> <span class="o">=</span> <span class="pn">[</span>
    <span onmouseout="hideTip(event, 'fs55', 111)" onmouseover="showTip(event, 'fs55', 111)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 112)" onmouseover="showTip(event, 'fs56', 112)" class="id">create</span> <span class="s">&quot;pep1&quot;</span> <span class="pn">(</span><span class="s">&quot;AAGECGEK&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs55', 113)" onmouseover="showTip(event, 'fs55', 113)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 114)" onmouseover="showTip(event, 'fs56', 114)" class="id">create</span> <span class="s">&quot;pep2&quot;</span> <span class="pn">(</span><span class="s">&quot;AAGEGEK&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs55', 115)" onmouseover="showTip(event, 'fs55', 115)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 116)" onmouseover="showTip(event, 'fs56', 116)" class="id">create</span> <span class="s">&quot;pep3&quot;</span> <span class="pn">(</span><span class="s">&quot;AAAGECGEK&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs55', 117)" onmouseover="showTip(event, 'fs55', 117)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 118)" onmouseover="showTip(event, 'fs56', 118)" class="id">create</span> <span class="s">&quot;pep4&quot;</span> <span class="pn">(</span><span class="s">&quot;AAGECGEL&quot;</span><span class="pn">)</span>
<span class="pn">]</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs30', 119)" onmouseover="showTip(event, 'fs30', 119)" class="id">cw</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs57', 120)" onmouseover="showTip(event, 'fs57', 120)" class="id">AlignSequences</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs54', 121)" onmouseover="showTip(event, 'fs54', 121)" class="id">sequences</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs58', 122)" onmouseover="showTip(event, 'fs58', 122)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs59', 123)" onmouseover="showTip(event, 'fs59', 123)" class="id">empty</span><span class="pn">)</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">{ MetaData = { Header = seq ['C'; 'L'; 'U'; 'S'; ...]
ConservationInfo = seq [' '; '*'; '*'; '*'; ...] }
    Sequences =
     [{ Tag = "pep1"
        Sequence = seq ['-'; 'A'; 'A'; 'G'; ...] };
      { Tag = "pep2"
        Sequence = seq ['-'; 'A'; 'A'; 'G'; ...] };
      { Tag = "pep3"
        Sequence = seq ['A'; 'A'; 'A'; 'G'; ...] };
      { Tag = "pep4"
        Sequence = seq ['-'; 'A'; 'A'; 'G'; ...] }] }
</code></pre></td></tr></table>

        <div class="fsdocs-tip" id="fs1">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs2">namespace BioFSharp.Algorithm</div>
<div class="fsdocs-tip" id="fs3">module PairwiseAlignment

from BioFSharp.Algorithm<br /><em>&lt;summary&gt;
Contains functions for evaluating the best possible alignments for 2 Sequences
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs4">val aaScoring: (AminoAcidSymbols.AminoAcidSymbol -&gt; AminoAcidSymbols.AminoAcidSymbol -&gt; int)</div>
<div class="fsdocs-tip" id="fs5">module ScoringMatrix

from BioFSharp.Algorithm<br /><em>&lt;summary&gt;
Contains functions for using the included similarity matrices. These assign a score to every pair of aminoacids/nucleotides and therefore rate the probability of their substitution. The Scoring Matrices are generally used for alignments.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs6">val getScoringMatrixAminoAcid: scoringMatrixType: ScoringMatrix.ScoringMatrixAminoAcid -&gt; (AminoAcidSymbols.AminoAcidSymbol -&gt; AminoAcidSymbols.AminoAcidSymbol -&gt; int)<br /><em>&lt;summary&gt;
creates a scoring function for amino acids out of a scoring matrix
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs7">type ScoringMatrixAminoAcid =
  | BLOSUM45
  | BLOSUM50
  | BLOSUM62
  | BLOSUM80
  | PAM30
  | PAM70
  | PAM250
  static member toFileName: (ScoringMatrixAminoAcid -&gt; string)<br /><em>&lt;summary&gt;
Union case of implemented amino acid scoring matrices with the given reference to its place in the library. Use the &quot;getScoringMatrixAminoAcid&quot; function to obtain a simple mapping function for every amino acid pair
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs8">union case ScoringMatrix.ScoringMatrixAminoAcid.BLOSUM62: ScoringMatrix.ScoringMatrixAminoAcid</div>
<div class="fsdocs-tip" id="fs9">val nucScoring: (Nucleotides.Nucleotide -&gt; Nucleotides.Nucleotide -&gt; int)</div>
<div class="fsdocs-tip" id="fs10">val getScoringMatrixNucleotide: scoringMatrixType: ScoringMatrix.ScoringMatrixNucleotide -&gt; (Nucleotides.Nucleotide -&gt; Nucleotides.Nucleotide -&gt; int)<br /><em>&lt;summary&gt;
creates a scoring function for nucleotides out of a scoring matrix
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs11">type ScoringMatrixNucleotide =
  | EDNA
  | Default
  static member toFileName: (ScoringMatrixNucleotide -&gt; string)<br /><em>&lt;summary&gt;
Union case of implemented nucleotide scoring matrices with the given reference to its place in the library. Use the &quot;getScoringMatrixNucleotide&quot; function to obtain a simple mapping function for every nucleotide pair
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs12">union case ScoringMatrix.ScoringMatrixNucleotide.EDNA: ScoringMatrix.ScoringMatrixNucleotide</div>
<div class="fsdocs-tip" id="fs13">val aaCosts: Costs&lt;AminoAcidSymbols.AminoAcidSymbol&gt;</div>
<div class="fsdocs-tip" id="fs14">val nucCosts: Costs&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs15">val aaSeq1: BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;</div>
<div class="fsdocs-tip" id="fs16">Multiple items<br />module BioArray

from BioFSharp.BioCollectionsExtensions<br /><br />--------------------<br />module BioArray

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioArray type and its according functions. The BioArray type is an array of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs17">val ofAminoAcidSymbolString: s: #(char seq) -&gt; BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;<br /><em>&lt;summary&gt;
 Generates amino acid symbol sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs18">val aaSeq2: BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;</div>
<div class="fsdocs-tip" id="fs19">val nucSeq1: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs20">val ofNucleotideString: s: #(char seq) -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Generates nucleotide sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs21">val nucSeq2: BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs22">val scoring: a: &#39;a -&gt; b: &#39;a -&gt; int (requires equality)</div>
<div class="fsdocs-tip" id="fs23">val a: &#39;a (requires equality)</div>
<div class="fsdocs-tip" id="fs24">val b: &#39;a (requires equality)</div>
<div class="fsdocs-tip" id="fs25">val costs: Costs&lt;&#39;a&gt; (requires equality)</div>
<div class="fsdocs-tip" id="fs26">val intSeq1: int array</div>
<div class="fsdocs-tip" id="fs27">val intSeq2: int array</div>
<div class="fsdocs-tip" id="fs28">namespace BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs29">module ClustalOWrapper

from BioFSharp.IO<br /><em>&lt;summary&gt;
Wrapper and its helpers for Clustal Omega multiple alignment tools
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs30">val cw: ClustalOWrapper</div>
<div class="fsdocs-tip" id="fs31">Multiple items<br />module ClustalOWrapper

from BioFSharp.IO<br /><em>&lt;summary&gt;
Wrapper and its helpers for Clustal Omega multiple alignment tools
&lt;/summary&gt;</em><br /><br />--------------------<br />type ClustalOWrapper =
  new: ?rootPath: string -&gt; ClustalOWrapper
  member AlignFromFile: inputPath: Input * outputPath: string * parameters: ClustalParams seq * ?name: string -&gt; unit
  member AlignSequences: input: TaggedSequence&lt;string,char&gt; seq * parameters: ClustalParams seq * ?name: string -&gt; Alignment&lt;TaggedSequence&lt;string,char&gt;,AlignmentInfo&gt;<br /><em>&lt;summary&gt;
 A wrapper to perform Clustal Omega alignment tasks    
&lt;/summary&gt;</em><br /><br />--------------------<br />new: ?rootPath: string -&gt; ClustalOWrapper</div>
<div class="fsdocs-tip" id="fs32">val inputPath: Input</div>
<div class="fsdocs-tip" id="fs33">type Input =
  | SequenceFile of string
  | TwoProfiles of string * string
  | SequenceFileAndProfile of string * string
  | SequenceFileAndHMM of string * string<br /><em>&lt;summary&gt;
Specify the type of input and assign file path
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs34">union case Input.SequenceFile: string -&gt; Input<br /><em>&lt;summary&gt;
Use this option to make a multiple alignment from a set of sequences. A sequence file must contain more than one sequence (at least two sequences).
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs35">val outputPath: string</div>
<div class="fsdocs-tip" id="fs36">val inputModifier: Parameters.ClustalParams</div>
<div class="fsdocs-tip" id="fs37">module Parameters

from BioFSharp.IO.ClustalOWrapper<br /><em>&lt;summary&gt;
Contains modifier parameter type for Clustal Omega wrapper
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs38">type ClustalParams =
  | Input of InputCustom seq
  | Output of OutputCustom seq
  | Clustering of ClusteringCustom seq
  | Iteration of IterationCustom seq
  | Limits of LimitsCustom seq
  | Miscallaneous of MiscallaneousCustom seq<br /><em>&lt;summary&gt;
Collection of parameters for specifying clustalo alignment
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs39">union case Parameters.ClustalParams.Input: Parameters.InputCustom seq -&gt; Parameters.ClustalParams<br /><em>&lt;summary&gt;
 Specify input parameters
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs40">type InputCustom =
  | Format of FileFormat
  | Dealign
  | IsProfile
  | SeqType of SeqType<br /><em>&lt;summary&gt;
Optional modifiers for input
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs41">union case Parameters.InputCustom.Format: Parameters.FileFormat -&gt; Parameters.InputCustom<br /><em>&lt;summary&gt;
Forced sequence input file format (default: auto)
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs42">type FileFormat =
  | FastA
  | Clustal
  | MSF
  | Phylip
  | Selex
  | Stockholm
  | Vienna<br /><em>&lt;summary&gt;
Input file format
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs43">union case Parameters.FileFormat.FastA: Parameters.FileFormat<br /><em>&lt;summary&gt;
FastA file format
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs44">val outputModifier: Parameters.ClustalParams</div>
<div class="fsdocs-tip" id="fs45">union case Parameters.ClustalParams.Output: Parameters.OutputCustom seq -&gt; Parameters.ClustalParams<br /><em>&lt;summary&gt;
 Specify output parameters
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs46">type OutputCustom =
  | Format of FileFormat
  | ResidueNumber
  | Wrap of int
  | OutputOrderAsTree<br /><em>&lt;summary&gt;
Optional modifiers for input
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs47">union case Parameters.OutputCustom.Format: Parameters.FileFormat -&gt; Parameters.OutputCustom<br /><em>&lt;summary&gt;
	MSA output file format (default: fasta)
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs48">union case Parameters.FileFormat.Clustal: Parameters.FileFormat<br /><em>&lt;summary&gt;
Clustal file format
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs49">val forceModifier: Parameters.ClustalParams</div>
<div class="fsdocs-tip" id="fs50">union case Parameters.ClustalParams.Miscallaneous: Parameters.MiscallaneousCustom seq -&gt; Parameters.ClustalParams<br /><em>&lt;summary&gt;
 Specify miscallaneous parameters
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs51">type MiscallaneousCustom =
  | Auto
  | Threads of int
  | Log of string
  | VerboseLevel of int
  | Version
  | LongVersion
  | Force<br /><em>&lt;summary&gt;
Optional, miscallaneous modifiers 
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs52">union case Parameters.MiscallaneousCustom.Force: Parameters.MiscallaneousCustom<br /><em>&lt;summary&gt;
 Force file overwriting
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs53">member ClustalOWrapper.AlignFromFile: inputPath: Input * outputPath: string * parameters: Parameters.ClustalParams seq * ?name: string -&gt; unit</div>
<div class="fsdocs-tip" id="fs54">val sequences: TaggedSequence&lt;string,char&gt; list</div>
<div class="fsdocs-tip" id="fs55">type TaggedSequence&lt;&#39;T,&#39;S&gt; =
  {
    Tag: &#39;T
    Sequence: &#39;S seq
  }
  static member create: tag: &#39;T -&gt; sequence: &#39;S seq -&gt; TaggedSequence&lt;&#39;T,&#39;S&gt;
  static member mapSequence: mapping: (&#39;S seq -&gt; &#39;M seq) -&gt; ts: TaggedSequence&lt;&#39;T,&#39;S&gt; -&gt; TaggedSequence&lt;&#39;T,&#39;M&gt;
  static member mapTag: mapping: (&#39;T -&gt; &#39;U) -&gt; ts: TaggedSequence&lt;&#39;T,&#39;S&gt; -&gt; TaggedSequence&lt;&#39;U,&#39;S&gt;<br /><em>&lt;summary&gt;
 Record of a sequence and its tag
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs56">static member TaggedSequence.create: tag: &#39;T -&gt; sequence: &#39;S seq -&gt; TaggedSequence&lt;&#39;T,&#39;S&gt;</div>
<div class="fsdocs-tip" id="fs57">member ClustalOWrapper.AlignSequences: input: TaggedSequence&lt;string,char&gt; seq * parameters: Parameters.ClustalParams seq * ?name: string -&gt; Alignment.Alignment&lt;TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt;</div>
<div class="fsdocs-tip" id="fs58">module Seq

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs59">val empty&lt;&#39;T&gt; : &#39;T seq</div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#Pairwise-Alignment">
      Pairwise Alignment
    </a>
  </li>
  <li class="level-2">
    <a href="#Aligning-aminoacid-and-nucleotide-sequences">
      Aligning aminoacid and nucleotide sequences
    </a>
  </li>
  <li class="level-2">
    <a href="#Aligning-anything-else">
      Aligning anything else
    </a>
  </li>
  <li class="level-1">
    <a href="#Multiple-sequence-alignment-with-ClustalOmega">
      Multiple sequence alignment with ClustalOmega
    </a>
  </li>
  <li class="level-2">
    <a href="#Aligning-sequences-from-files">
      Aligning sequences from files
    </a>
  </li>
  <li class="level-2">
    <a href="#Aligning-sequences-directly-in-F-Interactive">
      Aligning sequences directly in F# Interactive
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>