<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="GFF3" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="GFF3">
    
    <title>GFF3 | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.1/es5/tex-mml-chtml.js"></script>
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/biocollections.html">
    Biocollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header active">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header ">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/biocollections.html">
    Biocollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header active">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="GFF3-parsing" class="anchor" href="#GFF3-parsing">GFF3 parsing</a></h1>
<p><a href="https://mybinder.org/v2/gh/CSBiology/BioFSharp/gh-pages?filepath=03_03_gff3_parsing.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-binder.svg" alt="Binder" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.fsx"><img src="https://csbiology.github.io/BioFSharp/img/badge-script.svg" alt="Script" /></a>&emsp;
<a href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.ipynb"><img src="https://csbiology.github.io/BioFSharp/img/badge-notebook.svg" alt="Notebook" /></a></p>
<p><em>Summary:</em> This example shows how to parse and write gff3 formatted files with BioFSharp</p>
<h2><a name="Generic-Feature-Format-Version-3" class="anchor" href="#Generic-Feature-Format-Version-3">Generic Feature Format Version 3</a></h2>
<p>In GFF3 files every line represents one genomic feature with nine tab-delimited fields, whereas unlimited key-value pairs can be stored in field 9.
It is possible to link multiple features to genomic units using the 'Parent tag'.</p>
<p>In the following you can see a GFF file example (modified version of <a href="http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff">saccharomyces_cerevisiae.gff</a>):</p>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">##gff-version 3
# date Mon Feb  7 19:35:06 2005
chrI  SGD  gene  335  649  .  +  .  ID=YAL069W;Name=YAL069W;Ontology_term=GO:0000004,GO:0005554,GO:0008372;Note=Hypothetical%20ORF;dbxref=SGD:S000002143;orf_classification=Dubious
chrI  SGD  CDS  335  649  .  +  0  Parent=YAL069W;Name=YAL069W;Ontology_term=GO:0000004,GO:0005554,GO:0008372;Note=Hypothetical%20ORF;dbxref=SGD:S000002143;orf_classification=Dubious
###
##FASTA
&gt;chrI
CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACACATCCTAACACTACCCTAAC
ACAGCCCTAATCTAACCCTGGCCAACCTGTCTCTCAACTTACCCTCCATTACCCTGCCTCCACTCGTTACCCTGTCCCAT
TCAACCATACCACTCCGAACCACCATCCATCCCTCTACTTACTACCACTCACCCACCGTTACCCTCCAATTACCCATATC
CAACCCACTGCCACTTACCCTACCATTACCCTACCATCCACCATGACCTACTCACCATACTGTTCTTCTACCCACCATAT
TGAAACGCTAACAAATGATCGTAAATAACACACACGTGCTTACCCTACCACTTTATACCACCACCACATGCCATACTCAC
CCTCACTTGTATACTGATTTTACGTACGCACACGGATGCTACAGTATATACCATCTCAAACTTACCCTACTCTCAGATTC
CACTTCACTCCATGGCCCATCTCTCACTGAATCAGTACCAAATGCACTCACATCATTATGCACGGCACTTGCCTCAGCGG
TCTATACCCTGTGCCATTTACCCATAACGCCCATCATTATCCACATTTTGATATCTATATCTCATTCGGCGGTCCCAAAT
ATTGTATAACTGCCCTTAATACATACGTTATACCACTTTTGCACCATATACTTACCACTCCATTTATATACACTTATGTC
AATATTACAGAAAAATCCCCACAAAAATCACCTAAACATAAAAATATTCTACTTTTCAACAATAATACATAAACATATTG
</code></pre></td></tr></table>
<p>Directives (marked with "##[...]") provide additional information like the gff-version which has to be the first line of each file ("##gff-version 3[...]").
Comment lines have to start with a single "#[...]". It is possible that sequences in FastA format are attached at the end of the file. This has to be announced by a "##FASTA" directive line.</p>
<p>For further information visit <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3-Specifications</a>.</p>
<h2><a name="Reading-GFF3-files" class="anchor" href="#Reading-GFF3-files">Reading GFF3 files</a></h2>
<p>To read in a gff you have to insert a filepath and optionally a FastA converter. For further information about FastA check the <a href="https://csbiology.github.io/BioFSharp/FastA.html">FastA section</a>
or visit <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="id">IO</span>

<span class="c">//path of the input file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="id">filepathGFF</span> <span class="o">=</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gff3Example.gff&quot;</span><span class="pn">)</span>

<span class="c">//reads from file to seq of GFFLines</span>
<span class="c">//If no FASTA Sequence is included you directly can use GFF3.fromFileWithoutFasta [filepathGFF].</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="id">features</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="id">fromFile</span> <span onmouseout="hideTip(event, 'fs1', 8)" onmouseover="showTip(event, 'fs1', 8)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="id">ofNucleotideString</span> <span onmouseout="hideTip(event, 'fs3', 11)" onmouseover="showTip(event, 'fs3', 11)" class="id">filepathGFF</span> 
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">seq
[Directive "##gff-version 3"; Comment "# date Mon Feb  7 19:35:06 2005";
 GFFEntryLine
   { Seqid = "chrI"
     Source = "SGD"
     Feature = "gene"
     StartPos = 335
     EndPos = 649
     Score = nan
     Strand = '+'
     Phase = -1
     Attributes =
      map
        [("ID", ["YAL069W"]); ("Name", ["YAL069W"]);
         ("Note", ["Hypothetical%20ORF"]);
         ("Ontology_term", ["GO:0000004"; "GO:0005554"; "GO:0008372"]);
         ("dbxref", ["SGD:S000002143"]);
         ("orf_classification", ["Dubious"])]
     Supplement = [|"No supplement"|] };
 GFFEntryLine{
     ...
]
</code></pre></td></tr></table>
<h2><a name="How-to-use-GFF3SanityCheck" class="anchor" href="#How-to-use-GFF3SanityCheck">How to use GFF3SanityCheck</a></h2>
<p>The GFF3SanityCheck prints wether your GFF3 file is valid or not. It returns all specified errors including the lines in which they occured.
In contrast to GFF2 the field 3 (type, feature or method) of a GFF3 entry is restricted to terms defined by the sequence ontology (SO) so this validator is able to check if the entry is a valid SO term.
You can find new versions of the SO at (<a href="https://sourceforge.net/projects/song/files/SO_Feature_Annotation).">https://sourceforge.net/projects/song/files/SO_Feature_Annotation).</a></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//to validate the GFF file without SOTerm verification use this function and only insert the filepath</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 12)" onmouseover="showTip(event, 'fs9', 12)" class="id">featuresSanityCheck</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 13)" onmouseover="showTip(event, 'fs5', 13)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 14)" onmouseover="showTip(event, 'fs10', 14)" class="id">sanityCheck</span> <span onmouseout="hideTip(event, 'fs3', 15)" onmouseover="showTip(event, 'fs3', 15)" class="id">filepathGFF</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//path, name and version of the &#39;Sequence Ontology terms&#39;-file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 16)" onmouseover="showTip(event, 'fs11', 16)" class="id">filepathSO_Terms</span> <span class="o">=</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Sequence_Ontology_Terms_2_5_3.txt&quot;</span><span class="pn">)</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//to validate the gff file insert filepath</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 17)" onmouseover="showTip(event, 'fs12', 17)" class="id">featuresSanityCheckWithSOTerm</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 18)" onmouseover="showTip(event, 'fs5', 18)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 19)" onmouseover="showTip(event, 'fs13', 19)" class="id">sanityCheckWithSOTerm</span> <span onmouseout="hideTip(event, 'fs11', 20)" onmouseover="showTip(event, 'fs11', 20)" class="id">filepathSO_Terms</span> <span onmouseout="hideTip(event, 'fs3', 21)" onmouseover="showTip(event, 'fs3', 21)" class="id">filepathGFF</span>
</code></pre>
<h2><a name="How-to-use-GFF3RelationshipSearch" class="anchor" href="#How-to-use-GFF3RelationshipSearch">How to use GFF3RelationshipSearch</a></h2>
<p>You also can do a simple search for "Parent - child of" relationships giving back all genomic features which contain the searchterm in <strong>ID/Id</strong> or <strong>Parent</strong> field.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Term to search for:</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 22)" onmouseover="showTip(event, 'fs14', 22)" class="id">searchterm</span> <span class="o">=</span> <span class="s">&quot;YAL069W&quot;</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///with this function you can search features which are related to the searchterm</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 23)" onmouseover="showTip(event, 'fs15', 23)" class="id">gffExampleSearch</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 24)" onmouseover="showTip(event, 'fs5', 24)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="id">relationshipSearch</span> <span onmouseout="hideTip(event, 'fs4', 26)" onmouseover="showTip(event, 'fs4', 26)" class="id">features</span> <span onmouseout="hideTip(event, 'fs14', 27)" onmouseover="showTip(event, 'fs14', 27)" class="id">searchterm</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">seq
[{ Seqid = "chrI"
   Source = "SGD"
   Feature = "gene"
   StartPos = 335
   EndPos = 649
   Score = nan
   Strand = '+'
   Phase = -1
   Attributes =
    map
      [("ID", ["YAL069W"]); ("Name", ["YAL069W"]);
       ("Note", ["Hypothetical%20ORF"]);
       ("Ontology_term", ["GO:0000004"; "GO:0005554"; "GO:0008372"]);
       ("dbxref", ["SGD:S000002143"]); ("orf_classification", ["Dubious"])]
   Supplement = [|"No supplement"|] };
 { Seqid = "chrI"
   Source = "SGD"
   Feature = "CDS"
   StartPos = 335
   EndPos = 649
   Score = nan
   Strand = '+'
   Phase = 0
   Attributes =
    map
      [("Name", ["YAL069W"]); ("Note", ["Hypothetical%20ORF"]);
       ("Ontology_term", ["GO:0000004"; "GO:0005554"; "GO:0008372"]);
       ("Parent", ["YAL069W"]); ("dbxref", ["SGD:S000002143"]);
       ("orf_classification", ["Dubious"])]
   Supplement = [|"No supplement"|] }]
</code></pre></td></tr></table>
<h2><a name="Writing-GFF3-files" class="anchor" href="#Writing-GFF3-files">Writing GFF3 files</a></h2>
<p>In order to write a sequence of (GFFLine<_>) into a file use the following function.
If FastA sequences are included they are appended by a FastA writer described in the <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<p><em>Note: The order of key value pairs in field 9 (attributes) may be changed.</em></p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FASTA converter and a destination filepath and writes it into a .gff. Hint: Use converter = id if no FastA sequence is included.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 28)" onmouseover="showTip(event, 'fs17', 28)" class="id">gffExampleWrite</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 29)" onmouseover="showTip(event, 'fs5', 29)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 30)" onmouseover="showTip(event, 'fs18', 30)" class="id">write</span> <span onmouseout="hideTip(event, 'fs19', 31)" onmouseover="showTip(event, 'fs19', 31)" class="m">BioItem</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 32)" onmouseover="showTip(event, 'fs20', 32)" class="id">symbol</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gffExampleWrite.gff&quot;</span><span class="pn">)</span> <span onmouseout="hideTip(event, 'fs4', 33)" onmouseover="showTip(event, 'fs4', 33)" class="id">features</span>
</code></pre>
<h2><a name="Example-Sequence-of-CDS" class="anchor" href="#Example-Sequence-of-CDS">Example: Sequence of CDS</a></h2>
<p>If a FastA file is included you can look up the sequence of a CDS feature using the following function.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 34)" onmouseover="showTip(event, 'fs21', 34)" class="id">firstCDS</span> <span class="o">=</span> 
    <span class="c">//get GFFEntries</span>
    <span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 35)" onmouseover="showTip(event, 'fs22', 35)" class="fn">filteredGFFEntries</span> <span class="o">=</span> 
        <span onmouseout="hideTip(event, 'fs4', 36)" onmouseover="showTip(event, 'fs4', 36)" class="id">features</span> 
        <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs23', 37)" onmouseover="showTip(event, 'fs23', 37)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 38)" onmouseover="showTip(event, 'fs24', 38)" class="id">choose</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs25', 39)" onmouseover="showTip(event, 'fs25', 39)" class="fn">x</span> <span class="k">-&gt;</span>    
            <span class="k">match</span> <span onmouseout="hideTip(event, 'fs25', 40)" onmouseover="showTip(event, 'fs25', 40)" class="fn">x</span> <span class="k">with</span>
            <span class="pn">|</span> <span onmouseout="hideTip(event, 'fs5', 41)" onmouseover="showTip(event, 'fs5', 41)" class="m">GFF3</span><span class="pn">.</span><span class="id">GFFEntryLine</span> <span onmouseout="hideTip(event, 'fs25', 42)" onmouseover="showTip(event, 'fs25', 42)" class="id">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs26', 43)" onmouseover="showTip(event, 'fs26', 43)" class="id">Some</span> <span onmouseout="hideTip(event, 'fs25', 44)" onmouseover="showTip(event, 'fs25', 44)" class="id">x</span>
            <span class="pn">|</span> <span class="id">_</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs27', 45)" onmouseover="showTip(event, 'fs27', 45)" class="id">None</span><span class="pn">)</span>

    <span class="c">//get all CDS features</span>
    <span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 46)" onmouseover="showTip(event, 'fs28', 46)" class="fn">filteredCDSFeatures</span> <span class="o">=</span>
        <span onmouseout="hideTip(event, 'fs22', 47)" onmouseover="showTip(event, 'fs22', 47)" class="fn">filteredGFFEntries</span>
        <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs23', 48)" onmouseover="showTip(event, 'fs23', 48)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 49)" onmouseover="showTip(event, 'fs29', 49)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs30', 50)" onmouseover="showTip(event, 'fs30', 50)" class="fn">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs30', 51)" onmouseover="showTip(event, 'fs30', 51)" class="fn">x</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 52)" onmouseover="showTip(event, 'fs31', 52)" class="id">Feature</span> <span class="o">=</span> <span class="s">&quot;CDS&quot;</span><span class="pn">)</span>

    <span onmouseout="hideTip(event, 'fs28', 53)" onmouseover="showTip(event, 'fs28', 53)" class="fn">filteredCDSFeatures</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs23', 54)" onmouseover="showTip(event, 'fs23', 54)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs32', 55)" onmouseover="showTip(event, 'fs32', 55)" class="id">head</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs33', 56)" onmouseover="showTip(event, 'fs33', 56)" class="id">firstCDSSequence</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 57)" onmouseover="showTip(event, 'fs5', 57)" class="m">GFF3</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs34', 58)" onmouseover="showTip(event, 'fs34', 58)" class="id">getSequence</span> <span onmouseout="hideTip(event, 'fs21', 59)" onmouseover="showTip(event, 'fs21', 59)" class="id">firstCDS</span> <span onmouseout="hideTip(event, 'fs4', 60)" onmouseover="showTip(event, 'fs4', 60)" class="id">features</span>
</code></pre>
<table class="pre"><tr><td class="snippet"><pre class="fssnip"><code lang="text">seq [A; T; G; A; ...]
</code></pre></td></tr></table>

        <div class="fsdocs-tip" id="fs1">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs2">namespace BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs3">val filepathGFF: string</div>
<div class="fsdocs-tip" id="fs4">val features: FileFormats.GFF3.GFFLine&lt;Nucleotides.Nucleotide&gt; seq</div>
<div class="fsdocs-tip" id="fs5">module GFF3

from BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs6">val fromFile: fastAconverter: (char seq -&gt; #(&#39;a0 seq)) -&gt; filepath: string -&gt; FileFormats.GFF3.GFFLine&lt;&#39;a0&gt; seq<br /><em>&lt;summary&gt;
reads in a file and gives a GFFLine&amp;lt;&#39;a&amp;gt; list. If file contains a FastA sequence it is converted to FastA.FastaItem with given converter. (Use &#39;id&#39; as converter if no FastA is required).
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs7">module BioArray

from BioFSharp<br /><em>&lt;summary&gt;
This module contains the BioArray type and its according functions. The BioArray type is an array of objects using the IBioItem interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs8">val ofNucleotideString: s: #(char seq) -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><em>&lt;summary&gt;
 Generates nucleotide sequence of one-letter-code raw string
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs9">val featuresSanityCheck: unit</div>
<div class="fsdocs-tip" id="fs10">val sanityCheck: filepath: string -&gt; unit</div>
<div class="fsdocs-tip" id="fs11">val filepathSO_Terms: string</div>
<div class="fsdocs-tip" id="fs12">val featuresSanityCheckWithSOTerm: unit</div>
<div class="fsdocs-tip" id="fs13">val sanityCheckWithSOTerm: so_TermsPath: string -&gt; filepath: string -&gt; unit<br /><em>&lt;summary&gt;
Validates GFF3 file. Prints all appearances of errors with line index. If no (SO)FA check is needed set &quot;&quot; as so_TermsPath.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs14">val searchterm: string<br /><em>Term to search for:</em></div>
<div class="fsdocs-tip" id="fs15">val gffExampleSearch: FileFormats.GFF3.GFFEntry seq<br /><em>with this function you can search features which are related to the searchterm</em></div>
<div class="fsdocs-tip" id="fs16">val relationshipSearch: gffList: FileFormats.GFF3.GFFLine&lt;&#39;a&gt; seq -&gt; searchterm: string -&gt; FileFormats.GFF3.GFFEntry seq<br /><em>&lt;summary&gt;
Searches for a term and gives a list of all features of which the searchterm is the mainfeature (ID) or a child of it (Parent).
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs17">val gffExampleWrite: unit<br /><em>Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FASTA converter and a destination filepath and writes it into a .gff. Hint: Use converter = id if no FastA sequence is included.</em></div>
<div class="fsdocs-tip" id="fs18">val write: converter: (&#39;a -&gt; char) -&gt; path: string -&gt; input: FileFormats.GFF3.GFFLine&lt;&#39;a&gt; seq -&gt; unit (requires &#39;a :&gt; &#39;a0 seq)<br /><em>&lt;summary&gt;
writes GFF lines to file. Hint: Use id as converter if no FASTA sequence is included.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs19">module BioItem

from BioFSharp<br /><em>&lt;summary&gt;
 Basic functions on IBioItems interface
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs20">val symbol: bItem: #IBioItem -&gt; char<br /><em>&lt;summary&gt;
 Returns the symbol of the bio item
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs21">val firstCDS: FileFormats.GFF3.GFFEntry</div>
<div class="fsdocs-tip" id="fs22">val filteredGFFEntries: FileFormats.GFF3.GFFEntry seq</div>
<div class="fsdocs-tip" id="fs23">module Seq

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs24">val choose: chooser: (&#39;T -&gt; &#39;U option) -&gt; source: &#39;T seq -&gt; &#39;U seq</div>
<div class="fsdocs-tip" id="fs25">val x: FileFormats.GFF3.GFFLine&lt;Nucleotides.Nucleotide&gt;</div>
<div class="fsdocs-tip" id="fs26">union case Option.Some: Value: &#39;T -&gt; Option&lt;&#39;T&gt;</div>
<div class="fsdocs-tip" id="fs27">union case Option.None: Option&lt;&#39;T&gt;</div>
<div class="fsdocs-tip" id="fs28">val filteredCDSFeatures: FileFormats.GFF3.GFFEntry seq</div>
<div class="fsdocs-tip" id="fs29">val filter: predicate: (&#39;T -&gt; bool) -&gt; source: &#39;T seq -&gt; &#39;T seq</div>
<div class="fsdocs-tip" id="fs30">val x: FileFormats.GFF3.GFFEntry</div>
<div class="fsdocs-tip" id="fs31">FileFormats.GFF3.GFFEntry.Feature: string<br /><em>&lt;summary&gt;
feature, type or method; has to be a term from SO or SO accession number
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs32">val head: source: &#39;T seq -&gt; &#39;T</div>
<div class="fsdocs-tip" id="fs33">val firstCDSSequence: Nucleotides.Nucleotide seq</div>
<div class="fsdocs-tip" id="fs34">val getSequence: cDSfeature: FileFormats.GFF3.GFFEntry -&gt; gFFFile: FileFormats.GFF3.GFFLine&lt;&#39;a&gt; seq -&gt; &#39;a seq (requires &#39;a :&gt; &#39;a0 seq)<br /><em>&lt;summary&gt;
if a FastA sequence is included this function searches the features corresponding sequence
&lt;/summary&gt;</em></div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#GFF3-parsing">
      GFF3 parsing
    </a>
  </li>
  <li class="level-2">
    <a href="#Generic-Feature-Format-Version-3">
      Generic Feature Format Version 3
    </a>
  </li>
  <li class="level-2">
    <a href="#Reading-GFF3-files">
      Reading GFF3 files
    </a>
  </li>
  <li class="level-2">
    <a href="#How-to-use-GFF3SanityCheck">
      How to use GFF3SanityCheck
    </a>
  </li>
  <li class="level-2">
    <a href="#How-to-use-GFF3RelationshipSearch">
      How to use GFF3RelationshipSearch
    </a>
  </li>
  <li class="level-2">
    <a href="#Writing-GFF3-files">
      Writing GFF3 files
    </a>
  </li>
  <li class="level-2">
    <a href="#Example-Sequence-of-CDS">
      Example: Sequence of CDS
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>