<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Pairwise Alignment
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Pairwise-Alignment" class="anchor" href="#Pairwise-Alignment">Pairwise Alignment</a></h1>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/Algorithm/PairwiseAlignment.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/HLWeil">&#128366;view author of this tutorial</a>
<br><br>
<p>In this short tutorial, the usage of the Pairwise alignment-implementation is demonstrated.
For global alignments, the <b>NeedlemanWunsch</b>-algorithm is used. For local alignments, the <b>SmithWaterman</b>-algorithm is used.
For both implementations, the gapvalues are evaluated using the <b>affine</b> gapscoremodel.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="prep">#r</span> <span class="s">&quot;BioFSharp.dll&quot;</span>
<span class="k">open</span> <span class="i">BioFSharp</span>
<span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">Algorithm</span>
<span class="k">open</span> <span class="i">PairwiseAlignment</span>
</code></pre></td>
</tr>
</table>
<div class="box">
<h2><a name="Aligning-aminoacid-and-nucleotide-sequences" class="anchor" href="#Aligning-aminoacid-and-nucleotide-sequences">Aligning aminoacid- and nucleotide-sequences</a></h2>
<p>For defining the scores of matching and missmatching characters, the <b>scoring</b>-function is defined. In the case of Aminoacid- or Nucleotide-sequence-alignments, the integrated substitution-matrices can be used.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span class="i">aaScoring</span> <span class="o">=</span> <span class="i">ScoringMatrix</span><span class="o">.</span><span class="i">getScoringMatrixAminoAcid</span> <span class="i">ScoringMatrix</span><span class="o">.</span><span class="i">ScoringMatrixAminoAcid</span><span class="o">.</span><span class="i">BLOSUM62</span>

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span class="i">nucScoring</span> <span class="o">=</span> <span class="i">ScoringMatrix</span><span class="o">.</span><span class="i">getScoringMatrixNucleotide</span>  <span class="i">ScoringMatrix</span><span class="o">.</span><span class="i">ScoringMatrixNucleotide</span><span class="o">.</span><span class="i">EDNA</span>
</code></pre></td>
</tr>
</table>
<p>In order to align two sequences, not only the values for substitutions, but also the <b>costs</b> for gaps are needed. In this implementation, an affine gap-penalty is realized. An affine gap penalty weights continuations of gaps different than the opening of a gap.
The scoring function (in this case a substitution matrix) and the two gap penalty values are combined into the <b>Costs</b>-type.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span class="i">aaCosts</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">5</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span class="i">aaScoring</span> 
    }

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span class="i">nucCosts</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span class="i">nucScoring</span> 
    }
</code></pre></td>
</tr>
</table>
<p>The alignment functions use <b>Arrays</b> as input. The Elements can be of any type, but require type equality. Also they need to have type equality with the input of the scoring function.
Both the global and local alignment algorithms take the same parameters (costs,firstSequence,secondSequence) and return the same format.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span class="i">aaSeq1</span> <span class="o">=</span> <span class="s">&quot;ACDM&quot;</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidSymbolString</span>
<span class="k">let</span> <span class="i">aaSeq2</span> <span class="o">=</span> <span class="s">&quot;MAACEDM&quot;</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidSymbolString</span>

<span class="k">let</span> <span class="i">globalAAAlignment</span> <span class="o">=</span> <span class="i">NeedlemanWunsch</span><span class="o">.</span><span class="i">runAminoAcidSymbol</span> <span class="i">aaCosts</span> <span class="i">aaSeq1</span> <span class="i">aaSeq2</span>
<span class="k">let</span> <span class="i">localAAAlignment</span> <span class="o">=</span> <span class="i">SmithWaterman</span><span class="o">.</span><span class="i">runAminoAcidSymbol</span> <span class="i">aaCosts</span> <span class="i">aaSeq1</span> <span class="i">aaSeq2</span>
</code></pre></td>
</tr>
</table>
<p>This returns:<br />
<em>The field MetaData carries the score of the alignment</em>
<pre></p>
<p>val globalAAAlignment :
Alignment.Alignment<AminoAcidSymbols.AminoAcidSymbol list,Score> =
{MetaData = 12;
AlignedSequences = [[-; -; A; C; -; D; M]; [M; A; A; C; E; D; M]];}</p>
<p>val localAAAlignment :
Alignment.Alignment<AminoAcidSymbols.AminoAcidSymbol list,Score> =
{MetaData = 19;
AlignedSequences = [[A; C; -; D; M]; [A; C; E; D; M]];}
</pre></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For nucleotides</span>
<span class="k">let</span> <span class="i">nucSeq1</span> <span class="o">=</span> <span class="s">&quot;ATGA&quot;</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofNucleotideString</span>
<span class="k">let</span> <span class="i">nucSeq2</span> <span class="o">=</span> <span class="s">&quot;BATVAWG&quot;</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofNucleotideString</span>

<span class="k">let</span> <span class="i">globalNucAlignment</span> <span class="o">=</span> <span class="i">NeedlemanWunsch</span><span class="o">.</span><span class="i">runNucleotide</span> <span class="i">nucCosts</span> <span class="i">nucSeq1</span> <span class="i">nucSeq2</span>
<span class="k">let</span> <span class="i">localNucAlignment</span> <span class="o">=</span> <span class="i">SmithWaterman</span><span class="o">.</span><span class="i">runNucleotide</span> <span class="i">nucCosts</span> <span class="i">nucSeq1</span> <span class="i">nucSeq2</span>
</code></pre></td>
</tr>
</table>
</div>
<div class="box">
<h2><a name="Aligning-anything-else" class="anchor" href="#Aligning-anything-else">Aligning anything else</a></h2>
<p>This implementation was aimed to be as generic as possible. To achieve this, the scoring function can be designed at will, the only constraints being the need for two input variables and the type equality.<br />
Also besides the alignment functions which only take BioItems as input and represent the gaps by the appropriate gaps of type BioItem. There is also a generic alignment function <code>runGeneric</code> which returns <code>lists of options</code>, where None represents gaps. Therefore any input type can be used, given it matches the cost function.</p>
<p>For example, one could use a simple "if else" equality function to match nucleotides</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">scoring</span> <span class="i">a</span> <span class="i">b</span> <span class="o">=</span> 
    <span class="k">if</span> <span class="i">a</span> <span class="o">=</span> <span class="i">b</span> <span class="k">then</span> <span class="n">2</span>
    <span class="k">else</span> <span class="o">-</span><span class="n">2</span>

<span class="k">let</span> <span class="i">costs</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span class="i">scoring</span>
    }


<span class="k">let</span> <span class="i">globalAlignmentNuc</span> <span class="o">=</span> <span class="i">NeedlemanWunsch</span><span class="o">.</span><span class="i">runGeneric</span> <span class="i">costs</span> <span class="i">nucSeq1</span> <span class="i">nucSeq2</span>
<span class="k">let</span> <span class="i">localAlignmentNuc</span> <span class="o">=</span> <span class="i">SmithWaterman</span><span class="o">.</span><span class="i">runGeneric</span> <span class="i">costs</span> <span class="i">nucSeq1</span> <span class="i">nucSeq2</span>
</code></pre></td>
</tr>
</table>
<p>This returns:<br />
<pre></p>
<p>val globalAlignmentNuc :
Alignment.Alignment<Nucleotides.Nucleotide option list,Score> =
{MetaData = -1;
AlignedSequences =
[[None; Some A; Some T; Some G; Some A; None; None];
[Some B; Some A; Some T; Some V; Some A; Some W; Some G]];}</p>
<p>val localAlignmentNuc :
Alignment.Alignment<Nucleotides.Nucleotide option list,Score> =
{MetaData = 1;
AlignedSequences =
[[Some A; Some T; Some G; Some A]; [Some A; Some T; Some V; Some A]];}
</pre></p>
<p>or also Integers:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">intSeq1</span> <span class="o">=</span> [|<span class="n">1</span>;<span class="n">2</span>;<span class="n">3</span>;<span class="n">4</span>;<span class="n">5</span>|]
<span class="k">let</span> <span class="i">intSeq2</span> <span class="o">=</span> [|<span class="n">1</span>;<span class="n">1</span>;<span class="n">2</span>;<span class="n">4</span>;<span class="n">6</span>;<span class="n">7</span>;|]

<span class="k">let</span> <span class="i">globalAlignmentInt</span> <span class="o">=</span> <span class="i">NeedlemanWunsch</span><span class="o">.</span><span class="i">runGeneric</span> <span class="i">costs</span> <span class="i">intSeq1</span> <span class="i">intSeq2</span>
<span class="k">let</span> <span class="i">localAlignmentInt</span> <span class="o">=</span> <span class="i">SmithWaterman</span><span class="o">.</span><span class="i">runGeneric</span> <span class="i">costs</span> <span class="i">intSeq1</span> <span class="i">intSeq2</span>
</code></pre></td>
</tr>
</table>
</div>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                <li>
                    <a href="/BioFSharp/BioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF3.html">GFF format</a></li>
                <li><a href="/BioFSharp/GenBank.html">GenBank</a></li>
                <li>
                    <a href="/BioFSharp/Clustal.html">Clustal</a>
                </li>        
                <li>
                    <a href="/BioFSharp/Newick.html">Newick</a>
                </li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/CSV.html">Delimited files</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                
                <li class="nav-header">Documentation</li>
                <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
