<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Alignment</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<table class="HeadAPI">
<td class="Head"><h1>Pairwise Alignment</h1></td>
<td class="API">
    <a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/biofsharp-algorithm-pairwisealignment.html" >&#128194;View module documentation</a>
<td>
</table>
<p>In this short tutorial, the usage of the Pairwise alignment-implementation is demonstrated.
For global alignments, the <b>NeedlemanWunsch</b>-algorithm is used. For local alignments, the <b>SmithWaterman</b>-algorithm is used.
For both implementations, the gapvalues are evaluated using the <b>affine</b> gapscoremodel.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="prep">#r</span> <span class="s">&quot;BioFSharp.dll&quot;</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="i">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="i">BioFSharp</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="i">Algorithm</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="i">PairwiseAlignment</span>
</code></pre></td>
</tr>
</table>
<div class="box">
<h2><a name="Aligning-aminoacid-and-nucleotide-sequences" class="anchor" href="#Aligning-aminoacid-and-nucleotide-sequences">Aligning aminoacid- and nucleotide-sequences</a></h2>
<p>For defining the scores of matching and missmatching characters, the <b>scoring</b>-function is defined. In the case of Aminoacid- or Nucleotide-sequence-alignments, the integrated substitution-matrices can be used.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="f">aaScoring</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="t">ScoringMatrix</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="f">getScoringMatrixAminoAcid</span> <span onmouseout="hideTip(event, 'fs5', 8)" onmouseover="showTip(event, 'fs5', 8)" class="t">ScoringMatrix</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="t">ScoringMatrixAminoAcid</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="p">BLOSUM62</span>

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 11)" onmouseover="showTip(event, 'fs9', 11)" class="f">nucScoring</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 12)" onmouseover="showTip(event, 'fs5', 12)" class="t">ScoringMatrix</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs10', 13)" onmouseover="showTip(event, 'fs10', 13)" class="f">getScoringMatrixNucleotide</span>  <span onmouseout="hideTip(event, 'fs5', 14)" onmouseover="showTip(event, 'fs5', 14)" class="t">ScoringMatrix</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs11', 15)" onmouseover="showTip(event, 'fs11', 15)" class="t">ScoringMatrixNucleotide</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs12', 16)" onmouseover="showTip(event, 'fs12', 16)" class="p">EDNA</span>
</code></pre></td>
</tr>
</table>
<p>In order to align two sequences, not only the values for substitutions, but also the <b>costs</b> for gaps are needed. In this implementation, an affine gap-penalty is realized. An affine gap penalty weights continuations of gaps different than the opening of a gap.
The scoring function (in this case a substitution matrix) and the two gap penalty values are combined into the <b>Costs</b>-type.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs13', 17)" onmouseover="showTip(event, 'fs13', 17)" class="i">aaCosts</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">5</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 18)" onmouseover="showTip(event, 'fs4', 18)" class="f">aaScoring</span> 
    }

<span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 19)" onmouseover="showTip(event, 'fs14', 19)" class="i">nucCosts</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs9', 20)" onmouseover="showTip(event, 'fs9', 20)" class="f">nucScoring</span> 
    }
</code></pre></td>
</tr>
</table>
<p>The alignment functions use <b>Arrays</b> as input. The Elements can be of any type, but require type equality. Also they need to have type equality with the input of the scoring function.
Both the global and local alignment algorithms take the same parameters (costs,firstSequence,secondSequence) and return the same format.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For aminoacids</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 21)" onmouseover="showTip(event, 'fs15', 21)" class="i">aaSeq1</span> <span class="o">=</span> <span class="s">&quot;ACDM&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 22)" onmouseover="showTip(event, 'fs16', 22)" class="t">BioArray</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs17', 23)" onmouseover="showTip(event, 'fs17', 23)" class="f">ofAminoAcidSymbolString</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 24)" onmouseover="showTip(event, 'fs18', 24)" class="i">aaSeq2</span> <span class="o">=</span> <span class="s">&quot;MAACEDM&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="t">BioArray</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs17', 26)" onmouseover="showTip(event, 'fs17', 26)" class="f">ofAminoAcidSymbolString</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 27)" onmouseover="showTip(event, 'fs19', 27)" class="i">globalAAAlignment</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs20', 28)" onmouseover="showTip(event, 'fs20', 28)" class="t">NeedlemanWunsch</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs21', 29)" onmouseover="showTip(event, 'fs21', 29)" class="f">runAminoAcidSymbol</span> <span onmouseout="hideTip(event, 'fs13', 30)" onmouseover="showTip(event, 'fs13', 30)" class="i">aaCosts</span> <span onmouseout="hideTip(event, 'fs15', 31)" onmouseover="showTip(event, 'fs15', 31)" class="i">aaSeq1</span> <span onmouseout="hideTip(event, 'fs18', 32)" onmouseover="showTip(event, 'fs18', 32)" class="i">aaSeq2</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 33)" onmouseover="showTip(event, 'fs22', 33)" class="i">localAAAlignment</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs23', 34)" onmouseover="showTip(event, 'fs23', 34)" class="t">SmithWaterman</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs24', 35)" onmouseover="showTip(event, 'fs24', 35)" class="f">runAminoAcidSymbol</span> <span onmouseout="hideTip(event, 'fs13', 36)" onmouseover="showTip(event, 'fs13', 36)" class="i">aaCosts</span> <span onmouseout="hideTip(event, 'fs15', 37)" onmouseover="showTip(event, 'fs15', 37)" class="i">aaSeq1</span> <span onmouseout="hideTip(event, 'fs18', 38)" onmouseover="showTip(event, 'fs18', 38)" class="i">aaSeq2</span>
</code></pre></td>
</tr>
</table>
<p>This returns:<br />
<em>The field MetaData carries the score of the alignment</em>
<pre></p>
<p>val globalAAAlignment :
Alignment.Alignment<AminoAcidSymbols.AminoAcidSymbol list,Score> =
{MetaData = 12;
AlignedSequences = [[-; -; A; C; -; D; M]; [M; A; A; C; E; D; M]];}</p>
<p>val localAAAlignment :
Alignment.Alignment<AminoAcidSymbols.AminoAcidSymbol list,Score> =
{MetaData = 19;
AlignedSequences = [[A; C; -; D; M]; [A; C; E; D; M]];}
</pre></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//For nucleotides</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs25', 39)" onmouseover="showTip(event, 'fs25', 39)" class="i">nucSeq1</span> <span class="o">=</span> <span class="s">&quot;ATGA&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 40)" onmouseover="showTip(event, 'fs16', 40)" class="t">BioArray</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs26', 41)" onmouseover="showTip(event, 'fs26', 41)" class="f">ofNucleotideString</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 42)" onmouseover="showTip(event, 'fs27', 42)" class="i">nucSeq2</span> <span class="o">=</span> <span class="s">&quot;BATVAWG&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs16', 43)" onmouseover="showTip(event, 'fs16', 43)" class="t">BioArray</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs26', 44)" onmouseover="showTip(event, 'fs26', 44)" class="f">ofNucleotideString</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 45)" onmouseover="showTip(event, 'fs28', 45)" class="i">globalNucAlignment</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs20', 46)" onmouseover="showTip(event, 'fs20', 46)" class="t">NeedlemanWunsch</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs29', 47)" onmouseover="showTip(event, 'fs29', 47)" class="f">runNucleotide</span> <span onmouseout="hideTip(event, 'fs14', 48)" onmouseover="showTip(event, 'fs14', 48)" class="i">nucCosts</span> <span onmouseout="hideTip(event, 'fs25', 49)" onmouseover="showTip(event, 'fs25', 49)" class="i">nucSeq1</span> <span onmouseout="hideTip(event, 'fs27', 50)" onmouseover="showTip(event, 'fs27', 50)" class="i">nucSeq2</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs30', 51)" onmouseover="showTip(event, 'fs30', 51)" class="i">localNucAlignment</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs23', 52)" onmouseover="showTip(event, 'fs23', 52)" class="t">SmithWaterman</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs31', 53)" onmouseover="showTip(event, 'fs31', 53)" class="f">runNucleotide</span> <span onmouseout="hideTip(event, 'fs14', 54)" onmouseover="showTip(event, 'fs14', 54)" class="i">nucCosts</span> <span onmouseout="hideTip(event, 'fs25', 55)" onmouseover="showTip(event, 'fs25', 55)" class="i">nucSeq1</span> <span onmouseout="hideTip(event, 'fs27', 56)" onmouseover="showTip(event, 'fs27', 56)" class="i">nucSeq2</span>
</code></pre></td>
</tr>
</table>
</div>
<div class="box">
<h2><a name="Aligning-anything-else" class="anchor" href="#Aligning-anything-else">Aligning anything else</a></h2>
<p>This implementation was aimed to be as generic as possible. To achieve this, the scoring function can be designed at will, the only constraints being the need for two input variables and the type equality.<br />
Also besides the alignment functions which only take BioItems as input and represent the gaps by the appropriate gaps of type BioItem. There is also a generic alignment function <code>runGeneric</code> which returns <code>lists of options</code>, where None represents gaps. Therefore any input type can be used, given it matches the cost function.</p>
<p>For example, one could use a simple "if else" equality function to match nucleotides</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs32', 57)" onmouseover="showTip(event, 'fs32', 57)" class="f">scoring</span> <span onmouseout="hideTip(event, 'fs33', 58)" onmouseover="showTip(event, 'fs33', 58)" class="i">a</span> <span onmouseout="hideTip(event, 'fs34', 59)" onmouseover="showTip(event, 'fs34', 59)" class="i">b</span> <span class="o">=</span> 
    <span class="k">if</span> <span onmouseout="hideTip(event, 'fs33', 60)" onmouseover="showTip(event, 'fs33', 60)" class="i">a</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 61)" onmouseover="showTip(event, 'fs34', 61)" class="i">b</span> <span class="k">then</span> <span class="n">2</span>
    <span class="k">else</span> <span class="o">-</span><span class="n">2</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs35', 62)" onmouseover="showTip(event, 'fs35', 62)" class="i">costs</span> <span class="o">=</span> {
    <span class="i">Open</span> <span class="o">=</span> <span class="o">-</span><span class="n">2</span>
    <span class="i">Continuation</span> <span class="o">=</span> <span class="o">-</span><span class="n">1</span>
    <span class="i">Similarity</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs32', 63)" onmouseover="showTip(event, 'fs32', 63)" class="f">scoring</span>
    }


<span class="k">let</span> <span onmouseout="hideTip(event, 'fs36', 64)" onmouseover="showTip(event, 'fs36', 64)" class="i">globalAlignmentNuc</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs20', 65)" onmouseover="showTip(event, 'fs20', 65)" class="t">NeedlemanWunsch</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs37', 66)" onmouseover="showTip(event, 'fs37', 66)" class="f">runGeneric</span> <span onmouseout="hideTip(event, 'fs35', 67)" onmouseover="showTip(event, 'fs35', 67)" class="i">costs</span> <span onmouseout="hideTip(event, 'fs25', 68)" onmouseover="showTip(event, 'fs25', 68)" class="i">nucSeq1</span> <span onmouseout="hideTip(event, 'fs27', 69)" onmouseover="showTip(event, 'fs27', 69)" class="i">nucSeq2</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs38', 70)" onmouseover="showTip(event, 'fs38', 70)" class="i">localAlignmentNuc</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs23', 71)" onmouseover="showTip(event, 'fs23', 71)" class="t">SmithWaterman</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs39', 72)" onmouseover="showTip(event, 'fs39', 72)" class="f">runGeneric</span> <span onmouseout="hideTip(event, 'fs35', 73)" onmouseover="showTip(event, 'fs35', 73)" class="i">costs</span> <span onmouseout="hideTip(event, 'fs25', 74)" onmouseover="showTip(event, 'fs25', 74)" class="i">nucSeq1</span> <span onmouseout="hideTip(event, 'fs27', 75)" onmouseover="showTip(event, 'fs27', 75)" class="i">nucSeq2</span>
</code></pre></td>
</tr>
</table>
<p>This returns:<br />
<pre></p>
<p>val globalAlignmentNuc :
Alignment.Alignment<Nucleotides.Nucleotide option list,Score> =
{MetaData = -1;
AlignedSequences =
[[None; Some A; Some T; Some G; Some A; None; None];
[Some B; Some A; Some T; Some V; Some A; Some W; Some G]];}</p>
<p>val localAlignmentNuc :
Alignment.Alignment<Nucleotides.Nucleotide option list,Score> =
{MetaData = 1;
AlignedSequences =
[[Some A; Some T; Some G; Some A]; [Some A; Some T; Some V; Some A]];}
</pre></p>
<p>or also Integers:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs40', 76)" onmouseover="showTip(event, 'fs40', 76)" class="i">intSeq1</span> <span class="o">=</span> [|<span class="n">1</span>;<span class="n">2</span>;<span class="n">3</span>;<span class="n">4</span>;<span class="n">5</span>|]
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs41', 77)" onmouseover="showTip(event, 'fs41', 77)" class="i">intSeq2</span> <span class="o">=</span> [|<span class="n">1</span>;<span class="n">1</span>;<span class="n">2</span>;<span class="n">4</span>;<span class="n">6</span>;<span class="n">7</span>;|]

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs42', 78)" onmouseover="showTip(event, 'fs42', 78)" class="i">globalAlignmentInt</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs20', 79)" onmouseover="showTip(event, 'fs20', 79)" class="t">NeedlemanWunsch</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs37', 80)" onmouseover="showTip(event, 'fs37', 80)" class="f">runGeneric</span> <span onmouseout="hideTip(event, 'fs35', 81)" onmouseover="showTip(event, 'fs35', 81)" class="i">costs</span> <span onmouseout="hideTip(event, 'fs40', 82)" onmouseover="showTip(event, 'fs40', 82)" class="i">intSeq1</span> <span onmouseout="hideTip(event, 'fs41', 83)" onmouseover="showTip(event, 'fs41', 83)" class="i">intSeq2</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs43', 84)" onmouseover="showTip(event, 'fs43', 84)" class="i">localAlignmentInt</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs23', 85)" onmouseover="showTip(event, 'fs23', 85)" class="t">SmithWaterman</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs39', 86)" onmouseover="showTip(event, 'fs39', 86)" class="f">runGeneric</span> <span onmouseout="hideTip(event, 'fs35', 87)" onmouseover="showTip(event, 'fs35', 87)" class="i">costs</span> <span onmouseout="hideTip(event, 'fs40', 88)" onmouseover="showTip(event, 'fs40', 88)" class="i">intSeq1</span> <span onmouseout="hideTip(event, 'fs41', 89)" onmouseover="showTip(event, 'fs41', 89)" class="i">intSeq2</span>
</code></pre></td>
</tr>
</table>
</div>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace BioFSharp.Algorithm</div>
<div class="tip" id="fs3">module PairwiseAlignment<br /><br />from BioFSharp.Algorithm</div>
<div class="tip" id="fs4">val aaScoring : (AminoAcidSymbols.AminoAcidSymbol -&gt; AminoAcidSymbols.AminoAcidSymbol -&gt; int)<br /><br />Full name: Alignment.aaScoring</div>
<div class="tip" id="fs5">module ScoringMatrix<br /><br />from BioFSharp.Algorithm</div>
<div class="tip" id="fs6">val getScoringMatrixAminoAcid : scoringMatrixType:ScoringMatrix.ScoringMatrixAminoAcid -&gt; (AminoAcidSymbols.AminoAcidSymbol -&gt; AminoAcidSymbols.AminoAcidSymbol -&gt; int)<br /><br />Full name: BioFSharp.Algorithm.ScoringMatrix.getScoringMatrixAminoAcid</div>
<div class="tip" id="fs7">type ScoringMatrixAminoAcid =<br />&#160;&#160;| BLOSUM45<br />&#160;&#160;| BLOSUM50<br />&#160;&#160;| BLOSUM62<br />&#160;&#160;| BLOSUM80<br />&#160;&#160;| PAM30<br />&#160;&#160;| PAM70<br />&#160;&#160;| PAM250<br />&#160;&#160;static member toFileName : (ScoringMatrixAminoAcid -&gt; string)<br /><br />Full name: BioFSharp.Algorithm.ScoringMatrix.ScoringMatrixAminoAcid</div>
<div class="tip" id="fs8">union case ScoringMatrix.ScoringMatrixAminoAcid.BLOSUM62: ScoringMatrix.ScoringMatrixAminoAcid</div>
<div class="tip" id="fs9">val nucScoring : (Nucleotides.Nucleotide -&gt; Nucleotides.Nucleotide -&gt; int)<br /><br />Full name: Alignment.nucScoring</div>
<div class="tip" id="fs10">val getScoringMatrixNucleotide : scoringMatrixType:ScoringMatrix.ScoringMatrixNucleotide -&gt; (Nucleotides.Nucleotide -&gt; Nucleotides.Nucleotide -&gt; int)<br /><br />Full name: BioFSharp.Algorithm.ScoringMatrix.getScoringMatrixNucleotide</div>
<div class="tip" id="fs11">type ScoringMatrixNucleotide =<br />&#160;&#160;| EDNA<br />&#160;&#160;| Default<br />&#160;&#160;static member toFileName : (ScoringMatrixNucleotide -&gt; string)<br /><br />Full name: BioFSharp.Algorithm.ScoringMatrix.ScoringMatrixNucleotide</div>
<div class="tip" id="fs12">union case ScoringMatrix.ScoringMatrixNucleotide.EDNA: ScoringMatrix.ScoringMatrixNucleotide</div>
<div class="tip" id="fs13">val aaCosts : Costs&lt;AminoAcidSymbols.AminoAcidSymbol&gt;<br /><br />Full name: Alignment.aaCosts</div>
<div class="tip" id="fs14">val nucCosts : Costs&lt;Nucleotides.Nucleotide&gt;<br /><br />Full name: Alignment.nucCosts</div>
<div class="tip" id="fs15">val aaSeq1 : BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;<br /><br />Full name: Alignment.aaSeq1</div>
<div class="tip" id="fs16">module BioArray<br /><br />from BioFSharp</div>
<div class="tip" id="fs17">val ofAminoAcidSymbolString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;<br /><br />Full name: BioFSharp.BioArray.ofAminoAcidSymbolString</div>
<div class="tip" id="fs18">val aaSeq2 : BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt;<br /><br />Full name: Alignment.aaSeq2</div>
<div class="tip" id="fs19">val globalAAAlignment : Alignment.Alignment&lt;AminoAcidSymbols.AminoAcidSymbol list,Score&gt;<br /><br />Full name: Alignment.globalAAAlignment</div>
<div class="tip" id="fs20">module NeedlemanWunsch<br /><br />from BioFSharp.Algorithm.PairwiseAlignment</div>
<div class="tip" id="fs21">val runAminoAcidSymbol : costs:Costs&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; fstSeq:BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; sndSeq:BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; Alignment.Alignment&lt;AminoAcidSymbols.AminoAcidSymbol list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.NeedlemanWunsch.runAminoAcidSymbol</div>
<div class="tip" id="fs22">val localAAAlignment : Alignment.Alignment&lt;AminoAcidSymbols.AminoAcidSymbol list,Score&gt;<br /><br />Full name: Alignment.localAAAlignment</div>
<div class="tip" id="fs23">module SmithWaterman<br /><br />from BioFSharp.Algorithm.PairwiseAlignment</div>
<div class="tip" id="fs24">val runAminoAcidSymbol : costs:Costs&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; fstSeq:BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; sndSeq:BioArray.BioArray&lt;AminoAcidSymbols.AminoAcidSymbol&gt; -&gt; Alignment.Alignment&lt;AminoAcidSymbols.AminoAcidSymbol list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.SmithWaterman.runAminoAcidSymbol</div>
<div class="tip" id="fs25">val nucSeq1 : BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><br />Full name: Alignment.nucSeq1</div>
<div class="tip" id="fs26">val ofNucleotideString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><br />Full name: BioFSharp.BioArray.ofNucleotideString</div>
<div class="tip" id="fs27">val nucSeq2 : BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;<br /><br />Full name: Alignment.nucSeq2</div>
<div class="tip" id="fs28">val globalNucAlignment : Alignment.Alignment&lt;Nucleotides.Nucleotide list,Score&gt;<br /><br />Full name: Alignment.globalNucAlignment</div>
<div class="tip" id="fs29">val runNucleotide : costs:Costs&lt;Nucleotides.Nucleotide&gt; -&gt; fstSeq:BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; sndSeq:BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; Alignment.Alignment&lt;Nucleotides.Nucleotide list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.NeedlemanWunsch.runNucleotide</div>
<div class="tip" id="fs30">val localNucAlignment : Alignment.Alignment&lt;Nucleotides.Nucleotide list,Score&gt;<br /><br />Full name: Alignment.localNucAlignment</div>
<div class="tip" id="fs31">val runNucleotide : costs:Costs&lt;Nucleotides.Nucleotide&gt; -&gt; fstSeq:BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; sndSeq:BioArray.BioArray&lt;Nucleotides.Nucleotide&gt; -&gt; Alignment.Alignment&lt;Nucleotides.Nucleotide list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.SmithWaterman.runNucleotide</div>
<div class="tip" id="fs32">val scoring : a:&#39;a -&gt; b:&#39;a -&gt; int (requires equality)<br /><br />Full name: Alignment.scoring</div>
<div class="tip" id="fs33">val a : &#39;a (requires equality)</div>
<div class="tip" id="fs34">val b : &#39;a (requires equality)</div>
<div class="tip" id="fs35">val costs : Costs&lt;&#39;a&gt; (requires equality)<br /><br />Full name: Alignment.costs</div>
<div class="tip" id="fs36">val globalAlignmentNuc : Alignment.Alignment&lt;Nucleotides.Nucleotide option list,Score&gt;<br /><br />Full name: Alignment.globalAlignmentNuc</div>
<div class="tip" id="fs37">val runGeneric : costs:Costs&lt;&#39;a&gt; -&gt; fstSeq:&#39;a [] -&gt; sndSeq:&#39;a [] -&gt; Alignment.Alignment&lt;&#39;a option list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.NeedlemanWunsch.runGeneric</div>
<div class="tip" id="fs38">val localAlignmentNuc : Alignment.Alignment&lt;Nucleotides.Nucleotide option list,Score&gt;<br /><br />Full name: Alignment.localAlignmentNuc</div>
<div class="tip" id="fs39">val runGeneric : costs:Costs&lt;&#39;a&gt; -&gt; fstSeq:&#39;a [] -&gt; sndSeq:&#39;a [] -&gt; Alignment.Alignment&lt;&#39;a option list,Score&gt;<br /><br />Full name: BioFSharp.Algorithm.PairwiseAlignment.SmithWaterman.runGeneric</div>
<div class="tip" id="fs40">val intSeq1 : int []<br /><br />Full name: Alignment.intSeq1</div>
<div class="tip" id="fs41">val intSeq2 : int []<br /><br />Full name: Alignment.intSeq2</div>
<div class="tip" id="fs42">val globalAlignmentInt : Alignment.Alignment&lt;int option list,Score&gt;<br /><br />Full name: Alignment.globalAlignmentInt</div>
<div class="tip" id="fs43">val localAlignmentInt : Alignment.Alignment&lt;int option list,Score&gt;<br /><br />Full name: Alignment.localAlignmentInt</div>

        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                <li>
                    <a href="/BioFSharp/BioItem.html">BioItems</a>
                </li>
                <li><a href="/BioFSharp/BioCollections.html">BioCollections</a></li>
                <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li>
                    <a href="/BioFSharp/Clustal.html">Clustal</a>
                </li>
                <li><a href="/BioFSharp/CSV.html">CSV / Tabular</a></li>
                <li>
                    <a href="/BioFSharp/FastA.html">FastA</a>
                </li>   
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li>
                    <a href="/BioFSharp/GenBank.html">GenBank</a>
                </li>  
                <li>
                    <a href="/BioFSharp/GFF3.html">GFF3</a>
                </li>    
                <li>
                    <a href="/BioFSharp/Newick.html">Newick</a>
                </li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                
                <li class="nav-header">Documentation</li>
                <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
