<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>BioCollections</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="An open source bioinformatics toolbox written in F#. <https://csbiology.github.io/BioFSharp/>"/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <meta http-equiv="X-UA-Compatible" content="IE=11" >
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" /> <!--/BioFSharp-->
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" /> <!--/BioFSharp-->
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script> <!--/BioFSharp-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
	<body>
		<div class="container">
		  <div class="masthead">
			<ul class="nav nav-pills pull-right">
			  <li><a href="http://fsharp.org">fsharp.org</a></li>
			  <li><a href="https://github.com/CSBiology/BioFSharp">github page</a></li>
			</ul>
			<h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
		  </div>
		  <hr />
		  <div class="row">
			<div class="span9" id="main">
			  
<table class="HeadAPI">
<td class="Head"><h1>BioCollections</h1></td>
<td class="API">
    <a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/biofsharp-bioseq.html" >&#128194;View BioSeq documentation</a>
</td>
</table>
Analogous to the build-in collections BioFSharp provides BioSeq, BioList and BioArray for individual collection specific optimized operations. 
The easiest way to create them are the `ofBioItemString` -functions
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 5)" onmouseover="showTip(event, 'fs5', 5)" class="id">s1</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 6)" onmouseover="showTip(event, 'fs6', 6)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 7)" onmouseover="showTip(event, 'fs7', 7)" class="id">ofAminoAcidString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs8', 8)" onmouseover="showTip(event, 'fs8', 8)" class="id">s2</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 9)" onmouseover="showTip(event, 'fs9', 9)" class="m">BioList</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 10)" onmouseover="showTip(event, 'fs10', 10)" class="id">ofAminoAcidSymbolString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 11)" onmouseover="showTip(event, 'fs11', 11)" class="id">s3</span> <span class="o">=</span> <span class="s">&quot;TAGCAT&quot;</span>  <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 12)" onmouseover="showTip(event, 'fs12', 12)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 13)" onmouseover="showTip(event, 'fs13', 13)" class="id">ofNucleotideString</span> 
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Peptide represented as a Bioseq</span>
<span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 14)" onmouseover="showTip(event, 'fs6', 14)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 15)" onmouseover="showTip(event, 'fs7', 15)" class="id">ofAminoAcidString</span> 

<span class="c">///Peptide represented as a BioList</span>
<span class="s">&quot;PEPTIDE&quot;</span><span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs9', 16)" onmouseover="showTip(event, 'fs9', 16)" class="m">BioList</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 17)" onmouseover="showTip(event, 'fs10', 17)" class="id">ofAminoAcidSymbolString</span> 

<span class="c">///Nucleotide sequence represented as a BioArray</span>
<span class="s">&quot;TAGCAT&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 18)" onmouseover="showTip(event, 'fs12', 18)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 19)" onmouseover="showTip(event, 'fs13', 19)" class="id">ofNucleotideString</span> 
</code></pre></td>
</tr>
</table>
<p>Resulting BioSeq containing our peptide:</p>
<table class="pre"><tr><td><pre><code>seq [Pro; Glu; Pro; Thr; ...]</code></pre></td></tr></table>
<p>Resulting BioList containing our peptide:</p>
<table class="pre"><tr><td><pre><code>seq [P; E; P; T; ...]</code></pre></td></tr></table>
<p>Resulting BioArray containing our oligonucleotide:</p>
<table class="pre"><tr><td><pre><code>[|T; A; G; C; A; T|]</code></pre></td></tr></table>
<h2><a name="Basics" class="anchor" href="#Basics">Basics</a></h2>
<p>Some functions which might be needed regularly are defined to work with nucleotides and amino acids:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 20)" onmouseover="showTip(event, 'fs14', 20)" class="id">myPeptide</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 21)" onmouseover="showTip(event, 'fs6', 21)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 22)" onmouseover="showTip(event, 'fs7', 22)" class="id">ofAminoAcidString</span> 
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>seq [Pro; Glu; Pro; Thr; ...]</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 23)" onmouseover="showTip(event, 'fs15', 23)" class="id">myPeptideFormula</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 24)" onmouseover="showTip(event, 'fs6', 24)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="id">toFormula</span> <span onmouseout="hideTip(event, 'fs14', 26)" onmouseover="showTip(event, 'fs14', 26)" class="id">myPeptide</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs17', 27)" onmouseover="showTip(event, 'fs17', 27)" class="m">Formula</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 28)" onmouseover="showTip(event, 'fs18', 28)" class="id">toString</span> 
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"C34.00 H51.00 N7.00 O14.00"</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 29)" onmouseover="showTip(event, 'fs19', 29)" class="id">myPeptideMass</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 30)" onmouseover="showTip(event, 'fs6', 30)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 31)" onmouseover="showTip(event, 'fs20', 31)" class="id">toAverageMass</span> <span onmouseout="hideTip(event, 'fs14', 32)" onmouseover="showTip(event, 'fs14', 32)" class="id">myPeptide</span> 
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>781.810317</code></pre></td></tr></table>
<h2><a name="AminoAcids" class="anchor" href="#AminoAcids">AminoAcids</a></h2>
<h3><a name="Digestion" class="anchor" href="#Digestion">Digestion</a></h3>
<p>BioFSharp also comes equipped with a set of tools aimed at cutting apart amino acid sequences. To demonstrate the usage, we'll throw some <code>trypsin</code> at the small RuBisCO subunit of <em>Arabidopos thaliana</em>:<br />
In the first step, we define our input sequence and the protease we want to use.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs21', 33)" onmouseover="showTip(event, 'fs21', 33)" class="id">RBCS</span> <span class="o">=</span> 
    <span class="s">&quot;&quot;&quot;MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT&quot;&quot;&quot;</span> 
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 34)" onmouseover="showTip(event, 'fs12', 34)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 35)" onmouseover="showTip(event, 'fs22', 35)" class="id">ofAminoAcidString</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 36)" onmouseover="showTip(event, 'fs23', 36)" class="id">trypsin</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs24', 37)" onmouseover="showTip(event, 'fs24', 37)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs25', 38)" onmouseover="showTip(event, 'fs25', 38)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 39)" onmouseover="showTip(event, 'fs26', 39)" class="id">getProteaseBy</span> <span class="s">&quot;Trypsin&quot;</span>
</code></pre></td>
</tr>
</table>
<p>With these two things done, digesting the protein is a piece of cake. For doing this, just use the <code>digest</code> function.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 40)" onmouseover="showTip(event, 'fs27', 40)" class="id">digestedRBCS</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs24', 41)" onmouseover="showTip(event, 'fs24', 41)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 42)" onmouseover="showTip(event, 'fs28', 42)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 43)" onmouseover="showTip(event, 'fs29', 43)" class="id">digest</span> <span onmouseout="hideTip(event, 'fs23', 44)" onmouseover="showTip(event, 'fs23', 44)" class="id">trypsin</span> <span class="n">0</span> <span onmouseout="hideTip(event, 'fs21', 45)" onmouseover="showTip(event, 'fs21', 45)" class="id">RBCS</span> 
</code></pre></td>
</tr>
</table>
<br>
<button type="button" class="btn" data-toggle="collapse" data-target="#digested">Show/Hide digestion result</button>
<div id="digested" class="collapse digested">
<pre>
val digestedRBCS : Digestion.DigestedPeptide [] =
  [|{ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 1;
     MissCleavageEnd = 28;
     PepSequence =
      [Met; Ala; Ser; Ser; Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser;
       Pro; Ala; Gln; Ala; Thr; Met; Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 29;
     MissCleavageEnd = 37;
     PepSequence = [Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 38;
     MissCleavageEnd = 38;
     PepSequence = [Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 39;
     MissCleavageEnd = 52;
     PepSequence =
      [Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser; Asn; Gly; Gly; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 53;
     MissCleavageEnd = 64;
     PepSequence =
      [Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 65;
     MissCleavageEnd = 65;
     PepSequence = [Lys];}; {ProteinID = 0;
                             MissCleavages = 0;
                             MissCleavageStart = 66;
                             MissCleavageEnd = 66;
                             PepSequence = [Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 67;
     MissCleavageEnd = 83;
     PepSequence =
      [Phe; Glu; Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser; Glu;
       Leu; Ala; Lys];}; {ProteinID = 0;
                          MissCleavages = 0;
                          MissCleavageStart = 84;
                          MissCleavageEnd = 90;
                          PepSequence = [Glu; Val; Asp; Tyr; Leu; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 91;
     MissCleavageEnd = 92;
     PepSequence = [Asn; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 93;
     MissCleavageEnd = 108;
     PepSequence =
      [Trp; Ile; Pro; Cys; Val; Glu; Phe; Glu; Leu; Glu; His; Gly; Phe; Val;
       Tyr; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 109;
     MissCleavageEnd = 120;
     PepSequence =
      [Glu; His; Gly; Asn; Ser; Pro; Gly; Tyr; Tyr; Asp; Gly; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 121;
     MissCleavageEnd = 126;
     PepSequence = [Tyr; Trp; Thr; Met; Trp; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 127;
     MissCleavageEnd = 140;
     PepSequence =
      [Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp; Ser; Ala; Gln; Val; Leu; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 141;
     MissCleavageEnd = 146;
     PepSequence = [Glu; Val; Glu; Glu; Cys; Lys];}; {ProteinID = 0;
                                                      MissCleavages = 0;
                                                      MissCleavageStart = 147;
                                                      MissCleavageEnd = 147;
                                                      PepSequence = [Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 148;
     MissCleavageEnd = 155;
     PepSequence = [Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 156;
     MissCleavageEnd = 163;
     PepSequence = [Ile; Ile; Gly; Phe; Asp; Asn; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 164;
     MissCleavageEnd = 180;
     PepSequence =
      [Gln; Val; Gln; Cys; Ile; Ser; Phe; Ile; Ala; Tyr; Lys; Pro; Pro; Ser;
       Phe; Thr; Gly];}|]
</pre>
<button type="button" class="btn" data-toggle="collapse" data-target="#digested">Hide again</button>  
</div>
<br>
In reality, proteases don't always completely cut the protein down. Instead, some sites stay intact and should be considered for in silico analysis. 
This can easily be done with the `concernMissCleavages` function. It takes the minimum and maximum amount of misscleavages you want to have and also the digested protein. As a result you get all possible combinations arising from this information.
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs30', 46)" onmouseover="showTip(event, 'fs30', 46)" class="id">digestedRBCS&#39;</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs24', 47)" onmouseover="showTip(event, 'fs24', 47)" class="m">Digestion</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs28', 48)" onmouseover="showTip(event, 'fs28', 48)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 49)" onmouseover="showTip(event, 'fs31', 49)" class="id">concernMissCleavages</span> <span class="n">0</span> <span class="n">2</span> <span onmouseout="hideTip(event, 'fs27', 50)" onmouseover="showTip(event, 'fs27', 50)" class="id">digestedRBCS</span>
</code></pre></td>
</tr>
</table>
<button type="button" class="btn" data-toggle="collapse" data-target="#digested2">Show/Hide digestion result</button>
<div id="digested2" class="collapse digested2">
<pre>
val digestedRBCS' : Digestion.DigestedPeptide [] =
  [|{ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 164;
     MissCleavageEnd = 180;
     PepSequence =
      [Gln; Val; Gln; Cys; Ile; Ser; Phe; Ile; Ala; Tyr; Lys; Pro; Pro; Ser;
       Phe; Thr; Gly];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 156;
     MissCleavageEnd = 163;
     PepSequence = [Ile; Ile; Gly; Phe; Asp; Asn; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 148;
     MissCleavageEnd = 155;
     PepSequence = [Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 147;
     MissCleavageEnd = 147;
     PepSequence = [Lys];}; {ProteinID = 0;
                             MissCleavages = 0;
                             MissCleavageStart = 141;
                             MissCleavageEnd = 146;
                             PepSequence = [Glu; Val; Glu; Glu; Cys; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 127;
     MissCleavageEnd = 140;
     PepSequence =
      [Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp; Ser; Ala; Gln; Val; Leu; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 121;
     MissCleavageEnd = 126;
     PepSequence = [Tyr; Trp; Thr; Met; Trp; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 109;
     MissCleavageEnd = 120;
     PepSequence =
      [Glu; His; Gly; Asn; Ser; Pro; Gly; Tyr; Tyr; Asp; Gly; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 93;
     MissCleavageEnd = 108;
     PepSequence =
      [Trp; Ile; Pro; Cys; Val; Glu; Phe; Glu; Leu; Glu; His; Gly; Phe; Val;
       Tyr; Arg];}; {ProteinID = 0;
                     MissCleavages = 0;
                     MissCleavageStart = 91;
                     MissCleavageEnd = 92;
                     PepSequence = [Asn; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 84;
     MissCleavageEnd = 90;
     PepSequence = [Glu; Val; Asp; Tyr; Leu; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 67;
     MissCleavageEnd = 83;
     PepSequence =
      [Phe; Glu; Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser; Glu;
       Leu; Ala; Lys];}; {ProteinID = 0;
                          MissCleavages = 0;
                          MissCleavageStart = 66;
                          MissCleavageEnd = 66;
                          PepSequence = [Lys];}; {ProteinID = 0;
                                                  MissCleavages = 0;
                                                  MissCleavageStart = 65;
                                                  MissCleavageEnd = 65;
                                                  PepSequence = [Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 53;
     MissCleavageEnd = 64;
     PepSequence =
      [Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 39;
     MissCleavageEnd = 52;
     PepSequence =
      [Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser; Asn; Gly; Gly; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 38;
     MissCleavageEnd = 38;
     PepSequence = [Lys];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 29;
     MissCleavageEnd = 37;
     PepSequence = [Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 0;
     MissCleavageStart = 1;
     MissCleavageEnd = 28;
     PepSequence =
      [Met; Ala; Ser; Ser; Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser;
       Pro; Ala; Gln; Ala; Thr; Met; Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 156;
     MissCleavageEnd = 180;
     PepSequence =
      [Ile; Ile; Gly; Phe; Asp; Asn; Thr; Arg; Gln; Val; Gln; Cys; Ile; Ser;
       Phe; Ile; Ala; Tyr; Lys; Pro; Pro; Ser; Phe; Thr; Gly];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 148;
     MissCleavageEnd = 163;
     PepSequence =
      [Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg; Ile; Ile; Gly; Phe; Asp; Asn;
       Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 147;
     MissCleavageEnd = 155;
     PepSequence = [Lys; Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 141;
     MissCleavageEnd = 147;
     PepSequence = [Glu; Val; Glu; Glu; Cys; Lys; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 127;
     MissCleavageEnd = 146;
     PepSequence =
      [Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp; Ser; Ala; Gln; Val; Leu; Lys;
       Glu; Val; Glu; Glu; Cys; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 121;
     MissCleavageEnd = 140;
     PepSequence =
      [Tyr; Trp; Thr; Met; Trp; Lys; Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp;
       Ser; Ala; Gln; Val; Leu; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 109;
     MissCleavageEnd = 126;
     PepSequence =
      [Glu; His; Gly; Asn; Ser; Pro; Gly; Tyr; Tyr; Asp; Gly; Arg; Tyr; Trp;
       Thr; Met; Trp; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 93;
     MissCleavageEnd = 120;
     PepSequence =
      [Trp; Ile; Pro; Cys; Val; Glu; Phe; Glu; Leu; Glu; His; Gly; Phe; Val;
       Tyr; Arg; Glu; His; Gly; Asn; Ser; Pro; Gly; Tyr; Tyr; Asp; Gly; Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 91;
     MissCleavageEnd = 108;
     PepSequence =
      [Asn; Lys; Trp; Ile; Pro; Cys; Val; Glu; Phe; Glu; Leu; Glu; His; Gly;
       Phe; Val; Tyr; Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 84;
     MissCleavageEnd = 92;
     PepSequence = [Glu; Val; Asp; Tyr; Leu; Ile; Arg; Asn; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 67;
     MissCleavageEnd = 90;
     PepSequence =
      [Phe; Glu; Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser; Glu;
       Leu; Ala; Lys; Glu; Val; Asp; Tyr; Leu; Ile; Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 66;
     MissCleavageEnd = 83;
     PepSequence =
      [Lys; Phe; Glu; Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser;
       Glu; Leu; Ala; Lys];}; {ProteinID = 0;
                               MissCleavages = 1;
                               MissCleavageStart = 65;
                               MissCleavageEnd = 66;
                               PepSequence = [Lys; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 53;
     MissCleavageEnd = 65;
     PepSequence =
      [Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 39;
     MissCleavageEnd = 64;
     PepSequence =
      [Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser; Asn; Gly; Gly; Arg;
       Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 38;
     MissCleavageEnd = 52;
     PepSequence =
      [Lys; Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser; Asn; Gly; Gly;
       Arg];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 29;
     MissCleavageEnd = 38;
     PepSequence = [Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr; Arg; Lys];};
    {ProteinID = 0;
     MissCleavages = 1;
     MissCleavageStart = 1;
     MissCleavageEnd = 37;
     PepSequence =
      [Met; Ala; Ser; Ser; Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser;
       Pro; Ala; Gln; Ala; Thr; Met; Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys;
       Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 148;
     MissCleavageEnd = 180;
     PepSequence =
      [Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg; Ile; Ile; Gly; Phe; Asp; Asn;
       Thr; Arg; Gln; Val; Gln; Cys; Ile; Ser; Phe; Ile; Ala; Tyr; Lys; Pro;
       Pro; Ser; Phe; Thr; Gly];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 147;
     MissCleavageEnd = 163;
     PepSequence =
      [Lys; Glu; Tyr; Pro; Asn; Ala; Phe; Ile; Arg; Ile; Ile; Gly; Phe; Asp;
       Asn; Thr; Arg];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 141;
     MissCleavageEnd = 155;
     PepSequence =
      [Glu; Val; Glu; Glu; Cys; Lys; Lys; Glu; Tyr; Pro; Asn; Ala; Phe; Ile;
       Arg];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 127;
     MissCleavageEnd = 147;
     PepSequence =
      [Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp; Ser; Ala; Gln; Val; Leu; Lys;
       Glu; Val; Glu; Glu; Cys; Lys; Lys];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 121;
     MissCleavageEnd = 146;
     PepSequence =
      [Tyr; Trp; Thr; Met; Trp; Lys; Leu; Pro; Leu; Phe; Gly; Cys; Thr; Asp;
       Ser; Ala; Gln; Val; Leu; Lys; Glu; Val; Glu; Glu; Cys; Lys];};
    {ProteinID = 0;
     MissCleavages = 2;
     MissCleavageStart = 109;
     MissCleavageEnd = 140;
     PepSequence = [Glu; His; Gly; Asn; Ser; Pro; ...];}; ...|]
</pre>
<button type="button" class="btn" data-toggle="collapse" data-target="#digested2">Hide again</button>  
</div>
<br>
<h2><a name="Nucleotides" class="anchor" href="#Nucleotides">Nucleotides</a></h2>
<p><img src="img/Nucleotides.svg" alt="Nucleotides1" /></p>
<p><strong>Figure 1: Selection of covered nucleotide operations</strong> (A) Bilogical principle. (B) Workflow with <code>BioSeq</code>. (C) Other covered functionalities.</p>
<p>Let's imagine you have a given gene sequence and want to find out what the according protein might look like.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs32', 51)" onmouseover="showTip(event, 'fs32', 51)" class="id">myGene</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 52)" onmouseover="showTip(event, 'fs6', 52)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 53)" onmouseover="showTip(event, 'fs33', 53)" class="id">ofNucleotideString</span> <span class="s">&quot;ATGGCTAGATCGATCGATCGGCTAACGTAA&quot;</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>seq [A; T; G; G; ...]</code></pre></td></tr></table>
<p>Yikes! Unfortunately we got the 5'-3' coding strand. For proper transcription we should get the complementary strand first:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs34', 54)" onmouseover="showTip(event, 'fs34', 54)" class="id">myProperGene</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs35', 55)" onmouseover="showTip(event, 'fs35', 55)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs36', 56)" onmouseover="showTip(event, 'fs36', 56)" class="id">map</span> <span onmouseout="hideTip(event, 'fs37', 57)" onmouseover="showTip(event, 'fs37', 57)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs38', 58)" onmouseover="showTip(event, 'fs38', 58)" class="id">complement</span> <span onmouseout="hideTip(event, 'fs32', 59)" onmouseover="showTip(event, 'fs32', 59)" class="id">myGene</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>seq [T; A; C; C; ...]</code></pre></td></tr></table>
<p>Now let's transcribe and translate it:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs39', 60)" onmouseover="showTip(event, 'fs39', 60)" class="id">myTranslatedGene</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs34', 61)" onmouseover="showTip(event, 'fs34', 61)" class="id">myProperGene</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 62)" onmouseover="showTip(event, 'fs6', 62)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs40', 63)" onmouseover="showTip(event, 'fs40', 63)" class="id">transcribeTemplateStrand</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 64)" onmouseover="showTip(event, 'fs6', 64)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 65)" onmouseover="showTip(event, 'fs41', 65)" class="id">translate</span> <span class="n">0</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>Error: The type initializer for '&lt;StartupCode$BioFSharp&gt;.$Nucleotides' threw an exception.</code></pre></td></tr></table>
<p>Of course, if your input sequence originates from the coding strand, you can directly transcribe it to mRNA since the
only difference between the coding strand and the mRNA is the replacement of 'T' by 'U' (Figure 1B)</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs42', 66)" onmouseover="showTip(event, 'fs42', 66)" class="id">myTranslatedGeneFromCodingStrand</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs32', 67)" onmouseover="showTip(event, 'fs32', 67)" class="id">myGene</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 68)" onmouseover="showTip(event, 'fs6', 68)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs43', 69)" onmouseover="showTip(event, 'fs43', 69)" class="id">transcribeCodingStrand</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs6', 70)" onmouseover="showTip(event, 'fs6', 70)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 71)" onmouseover="showTip(event, 'fs41', 71)" class="id">translate</span> <span class="n">0</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>Error: The type initializer for '&lt;StartupCode$BioFSharp&gt;.$Nucleotides' threw an exception.</code></pre></td></tr></table>
<p>Other Nucleotide conversion operations are also covered:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 72)" onmouseover="showTip(event, 'fs44', 72)" class="id">mySmallGene</span>      <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 73)" onmouseover="showTip(event, 'fs6', 73)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs33', 74)" onmouseover="showTip(event, 'fs33', 74)" class="id">ofNucleotideString</span>  <span class="s">&quot;ATGTTCCGAT&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs45', 75)" onmouseover="showTip(event, 'fs45', 75)" class="id">smallGeneRev</span>     <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 76)" onmouseover="showTip(event, 'fs6', 76)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 77)" onmouseover="showTip(event, 'fs46', 77)" class="id">reverse</span> <span onmouseout="hideTip(event, 'fs44', 78)" onmouseover="showTip(event, 'fs44', 78)" class="id">mySmallGene</span> 
<span class="c">//Original: ATGTTCCGAT</span>
<span class="c">//Output:   TAGCCTTGTA</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs47', 79)" onmouseover="showTip(event, 'fs47', 79)" class="id">smallGeneComp</span>    <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 80)" onmouseover="showTip(event, 'fs6', 80)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs48', 81)" onmouseover="showTip(event, 'fs48', 81)" class="id">complement</span> <span onmouseout="hideTip(event, 'fs44', 82)" onmouseover="showTip(event, 'fs44', 82)" class="id">mySmallGene</span>
<span class="c">//Original: ATGTTCCGAT</span>
<span class="c">//Output:   TACAAGGCTA</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs49', 83)" onmouseover="showTip(event, 'fs49', 83)" class="id">smallGeneRevComp</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs6', 84)" onmouseover="showTip(event, 'fs6', 84)" class="m">BioSeq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs50', 85)" onmouseover="showTip(event, 'fs50', 85)" class="id">reverseComplement</span> <span onmouseout="hideTip(event, 'fs44', 86)" onmouseover="showTip(event, 'fs44', 86)" class="id">mySmallGene</span>
<span class="c">//Original: ATGTTCCGAT</span>
<span class="c">//Reverse:  TAGCCTTGTA</span>
<span class="c">//Output:   ATCGGAACAT</span>
</code></pre></td>
</tr>
</table>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">Multiple items<br />namespace FSharp<br /><br />--------------------<br />namespace Microsoft.FSharp</div>
<div class="tip" id="fs3">namespace FSharp.Plotly</div>
<div class="tip" id="fs4">namespace FSharpAux</div>
<div class="tip" id="fs5">val s1 : BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs6">module BioSeq<br /><br />from BioFSharp</div>
<div class="tip" id="fs7">val ofAminoAcidString : s:#seq&lt;char&gt; -&gt; BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs8">val s2 : BioList.BioList&lt;AminoAcidSymbols.AminoAcidSymbol&gt;</div>
<div class="tip" id="fs9">module BioList<br /><br />from BioFSharp</div>
<div class="tip" id="fs10">val ofAminoAcidSymbolString : s:#seq&lt;char&gt; -&gt; BioList.BioList&lt;AminoAcidSymbols.AminoAcidSymbol&gt;</div>
<div class="tip" id="fs11">val s3 : BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs12">module BioArray<br /><br />from BioFSharp</div>
<div class="tip" id="fs13">val ofNucleotideString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs14">val myPeptide : BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;<br /><em><br /><br />Nucleotide sequence represented as a BioArray</em></div>
<div class="tip" id="fs15">val myPeptideFormula : string</div>
<div class="tip" id="fs16">val toFormula : bs:seq&lt;#IBioItem&gt; -&gt; Formula.Formula</div>
<div class="tip" id="fs17">module Formula<br /><br />from BioFSharp</div>
<div class="tip" id="fs18">val toString : f:Formula.Formula -&gt; string</div>
<div class="tip" id="fs19">val myPeptideMass : float</div>
<div class="tip" id="fs20">val toAverageMass : bs:seq&lt;#IBioItem&gt; -&gt; float</div>
<div class="tip" id="fs21">val RBCS : BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs22">val ofAminoAcidString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs23">val trypsin : Digestion.Protease</div>
<div class="tip" id="fs24">module Digestion<br /><br />from BioFSharp</div>
<div class="tip" id="fs25">module Table<br /><br />from BioFSharp.Digestion</div>
<div class="tip" id="fs26">val getProteaseBy : name:string -&gt; Digestion.Protease</div>
<div class="tip" id="fs27">val digestedRBCS : Digestion.DigestedPeptide&lt;int&gt; []</div>
<div class="tip" id="fs28">module BioArray<br /><br />from BioFSharp.Digestion</div>
<div class="tip" id="fs29">val digest : protease:Digestion.Protease -&gt; proteinID:int -&gt; aas:AminoAcids.AminoAcid [] -&gt; Digestion.DigestedPeptide&lt;int&gt; []</div>
<div class="tip" id="fs30">val digestedRBCS&#39; : Digestion.DigestedPeptide&lt;int&gt; []</div>
<div class="tip" id="fs31">val concernMissCleavages : minMissCleavages:int -&gt; maxMisscleavages:int -&gt; digestedPeptidesA:Digestion.DigestedPeptide&lt;&#39;a&gt; [] -&gt; Digestion.DigestedPeptide&lt;&#39;a&gt; [] (requires equality)</div>
<div class="tip" id="fs32">val myGene : BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs33">val ofNucleotideString : s:#seq&lt;char&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs34">val myProperGene : seq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs35">Multiple items<br />module Seq<br /><br />from FSharpAux<br /><br />--------------------<br />module Seq<br /><br />from FSharp.Plotly<br /><br />--------------------<br />module Seq<br /><br />from Microsoft.FSharp.Collections</div>
<div class="tip" id="fs36">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; source:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;U&gt;</div>
<div class="tip" id="fs37">module Nucleotides<br /><br />from BioFSharp</div>
<div class="tip" id="fs38">val complement : nuc:Nucleotides.Nucleotide -&gt; Nucleotides.Nucleotide</div>
<div class="tip" id="fs39">val myTranslatedGene : BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs40">val transcribeTemplateStrand : nucs:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs41">val translate : nucleotideOffset:int -&gt; rnaSeq:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs42">val myTranslatedGeneFromCodingStrand : BioSeq.BioSeq&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs43">val transcribeCodingStrand : nucs:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs44">val mySmallGene : BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs45">val smallGeneRev : BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs46">val reverse : nucs:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs47">val smallGeneComp : BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs48">val complement : nucs:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs49">val smallGeneRevComp : BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>
<div class="tip" id="fs50">val reverseComplement : nucs:seq&lt;Nucleotides.Nucleotide&gt; -&gt; BioSeq.BioSeq&lt;Nucleotides.Nucleotide&gt;</div>

			</div>
			<div class="span3" >
			  <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
              <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                  <li class="nav-header">BioFSharp</li>
                  <li><a href="/BioFSharp/index.html">Home page</a></li>
                  <li class="divider"></li>
                  <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                  
                  <li class="nav-header">BioFSharp core</li>
                  <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                  <li>
                      <a href="/BioFSharp/BioItem.html">BioItems</a>
                  </li>
                  <li><a href="/BioFSharp/BioCollections.html">BioCollections</a></li>
                  <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                  <li class="nav-header">Algorithm</li>
                  <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                  <li><a href="/BioFSharp/ClustalOWrapper.html">Clustal Omega</a></li>
                  <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                  <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                  <li>
                      <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                  </li>
                  <li class="nav-header">File I/O</li>
                  <li>
                      <a href="/BioFSharp/Clustal.html">Clustal</a>
                  </li>
                  <li><a href="/BioFSharp/CSV.html">CSV / Tabular</a></li>
                  <li>
                      <a href="/BioFSharp/FastA.html">FastA</a>
                  </li>
                  <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                  <li>
                      <a href="/BioFSharp/GenBank.html">GenBank</a>
                  </li>
                  <li>
                      <a href="/BioFSharp/GFF3.html">GFF3</a>
                  </li>
                  <li>
                      <a href="/BioFSharp/Newick.html">Newick</a>
                  </li>
                  <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                  <li>
                      <a href="/BioFSharp/FSIPrinters.html">FSIPrinters</a>
                  </li>
                  <li class="nav-header">Web I/O</li>
                  <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                  <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                  <li class="nav-header">Biostatistics</li>
                  <li><a href="/BioFSharp/GSEA.html">Ontology enrichment</a></li>
                  <li class="nav-header">Biovisualization</li>
                  <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                  <li class="nav-header">GPU Computing</li>
                  <li><a href="/BioFSharp/ParallelAlignment.html">Sequence Alignment</a></li>
                  <li class="nav-header">Documentation</li>
                  <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                  
              </ul>
			</div>
		  </div>
		</div>
		<a href="https://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
	  </body>
  </html>
