<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>BioItems
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="BioItems" class="anchor" href="#BioItems">BioItems</a></h1>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/AminoAcids.fs">&lt;/&gt;view source code</a>
<a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/biofsharp-aminoacids.html" >&#128194;View module documentation</a>
(AminoAcids)
<br><br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/Nucleotides.fs">&lt;/&gt;view source code</a>
<a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/biofsharp-nucleotides.html" >&#128194;View module documentation</a>
(Nucleotides)
<br><br>
<a id="Author" href="https://github.com/HLWeil">&#128366;view author of this tutorial</a>
<br><br>
Often, dealing with similar problems separately results in different approaches. In a programming background, this might make things needlessly complex. Therefore in BioFSharp nucleotides and amino acids are based on the same structural scaffold, leading to a consistent way of working with them. This can come in handy especially when working with their formulas.  
<h2><a name="Basics" class="anchor" href="#Basics">Basics</a></h2>
<p>First let's define some amino acids and nucleotides</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">ala</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Ala</span>
<span class="k">let</span> <span class="i">lys</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Lys</span>
<span class="k">let</span> <span class="i">glu</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Glu</span>
<span class="k">let</span> <span class="i">ter</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Ter</span>

<span class="k">let</span> <span class="i">g</span> <span class="o">=</span> <span class="i">Nucleotides</span><span class="o">.</span><span class="i">G</span>
<span class="k">let</span> <span class="i">t</span> <span class="o">=</span> <span class="i">Nucleotides</span><span class="o">.</span><span class="i">T</span>
<span class="k">let</span> <span class="i">gap</span> <span class="o">=</span> <span class="i">Nucleotides</span><span class="o">.</span><span class="i">Gap</span>
</code></pre></td>
</tr>
</table>
<p>Many functions are similar for AminoAcids and Nucleotides</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">AminoAcids</span><span class="o">.</span><span class="i">name</span> <span class="i">ala</span> <span class="c">// val it : string = &quot;Alanin&quot;</span>
<span class="i">Nucleotides</span><span class="o">.</span><span class="i">name</span> <span class="i">g</span> <span class="c">// val it : string = &quot;Guanine&quot;</span>

<span class="i">AminoAcids</span><span class="o">.</span><span class="i">formula</span> <span class="i">lys</span> <span class="o">|&gt;</span> <span class="i">Formula</span><span class="o">.</span><span class="i">toString</span> <span class="c">// val it : string = &quot;C6.00 H12.00 N2.00 O1.00 &quot;</span>
<span class="i">Nucleotides</span><span class="o">.</span><span class="i">formula</span> <span class="i">t</span> <span class="o">|&gt;</span> <span class="i">Formula</span><span class="o">.</span><span class="i">toString</span> <span class="c">// val it : string = &quot;C10.00 H14.00 N2.00 O5.00 &quot;</span>

<span class="i">AminoAcids</span><span class="o">.</span><span class="i">isTerminator</span> <span class="i">glu</span> <span class="c">// val it : bool = false</span>
<span class="i">AminoAcids</span><span class="o">.</span><span class="i">isTerminator</span> <span class="i">ter</span> <span class="c">// val it : bool = true</span>
</code></pre></td>
</tr>
</table>
<p>Nucleotides and AminoAcids in FSharp are represented as Union cases. This makes applying functions selectively very easy.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">filterLysine</span> <span class="i">aa</span> <span class="o">=</span> 
    <span class="k">match</span> <span class="i">aa</span> <span class="k">with</span>
    | <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Lys</span> <span class="k">-&gt;</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">Gap</span>
    | _ <span class="k">-&gt;</span> <span class="i">aa</span>

<span class="i">filterLysine</span> <span class="i">ala</span> <span class="c">// val it : AminoAcids.AminoAcid = Ala</span>
<span class="i">filterLysine</span> <span class="i">lys</span> <span class="c">// val it : AminoAcids.AminoAcid = Gap</span>
</code></pre></td>
</tr>
</table>
<p>Of course some functions like these are already defined. Let's use a predefined function to find charged amino acids.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
<span class="l">18: </span>
<span class="l">19: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">giveMePositiveAAs</span> <span class="i">aminoAcid</span> <span class="o">=</span> 
    <span class="k">match</span> <span class="i">aminoAcid</span> <span class="k">with</span>
    | <span class="i">a</span> <span class="k">when</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">isPosCharged</span> <span class="i">a</span> <span class="k">-&gt;</span> 
        <span class="i">printfn</span> 
            <span class="s">&quot;Hey, how are you? I am %s, but my friends call me %c. I&#39;m usually in a positive mood&quot;</span>
            (<span class="i">AminoAcids</span><span class="o">.</span><span class="i">name</span> <span class="i">a</span>)
            (<span class="i">AminoAcids</span><span class="o">.</span><span class="i">symbol</span> <span class="i">a</span>)

    | <span class="i">a</span> <span class="k">when</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">isNegCharged</span> <span class="i">a</span> <span class="k">-&gt;</span> 
        <span class="i">printfn</span> 
            <span class="s">&quot;I am %s, short: %c. I&#39;m usually in a negative mood&quot;</span>
            (<span class="i">AminoAcids</span><span class="o">.</span><span class="i">name</span> <span class="i">a</span>)
            (<span class="i">AminoAcids</span><span class="o">.</span><span class="i">symbol</span> <span class="i">a</span>)

    | _ <span class="k">-&gt;</span> <span class="i">printfn</span> <span class="s">&quot;Just strolling around, minding my own business.&quot;</span>

<span class="i">giveMePositiveAAs</span> <span class="i">ala</span>
<span class="i">giveMePositiveAAs</span> <span class="i">lys</span>
<span class="i">giveMePositiveAAs</span> <span class="i">glu</span>
</code></pre></td>
</tr>
</table>
<pre>
Just strolling around, minding my own business.
val it : unit = ()
Hey, how are you? I am Lysine, but my friends call me K. I'm usually in a positive mood
val it : unit = ()
I am Glutamic Acid, short: E. I'm usually in a negative mood
val it : unit = ()
</pre>  
<h2><a name="AminoAcids" class="anchor" href="#AminoAcids">AminoAcids</a></h2>
<h3><a name="Modification" class="anchor" href="#Modification">Modification</a></h3>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/ModificationInfo.fs">&lt;/&gt;view source code</a>
<a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/biofsharp-modificationinfo.html" >&#128194;View module documentation</a>
<br>
What makes working on Amino Acids with BioFSharp truly powerful is the ability to easily modify AminoAcids, even altering their mass and formula. In the following example we try to find out the mass of a phosphorylated Serine. Applications like these might be quite usefull for identification of peptides in mass spectrometry.  
As a first step, we make ourselves an easy helper function which returns the formula of an AminoAcid as string. Afterwards we start out by creating a Serine.
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//easy helper function</span>
<span class="k">let</span> <span class="i">getF</span> (<span class="i">aa</span><span class="o">:</span><span class="i">AminoAcids</span><span class="o">.</span><span class="i">AminoAcid</span>) <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">formula</span> <span class="i">aa</span> <span class="o">|&gt;</span> <span class="i">Formula</span><span class="o">.</span><span class="i">toString</span>

<span class="c">///Serine -H20</span>
<span class="k">let</span> <span class="i">mySerine</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">AminoAcid</span><span class="o">.</span><span class="i">Ser</span>

<span class="i">getF</span> <span class="i">mySerine</span> <span class="c">//val it : string = &quot;C3.00 H5.00 N1.00 O2.00 &quot;</span>
</code></pre></td>
</tr>
</table>
<p>As you can see by the formula. Our Serine is missing two H and an O. In BioFSharp, all Amino Acids are dehydrolysed by default, because it is assumed that the user will use collections representing a peptide, rather than single Amino Acids. For our cause we want serine in hydrolysed form. An easy way to achieve this is to modify it. An addition of H2O is quite common and therefore premade:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Hydrolysed serine</span>
<span class="k">let</span> <span class="i">hydroSerine</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">setModification</span> <span class="i">ModificationInfo</span><span class="o">.</span><span class="i">Table</span><span class="o">.</span><span class="i">H2O</span> <span class="i">mySerine</span>

<span class="i">getF</span> <span class="i">hydroSerine</span> <span class="c">//val it : string = &quot;C3.00 H7.00 N1.00 O3.00 &quot;</span>
</code></pre></td>
</tr>
</table>
<p>So far so good. Now let's add the phosphate. For this we first create a function which alters the formula of a given molecule in the way a phosphorylation would. In the second step we create a modification resembling a phosphorylation of a residual. At last we modify our Serine with this modification.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Phosphorylation of OH-Groups adds PO3 to formula and removes one H</span>
<span class="k">let</span> <span class="i">phosporylate</span> <span class="i">formula</span> <span class="o">=</span>  
    <span class="i">Formula</span><span class="o">.</span><span class="i">add</span> (<span class="i">Formula</span><span class="o">.</span><span class="i">parseFormulaString</span> <span class="s">&quot;PO3&quot;</span>) <span class="i">formula</span>
    <span class="o">|&gt;</span> <span class="i">Formula</span><span class="o">.</span><span class="i">substract</span> (<span class="i">Formula</span><span class="o">.</span><span class="i">parseFormulaString</span> <span class="s">&quot;H&quot;</span>)

<span class="c">//We create a modification at the residual called phosphorylation</span>
<span class="k">let</span> <span class="i">phosphorylation</span> <span class="o">=</span> <span class="i">ModificationInfo</span><span class="o">.</span><span class="i">createModification</span> <span class="s">&quot;Phosphorylation&quot;</span> <span class="i">ModificationInfo</span><span class="o">.</span><span class="i">ModLocation</span><span class="o">.</span><span class="i">Residual</span> <span class="i">phosporylate</span>

<span class="c">///phosphorylated Serine</span>
<span class="k">let</span> <span class="i">phosphoSerine</span> <span class="o">=</span> <span class="i">AminoAcids</span><span class="o">.</span><span class="i">setModification</span> <span class="i">phosphorylation</span> <span class="i">hydroSerine</span>

<span class="i">getF</span> <span class="i">phosphoSerine</span> <span class="c">//val it : string = &quot;P1.00 C3.00 H6.00 N1.00 O6.00 &quot;</span>
</code></pre></td>
</tr>
</table>
<p>As you can see the Serine is phosphorylated just as we wanted. Our inital aim was to check the mass, this can be done quite easily:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">AminoAcids</span><span class="o">.</span><span class="i">averageMass</span> <span class="i">mySerine</span> <span class="c">//val it : float = 87.077575</span>
<span class="i">AminoAcids</span><span class="o">.</span><span class="i">averageMass</span> <span class="i">phosphoSerine</span> <span class="c">//val it : float = 183.056884</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Nucleotides" class="anchor" href="#Nucleotides">Nucleotides</a></h2>
<p>As working with nucleotides is usually focused on the sequence of the bases, rather than how they actually look like, the list of nucleotide specific functions would be quite short. But there are some helper like:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">myAdenine</span> <span class="o">=</span> <span class="i">Nucleotides</span><span class="o">.</span><span class="i">A</span> <span class="c">// val myAdenine : Nucleotides.Nucleotide = A</span>
<span class="k">let</span> <span class="i">myThymine</span> <span class="o">=</span> <span class="i">Nucleotides</span><span class="o">.</span><span class="i">complement</span> <span class="i">myAdenine</span> <span class="c">// val myThymine : Nucleotides.Nucleotide = T</span>

<span class="i">Nucleotides</span><span class="o">.</span><span class="i">replaceTbyU</span> <span class="i">myAdenine</span> <span class="c">// val it : Nucleotides.Nucleotide = A</span>
<span class="i">Nucleotides</span><span class="o">.</span><span class="i">replaceTbyU</span> <span class="i">myThymine</span> <span class="c">// val it : Nucleotides.Nucleotide = U</span>
</code></pre></td>
</tr>
</table>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                <li>
                    <a href="/BioFSharp/BioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF3.html">GFF format</a></li>
                <li><a href="/BioFSharp/GenBank.html">GenBank</a></li>
                <li>
                    <a href="/BioFSharp/Clustal.html">Clustal</a>
                </li>        
                <li>
                    <a href="/BioFSharp/Newick.html">Newick</a>
                </li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/CSV.html">Delimited files</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                
                <li class="nav-header">Documentation</li>
                <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
