<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Introducing BlastWrapper
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Introducing-BlastWrapper" class="anchor" href="#Introducing-BlastWrapper">Introducing BlastWrapper</a></h1>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/BlastWrapper.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/kMutagene">&#128366;view author of this tutorial</a>
<br><br>
BlastWrapper is a tool for performing different tasks in NCBI BLAST console applications (version 2.2.31+).
It is able to create BLAST databases and perform **blastN** or **blastP** queries, while providing a way to set
output parameter for creating a custom output format.
Official documentation for all BLAST applications can be found [here](http://www.ncbi.nlm.nih.gov/books/NBK279690).
<p>For the purpose of this tutorial, we will build a protein database using a <code>.fastA</code> file containing chloroplast proteins
of Chlamydomonas reinhardtii included in BioFSharp/docs/content/data.</p>
<p>Our query protein for the subsequent BLAST search will be the <a href="http://www.ncbi.nlm.nih.gov/protein/7525013?report=fasta">photosystem II protein D1</a> from Arabidopsis thaliana chloroplast.</p>
<h1><a name="How-to-use-BlastWrapper" class="anchor" href="#How-to-use-BlastWrapper">How to use BlastWrapper</a></h1>
<h2><a name="Creation-of-a-BLAST-database" class="anchor" href="#Creation-of-a-BLAST-database">Creation of a BLAST database</a></h2>
<p>We will use the minimal amount of parameters needed to create a BLAST database from an input file.
The created database files will have the same name as the input file and will be located in the same folder.
However, there are many parameters you can use to specify your database. Please refer to the <a href="http://www.ncbi.nlm.nih.gov/books/NBK279675/">NCBI user manual</a> for more information.</p>
<p>First, lets specify the path of our input and the type of our database. Use a string for the input path and the provided <code>MakeDbParams</code> type
for every other parameter.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span>
<span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span>
<span class="k">open</span> <span class="i">BlastNCBI</span>
<span class="k">open</span> <span class="i">Parameters</span>

<span class="c">///path and name of the input file/output database. </span>
<span class="k">let</span> <span class="i">inputFile</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Chlamy_Cp.fastA&quot;</span>)

<span class="c">///defines database type (in this case: a protein database)</span>
<span class="k">let</span> <span class="i">typeOfDatabase</span> <span class="o">=</span> <span class="i">Parameters</span><span class="o">.</span><span class="i">MakeDbParams</span><span class="o">.</span><span class="i">DbType</span> <span class="i">Parameters</span><span class="o">.</span><span class="i">Protein</span>
</code></pre></td>
</tr>
</table>
<p>The wrapper needs to know the path of the ncbi applications.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///the path of the /bin folder where the BLAST applications are located</span>
<span class="k">let</span> <span class="i">ncbiPath</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/../../lib/ncbi-blast/bin&quot;</span>)
</code></pre></td>
</tr>
</table>
<p>We now provide the wrapper our ncbi path, the input path and a sequence of parameters (containing just one parameter in this case, the database type).</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">BlastWrapper</span>(<span class="i">ncbiPath</span>)<span class="o">.</span><span class="i">makeblastdb</span> <span class="i">inputFile</span> ([<span class="i">typeOfDatabase</span>;] <span class="o">|&gt;</span> <span class="i">seq</span><span class="o">&lt;</span><span class="i">Parameters</span><span class="o">.</span><span class="i">MakeDbParams</span><span class="o">&gt;</span>)
</code></pre></td>
</tr>
</table>
<p>This creates 3 new files in our directory:
<code>Chlamy_Cp.fastA.phr</code>,<code>Chlamy_Cp.fastA.pin</code> and <code>Chlamy_Cp.fastA.psq</code>.</p>
<p>We have sucesssfully created our search database.</p>
<h2><a name="Creating-a-fastA-file-from-an-aminoacid-string" class="anchor" href="#Creating-a-fastA-file-from-an-aminoacid-string">Creating a .fastA file from an aminoacid string</a></h2>
<p><em>Note: this step is not necessary if you want to use an already existing file as query. If this is the case, skip to step 3.</em></p>
<p>First, lets specify a string with our aminoacid sequence and convert it to a <code>BioSeq</code>.
For more information about <code>BioSeq</code>, please refer to this <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-bioseq.html">documentation</a></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Raw string of the aminoacid sequence of our query protein</span>
<span class="k">let</span> <span class="i">aminoacidString</span> <span class="o">=</span> <span class="s">&quot;MTAILERRESESLWGRFCNWITSTENRLYIGWFGVLMIPTLLTATSVFIIAFIAAPPVDIDGIREPVSGS</span>
<span class="s">LLYGNNIISGAIIPTSAAIGLHFYPIWEAASVDEWLYNGGPYELIVLHFLLGVACYMGREWELSFRLGMR</span>
<span class="s">PWIAVAYSAPVAAATAVFLIYPIGQGSFSDGMPLGISGTFNFMIVFQAEHNILMHPFHMLGVAGVFGGSL</span>
<span class="s">FSAMHGSLVTSSLIRETTENESANEGYRFGQEEETYNIVAAHGYFGRLIFQYASFNNSRSLHFFLAAWPV</span>
<span class="s">VGIWFTALGISTMAFNLNGFNFNQSVVDSQGRVINTWADIINRANLGMEVMHERNAHNFPLDLAAVEAPS</span>
<span class="s">TNG&quot;</span>
<span class="c">///header for the .fastA file</span>
<span class="k">let</span> <span class="i">header</span> <span class="o">=</span> <span class="s">&quot;&gt;gi|7525013|ref|NP_051039.1| photosystem II protein D1 (chloroplast) [Arabidopsis thaliana]&quot;</span>

<span class="c">///Query sequency represented as a sequence of `AminoAcid` one of BioFSharp&#39;s `BioItems`</span>
<span class="k">let</span> <span class="i">querySequence</span> <span class="o">=</span> <span class="i">BioSeq</span><span class="o">.</span><span class="i">ofAminoAcidString</span> <span class="i">aminoacidString</span>
</code></pre></td>
</tr>
</table>
<p>We will now use BioFSharp's <code>FastA</code> library to create a <code>FASTA</code> item and write it to a file.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///path and name of the query file</span>
<span class="k">let</span> <span class="i">queryFastaPath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/testQuery.fastA&quot;</span>

<span class="c">///FastaItem containing header string and query sequence</span>
<span class="k">let</span> <span class="i">queryFastaItem</span> <span class="o">=</span> <span class="i">FastA</span><span class="o">.</span><span class="i">createFastaItem</span> <span class="i">header</span> <span class="i">querySequence</span>
</code></pre></td>
</tr>
</table>
<p>To create our <code>.fastA</code> file, we need to use the <code>BioItem.symbol</code> converter, which will convert the 3 letter code of the aminoacids in our biosequence
to the one letter symbol (eg. Met -&gt; M)</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">FastA</span><span class="o">.</span><span class="i">write</span> <span class="i">BioItem</span><span class="o">.</span><span class="i">symbol</span> <span class="i">queryFastaPath</span> [<span class="i">queryFastaItem</span>;] 
</code></pre></td>
</tr>
</table>
<h2><a name="Performing-the-BLAST-search" class="anchor" href="#Performing-the-BLAST-search">Performing the BLAST search</a></h2>
<p>We have created our search database and the query we want to find. Before we can perform the actual search, we need to define the BLAST prameters.</p>
<p><em>Note: custom output formats can only be specified for output types <code>CSV</code>, <code>tabular</code> and <code>tabular with comments</code>. For more information, check
the <a href="http://www.ncbi.nlm.nih.gov/books/NBK279675/">options for the command-line applicaions</a></em></p>
<p>First, lets specify the overall output type. This will define the outline of our output. We want our output to be in tabular form, with added information
in the form of comments.</p>
<p><em>Note: when not specified otherwise, the output type will be <code>pairwise</code></em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///overall outline of the output </span>
<span class="k">let</span> <span class="i">outputType</span> <span class="o">=</span> <span class="i">OutputType</span><span class="o">.</span><span class="i">TabularWithComments</span>
</code></pre></td>
</tr>
</table>
<p>We have a large selection of parameters that we can include in the output.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///a sequence of custom output format parameters</span>
<span class="k">let</span> <span class="i">outputFormat</span><span class="o">=</span> 
    
    [   
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Query_SeqId</span>; 
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Subject_SeqId</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Query_Length</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Subject_Length</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">AlignmentLength</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">MismatchCount</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">IdentityCount</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">PositiveScoringMatchCount</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Evalue</span>;
        <span class="i">OutputCustom</span><span class="o">.</span><span class="i">Bitscore</span>;
    ] 
        <span class="o">|&gt;</span> <span class="i">List</span><span class="o">.</span><span class="i">toSeq</span>
</code></pre></td>
</tr>
</table>
<p>Finally, we create a <code>BlastParam</code> of the type <code>OutputTypeCustom</code> from a touple of <code>outputType</code> and <code>outputFormat</code>.</p>
<p><em>Note: No touple required if you want to use the default output format. If this is the case,
just create a <code>BlastParam</code> of type <code>OutputType</code>.</em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///The final output format</span>
<span class="k">let</span> <span class="i">customOutputFormat</span> <span class="o">=</span> <span class="i">OutputTypeCustom</span>(<span class="i">outputType</span> , <span class="i">outputFormat</span>)
</code></pre></td>
</tr>
</table>
<p>We now have everything set up to perform the BLAST search. As we are talking about proteins, we will use blastP. The parameters needed for the Wrapper function are:</p>
<ul>
<li>path of the ncbi/bin folder</li>
<li>path and name of the search database</li>
<li>path and name of the query</li>
<li>path and name of our output file</li>
<li>a sequence of BLAST parameters, containing any parameters additional to the above (like our customOutputFormat)</li>
</ul>
<p><em>Note: in this case we can use the string <code>inputFile</code> that we used above for creating our database, as we did not specify another path or name for our database. Adjust accordingly if
done otherwise</em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///output file of the BLAST search</span>
<span class="k">let</span> <span class="i">outputPath</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Output.txt&quot;</span>) 

<span class="i">BlastWrapper</span>(<span class="i">ncbiPath</span>)<span class="o">.</span><span class="i">blastP</span> <span class="i">inputFile</span> <span class="i">queryFastaPath</span> <span class="i">outputPath</span> ([<span class="i">customOutputFormat</span>;] <span class="o">|&gt;</span> <span class="i">seq</span><span class="o">&lt;</span><span class="i">BlastParams</span><span class="o">&gt;</span>)
</code></pre></td>
</tr>
</table>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li>
                    <a href="/BioFSharp/IBioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/SeqProperties.html">Sequence Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF.html">GFF format</a></li>
                <li><a href="/BioFSharp/Genebank.html">Genebank</a></li>
                <li><a href="/BioFSharp/Clustal.html">Clustal</a></li>
                <li><a href="/BioFSharp/Newick.html">Newick</a></li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/GESA.html">Web logo</a></li>

                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
