<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>FastA format I/O
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="FastA-format-I-O" class="anchor" href="#FastA-format-I-O">FastA format I/O</a></h1>
<p>One of the various biology-associated file formats that can be manipulated using BioFSharp is the FastA format.
The FastA format can be used to represent sequences of amino acids or nucleotides written in single-letter code.
<br>
>sp|P19532| ribosomal protein L20 GN=rpl20 PE=rpl20.p01
MTRVKRGNVSRKRHKKILNMSKGFRGAASTLFRTANQQNMKALRYSYRNRRQKKRDFRRM
WITRVNSAVRRYGLNYSEFMNYLKTHKIQLNRKVIAQLSICDPEAFMQLLLF*
<br>
One sequence constists of two parts: The first line (Header) starting with a "&gt;" is followed by a sequence identification code which should represent a unique description of the sequence.
Subsequent lines contain the sequence itself, which is separated into chunks of 60 to 80 characters per line.
For further information about the format please visit <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=BlastHelp">NCBI - FASTA</a>.</p>
<h2><a name="Reading-FastA-files" class="anchor" href="#Reading-FastA-files">Reading FastA files</a></h2>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
<span class="l">9: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span>
<span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span>

<span class="k">let</span> <span class="i">fileDir</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/&quot;</span>  <span class="c">//FASTAExample1.fasta&quot;</span>

<span class="c">//reads from file to an array of FastaItems.</span>
<span class="k">let</span> <span class="i">sequences</span> <span class="o">=</span> 
    <span class="i">fileDir</span> <span class="o">+</span> <span class="s">&quot;Chlamy_Cp.fastA&quot;</span>
    <span class="o">|&gt;</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>
</code></pre></td>
</tr>
</table>
<p>Analogously it is possible to directly read compressed fastA files:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">sequences2</span> <span class="o">=</span> 
    <span class="i">fileDir</span> <span class="o">+</span> <span class="s">&quot;Chlamy_Cp.gz&quot;</span>
    <span class="o">|&gt;</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>
</code></pre></td>
</tr>
</table>
<p>In both cases it is worth noticing that <code>BioArray.ofAminoAcidString</code> can be replaced by any converter function.
The converter maps from the sequence of character to either amino acid or nucleotide sequences. Therefore use <code>BioArray.ofAminoAcidString</code> for petide and <code>BioArray.ofNucleotideString</code> for gene FastA files, respectively.
It is of course also possible to introduce any converter function.</p>
<h2><a name="Writing-FastA-files" class="anchor" href="#Writing-FastA-files">Writing FastA files</a></h2>
<p>In order to write a collection of sequences (<code>FastaItem&lt;_&gt;</code>) into a file use the following function.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">sequences</span>
<span class="o">|&gt;</span> <span class="i">FastA</span><span class="o">.</span><span class="i">write</span> <span class="i">BioItem</span><span class="o">.</span><span class="i">symbol</span> (<span class="i">fileDir</span> <span class="o">+</span> <span class="s">&quot;FASTAExample3.fasta&quot;</span>)  
</code></pre></td>
</tr>
</table>
<h2><a name="Example-protein-length-distribution" class="anchor" href="#Example-protein-length-distribution">Example: protein length distribution</a></h2>
<p>With the FastA reader it is straightforward to access the protein length distribution:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">sequencesLength</span> <span class="o">=</span> 
    <span class="i">fileDir</span> <span class="o">+</span> <span class="s">&quot;Chlamy_Cp.fastA&quot;</span>
    <span class="o">|&gt;</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>
    <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">map</span> (<span class="k">fun</span> <span class="i">item</span> <span class="k">-&gt;</span> <span class="i">item</span><span class="o">.</span><span class="i">Sequence</span><span class="o">.</span><span class="i">Length</span>)
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">Chart</span><span class="o">.</span><span class="i">Histogram</span> <span class="i">sequencesLength</span>
<span class="o">|&gt;</span> <span class="i">Chart</span><span class="o">.</span><span class="i">withX_AxisStyle</span>(<span class="s">&quot;length&quot;</span>)
<span class="o">|&gt;</span> <span class="i">Chart</span><span class="o">.</span><span class="i">withY_AxisStyle</span>(<span class="s">&quot;#count&quot;</span>)
</code></pre></td>
</tr>
</table>
<div id="b423d30d-ab6f-4479-bc7c-f4cb80a565ce" style="width: 700px; height: 500px;"><!-- Plotly chart will be drawn inside this DIV --></div>        
  <script>
    var data = [{"type":"histogram","x":[161,688,47,419,113,525,216,294,38,96,279,93,137,123,180,154,752,258,353,353,34,142,169,101,35,63,354,89,45,32,509,738,911,138,198,192,83,627,823,45,39,38,713,3121,736,476,509,38,501,83,176,131,470,492,1996,134,42,239,43,353,2972,462,82,33,490,589,382,444,547,506,383,163,293,369],"marker":{}}];
    var layout = {"xaxis":{"title":"length"},"yaxis":{"title":"#count"}};
    Plotly.newPlot('b423d30d-ab6f-4479-bc7c-f4cb80a565ce', data, layout);
  </script>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                <li>
                    <a href="/BioFSharp/BioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF3.html">GFF format</a></li>
                <li><a href="/BioFSharp/GenBank.html">GenBank</a></li>
                <li>
                    <a href="/BioFSharp/Clustal.html">Clustal</a>
                </li>        
                <li>
                    <a href="/BioFSharp/Newick.html">Newick</a>
                </li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/CSV.html">Delimited files</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                
                <li class="nav-header">Documentation</li>
                <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
