<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Pretty Printers
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="An open source bioinformatics toolbox written in F#. <https://csbiology.github.io/BioFSharp/>"/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <meta http-equiv="X-UA-Compatible" content="IE=11" >
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" /> <!--/BioFSharp-->
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" /> <!--/BioFSharp-->
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script> <!--/BioFSharp-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
	<body>
		<div class="container">
		  <div class="masthead">
			<ul class="nav nav-pills pull-right">
			  <li><a href="http://fsharp.org">fsharp.org</a></li>
			  <li><a href="https://github.com/CSBiology/BioFSharp">github page</a></li>
			</ul>
			<h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
		  </div>
		  <hr />
		  <div class="row">
			<div class="span9" id="main">
			  
<h1><a name="Pretty-Printers" class="anchor" href="#Pretty-Printers">Pretty Printers</a></h1>
<p>BioFSharp is designed to work well with the FSI, as data evaluation usually involves a lot of scripting. The <code>BioFSharp.IO.FSIPrinters</code> module offers
a bunch of functions to view our data types in a structured string format. This helps avoiding to save interim results for the purpose of
visual investigation. To use these printers, use the <code>fsi.AddPrinter</code> function to register the desired printer. This will override the default
printing behaviour of the respective type in the FSI.</p>
<p>We also provide a <code>BioFSharp.IO.fsx</code> convenience script in our nuget packages, that registers all (except modification printers) printers to the FSI.
just <code>#load</code> the script and you get all the printing goodness.</p>
<p>However, if you want to selectively register FSI printers,here are the currently implemented ones:</p>
<h2><a name="BioItems" class="anchor" href="#BioItems">BioItems</a></h2>
<p>There are two options when printing BioItems concerning modifications:</p>
<ul>
<li>
<p><code>prettyPrintBioItem</code> will not directly print modifications, but indicates them by printing the symbol of the modified item in lowercase.
This will not display all available information, but in return keep collections nicely formatted, as you can see later.</p>
</li>
<li><p><code>prettyPrintBioItemWithModifications</code> will print all modifications on BioItems in square brackets appended to the modified items symbol</p></li>
</ul>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="id">aa</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="m">AminoAcids</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="id">Ala</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="id">nuc</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs7', 8)" onmouseover="showTip(event, 'fs7', 8)" class="m">Nucleotides</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 9)" onmouseover="showTip(event, 'fs8', 9)" class="id">T</span>

<span class="c">//a twofold modified Alanine</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 10)" onmouseover="showTip(event, 'fs9', 10)" class="id">modifiedAA</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs3', 11)" onmouseover="showTip(event, 'fs3', 11)" class="id">aa</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs4', 12)" onmouseover="showTip(event, 'fs4', 12)" class="m">AminoAcids</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 13)" onmouseover="showTip(event, 'fs10', 13)" class="id">setModification</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs11', 14)" onmouseover="showTip(event, 'fs11', 14)" class="m">ModificationInfo</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 15)" onmouseover="showTip(event, 'fs12', 15)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 16)" onmouseover="showTip(event, 'fs13', 16)" class="id">N15</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs4', 17)" onmouseover="showTip(event, 'fs4', 17)" class="m">AminoAcids</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 18)" onmouseover="showTip(event, 'fs10', 18)" class="id">setModification</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs11', 19)" onmouseover="showTip(event, 'fs11', 19)" class="m">ModificationInfo</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 20)" onmouseover="showTip(event, 'fs12', 20)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs14', 21)" onmouseover="showTip(event, 'fs14', 21)" class="id">CO</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<p>Standard console output:</p>
<table class="pre"><tr><td><pre><code>Ala</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>T</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>Mod
  (Ala,
   [{ Name = "CO addition"
      IsBiological = true
      Location = Cterm
      Modify = &lt;fun:createModificationWithAdd@56&gt; }; { Name = "#N15"
                                                       IsBiological = true
                                                       Location = Isotopic
                                                       Modify = &lt;fun:N15@89&gt; }])</code></pre></td></tr></table>
<p>Console output using <code>prettyPrintBioItem</code>:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printer</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 46)" onmouseover="showTip(event, 'fs16', 46)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs17', 47)" onmouseover="showTip(event, 'fs17', 47)" class="id">prettyPrintBioItem</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"A"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"T"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"a"</code></pre></td></tr></table>
<p>Console output using <code>prettyPrintBioItemWithModifications</code>:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printer</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 48)" onmouseover="showTip(event, 'fs16', 48)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs21', 49)" onmouseover="showTip(event, 'fs21', 49)" class="id">prettyPrintBioItemWithModifications</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"A"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"T"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"A[CO addition;#N15]"</code></pre></td></tr></table>
<h2><a name="BioCollections" class="anchor" href="#BioCollections">BioCollections</a></h2>
<p>The pretty printers for BioCollections (<code>BioSeq</code>,<code>BioArray</code>, and <code>BioList</code>) show the contents of the collections as
6 x 10 space-sepearated item blocks per line, preceeded by an index indicator. This is similar to the ORIGIN section of a genebank file.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//an amino acid sequence</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs24', 50)" onmouseover="showTip(event, 'fs24', 50)" class="id">largerSequence</span> <span class="o">=</span> 
    <span class="s">&quot;&quot;&quot;MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT</span>
<span class="s">    MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT</span>
<span class="s">    MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT&quot;&quot;&quot;</span> 
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs25', 51)" onmouseover="showTip(event, 'fs25', 51)" class="m">BioArray</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 52)" onmouseover="showTip(event, 'fs26', 52)" class="id">ofAminoAcidString</span>

<span class="c">//the same amino acid sequence with a 15N labeled Alanine at position 4</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 53)" onmouseover="showTip(event, 'fs27', 53)" class="id">modifiedSequence</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs28', 54)" onmouseover="showTip(event, 'fs28', 54)" class="m">Array</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs29', 55)" onmouseover="showTip(event, 'fs29', 55)" class="id">copy</span> <span onmouseout="hideTip(event, 'fs24', 56)" onmouseover="showTip(event, 'fs24', 56)" class="id">largerSequence</span>
<span onmouseout="hideTip(event, 'fs27', 57)" onmouseover="showTip(event, 'fs27', 57)" class="id">modifiedSequence</span><span class="m">.</span><span class="pn">[</span><span class="n">3</span><span class="pn">]</span> <span class="k">&lt;-</span> <span onmouseout="hideTip(event, 'fs4', 58)" onmouseover="showTip(event, 'fs4', 58)" class="m">AminoAcids</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs10', 59)" onmouseover="showTip(event, 'fs10', 59)" class="id">setModification</span> <span onmouseout="hideTip(event, 'fs11', 60)" onmouseover="showTip(event, 'fs11', 60)" class="m">ModificationInfo</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 61)" onmouseover="showTip(event, 'fs12', 61)" class="m">Table</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 62)" onmouseover="showTip(event, 'fs13', 62)" class="id">N15</span> <span onmouseout="hideTip(event, 'fs4', 63)" onmouseover="showTip(event, 'fs4', 63)" class="m">AminoAcids</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs5', 64)" onmouseover="showTip(event, 'fs5', 64)" class="id">Ala</span>
</code></pre></td>
</tr>
</table>
<p>Standard console output:</p>
<table class="pre"><tr><td><pre><code>[|Met; Ala; Ser; Ser; Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser; Pro; Ala;
  Gln; Ala; Thr; Met; Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys; Ser; Ser; Ala; Ala;
  Phe; Pro; Ala; Thr; Arg; Lys; Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser;
  Asn; Gly; Gly; Arg; Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys;
  Lys; Lys; Phe; Glu; Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser; Glu;
  Leu; Ala; Lys; Glu; Val; Asp; Tyr; Leu; Ile; Arg; Asn; Lys; Trp; Ile; Pro; Cys;
  Val; Glu; Phe; Glu; ...|]</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>[|Met; Ala; Ser;
  Mod (Ala,[{ Name = "#N15"
            IsBiological = true
            Location = Isotopic
            Modify = &lt;fun:N15@89&gt; }]);
  Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser; Pro; Ala; Gln; Ala; Thr; Met;
  Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys; Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr;
  Arg; Lys; Ala; Asn; Asn; Asp; Ile; Thr; Ser; Ile; Thr; Ser; Asn; Gly; Gly; Arg;
  Val; Asn; Cys; Met; Gln; Val; Trp; Pro; Pro; Ile; Gly; Lys; Lys; Lys; Phe; Glu;
  Thr; Leu; Ser; Tyr; Leu; Pro; Asp; Leu; Thr; Asp; Ser; Glu; Leu; Ala; Lys; Glu;
  Val; Asp; Tyr; Leu; Ile; Arg; Asn; Lys; Trp; Ile; Pro; Cys; Val; Glu; Phe; Glu;
  ...|]</code></pre></td></tr></table>
<p>Console output using <code>prettyPrintBioCollection</code>:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printer</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 81)" onmouseover="showTip(event, 'fs16', 81)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs31', 82)" onmouseover="showTip(event, 'fs31', 82)" class="id">prettyPrintBioCollection</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"
         1  MASSMLSSAT MVASPAQATM VAPFNGLKSS AAFPATRKAN NDITSITSNG GRVNCMQVWP
        61  PIGKKKFETL SYLPDLTDSE LAKEVDYLIR NKWIPCVEFE LEHGFVYREH GNSPGYYDGR
       121  YWTMWKLPLF GCTDSAQVLK EVEECKKEYP NAFIRIIGFD NTRQVQCISF IAYKPPSFTM
       181  ASSMLSSATM VASPAQATMV APFNGLKSSA AFPATRKANN DITSITSNGG RVNCMQVWPP
       241  IGKKKFETLS YLPDLTDSEL AKEVDYLIRN KWIPCVEFEL EHGFVYREHG NSPGYYDGRY
       301  WTMWKLPLFG CTDSAQVLKE VEECKKEYPN AFIRIIGFDN TRQVQCISFI AYKPPSFTMA
       361  SSMLSSATMV ASPAQATMVA PFNGLKSSAA FPATRKANND ITSITSNGGR VNCMQVWPPI
       421  GKKKFETLSY LPDLTDSELA KEVDYLIRNK WIPCVEFELE HGFVYREHGN SPGYYDGRYW
       481  TMWKLPLFGC TDSAQVLKEV EECKKEYPNA FIRIIGFDNT RQVQCISFIA YKPPSFT
"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"
         1  MASaMLSSAT MVASPAQATM VAPFNGLKSS AAFPATRKAN NDITSITSNG GRVNCMQVWP
        61  PIGKKKFETL SYLPDLTDSE LAKEVDYLIR NKWIPCVEFE LEHGFVYREH GNSPGYYDGR
       121  YWTMWKLPLF GCTDSAQVLK EVEECKKEYP NAFIRIIGFD NTRQVQCISF IAYKPPSFTM
       181  ASSMLSSATM VASPAQATMV APFNGLKSSA AFPATRKANN DITSITSNGG RVNCMQVWPP
       241  IGKKKFETLS YLPDLTDSEL AKEVDYLIRN KWIPCVEFEL EHGFVYREHG NSPGYYDGRY
       301  WTMWKLPLFG CTDSAQVLKE VEECKKEYPN AFIRIIGFDN TRQVQCISFI AYKPPSFTMA
       361  SSMLSSATMV ASPAQATMVA PFNGLKSSAA FPATRKANND ITSITSNGGR VNCMQVWPPI
       421  GKKKFETLSY LPDLTDSELA KEVDYLIRNK WIPCVEFELE HGFVYREHGN SPGYYDGRYW
       481  TMWKLPLFGC TDSAQVLKEV EECKKEYPNA FIRIIGFDNT RQVQCISFIA YKPPSFT
"</code></pre></td></tr></table>
<p>When printing modifications as shown before with BioItems, the blocks are not aligned anymore, but more information is displayed.</p>
<p>Console output using <code>prettyPrintBioCollectionWithModifications</code>:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printer</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 83)" onmouseover="showTip(event, 'fs16', 83)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs34', 84)" onmouseover="showTip(event, 'fs34', 84)" class="id">prettyPrintBioCollectionWithModifications</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"
         1  MASSMLSSAT MVASPAQATM VAPFNGLKSS AAFPATRKAN NDITSITSNG GRVNCMQVWP
        61  PIGKKKFETL SYLPDLTDSE LAKEVDYLIR NKWIPCVEFE LEHGFVYREH GNSPGYYDGR
       121  YWTMWKLPLF GCTDSAQVLK EVEECKKEYP NAFIRIIGFD NTRQVQCISF IAYKPPSFTM
       181  ASSMLSSATM VASPAQATMV APFNGLKSSA AFPATRKANN DITSITSNGG RVNCMQVWPP
       241  IGKKKFETLS YLPDLTDSEL AKEVDYLIRN KWIPCVEFEL EHGFVYREHG NSPGYYDGRY
       301  WTMWKLPLFG CTDSAQVLKE VEECKKEYPN AFIRIIGFDN TRQVQCISFI AYKPPSFTMA
       361  SSMLSSATMV ASPAQATMVA PFNGLKSSAA FPATRKANND ITSITSNGGR VNCMQVWPPI
       421  GKKKFETLSY LPDLTDSELA KEVDYLIRNK WIPCVEFELE HGFVYREHGN SPGYYDGRYW
       481  TMWKLPLFGC TDSAQVLKEV EECKKEYPNA FIRIIGFDNT RQVQCISFIA YKPPSFT
"</code></pre></td></tr></table>
<table class="pre"><tr><td><pre><code>"
         1  MASA[#N15]MLSSAT MVASPAQATM VAPFNGLKSS AAFPATRKAN NDITSITSNG GRVNCMQVWP
        61  PIGKKKFETL SYLPDLTDSE LAKEVDYLIR NKWIPCVEFE LEHGFVYREH GNSPGYYDGR
       121  YWTMWKLPLF GCTDSAQVLK EVEECKKEYP NAFIRIIGFD NTRQVQCISF IAYKPPSFTM
       181  ASSMLSSATM VASPAQATMV APFNGLKSSA AFPATRKANN DITSITSNGG RVNCMQVWPP
       241  IGKKKFETLS YLPDLTDSEL AKEVDYLIRN KWIPCVEFEL EHGFVYREHG NSPGYYDGRY
       301  WTMWKLPLFG CTDSAQVLKE VEECKKEYPN AFIRIIGFDN TRQVQCISFI AYKPPSFTMA
       361  SSMLSSATMV ASPAQATMVA PFNGLKSSAA FPATRKANND ITSITSNGGR VNCMQVWPPI
       421  GKKKFETLSY LPDLTDSELA KEVDYLIRNK WIPCVEFELE HGFVYREHGN SPGYYDGRYW
       481  TMWKLPLFGC TDSAQVLKEV EECKKEYPNA FIRIIGFDNT RQVQCISFIA YKPPSFT
"</code></pre></td></tr></table>
<h2><a name="Clustal-formatted-alignments" class="anchor" href="#Clustal-formatted-alignments">Clustal formatted alignments:</a></h2>
<p>The <code>prettyPrintClustal</code> printer displays clustal alignments just as they would apper in a .asn file:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs36', 85)" onmouseover="showTip(event, 'fs36', 85)" class="id">path</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/clustalExample.asn&quot;</span>

<span class="c">//Clustal algnment read in from a clustal formated .asn file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs37', 86)" onmouseover="showTip(event, 'fs37', 86)" class="id">clustalAlignment</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs38', 87)" onmouseover="showTip(event, 'fs38', 87)" class="m">Clustal</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs39', 88)" onmouseover="showTip(event, 'fs39', 88)" class="id">ofFile</span> <span onmouseout="hideTip(event, 'fs36', 89)" onmouseover="showTip(event, 'fs36', 89)" class="id">path</span>
</code></pre></td>
</tr>
</table>
<p>Standard console output:</p>
<table class="pre"><tr><td><pre><code>{ MetaData = { Header = seq ['C'; 'L'; 'U'; 'S'; ...]
               ConservationInfo = seq ['*'; ' '; '*'; '*'; ...] }
  Sequences =
             [{ Tag = "C_reinhardtii"
                Sequence = seq ['M'; 'V'; 'P'; 'Q'; ...] };
              { Tag = "N_tabacum"
                Sequence = seq ['M'; 'S'; 'P'; 'Q'; ...] };
              { Tag = "A_thaliana"
                Sequence = seq ['M'; 'S'; 'P'; 'Q'; ...] };
              { Tag = "O_sativa"
                Sequence = seq ['M'; 'S'; 'P'; 'Q'; ...] }] }</code></pre></td></tr></table>
<p>Console output using <code>prettyPrintClustal</code>:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printer</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 94)" onmouseover="showTip(event, 'fs16', 94)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs41', 95)" onmouseover="showTip(event, 'fs41', 95)" class="id">prettyPrintClustal</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>"
CLUSTAL O(1.2.4) multiple sequence alignment

C_reinhardtii 
MVPQTETKAGAGFKAGVKDYRLTYYTPDYVVRDTDILAAFRMTPQLGVPPEECGAAVAAE
N_tabacum     
MSPQTETKASVGFKAGVKEYKLTYYTPEYQTKDTDILAAFRVTPQPGVPPEEAGAAVAAE
A_thaliana    
MSPQTETKASVGFKAGVKEYKLTYYTPEYETKDTDILAAFRVTPQPGVPPEEAGAAVAAE
O_sativa      
MSPQTETKASVGFKAGVKDYKLTYYTPEYETKDTDILAAFRVTPQPGVPPEEAGAAVAAE
              
* *******..*******:*:******:* .:*********:*** ******.*******

C_reinhardtii 
SSTGTWTTVWTDGLTSLDRYKGRCYDIEPVPGEDNQYIAYVAYPIDLFEEGSVTNMFTSI
N_tabacum     
SSTGTWTTVWTDGLTSLDRYKGRCYRIERVVGEKDQYIAYVAYPLDLFEEGSVTNMFTSI
A_thaliana    
SSTGTWTTVWTDGLTSLDRYKGRCYHIEPVPGEETQFIAYVAYPLDLFEEGSVTNMFTSI
O_sativa      
SSTGTWTTVWTDGLTSLDRYKGRCYHIEPVVGEDNQYIAYVAYPLDLFEEGSVTNMFTSI
              
************************* ** * **. *:*******:***************

C_reinhardtii 
VGNVFGFKALRALRLEDLRIPPAYVKTFVGPPHGIQVERDKLNKYGRGLLGCTIKPKLGL
N_tabacum     
VGNVFGFKALRALRLEDLRIPPAYVKTFQGPPHGIQVERDKLNKYGRPLLGCTIKPKLGL
A_thaliana    
VGNVFGFKALAALRLEDLRIPPAYTKTFQGPPHGIQVERDKLNKYGRPLLGCTIKPKLGL
O_sativa      
VGNVFGFKALRALRLEDLRIPPTYSKTFQGPPHGIQVERDKLNKYGRPLLGCTIKPKLGL
              
********** ***********:* *** ****************** ************

C_reinhardtii 
SAKNYGRAVYECLRGGLDFTKDDENVNSQPFMRWRDRFLFVAEAIYKAQAETGEVKGHYL
N_tabacum     
SAKNYGRAVYECLRGGLDFTKDDENVNSQPF-RWRDRFLFCAEALYKAQAETGEIKGHYL
A_thaliana    
SAKNYGRAVYECLRGGLDFTKDDENVNSQPFMRWRDRFLFCAEAIYKSQAETGEIKGHYL
O_sativa      
SAKNYGRACYECLRGGLDFTKDDENVNSQPFMRWRDRFVFCAEAIYKSQAETGEIKGHYL
              
******** ********************** ******:* ***:**:******:*****

C_reinhardtii 
NATAGTCEEMMKRAVCAKELGVPIIMHDYLTGGFTANTSLAIYCRDNGLLLHIHRAMHAV
N_tabacum     
NATAGTCEEMIKRAVFARELGVPIVMHDYLTGGFTANTSLAHYCRDNGLLLHIHRAMHAV
A_thaliana    
NATAGTCEEMIKRAVFARELGVPIVMHDYLTGGFTANTSLSHYCRDNGLLLHIHRAMHAV
O_sativa      
NATAGTCEEMIKRAVFARELGVPIVMHDYLTGGFTANTSLAHYCRDNGLLLHIHRAMHAV
              
**********:**** *:******:***************: ******************

C_reinhardtii 
IDRQRNHGIHFRVLAKALRMSGGDHLHSGTVVGKLEGEREVTLGFVDLMRDDYVEKDRSR
N_tabacum     
IDRQKNHGIHFRVLAKALRMSGGDHIHSGTVVGKLEGERDITLGFVDLLRDDFVEQDRSR
A_thaliana    
IDRQKNHGMHFRVLAKALRLSGGDHIHAGTVVGKLEGDRESTLGFVDLLRDDYVEKDRSR
O_sativa      
IDRQKNHGMHFRVLAKALRMSGGDHIHAGTVVGKLEGEREMTLGFVDLLRDDFIEKDRAR
              
****:***:**********:*****:*:*********:*: *******:***::*:**:*

C_reinhardtii 
GIYFTQDWCSMPGVMPVASGGIHVWHMPALVEIFGDDACLQFGGGTLGHPWGNAPGAAAN
N_tabacum     
GIYFTQDWVSLPGVLPVASGGIHVWHMPALTEIFGDDSVLQFGGGTLGHPWGNAPGAVAN
A_thaliana    
GIFFTQDWVSLPGVLPVASGGIHVWHMPALTEIFGDDSVLQFGGGTLGHPWGNAPGAVAN
O_sativa      
GIFFTQDWVSMPGVIPVASGGIHVWHMPALTEIFGDDSVLQFGGGTLGHPWGNAPGAAAN
              
**:***** *:***:***************.******: ******************.**

C_reinhardtii 
RVALEACTQARNEGRDLAREGGDVIRSACKWSPELAAACEVWKEIKFEFDTIDKL----
N_tabacum     
RVALEACVKARNEGRDLAQEGNEIIREACKWSPELAAACEVWKEIVFNFAAVDVLDK--
A_thaliana    
RVALEACVQARNEGRDLAVEGNEIIREACKWSPELAAACEVWKEITFNFPTIDKLDGQE
O_sativa      
RVALEACVQARNEGRDLAREGNEIIRSACKWSPELAAACEIWKAIKFEFEPVDKLDS--
              
*******.:********* **.::**.*************:** * *:*  :* *    
"</code></pre></td></tr></table>
<h2><a name="SOFT" class="anchor" href="#SOFT">SOFT</a></h2>
<p>There are 5 printers available:</p>
<p><code>prettyPrintGSE</code> and <code>prettyPrintGPL</code> format the top level types <code>SOFT.Series.GSE</code> and <code>SOFT.Platform.GPL</code> including associated record metadata.</p>
<p><code>prettyPrintSampleRecord</code>,<code>prettyPrintSeriesRecord</code>, and <code>prettyPrintPlatformRecord</code> format single record metadata.</p>
<p>All SOFT types are very large and hard to read from standard output, especially the nested top level types, so we will omitt the standard output for readability.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//register the desired printers</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 96)" onmouseover="showTip(event, 'fs16', 96)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs42', 97)" onmouseover="showTip(event, 'fs42', 97)" class="id">prettyPrintGPL</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 98)" onmouseover="showTip(event, 'fs16', 98)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs43', 99)" onmouseover="showTip(event, 'fs43', 99)" class="id">prettyPrintGSE</span><span class="pn">)</span>

<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 100)" onmouseover="showTip(event, 'fs16', 100)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs44', 101)" onmouseover="showTip(event, 'fs44', 101)" class="id">prettyPrintSampleRecord</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 102)" onmouseover="showTip(event, 'fs16', 102)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs45', 103)" onmouseover="showTip(event, 'fs45', 103)" class="id">prettyPrintSeriesRecord</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 104)" onmouseover="showTip(event, 'fs16', 104)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs46', 105)" onmouseover="showTip(event, 'fs46', 105)" class="id">prettyPrintPlatformRecord</span><span class="pn">)</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs47', 106)" onmouseover="showTip(event, 'fs47', 106)" class="id">gplPath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/GPL15922_family.soft&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs48', 107)" onmouseover="showTip(event, 'fs48', 107)" class="id">gpl15922</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs49', 108)" onmouseover="showTip(event, 'fs49', 108)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs50', 109)" onmouseover="showTip(event, 'fs50', 109)" class="m">Platform</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs51', 110)" onmouseover="showTip(event, 'fs51', 110)" class="id">fromFile</span> <span onmouseout="hideTip(event, 'fs47', 111)" onmouseover="showTip(event, 'fs47', 111)" class="id">gplPath</span>
</code></pre></td>
</tr>
</table>
<p>Console output using <code>prettyPrintGPL</code>:</p>
<table class="pre"><tr><td><pre><code>"
GEO PLATFORM RECORD GPL15922
============================

Title:          Illumina HiSeq 2000 (Chlamydomonas reinhardtii)

Organism(s):    Chlamydomonas reinhardtii

Description:    

Technology:     high-throughput sequencing

Contributor(s): 

Series(s):      GSE101944 =&gt; Transcriptome analysis of Chlamydomonas reinhardtii chloroplast and mitochondria
                GSE112394 =&gt; Transcriptomics analysis of the Chlamydomonas reinhardtii diurnal cycle.
                GSE40031 =&gt; Retrograde bilin signaling enables Chlamydomonas greening and phototrophic survival
                GSE44611 =&gt; Changes in membrane lipid composition during Iron depletion and starvation in Chlamydomonas reinhardtii
                GSE55253 =&gt; The Path to Triacylglyceride Obesity in the sta6 Strain of Chlamydomonas reinhardtii
                GSE58786 =&gt; RNAseq Analysis of Transcriptomic Response to Zn Resupply in Chlamydomonas reinhardtii
                GSE59629 =&gt; Lineage-Specific Chromatin Signatures Reveal a Master Lipid Switch in Microalgae
                GSE62690 =&gt; N6-Methyldeoxyadenosine Marks Active Transcription Start Sites in Chlamydomonas
                GSE71469 =&gt; Chlamydomonas diurnal transcriptome
                GSE77388 =&gt; Bilin-dependent photoacclimation in Chlamydomonas reinhardtii

Samples         GSM1087792 =&gt; C.reinhardtii _Fe_Long_0_hours
                GSM1087793 =&gt; C.reinhardtii _Fe_Long_0.5_hours
                GSM1087794 =&gt; C.reinhardtii _Fe_Long_1_hours
                GSM1087795 =&gt; C.reinhardtii _Fe_Long_2_hours
                GSM1087796 =&gt; C.reinhardtii _Fe_Long_4_hours
                GSM1087797 =&gt; C.reinhardtii _Fe_Long_8_hours
                GSM1087798 =&gt; C.reinhardtii _Fe_Long_12_hours
                GSM1087799 =&gt; C.reinhardtii _Fe_Long_24_hours
                GSM1087800 =&gt; C.reinhardtii _Fe_Long_48_hours
                GSM1332630 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 0h pre-wash
                GSM1332631 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 0h
                GSM1332632 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 0.5h
                GSM1332633 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 2h
                GSM1332634 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 4h
                GSM1332635 =&gt; Chlamydomonas reinhardtii CC-4349 minus N 8h
                (...and 208 more)

(For more Metadata, access the type directly)
            "</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs53', 116)" onmouseover="showTip(event, 'fs53', 116)" class="id">gsePath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/GSE71469_family.soft&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs54', 117)" onmouseover="showTip(event, 'fs54', 117)" class="id">gse71469</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs49', 118)" onmouseover="showTip(event, 'fs49', 118)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs55', 119)" onmouseover="showTip(event, 'fs55', 119)" class="m">Series</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs56', 120)" onmouseover="showTip(event, 'fs56', 120)" class="id">fromFile</span> <span onmouseout="hideTip(event, 'fs53', 121)" onmouseover="showTip(event, 'fs53', 121)" class="id">gsePath</span>
</code></pre></td>
</tr>
</table>
<p>Console output using <code>prettyPrintGSE</code>:</p>
<table class="pre"><tr><td><pre><code>"
GEO SERIES RECORD GSE71469
==========================

Type(s):        Expression profiling by high throughput sequencing

Platform(s):    GPL15922 =&gt; Illumina HiSeq 2000 (Chlamydomonas reinhardtii)

Title:          Chlamydomonas diurnal transcriptome

Contributor(s): James,G,Umen
                Sabeeha,S,Merchant
                Ian,K,Blaby
                James,M,Zones

Design:         Synchronous phototrophic cultures of Chlamydomonas reinhardtii were prepared in controlled diurnal bioreactor growth conditions and
                sampled in two 24-hour replicate time courses at 1-hour and 0.5-hour intervals for Illumina paired-end
                RNA-Seq. More than ~3 million paired end reads were obtained for each sample.

Summary:        We combined a highly synchronous photobioreactor culture system with frequent temporal sampling to characterize genome-wide
                periodic gene expression in Chlamydomonas.

Samples         GSM1835200 =&gt; 1_1
                GSM1835201 =&gt; 2_1
                GSM1835202 =&gt; 3_1
                GSM1835203 =&gt; 4_1
                GSM1835204 =&gt; 5_1
                GSM1835205 =&gt; 6_1
                GSM1835206 =&gt; 7_1
                GSM1835207 =&gt; 8_1
                GSM1835208 =&gt; 9_1
                GSM1835209 =&gt; 10_1
                GSM1835210 =&gt; 11_1
                GSM1835211 =&gt; 11.5_1
                GSM1835212 =&gt; 12_1
                GSM1835213 =&gt; 12.5_1
                GSM1835214 =&gt; 13_1
                (...and 41 more)

(For more Metadata, access the type directly)
            "</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs58', 126)" onmouseover="showTip(event, 'fs58', 126)" class="id">smplRecord</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs54', 127)" onmouseover="showTip(event, 'fs54', 127)" class="id">gse71469</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs49', 128)" onmouseover="showTip(event, 'fs49', 128)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs55', 129)" onmouseover="showTip(event, 'fs55', 129)" class="m">Series</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs59', 130)" onmouseover="showTip(event, 'fs59', 130)" class="id">getAssociatedSamples</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs60', 131)" onmouseover="showTip(event, 'fs60', 131)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs61', 132)" onmouseover="showTip(event, 'fs61', 132)" class="id">item</span> <span class="n">0</span>
</code></pre></td>
</tr>
</table>
<p>Console output using <code>prettyPrintSampleRecord</code>:</p>
<table class="pre"><tr><td><pre><code>"
METADATA FOR SAMPLE RECORD GSM1835200
=====================================

Title:              1_1

Type(s):            SRA

Organism(s)         [CHANNEL 1]

                    Chlamydomonas reinhardtii

Characteristics:    [CHANNEL 1]

                    time (zeitgeber): ZT1
                    protocol: light
                    replicate time course: Replicate 1
                    strain: 6145c (CC-2895)

Description:        

(For more Metadata, access the type directly)
            "</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs63', 137)" onmouseover="showTip(event, 'fs63', 137)" class="id">seriesRecord</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs48', 138)" onmouseover="showTip(event, 'fs48', 138)" class="id">gpl15922</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs49', 139)" onmouseover="showTip(event, 'fs49', 139)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs50', 140)" onmouseover="showTip(event, 'fs50', 140)" class="m">Platform</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs64', 141)" onmouseover="showTip(event, 'fs64', 141)" class="id">getAssociatedSeries</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs60', 142)" onmouseover="showTip(event, 'fs60', 142)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs61', 143)" onmouseover="showTip(event, 'fs61', 143)" class="id">item</span> <span class="n">0</span>
</code></pre></td>
</tr>
</table>
<p>Console output using <code>prettyPrintSeriesRecord</code>:</p>
<table class="pre"><tr><td><pre><code>"
GEO SERIES RECORD GSE101944
===========================

Type(s):        Expression profiling by high throughput sequencing

Title:          Transcriptome analysis of Chlamydomonas reinhardtii chloroplast and mitochondria

Contributor(s): Daniela,,Strenkert
                Sean,D,Gallaher

Design:         RNA-Seq libraries were prepared from Chlamydomonas reinhardtii sampled under 12 different conditions including: diurnal growth
                in light, diurnal growth in dark, +Fe, Ã¢â¬âFe, +Cu, Ã¢â¬âCu, mt+ vegetative, mtÃ¢â¬â vegetative, mt+
                gamete, mtÃ¢â¬â gamete, zygote, and post-germination.

Summary:        Here, we report on the transcriptome of the organelles of the micro-alga, Chlamydomonas reinhardtii, sampled
                under a number of different conditions. The preparation of the RNA-Seq libraries and their analysis
                were performed using protocols optimized for organellar transcripts. Samples include growth in media +/Ã¢â¬â Fe,
                growth in media +/Ã¢â¬â Cu, diurnal growth samples collected in dark and light, and the
                sexual cycle.

(For more Metadata, access the type directly)
            "</code></pre></td></tr></table>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs66', 148)" onmouseover="showTip(event, 'fs66', 148)" class="id">pltfrmRecord</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs54', 149)" onmouseover="showTip(event, 'fs54', 149)" class="id">gse71469</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs49', 150)" onmouseover="showTip(event, 'fs49', 150)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs55', 151)" onmouseover="showTip(event, 'fs55', 151)" class="m">Series</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs67', 152)" onmouseover="showTip(event, 'fs67', 152)" class="id">getAssociatedPlatforms</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs60', 153)" onmouseover="showTip(event, 'fs60', 153)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs61', 154)" onmouseover="showTip(event, 'fs61', 154)" class="id">item</span> <span class="n">0</span>
</code></pre></td>
</tr>
</table>
<p>Console output using <code>prettyPrintSampleRecord</code>:</p>
<table class="pre"><tr><td><pre><code>"
GEO PLATFORM RECORD GPL15922
============================

Title:          Illumina HiSeq 2000 (Chlamydomonas reinhardtii)

Organism(s):    Chlamydomonas reinhardtii

Description:    

Technology:     high-throughput sequencing

Contributor(s): 

(For more Metadata, access the type directly)
            "</code></pre></td></tr></table>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace BioFSharp.IO</div>
<div class="tip" id="fs3">val aa : AminoAcids.AminoAcid</div>
<div class="tip" id="fs4">module AminoAcids<br /><br />from BioFSharp</div>
<div class="tip" id="fs5">union case AminoAcids.AminoAcid.Ala: AminoAcids.AminoAcid</div>
<div class="tip" id="fs6">val nuc : Nucleotides.Nucleotide</div>
<div class="tip" id="fs7">module Nucleotides<br /><br />from BioFSharp</div>
<div class="tip" id="fs8">union case Nucleotides.Nucleotide.T: Nucleotides.Nucleotide</div>
<div class="tip" id="fs9">val modifiedAA : AminoAcids.AminoAcid</div>
<div class="tip" id="fs10">val setModification : md:ModificationInfo.Modification -&gt; aa:AminoAcids.AminoAcid -&gt; AminoAcids.AminoAcid</div>
<div class="tip" id="fs11">module ModificationInfo<br /><br />from BioFSharp</div>
<div class="tip" id="fs12">module Table<br /><br />from BioFSharp.ModificationInfo</div>
<div class="tip" id="fs13">val N15 : ModificationInfo.Modification</div>
<div class="tip" id="fs14">val CO : ModificationInfo.Modification</div>
<div class="tip" id="fs15">val aaPrint : string</div>
<div class="tip" id="fs16">module FSIPrinters<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs17">val prettyPrintBioItem : a:#IBioItem -&gt; string</div>
<div class="tip" id="fs18">val nPrint : string</div>
<div class="tip" id="fs19">val aaPrintWithMod : string</div>
<div class="tip" id="fs20">val aaPrint&#39; : string</div>
<div class="tip" id="fs21">val prettyPrintBioItemWithModifications : a:#IBioItem -&gt; string</div>
<div class="tip" id="fs22">val nPrint&#39; : string</div>
<div class="tip" id="fs23">val aaPrintWithMod&#39; : string</div>
<div class="tip" id="fs24">val largerSequence : BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs25">module BioArray<br /><br />from BioFSharp</div>
<div class="tip" id="fs26">val ofAminoAcidString : s:#seq&lt;char&gt; -&gt; BioArray.BioArray&lt;AminoAcids.AminoAcid&gt;</div>
<div class="tip" id="fs27">val modifiedSequence : AminoAcids.AminoAcid []</div>
<div class="tip" id="fs28">module Array<br /><br />from Microsoft.FSharp.Collections</div>
<div class="tip" id="fs29">val copy : array:&#39;T [] -&gt; &#39;T []</div>
<div class="tip" id="fs30">val colPrnt : string</div>
<div class="tip" id="fs31">val prettyPrintBioCollection : sequence:seq&lt;#IBioItem&gt; -&gt; string</div>
<div class="tip" id="fs32">val colPrntMod : string</div>
<div class="tip" id="fs33">val colPrnt&#39; : string</div>
<div class="tip" id="fs34">val prettyPrintBioCollectionWithModifications : sequence:seq&lt;#IBioItem&gt; -&gt; string</div>
<div class="tip" id="fs35">val colPrntMod&#39; : string</div>
<div class="tip" id="fs36">val path : string</div>
<div class="tip" id="fs37">val clustalAlignment : Alignment.Alignment&lt;TaggedSequence.TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt;</div>
<div class="tip" id="fs38">module Clustal<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs39">val ofFile : path:string -&gt; Alignment.Alignment&lt;TaggedSequence.TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt;</div>
<div class="tip" id="fs40">val clustalPrnt : string</div>
<div class="tip" id="fs41">val prettyPrintClustal : alignment:Alignment.Alignment&lt;TaggedSequence.TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt; -&gt; string</div>
<div class="tip" id="fs42">val prettyPrintGPL : gpl:Platform.GPL -&gt; string</div>
<div class="tip" id="fs43">val prettyPrintGSE : gse:Series.GSE -&gt; string</div>
<div class="tip" id="fs44">val prettyPrintSampleRecord : sample:SampleRecord -&gt; string</div>
<div class="tip" id="fs45">val prettyPrintSeriesRecord : series:SeriesRecord -&gt; string</div>
<div class="tip" id="fs46">val prettyPrintPlatformRecord : platform:PlatformRecord -&gt; string</div>
<div class="tip" id="fs47">val gplPath : string</div>
<div class="tip" id="fs48">val gpl15922 : Platform.GPL</div>
<div class="tip" id="fs49">module SOFT<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs50">module Platform<br /><br />from BioFSharp.IO.SOFT</div>
<div class="tip" id="fs51">val fromFile : path:string -&gt; Platform.GPL</div>
<div class="tip" id="fs52">val gplPrnt : string</div>
<div class="tip" id="fs53">val gsePath : string</div>
<div class="tip" id="fs54">val gse71469 : Series.GSE</div>
<div class="tip" id="fs55">module Series<br /><br />from BioFSharp.IO.SOFT</div>
<div class="tip" id="fs56">val fromFile : path:string -&gt; Series.GSE</div>
<div class="tip" id="fs57">val gsePrnt : string</div>
<div class="tip" id="fs58">val smplRecord : SampleRecord</div>
<div class="tip" id="fs59">val getAssociatedSamples : gse:Series.GSE -&gt; SampleRecord list</div>
<div class="tip" id="fs60">Multiple items<br />module List<br /><br />from Microsoft.FSharp.Collections<br /><br />--------------------<br />type List&lt;&#39;T&gt; =<br />&#160;&#160;| ( [] )<br />&#160;&#160;| ( :: ) of Head: &#39;T * Tail: &#39;T list<br />&#160;&#160;&#160;&#160;interface IReadOnlyList&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IReadOnlyCollection&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IEnumerable<br />&#160;&#160;&#160;&#160;interface IEnumerable&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;member GetSlice : startIndex:int option * endIndex:int option -&gt; &#39;T list<br />&#160;&#160;&#160;&#160;member Head : &#39;T<br />&#160;&#160;&#160;&#160;member IsEmpty : bool<br />&#160;&#160;&#160;&#160;member Item : index:int -&gt; &#39;T with get<br />&#160;&#160;&#160;&#160;member Length : int<br />&#160;&#160;&#160;&#160;member Tail : &#39;T list<br />&#160;&#160;&#160;&#160;...</div>
<div class="tip" id="fs61">val item : index:int -&gt; list:&#39;T list -&gt; &#39;T</div>
<div class="tip" id="fs62">val smplPrint : string</div>
<div class="tip" id="fs63">val seriesRecord : SeriesRecord</div>
<div class="tip" id="fs64">val getAssociatedSeries : gpl:Platform.GPL -&gt; SeriesRecord list</div>
<div class="tip" id="fs65">val seriesPrint : string</div>
<div class="tip" id="fs66">val pltfrmRecord : PlatformRecord</div>
<div class="tip" id="fs67">val getAssociatedPlatforms : gse:Series.GSE -&gt; PlatformRecord list</div>
<div class="tip" id="fs68">val pltfrmPrint : string</div>

			</div>
			<div class="span3" >
			  <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
              <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                  <li class="nav-header">BioFSharp</li>
                  <li><a href="/BioFSharp/index.html">Home page</a></li>
                  <li class="divider"></li>
                  <li><a href="/BioFSharp/release-notes.html">Release Notes</a></li>
                  <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                  
                  <li class="nav-header">BioFSharp core</li>
                  <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                  <li>
                      <a href="/BioFSharp/BioItem.html">BioItems</a>
                  </li>
                  <li><a href="/BioFSharp/BioCollections.html">BioCollections</a></li>
                  <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                  <li class="nav-header">BioContainers</li>
                  <li><a href="/BioFSharp/BioContainers.html">Introduction</a></li>
                  <li><a href="/BioFSharp/BioContainersDesignGuide.html">Container API Design Guide</a></li>
                  <li class="nav-header">Algorithm</li>
                  <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                  <li><a href="/BioFSharp/ClustalOWrapper.html">Clustal Omega</a></li>
                  <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                  <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                  <li>
                      <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                  </li>
                  <li class="nav-header">File I/O</li>
                  <li>
                      <a href="/BioFSharp/Clustal.html">Clustal</a>
                  </li>
                  <li><a href="/BioFSharp/CSV.html">CSV / Tabular</a></li>
                  <li>
                      <a href="/BioFSharp/FastA.html">FastA</a>
                  </li>
                  <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                  <li>
                      <a href="/BioFSharp/GenBank.html">GenBank</a>
                  </li>
                  <li><a href="/BioFSharp/SOFT.html">GEO Metadata (SOFT)</a></li>
                  <li>
                      <a href="/BioFSharp/GFF3.html">GFF3</a>
                  </li>
                  <li>
                      <a href="/BioFSharp/Newick.html">Newick</a>
                  </li>
                  <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                  <li>
                      <a href="/BioFSharp/FSIPrinters.html">FSIPrinters</a>
                          <ul class="nav nav-list">
                              <li><a href="/BioFSharp/FSIPrinters.html#BioItems">BioItems</a></li>
                              <li><a href="/BioFSharp/FSIPrinters.html#BioCollections">BioCollections</a></li>
                              <li><a href="/BioFSharp/FSIPrinters.html#Clustal-formatted-alignments">Clustal</a></li>
                              <li><a href="/BioFSharp/FSIPrinters.html#SOFT">GEO Metadata (SOFT)</a></li>
                          </ul>
                  </li>
                  <li class="nav-header">Web I/O</li>
                  <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                  <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                  <li class="nav-header">Biostatistics</li>
                  <li><a href="/BioFSharp/GSEA.html">Ontology enrichment</a></li>
                  <li class="nav-header">Biovisualization</li>
                  <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                  <li class="nav-header">GPU Computing</li>
                  <li><a href="/BioFSharp/ParallelAlignment.html">Sequence Alignment</a></li>
                  <li class="nav-header">Documentation</li>
                  <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                  
              </ul>
			</div>
		  </div>
		</div>
		<a href="https://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
	  </body>
  </html>
