<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Introducing GFF3Parser
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Introducing-GFF3Parser" class="anchor" href="#Introducing-GFF3Parser">Introducing GFF3Parser</a></h1>
<br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/GFF3Parser.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/bvenn">&#128366;view author of this tutorial</a>
<br><br>
The GFF3Parser is a tool to validate, read and write **GFF3** _(Generic Feature Format Version 3)_-Files.
<p>All examples are executed on a modified version of <a href="http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff">saccharomyces_cerevisiae.gff</a>.</p>
<h2><a name="Introducing-Generic-Feature-Format-Version-3" class="anchor" href="#Introducing-Generic-Feature-Format-Version-3">Introducing Generic Feature Format Version 3</a></h2>
<p>In GFF3 files every line represents one genomic feature with nine tab-delimited fields, whereas unlimited key-value pairs can be stored in field 9.
GFFEntries are implemented as a record type containing the following fields:
<br><br></p>
<div id="responsiveTable">
<table>
<thead>
<tr class="header">
<th><p>Field name</p></th>
<th><p>type</p></th>
<th><p>Description</p></th>
<th><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>Seqid</p></td>
<td><p><code>string</code></p></td>
<td><p>ID of the landmark</p></td>
<td><p>"chrI"</p></td>
</tr>
<tr class="even">
<td><p>Source</p></td>
<td><p><code>string</code></p></td>
<td><p>The source is a free text qualifier to describe the algorithm or operating procedure that generated this feature</p></td>
<td><p>"SGD"</p></td>
</tr>
<tr class="odd">
<td><p>Feature</p></td>
<td><p><code>string</code></p></td>
<td><p>also called method or type; should be a term from the Sequence Ontology or an SO accession number</p></td>
<td><p>"gene"</p></td>
</tr>
<tr class="even">
<td><p>StartPos</p></td>
<td><p><code>int</code></p></td>
<td><p>start coordinate of the feature (1-based interger coordinate)</p></td>
<td><p>335</p></td>
</tr>
<tr class="odd">
<td><p>EndPos</p></td>
<td><p><code>int</code></p></td>
<td><p>end coordinate of the feature (1-based interger coordinate)</p></td>
<td><p>649</p></td>
</tr>
<tr class="even">
<td><p>Score</p></td>
<td><p><code>float</code></p></td>
<td><p>The score is a not well defined floating point number representing different values depending of what you are looking at (nan  if no phase is avaliable)</p></td>
<td><p>nan</p></td>
</tr>
<tr class="odd">
<td><p>Strand</p></td>
<td><p><code>char</code></p></td>
<td><p>strand of the feature ('\000'  if no phase is avaliable)</p></td>
<td><p>'+'</p></td>
</tr>
<tr class="even">
<td><p>Phase</p></td>
<td><p><code>int</code></p></td>
<td><p>for features of type CDS; indicates where the feature begins with reference to the reading frame (-1 if no phase is avaliable)</p></td>
<td><p>-1</p></td>
</tr>
<tr class="odd">
<td><p>Attributes</p></td>
<td><p><code>Map&lt;string,string list&gt;</code></p></td>
<td><p>The field Attributes can contain multiple tag-value pairs, some with predefined meanings. Please check the specification website for further information.</p></td>
<td><p>seq [[ID, [YAL069W]];[Ontology_term, [GO:0000004; GO:0005554]]]</p></td>
</tr>
</tbody>
</table>

<p>| Supplement | <code>string []</code>               | If additional columns (e.g. counts) are present, they are written in this tab-delimited string array.                                                     | [|""|]                                                           |</p>
</div>
<br>
Below you can see an example GFF3 file.
<p><img src="@Root/img/GFF3.png" alt="GFF3" style="width:150px;margin:10px" />
<br>
<div id="responsiveTable"></p>
<table>
<thead>
<tr class="header">
<th><p>Line number</p></th>
<th><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>1</p></td>
<td><p>Declaration of file version</p></td>
</tr>
<tr class="even">
<td><p>2</p></td>
<td><p>Comment</p></td>
</tr>
<tr class="odd">
<td><p>4+5</p></td>
<td><p>GFF entry lines</p></td>
</tr>
<tr class="even">
<td><p>6</p></td>
<td><p>FastA directive</p></td>
</tr>
<tr class="odd">
<td><p>7+</p></td>
<td><p>Fasta Sequence</p></td>
</tr>
</tbody>
</table>

</div>
<br>
<p>Directives (marked with "##[...]") provide additional information like the gff-version which has to be the first line of each file ("##gff-version 3[...]").
Comment lines have to start with a single "#[...]". It is possible that sequences in FastA format are at the end of the file. This has to be announced by a "##FASTA" directive line.</p>
<p>For further information visit <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3-Specifications</a>.</p>
<h2><a name="How-to-use-GFF3Parser" class="anchor" href="#How-to-use-GFF3Parser">How to use GFF3Parser</a></h2>
<p>To read in a gff you have to insert a filepath and optional a FASTA converter. For further information about FASTA check the <a href="https://csbiology.github.io/BioFSharp/Readers.html#FastA">FASTA section</a> on this site
or visit <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span>

<span class="c">///path and name of the input file</span>
<span class="k">let</span> <span class="i">filepathGFF</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gff3Example.gff&quot;</span>)

<span class="c">///To read in a gff file you have to insert a FASTA converter and a source filepath. </span>
<span class="c">///If no FASTA Sequence is included you directly can use GFF3Parser.Reader.GFF3readerWithoutFasta [filepathGFF].</span>
<span class="k">let</span> <span class="i">gffExampleRead</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">GFF3reader</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">BioArray</span><span class="o">.</span><span class="i">ofNucleotideString</span> <span class="i">filepathGFF</span> 
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3SanityCheck" class="anchor" href="#How-to-use-GFF3SanityCheck">How to use GFF3SanityCheck</a></h2>
<p>The GFF3SanityCheck prints wheter your GFF3 file is valid or not. It returns all specified errors including the lines in which the errors occured.
In contrast to GFF2 the field 3 (type, feature or method) of a GFF3 entry is restricted to terms defined by the sequence ontology (SO) so this validator checks if the entry is a valid SO term.
You can find new versions of the SO at (<a href="https://sourceforge.net/projects/song/files/SO_Feature_Annotation).">https://sourceforge.net/projects/song/files/SO_Feature_Annotation).</a></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///to validate the gff file without SOTerm verification use this function and only insert filepath</span>
<span class="k">let</span> <span class="i">gffExampleSanityCheckSimple</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">sanityCheck</span> <span class="i">filepathGFF</span>

<span class="c">///path, name and version of the &#39;Sequence Ontology terms&#39;-file</span>
<span class="k">let</span> <span class="i">filepathSO_Terms</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Sequence_Ontology_Terms_2_5_3.txt&quot;</span>)

<span class="c">///to validate the gff file insert filepath</span>
<span class="k">let</span> <span class="i">gffExampleSanityCheck</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">sanityCheckWithSOTerm</span> <span class="i">filepathSO_Terms</span> <span class="i">filepathGFF</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3RelationshipSearch" class="anchor" href="#How-to-use-GFF3RelationshipSearch">How to use GFF3RelationshipSearch</a></h2>
<p>You also can do a simple search for "<em>Parent</em> - <em>child of</em>" relationships giving back all genomic features which contain the <em>searchterm</em> in <strong>ID/Id</strong> or <strong>Parent</strong> field.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Term to search for:</span>
<span class="k">let</span> <span class="i">searchterm</span> <span class="o">=</span> <span class="s">&quot;YAL069W&quot;</span>

<span class="c">///with this function you can search for a searchterm and you will get features which are related</span>
<span class="k">let</span> <span class="i">gffExampleSearch</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">relationshipSearch</span> <span class="i">gffExampleRead</span> <span class="i">searchterm</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3Writer" class="anchor" href="#How-to-use-GFF3Writer">How to use GFF3Writer</a></h2>
<p>This writer takes GFF lines in stacks of 20,000 items and write/append them to a given filepath.
If FastA sequences are included they are appended by a FastA writer described in the <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<p><em>Note: The order of key value pairs in field 9 (attributes) may be changed.</em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FASTA converter and a destination filepath and writes it into a .gff.</span>
<span class="k">let</span> <span class="i">gffExampleWrite</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">GFF3Writer</span> <span class="i">gffExampleRead</span> <span class="i">BioItem</span><span class="o">.</span><span class="i">symbol</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gffExampleWrite.gff&quot;</span>)
</code></pre></td>
</tr>
</table>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                <li>
                    <a href="/BioFSharp/BioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF3.html">GFF format</a></li>
                <li><a href="/BioFSharp/GenBank.html">GenBank</a></li>
                <li>
                    <a href="/BioFSharp/Clustal.html">Clustal</a>
                </li>        
                <li>
                    <a href="/BioFSharp/Newick.html">Newick</a>
                </li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/CSV.html">Delimited files</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                
                <li class="nav-header">Documentation</li>
                <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
