<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Introduction
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Introduction" class="anchor" href="#Introduction">Introduction</a></h1>
<p>After you have downloaded and set up BioFSharp (as described <a href="index.html#Setting-up-BioFSharp">here</a>), it is time to see it in action. Since we are programmers, let's say 'Hello World' !</p>
<p>The focus of this brief tutorial lies on introducing you to the BioFSharp workflow. We will build a protein from a raw string and look at some of it's properties.
Detailed information about the functions used can be either found in the specific tutorials in the sidebar or the API reference.</p>
<p>First, we reference and open BioFSharp.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="prep">#r</span><span class="s">&quot;BioFSharp.dll&quot;</span>

<span class="k">open</span> <span class="i">BioFSharp</span>
</code></pre></td>
</tr>
</table>
<h1><a name="The-Hello-World-Protein" class="anchor" href="#The-Hello-World-Protein">The Hello World Protein</a></h1>
<p>Time to say 'Hello World' the bioinformatician way. This is our raw string:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///string form of our hello world protein</span>
<span class="k">let</span> <span class="i">rawString</span> <span class="o">=</span> <span class="s">&quot;HELLOWORLD!&quot;</span>
</code></pre></td>
</tr>
</table>
<p>BioFSharp comes with various data structures for biological objects such as aminoacids. As you most likely know, you can abbreviate a
sequence of aminoacids with a three or oneletter code for each single aminoacid. We can convert a character to an aminoacid by using the
<code>charToOptionAminoAcid</code> function from the <code>BioItemsConverter</code> library. This will return us an option type, being either an aminoacid or <code>None</code>
if the caracter is not coding for an aminoacid.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">BioItemsConverter</span>

<span class="c">///valid character </span>
<span class="k">let</span> <span class="i">valid</span> <span class="o">=</span> <span class="i">OptionConverter</span><span class="o">.</span><span class="i">charToOptionAminoAcid</span> <span class="s">&#39;A&#39;</span>
<span class="c">///invalid character</span>
<span class="k">let</span> <span class="i">invalid</span> <span class="o">=</span> <span class="i">OptionConverter</span><span class="o">.</span><span class="i">charToOptionAminoAcid</span> <span class="s">&#39;?&#39;</span>
</code></pre></td>
</tr>
</table>
<p>Which results in 'A' being recognized as <code>Some Ala</code> and '?' as <code>None</code>.</p>
<p>To parse our entire string, we can use any of the BioCollections' <code>ofAminoAcidString</code> functions. For more information about BioSeq, BioList and BioArray go <a href="Biosequences.html">here</a></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Protein represented as a Bioseq</span>
<span class="k">let</span> <span class="i">parsedProtein1</span> <span class="o">=</span> <span class="i">rawString</span> <span class="o">|&gt;</span> <span class="i">BioSeq</span><span class="o">.</span><span class="i">ofAminoAcidString</span> 

<span class="c">///Protein represented as a BioList</span>
<span class="k">let</span> <span class="i">parsedProtein2</span> <span class="o">=</span> <span class="i">rawString</span> <span class="o">|&gt;</span> <span class="i">BioList</span><span class="o">.</span><span class="i">ofAminoAcidString</span> 

<span class="c">///Protein represented as a BioArray</span>
<span class="k">let</span> <span class="i">parsedProtein3</span> <span class="o">=</span> <span class="i">rawString</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span> 
</code></pre></td>
</tr>
</table>
<p>This yields us our Hello World protein. note that the '!' from the raw string is not contained in the sequence as it is not coding for an aminoacid.</p>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li>
                    <a href="/BioFSharp/IBioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/SeqProperties.html">Sequence Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF.html">GFF format</a></li>
                <li><a href="/BioFSharp/Genebank.html">Genebank</a></li>
                <li><a href="/BioFSharp/Clustal.html">Clustal</a></li>
                <li><a href="/BioFSharp/Newick.html">Newick</a></li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/GESA.html">Web logo</a></li>

                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
