<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Peptide retrieval by mass
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>

    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Peptide-retrieval-by-mass" class="anchor" href="#Peptide-retrieval-by-mass">Peptide retrieval by mass</a></h1>
<p>This part of the documentation aims to give you a brief overview of the functionality of SearchDB.fs. This Module contains functions<br />
to initialize a user configured peptide database and to subsequently perform common database operations such as inserts or lookUps.</p>
<h2><a name="Creating-a-peptide-database" class="anchor" href="#Creating-a-peptide-database">Creating a peptide database</a></h2>
<p>Prior to database creation the user has to specify the database parameters. The following examples will show this process in detail.
Some complex parameters will be discussed more specifically.</p>
<h2><a name="The-concept-of-peptide-modifications" class="anchor" href="#The-concept-of-peptide-modifications">The concept of peptide modifications</a></h2>
<p>In the library the term "Modification" is understood as literally any change to an amino acid formula. These changes can vary
from the exchange of an isotope (e.g. a N14 gets replaced with a N15) to an addition/deletion of chemical groups e.g. a phosphorylation.</p>
<h2><a name="Differentiation-of-fixed-variable-and-isotopic-modifications" class="anchor" href="#Differentiation-of-fixed-variable-and-isotopic-modifications">Differentiation of fixed, variable and isotopic modifications</a></h2>
<p>Creating a peptide database the user not only has to specify the modification itself but also its persistance. This is due to the fact
that the abundance of a modification can vary from one sample to another. This thought lead to a distinction between fixed, variable and
isotopic modifications during the construction of this library.</p>
<p>In a natural sample it is very unlikely for a modification like a phosphorylation to be present at every matching amino acid of every
peptide. The conclusion of this finding is that a database has to contain both, a modified and a unmodified version of the peptide simulating a
<strong>variable</strong> occurency of the modification coining the term <strong>"VariableModification"</strong>.</p>
<p>Contrary, further manipulations to the sample e.g. a selection of only N-terminally phosphorylated peptides can result in a sample
for which it can be assumed that roughly every peptide contains this modification. This would justify a creation of a peptide database with a
<strong>fixed</strong> N-terminal phosphorylation at every matching amino acid of every peptide. Hereby coining the term <strong>"FixedModification"</strong>.</p>
<hr />
<p>This expression constructs a new instance of the type SearchModification, which is needed to define Fixed- and VariableModifications
This type is inferred of the XMod ontology and the information stored in this type is needed to build instances of the type Modification.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
<span class="l">18: </span>
<span class="l">19: </span>
<span class="l">20: </span>
<span class="l">21: </span>
<span class="l">22: </span>
<span class="l">23: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="i">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="i">BioFSharp</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="i">Mz</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="i">ModificationInfo</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="i">AminoAcids</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="i">SearchDB</span>
<span class="c">/// Returns a instance of the type SearchModification</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="i">phosphorylation</span> <span class="o">=</span> {
    <span class="c">// Name of this SearchModification </span>
    <span class="i">Name</span><span class="o">=</span><span class="s">&quot;Phosphorylation&quot;</span> 
    <span class="c">// Xmod accession Number</span>
    <span class="i">Accession</span><span class="o">=</span><span class="s">&quot;21&quot;</span> 
    <span class="c">// A brief description of the SearchModification</span>
    <span class="i">Description</span><span class="o">=</span><span class="s">&quot;Addition of a phosphate group to the AA residue&quot;</span>
    <span class="c">// A string representation of the formula composition </span>
    <span class="i">Composition</span><span class="o">=</span> <span class="s">&quot;HO3P&quot;</span>
    <span class="c">// Specifices the ModLocation</span>
    <span class="i">Site</span><span class="o">=</span> [<span onmouseout="hideTip(event, 'fs7', 8)" onmouseover="showTip(event, 'fs7', 8)" class="p">Specific</span>(<span onmouseout="hideTip(event, 'fs8', 9)" onmouseover="showTip(event, 'fs8', 9)" class="p">Ser</span>,<span onmouseout="hideTip(event, 'fs9', 10)" onmouseover="showTip(event, 'fs9', 10)" class="t">ModLocation</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs10', 11)" onmouseover="showTip(event, 'fs10', 11)" class="i">Residual</span>); <span onmouseout="hideTip(event, 'fs7', 12)" onmouseover="showTip(event, 'fs7', 12)" class="p">Specific</span>(<span onmouseout="hideTip(event, 'fs11', 13)" onmouseover="showTip(event, 'fs11', 13)" class="p">Thr</span>,<span onmouseout="hideTip(event, 'fs9', 14)" onmouseover="showTip(event, 'fs9', 14)" class="t">ModLocation</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="i">Residual</span>); <span onmouseout="hideTip(event, 'fs7', 16)" onmouseover="showTip(event, 'fs7', 16)" class="p">Specific</span>(<span onmouseout="hideTip(event, 'fs12', 17)" onmouseover="showTip(event, 'fs12', 17)" class="p">Tyr</span>,<span onmouseout="hideTip(event, 'fs9', 18)" onmouseover="showTip(event, 'fs9', 18)" class="t">ModLocation</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs10', 19)" onmouseover="showTip(event, 'fs10', 19)" class="i">Residual</span>)]
    <span class="c">// Specifies if the Modification is additive or subtractive</span>
    <span class="i">MType</span><span class="o">=</span><span onmouseout="hideTip(event, 'fs13', 20)" onmouseover="showTip(event, 'fs13', 20)" class="t">SearchModType</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs14', 21)" onmouseover="showTip(event, 'fs14', 21)" class="p">Plus</span>
    <span class="c">// If a modified AminoAcid is shown via string representation this code is placed before</span>
    <span class="c">// the AminoAcid Symbol e.g. &quot;phS&quot;</span>
    <span class="i">XModCode</span><span class="o">=</span> <span class="s">&quot;ph&quot;</span>
    } 
</code></pre></td>
</tr>
</table>
<p>As mentioned before, there is a need to model modifications that affect the peptide sequences on a <strong>isotopic</strong> level. This could be understood
as a special kind of FixedModification with the additional constrain that there is no addition or deletion of a chemical group but a isotopic
modification affecting every amino acid. e.g. a exchange of every N14 to a N15. As one could expect, these features coined the term "IsotopicModification".</p>
<p>This expression returns a instance of the type SearchInfoIsotopic, which is needed to define IsotopicModifications. This type is needed if the user
plans to introduce a IsotopicModifications e.g a replacement of the N14 isotopes with heavier N15 isotopes.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns a instance of the type SearchInfoIsotopic </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 22)" onmouseover="showTip(event, 'fs15', 22)" class="i">isoMod</span> <span class="o">=</span> (<span onmouseout="hideTip(event, 'fs5', 23)" onmouseover="showTip(event, 'fs5', 23)" class="t">SearchDB</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs16', 24)" onmouseover="showTip(event, 'fs16', 24)" class="f">createSearchInfoIsotopic</span> <span class="s">&quot;N15&quot;</span> <span onmouseout="hideTip(event, 'fs17', 25)" onmouseover="showTip(event, 'fs17', 25)" class="t">Elements</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs18', 26)" onmouseover="showTip(event, 'fs18', 26)" class="t">Table</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs19', 27)" onmouseover="showTip(event, 'fs19', 27)" class="i">N</span> <span onmouseout="hideTip(event, 'fs17', 28)" onmouseover="showTip(event, 'fs17', 28)" class="t">Elements</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs18', 29)" onmouseover="showTip(event, 'fs18', 29)" class="t">Table</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs20', 30)" onmouseover="showTip(event, 'fs20', 30)" class="t">Heavy</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs21', 31)" onmouseover="showTip(event, 'fs21', 31)" class="i">N15</span>)
</code></pre></td>
</tr>
</table>
<h2><a name="Selecting-and-initializing-a-mass-function" class="anchor" href="#Selecting-and-initializing-a-mass-function">Selecting and initializing a mass function</a></h2>
<p>The following expression returns a instance of a massfunction. The user can choose between functions that either return
the monoisotopic or the average mass of types that inherit the IBioItem interface (e.g. AminoAcids).<br />
Moreover, each function exists in a "WhitMemP" version that uses a dictionary to "remember" (store) already computed masses.
This approach is especially useful when the user aims to do many mass computations subsequently, as it is the case when
building a new database. The usage of a dictionary is the reason for this function to be initialized prior to database
creation.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Returns a instance of the massfunction BioItem.initMonoisoMassWithMemP</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs22', 32)" onmouseover="showTip(event, 'fs22', 32)" class="f">massF</span> <span class="o">:</span> (<span onmouseout="hideTip(event, 'fs23', 33)" onmouseover="showTip(event, 'fs23', 33)" class="t">IBioItem</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs24', 34)" onmouseover="showTip(event, 'fs24', 34)" class="t">float</span>) <span class="o">=</span>  <span onmouseout="hideTip(event, 'fs25', 35)" onmouseover="showTip(event, 'fs25', 35)" class="t">BioItem</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs26', 36)" onmouseover="showTip(event, 'fs26', 36)" class="f">initMonoisoMassWithMemP</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Incorporation-of-all-parameters-in-a-single-record-type" class="anchor" href="#Incorporation-of-all-parameters-in-a-single-record-type">Incorporation of all parameters in a single record type</a></h2>
<p>This type contains all formerly designed Parameters and additional parameters.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
<span class="l">18: </span>
<span class="l">19: </span>
<span class="l">20: </span>
<span class="l">21: </span>
<span class="l">22: </span>
<span class="l">23: </span>
<span class="l">24: </span>
<span class="l">25: </span>
<span class="l">26: </span>
<span class="l">27: </span>
<span class="l">28: </span>
<span class="l">29: </span>
<span class="l">30: </span>
<span class="l">31: </span>
<span class="l">32: </span>
<span class="l">33: </span>
<span class="l">34: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns a instance of the type SearchDbParams</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 37)" onmouseover="showTip(event, 'fs27', 37)" class="i">paramTestN15</span> <span class="o">=</span> {
        <span class="i">Name</span><span class="o">=</span><span class="s">&quot;Creinhardtii_CpN15&quot;</span>
        <span class="c">// Path of db storage folder</span>
        <span class="i">DbFolder</span>            <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/&quot;</span>)
        <span class="c">// Path of the user selected fasta file  </span>
        <span class="i">FastaPath</span>           <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Chlamy_Cp.fastA&quot;</span>)
        <span class="c">// Function that specifies the conversion of the Fasta Header to the protein name inserted in the database</span>
        <span class="i">FastaHeaderToName</span>   <span class="o">=</span> <span onmouseout="hideTip(event, 'fs28', 38)" onmouseover="showTip(event, 'fs28', 38)" class="f">id</span>
        <span class="c">// Protease that is used during in silico digestion of the proteins</span>
        <span class="i">Protease</span>            <span class="o">=</span> <span onmouseout="hideTip(event, 'fs29', 39)" onmouseover="showTip(event, 'fs29', 39)" class="t">Digestion</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs30', 40)" onmouseover="showTip(event, 'fs30', 40)" class="t">Table</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs31', 41)" onmouseover="showTip(event, 'fs31', 41)" class="f">getProteaseBy</span> <span class="s">&quot;Trypsin&quot;</span>
        <span class="c">// Minimum number of missed cleavages</span>
        <span class="i">MinMissedCleavages</span>  <span class="o">=</span> <span class="n">0</span>
        <span class="c">// Maximum number of missed cleavages</span>
        <span class="i">MaxMissedCleavages</span>  <span class="o">=</span> <span class="n">3</span>
        <span class="c">// Maximum peptide mass inserted in the database</span>
        <span class="i">MaxMass</span>             <span class="o">=</span> <span class="n">15000.</span>
        <span class="c">// Mininmum peptide length </span>
        <span class="i">MinPepLength</span>        <span class="o">=</span> <span class="n">4</span>
        <span class="c">// Maximum pepide length</span>
        <span class="i">MaxPepLength</span>        <span class="o">=</span> <span class="n">65</span>
        <span class="c">// Contains a instance of SearchInfoIsotopic, set to None if no isotopic modification shall be introduced</span>
        <span class="i">IsotopicMod</span>           <span class="o">=</span> [<span onmouseout="hideTip(event, 'fs15', 42)" onmouseover="showTip(event, 'fs15', 42)" class="i">isoMod</span>] 
        <span class="c">// MassMode used during db creation</span>
        <span onmouseout="hideTip(event, 'fs32', 43)" onmouseover="showTip(event, 'fs32', 43)" class="i">MassMode</span>            <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 44)" onmouseover="showTip(event, 'fs5', 44)" class="t">SearchDB</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs32', 45)" onmouseover="showTip(event, 'fs32', 45)" class="t">MassMode</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs33', 46)" onmouseover="showTip(event, 'fs33', 46)" class="p">Monoisotopic</span>
        <span class="c">// Massfunction used during db creation</span>
        <span class="i">MassFunction</span>        <span class="o">=</span> <span onmouseout="hideTip(event, 'fs22', 47)" onmouseover="showTip(event, 'fs22', 47)" class="f">massF</span>
        <span class="c">// List of &quot;Fixedmodifications&quot;. </span>
        <span class="i">FixedMods</span>           <span class="o">=</span> [<span onmouseout="hideTip(event, 'fs6', 48)" onmouseover="showTip(event, 'fs6', 48)" class="i">phosphorylation</span>]
        <span class="c">// List of &quot;VariableModificaions&quot;.  </span>
        <span class="i">VariableMods</span>        <span class="o">=</span> []
        <span class="c">// Maximum number of variable modifications per peptide this constrain is used to limit the computational </span>
        <span class="i">VarModThreshold</span>     <span class="o">=</span> <span class="n">5</span>
        }
</code></pre></td>
</tr>
</table>
<p>The following example illustrates how a database lookUp can be initialized. The function SearchDB.getPeptideLookUpBy
takes the defined SearchDbParams and returns a function that takes two float values (representing masses) as input
parameters and returns a list of peptides each with a mass in the range of first to the second mass used as input
parameters.</p>
<p>SearchDB.getPeptideLookUpBy first checks if a database with the given parameters already exists before it starts the
database creation process. If there is a database matching the user given SearchParameters the process will be finished
nearly immediatly. If a new database has to be created the computation time will depend on the complexity of the given
parameters such as the size of the FASTA file, the number of given Modifications, the maxPepLength et cetera.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns a function that takes two masses as input parameters and returns a list of peptides (wrapped in the type LookUpResult)</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs34', 49)" onmouseover="showTip(event, 'fs34', 49)" class="f">n15DBLookUpBy</span> <span class="o">=</span>  
        <span onmouseout="hideTip(event, 'fs5', 50)" onmouseover="showTip(event, 'fs5', 50)" class="t">SearchDB</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs35', 51)" onmouseover="showTip(event, 'fs35', 51)" class="f">getPeptideLookUpBy</span> <span onmouseout="hideTip(event, 'fs27', 52)" onmouseover="showTip(event, 'fs27', 52)" class="i">paramTestN15</span>   
         
</code></pre></td>
</tr>
</table>
<p>This example shows how the pattern used before can be modified slightly to return a function that only needs one parameter as
input to return a LookUpResult list.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns a function that takes one inputMass and returns a list of peptides (wrapped in the type LookUpResult) in the range </span>
<span class="c">/// of the inputMass-0.5 to inputMass+0.5 </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs36', 53)" onmouseover="showTip(event, 'fs36', 53)" class="f">n15LookUpPeptideBy</span> <span onmouseout="hideTip(event, 'fs37', 54)" onmouseover="showTip(event, 'fs37', 54)" class="i">inputMass</span> <span class="o">=</span>
        <span class="k">let</span> <span onmouseout="hideTip(event, 'fs38', 55)" onmouseover="showTip(event, 'fs38', 55)" class="i">lowerBorder</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 56)" onmouseover="showTip(event, 'fs37', 56)" class="i">inputMass</span><span class="o">-</span><span class="n">0.5</span>
        <span class="k">let</span> <span onmouseout="hideTip(event, 'fs39', 57)" onmouseover="showTip(event, 'fs39', 57)" class="i">upperBorder</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 58)" onmouseover="showTip(event, 'fs37', 58)" class="i">inputMass</span><span class="o">+</span><span class="n">0.5</span> 
        <span onmouseout="hideTip(event, 'fs34', 59)" onmouseover="showTip(event, 'fs34', 59)" class="f">n15DBLookUpBy</span> <span onmouseout="hideTip(event, 'fs38', 60)" onmouseover="showTip(event, 'fs38', 60)" class="i">lowerBorder</span> <span onmouseout="hideTip(event, 'fs39', 61)" onmouseover="showTip(event, 'fs39', 61)" class="i">upperBorder</span>
</code></pre></td>
</tr>
</table>
<p>This function can be used as shown in the code snippet below. As a inputMass the monoisotopic mass of the peptide <em>ANLGMEVMHER</em> can be used as a input.
This peptide is encoded by a protein which can be found in the FASTA used for the database creation. Surprisingly the list you obtain will be empty because
the database does not contain any peptides in the given massRange.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// PeptideMass of the peptide known from the peptide *ANLGMEVMHER*. </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs40', 62)" onmouseover="showTip(event, 'fs40', 62)" class="i">peptideMass</span> <span class="o">=</span> <span class="n">1285.591258</span>

<span class="c">/// Returns the result of the peptide lookup </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs41', 63)" onmouseover="showTip(event, 'fs41', 63)" class="i">n15LookUpResult</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs36', 64)" onmouseover="showTip(event, 'fs36', 64)" class="f">n15LookUpPeptideBy</span> <span onmouseout="hideTip(event, 'fs40', 65)" onmouseover="showTip(event, 'fs40', 65)" class="i">peptideMass</span>
</code></pre></td>
</tr>
</table>
<p>The possibility that the database is empty can be ruled out by performing a lookUp with the mass 1000., which should return a list filled with
items of the type LookUpResult. The explanation to this behaviour can be found examining the SearchDbParameters bound to the name <strong>paramTestN15</strong>.
When looking at the parameters we see that we introduced a IsotopicModification. If the peptide was optained from a sample without this isotopic
modification, a second lookUp in a database lacking IsotopicModification might return a result list containing the peptide of interest.</p>
<p>As a first step a new instance of the type SearchDbParams has to be instantiated and a database has to be created:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
<span class="l">18: </span>
<span class="l">19: </span>
<span class="l">20: </span>
<span class="l">21: </span>
<span class="l">22: </span>
<span class="l">23: </span>
<span class="l">24: </span>
<span class="l">25: </span>
<span class="l">26: </span>
<span class="l">27: </span>
<span class="l">28: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns a instance of the type SearchDbParams</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs42', 66)" onmouseover="showTip(event, 'fs42', 66)" class="i">paramTestN14</span> <span class="o">=</span> {
        <span class="i">Name</span><span class="o">=</span><span class="s">&quot;Creinhardtii_CpN14UnMod&quot;</span>
        <span class="i">DbFolder</span>            <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/&quot;</span>)
        <span class="i">FastaPath</span>           <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Chlamy_Cp.fastA&quot;</span>)
        <span class="i">FastaHeaderToName</span>   <span class="o">=</span> <span onmouseout="hideTip(event, 'fs28', 67)" onmouseover="showTip(event, 'fs28', 67)" class="f">id</span>
        <span class="i">Protease</span>            <span class="o">=</span> <span onmouseout="hideTip(event, 'fs29', 68)" onmouseover="showTip(event, 'fs29', 68)" class="t">Digestion</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs30', 69)" onmouseover="showTip(event, 'fs30', 69)" class="t">Table</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs31', 70)" onmouseover="showTip(event, 'fs31', 70)" class="f">getProteaseBy</span> <span class="s">&quot;Trypsin&quot;</span>
        <span class="i">MinMissedCleavages</span>  <span class="o">=</span> <span class="n">0</span>
        <span class="i">MaxMissedCleavages</span>  <span class="o">=</span> <span class="n">3</span>
        <span class="i">MaxMass</span>             <span class="o">=</span> <span class="n">15000.</span>
        <span class="i">MinPepLength</span>        <span class="o">=</span> <span class="n">4</span>
        <span class="i">MaxPepLength</span>        <span class="o">=</span> <span class="n">65</span>
        <span class="i">IsotopicMod</span>           <span class="o">=</span> []
        <span onmouseout="hideTip(event, 'fs32', 71)" onmouseover="showTip(event, 'fs32', 71)" class="i">MassMode</span>            <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 72)" onmouseover="showTip(event, 'fs5', 72)" class="t">SearchDB</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs32', 73)" onmouseover="showTip(event, 'fs32', 73)" class="t">MassMode</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs33', 74)" onmouseover="showTip(event, 'fs33', 74)" class="p">Monoisotopic</span>
        <span class="i">MassFunction</span>        <span class="o">=</span> <span onmouseout="hideTip(event, 'fs22', 75)" onmouseover="showTip(event, 'fs22', 75)" class="f">massF</span>
        <span class="i">FixedMods</span>           <span class="o">=</span> [] 
        <span class="i">VariableMods</span>        <span class="o">=</span> []
        <span class="i">VarModThreshold</span>     <span class="o">=</span> <span class="n">5</span>
        }
<span class="c">/// Returns a function that takes two masses as input parameters and returns a list of peptides (wrapped in the type LookUpResult)</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs43', 76)" onmouseover="showTip(event, 'fs43', 76)" class="f">n14DBLookUpBy</span> <span class="o">=</span>  
        <span onmouseout="hideTip(event, 'fs5', 77)" onmouseover="showTip(event, 'fs5', 77)" class="t">SearchDB</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs35', 78)" onmouseover="showTip(event, 'fs35', 78)" class="f">getPeptideLookUpBy</span> <span onmouseout="hideTip(event, 'fs42', 79)" onmouseover="showTip(event, 'fs42', 79)" class="i">paramTestN14</span>   
<span class="c">/// Returns a function that takes one inputMass and returns a list of peptides (wrapped in the type LookUpResult) in the range </span>
<span class="c">/// of the inputMass-0.5 to inputMass+0.5 </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 80)" onmouseover="showTip(event, 'fs44', 80)" class="f">n14LookUpPeptideBy</span> <span onmouseout="hideTip(event, 'fs37', 81)" onmouseover="showTip(event, 'fs37', 81)" class="i">inputMass</span> <span class="o">=</span>
        <span class="k">let</span> <span onmouseout="hideTip(event, 'fs38', 82)" onmouseover="showTip(event, 'fs38', 82)" class="i">lowerBorder</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 83)" onmouseover="showTip(event, 'fs37', 83)" class="i">inputMass</span><span class="o">-</span><span class="n">0.5</span>
        <span class="k">let</span> <span onmouseout="hideTip(event, 'fs39', 84)" onmouseover="showTip(event, 'fs39', 84)" class="i">upperBorder</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs37', 85)" onmouseover="showTip(event, 'fs37', 85)" class="i">inputMass</span><span class="o">+</span><span class="n">0.5</span> 
        <span onmouseout="hideTip(event, 'fs43', 86)" onmouseover="showTip(event, 'fs43', 86)" class="f">n14DBLookUpBy</span> <span onmouseout="hideTip(event, 'fs38', 87)" onmouseover="showTip(event, 'fs38', 87)" class="i">lowerBorder</span> <span onmouseout="hideTip(event, 'fs39', 88)" onmouseover="showTip(event, 'fs39', 88)" class="i">upperBorder</span>
</code></pre></td>
</tr>
</table>
<p>After these steps the following lookup will result in list of lookupResults containing the peptide "ANLGMEVMHER".</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// Returns the result of the peptide lookup</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs45', 89)" onmouseover="showTip(event, 'fs45', 89)" class="i">n14LookUpResult</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs44', 90)" onmouseover="showTip(event, 'fs44', 90)" class="f">n14LookUpPeptideBy</span> <span onmouseout="hideTip(event, 'fs40', 91)" onmouseover="showTip(event, 'fs40', 91)" class="i">peptideMass</span>
</code></pre></td>
</tr>
</table>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace BioFSharp.Mz</div>
<div class="tip" id="fs3">module ModificationInfo<br /><br />from BioFSharp</div>
<div class="tip" id="fs4">module AminoAcids<br /><br />from BioFSharp</div>
<div class="tip" id="fs5">module SearchDB<br /><br />from BioFSharp.Mz</div>
<div class="tip" id="fs6">val phosphorylation : SearchModification<br /><br />Full name: PeptideLookUp.phosphorylation<br /><em><br /><br />&#160;Returns a instance of the type SearchModification</em></div>
<div class="tip" id="fs7">union case SearchModSite.Specific: AminoAcid * ModLocation -&gt; SearchModSite</div>
<div class="tip" id="fs8">union case AminoAcid.Ser: AminoAcid</div>
<div class="tip" id="fs9">type ModLocation =<br />&#160;&#160;|  Residual  =  0<br />&#160;&#160;|  Cterm  =  1<br />&#160;&#160;|  Nterm  =  2<br />&#160;&#160;|  ProteinCterm  =  3<br />&#160;&#160;|  ProteinNterm  =  4<br />&#160;&#160;|  Isotopic  =  5<br /><br />Full name: BioFSharp.ModificationInfo.ModLocation</div>
<div class="tip" id="fs10">ModLocation.Residual: ModLocation = 0</div>
<div class="tip" id="fs11">union case AminoAcid.Thr: AminoAcid</div>
<div class="tip" id="fs12">union case AminoAcid.Tyr: AminoAcid</div>
<div class="tip" id="fs13">type SearchModType =<br />&#160;&#160;| Minus<br />&#160;&#160;| Plus<br /><br />Full name: BioFSharp.Mz.SearchDB.SearchModType</div>
<div class="tip" id="fs14">union case SearchModType.Plus: SearchModType</div>
<div class="tip" id="fs15">val isoMod : SearchInfoIsotopic<br /><br />Full name: PeptideLookUp.isoMod<br /><em><br /><br />&#160;Returns a instance of the type SearchInfoIsotopic </em></div>
<div class="tip" id="fs16">val createSearchInfoIsotopic : name:string -&gt; sourceEl:Elements.Element -&gt; targetEl:Elements.Element -&gt; SearchInfoIsotopic<br /><br />Full name: BioFSharp.Mz.SearchDB.createSearchInfoIsotopic</div>
<div class="tip" id="fs17">module Elements<br /><br />from BioFSharp</div>
<div class="tip" id="fs18">module Table<br /><br />from BioFSharp.Elements</div>
<div class="tip" id="fs19">val N : Elements.Element<br /><br />Full name: BioFSharp.Elements.Table.N</div>
<div class="tip" id="fs20">module Heavy<br /><br />from BioFSharp.Elements.Table</div>
<div class="tip" id="fs21">val N15 : Elements.Element<br /><br />Full name: BioFSharp.Elements.Table.Heavy.N15</div>
<div class="tip" id="fs22">val massF : (IBioItem -&gt; float)<br /><br />Full name: PeptideLookUp.massF<br /><em><br /><br />Returns a instance of the massfunction BioItem.initMonoisoMassWithMemP</em></div>
<div class="tip" id="fs23">type IBioItem =<br />&#160;&#160;interface<br />&#160;&#160;&#160;&#160;abstract member Formula : Formula<br />&#160;&#160;&#160;&#160;abstract member Name : string<br />&#160;&#160;&#160;&#160;abstract member Symbol : char<br />&#160;&#160;&#160;&#160;abstract member isGap : bool<br />&#160;&#160;&#160;&#160;abstract member isTerminator : bool<br />&#160;&#160;end<br /><br />Full name: BioFSharp.IBioItem</div>
<div class="tip" id="fs24">Multiple items<br />val float : value:&#39;T -&gt; float (requires member op_Explicit)<br /><br />Full name: Microsoft.FSharp.Core.Operators.float<br /><br />--------------------<br />type float = System.Double<br /><br />Full name: Microsoft.FSharp.Core.float<br /><br />--------------------<br />type float&lt;&#39;Measure&gt; = float<br /><br />Full name: Microsoft.FSharp.Core.float&lt;_&gt;</div>
<div class="tip" id="fs25">module BioItem<br /><br />from BioFSharp</div>
<div class="tip" id="fs26">val initMonoisoMassWithMemP&lt;&#39;a (requires &#39;a :&gt; IBioItem)&gt; : (IBioItem -&gt; float)<br /><br />Full name: BioFSharp.BioItem.initMonoisoMassWithMemP</div>
<div class="tip" id="fs27">val paramTestN15 : SearchDbParams<br /><br />Full name: PeptideLookUp.paramTestN15<br /><em><br /><br />&#160;Returns a instance of the type SearchDbParams</em></div>
<div class="tip" id="fs28">val id : x:&#39;T -&gt; &#39;T<br /><br />Full name: Microsoft.FSharp.Core.Operators.id</div>
<div class="tip" id="fs29">module Digestion<br /><br />from BioFSharp</div>
<div class="tip" id="fs30">module Table<br /><br />from BioFSharp.Digestion</div>
<div class="tip" id="fs31">val getProteaseBy : name:string -&gt; Digestion.Protease<br /><br />Full name: BioFSharp.Digestion.Table.getProteaseBy</div>
<div class="tip" id="fs32">type MassMode =<br />&#160;&#160;| Average<br />&#160;&#160;| Monoisotopic<br />&#160;&#160;override ToString : unit -&gt; string<br /><br />Full name: BioFSharp.Mz.SearchDB.MassMode</div>
<div class="tip" id="fs33">union case MassMode.Monoisotopic: MassMode</div>
<div class="tip" id="fs34">val n15DBLookUpBy : (float -&gt; float -&gt; LookUpResult&lt;AminoAcid&gt; list)<br /><br />Full name: PeptideLookUp.n15DBLookUpBy<br /><em><br /><br />&#160;Returns a function that takes two masses as input parameters and returns a list of peptides (wrapped in the type LookUpResult)</em></div>
<div class="tip" id="fs35">val getPeptideLookUpBy : sdbParams:SearchDbParams -&gt; (float -&gt; float -&gt; LookUpResult&lt;AminoAcid&gt; list)<br /><br />Full name: BioFSharp.Mz.SearchDB.getPeptideLookUpBy</div>
<div class="tip" id="fs36">val n15LookUpPeptideBy : inputMass:float -&gt; LookUpResult&lt;AminoAcid&gt; list<br /><br />Full name: PeptideLookUp.n15LookUpPeptideBy<br /><em><br /><br />&#160;Returns a function that takes one inputMass and returns a list of peptides (wrapped in the type LookUpResult) in the range <br />&#160;of the inputMass-0.5 to inputMass+0.5 </em></div>
<div class="tip" id="fs37">val inputMass : float</div>
<div class="tip" id="fs38">val lowerBorder : float</div>
<div class="tip" id="fs39">val upperBorder : float</div>
<div class="tip" id="fs40">val peptideMass : float<br /><br />Full name: PeptideLookUp.peptideMass<br /><em><br /><br />&#160;PeptideMass of the peptide known from the peptide *ANLGMEVMHER*. </em></div>
<div class="tip" id="fs41">val n15LookUpResult : LookUpResult&lt;AminoAcid&gt; list<br /><br />Full name: PeptideLookUp.n15LookUpResult<br /><em><br /><br />&#160;Returns the result of the peptide lookup </em></div>
<div class="tip" id="fs42">val paramTestN14 : SearchDbParams<br /><br />Full name: PeptideLookUp.paramTestN14<br /><em><br /><br />&#160;Returns a instance of the type SearchDbParams</em></div>
<div class="tip" id="fs43">val n14DBLookUpBy : (float -&gt; float -&gt; LookUpResult&lt;AminoAcid&gt; list)<br /><br />Full name: PeptideLookUp.n14DBLookUpBy<br /><em><br /><br />&#160;Returns a function that takes two masses as input parameters and returns a list of peptides (wrapped in the type LookUpResult)</em></div>
<div class="tip" id="fs44">val n14LookUpPeptideBy : inputMass:float -&gt; LookUpResult&lt;AminoAcid&gt; list<br /><br />Full name: PeptideLookUp.n14LookUpPeptideBy<br /><em><br /><br />&#160;Returns a function that takes one inputMass and returns a list of peptides (wrapped in the type LookUpResult) in the range <br />&#160;of the inputMass-0.5 to inputMass+0.5 </em></div>
<div class="tip" id="fs45">val n14LookUpResult : LookUpResult&lt;AminoAcid&gt; list<br /><br />Full name: PeptideLookUp.n14LookUpResult<br /><em><br /><br />&#160;Returns the result of the peptide lookup</em></div>

        </div>
        <div class="span3">
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
          <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
            <li class="nav-header">BioFSharp</li>
            <li><a href="/BioFSharp/index.html">Home page</a></li>
            <li class="divider"></li>
            <li><a href="http://nuget.org/packages/BioFSharp">Get Library via NuGet</a></li>
            <li><a href="http://github.com/CSBiology/BioFSharp">Source Code on GitHub</a></li>
            <li><a href="/BioFSharp/license.html">License</a></li>
            <li><a href="/BioFSharp/release-notes.html">Release Notes</a></li>
            
            <li class="nav-header">Getting started</li>
              
              <li><a href="/BioFSharp/BioSequence.html">BioSequence</a></li>
              <li class="divider"></li>
              <li><a href="/BioFSharp/CentroidSpectra.html">Spectrum centroidization</a></li>
              <li><a href="/BioFSharp/ChargeStateDetermination.html">Charge state determination</a></li>
              <li><a href="/BioFSharp/PeptideLookUp.html">Peptide look up</a></li>
              <li><a href="/BioFSharp/PeptideIdentification.html">Peptide identification</a></li>
              <li class="divider"></li>
              <li><a href="/BioFSharp/BlastWrapper.html">Use NCBI-Blast</a></li>
            <li class="nav-header">Documentation</li>
            <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>
          </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
