<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Using Bio Format Readers
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Using-Bio-Format-Readers" class="anchor" href="#Using-Bio-Format-Readers">Using Bio Format Readers</a></h1>
<p>BioFSharp contains a set of readers for different biology-associated file formats. This documentation aims to give an introduction for them.</p>
<a name="FastQ"></a>
<h2><a name="FastQ" class="anchor" href="#FastQ">FastQ</a></h2>
<br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/FastQ.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/MikhayN">&#128366;view author of this tutorial</a>
<br><br>
<p>This module allows to parse FASTQ format data with original 4-lines entries into this record type</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// FastqItem record contains header, sequence, qualityheader, qualitysequence of one entry</span>
<span class="k">type</span> <span class="i">FastqItem</span><span class="o">&lt;</span><span class="o">&#39;</span><span class="i">a</span>,<span class="o">&#39;</span><span class="i">b</span><span class="o">&gt;</span> <span class="o">=</span> {
    <span class="i">Header</span>          <span class="o">:</span> <span class="i">string</span>
    <span class="i">Sequence</span>        <span class="o">:</span> <span class="o">&#39;</span><span class="i">a</span>
    <span class="i">QualityHeader</span>   <span class="o">:</span> <span class="i">string</span>
    <span class="i">QualitySequence</span> <span class="o">:</span> <span class="o">&#39;</span><span class="i">b</span>      
}
</code></pre></td>
</tr>
</table>
<p>To be able to use this parser you need to define two converter functions,
one example for each you can also find in our module, but you also may need to write your own.</p>
<p>We can convert sequence string to predefined option type of Amino Acids, using converter function
from our library 'BioFSharp.BioItemsConverter.OptionConverter'</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// get characters as sequence units</span>
<span class="k">let</span> <span class="i">converterToAA</span> <span class="i">string</span> <span class="o">=</span>
    <span class="i">string</span>
    <span class="o">|&gt;</span> <span class="i">String</span><span class="o">.</span><span class="i">toCharArray</span>
    <span class="o">|&gt;</span> <span class="i">Array</span><span class="o">.</span><span class="i">map</span> (<span class="i">BioFSharp</span><span class="o">.</span><span class="i">BioItemsConverter</span><span class="o">.</span><span class="i">OptionConverter</span><span class="o">.</span><span class="i">charToOptionAminoAcid</span>)
</code></pre></td>
</tr>
</table>
<p>If you have following possible values for quality sequence:
'!""#$%&amp;'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`abcdefghijklmnopqrstuvwxyz{|}~',
with Sanger format, that can encode a Phred quality score from 0 to 93 using ASCII 33 to 126,
then you can use our converting function:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// get Phred quality score</span>
<span class="k">let</span> <span class="i">qualityConvertFn</span> <span class="i">string</span> <span class="o">=</span>
    <span class="i">string</span>
    <span class="o">|&gt;</span> <span class="i">String</span><span class="o">.</span><span class="i">toCharArray</span>
    <span class="o">|&gt;</span> <span class="i">Array</span><span class="o">.</span><span class="i">map</span> (<span class="k">fun</span> <span class="i">i</span> <span class="k">-&gt;</span> <span class="i">i</span><span class="o">.</span><span class="i">GetHashCode</span>()<span class="o">-</span><span class="n">33</span>)
</code></pre></td>
</tr>
</table>
<p>And then you can easily use this module to read your FastQ file</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">yourFastqFile</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/FastQtest.fastq&quot;</span>)

<span class="k">let</span> <span class="i">FastQSequence</span> <span class="o">=</span> 
    <span class="i">FastQ</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">converterToAA</span> <span class="i">qualityConvertFn</span> <span class="i">yourFastqFile</span>
</code></pre></td>
</tr>
</table>
<hr>
<a name="FastA"></a>
<h2><a name="Introducing-FASTA-format" class="anchor" href="#Introducing-FASTA-format">Introducing FASTA format</a></h2>
<br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/FastA.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/bvenn">&#128366;view author of this tutorial</a>
<br><br>
<p>The FASTA format can be used to represent sequences of amino acids or nucleotides written in single-letter code
(for further description of single-letter codes visit <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-nucleotides-nucleotide.html">Nucleotides</a> or <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-aminoacids-aminoacid.html">AminoAcids</a> respectively).
One sequence constists of two parts: The first line (Header) starting with a "&gt;" is followed by a sequence identification code which should represent a unique description of the sequence.
Subsequent lines contain the sequence itself, which is separated into chunks of 60 to 80 characters per line.
For further information about the format please visit <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=BlastHelp">NCBI - FASTA</a>.</p>
<p>Sequences in FASTA format are implemented as a record type containing the following fields:
<br><br></p>
<div id="responsiveTable">
<table>
<thead>
<tr class="header">
<th><p>Field name</p></th>
<th><p>type</p></th>
<th><p>Description</p></th>
<th><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>Header</p></td>
<td><p><code>string</code></p></td>
<td><p>sequence identification code</p></td>
<td><p>"gi&amp;#124;0000&amp;#124;gb&amp;#124;accession&amp;#124;0001&amp;#124; example sequence"</p></td>
</tr>
<tr class="even">
<td><p>Sequence</p></td>
<td><p><code>'a</code></p></td>
<td><p>Sequence of nucleotides or aminoacids in type of choice</p></td>
<td><p>BioArray.BioArray<AminoAcids.AminoAcid></p></td>
</tr>
</tbody>
</table>

</div>
<br>>
<h2><a name="Creating-an-FastaItem" class="anchor" href="#Creating-an-FastaItem">Creating an FastaItem</a></h2>
<p>To create a FastaItem type you directly can use the provided <code>create</code> function.
<br></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span><span class="o">.</span><span class="i">FastA</span>

<span class="c">//creates a FastaItem out of a header and a sequence in given type</span>
<span class="k">let</span> <span class="i">exampleFasta</span> <span class="o">=</span> <span class="i">FastA</span><span class="o">.</span><span class="i">createFastaItem</span> <span class="s">&quot;gi|0000|gb|accession|0001| example sequence&quot;</span> ([|<span class="s">&#39;M&#39;</span>;<span class="s">&#39;F&#39;</span>;<span class="s">&#39;H&#39;</span>;<span class="s">&#39;F&#39;</span>;<span class="s">&#39;I&#39;</span>;<span class="s">&#39;L&#39;</span>;<span class="s">&#39;P&#39;</span>;<span class="s">&#39;F&#39;</span>;<span class="s">&#39;T&#39;</span>|] <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>)
</code></pre></td>
</tr>
</table>
<br>
<h2><a name="Reading-FastaItem" class="anchor" href="#Reading-FastaItem">Reading FastaItem</a></h2>
<p>You can convert the sequence of <code>chars</code> to predefined sequences of <code>AminoAcids</code> or <code>Nucleotides</code>, using converter function
from our library <a href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/BioList.fs">BioFSharp.BioList</a>,
<a href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/BioArray.fs">BioFSharp.BioArray</a>
or <a href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/BioSeq.fs">BioFSharp.BioSeq</a>.</p>
<h4><a name="There-are-multiple-options-to-read-in-a-FASTA-file" class="anchor" href="#There-are-multiple-options-to-read-in-a-FASTA-file">There are multiple options to read in a FASTA file:</a></h4>
<ol>
<li>You can read in the <code>.fasta</code> with the path of the file.</li>
</ol>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">filepathFASTA</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/FASTAExample1.fasta&quot;</span>)

<span class="c">//reads from file to an array of FastaItems.</span>
<span class="k">let</span> <span class="i">exampleFromFile</span> <span class="o">=</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span> <span class="i">filepathFASTA</span>
</code></pre></td>
</tr>
</table>
<br>
2. If the FASTA file is compressed in a gzip file you can read it using the same parameters as above and the following function.
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">filepathGzipFASTA</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/FASTAExample2.gz&quot;</span>)

<span class="c">//reads from gzip file to an List of FastaItems.</span>
<span class="k">let</span> <span class="i">exampleFromGzipFile</span>  <span class="o">=</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromGzipFile</span> <span class="i">BioList</span><span class="o">.</span><span class="i">ofAminoAcidString</span> <span class="i">filepathGzipFASTA</span>
</code></pre></td>
</tr>
</table>
<br>
3. To convert `seq<string>` to `seq<FastaItem>` you can use the following function.
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">FSharp</span><span class="o">.</span><span class="i">Care</span><span class="o">.</span><span class="i">IO</span>

<span class="k">let</span> <span class="i">stringSequence</span> <span class="o">=</span> <span class="i">FileIO</span><span class="o">.</span><span class="i">readFile</span>(<span class="i">filepathFASTA</span>)

<span class="c">//reads seq&lt;string&gt; to sequence of FastaItems</span>
<span class="k">let</span> <span class="i">exampleFromFileEnumerator</span> <span class="o">=</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFileEnumerator</span> <span class="i">BioSeq</span><span class="o">.</span><span class="i">ofAminoAcidString</span> <span class="i">stringSequence</span>
</code></pre></td>
</tr>
</table>
<br>
To write a `seq<FastaItem>` into a file use the following function.
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">seqFastaItem</span> <span class="o">=</span> <span class="i">seq</span> [<span class="i">exampleFasta</span>]

<span class="c">//writes a sequence of FastaItems into a .fasta file</span>
<span class="i">FastA</span><span class="o">.</span><span class="i">write</span> <span class="i">BioItem</span><span class="o">.</span><span class="i">symbol</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/FASTAExample3.fasta&quot;</span>) <span class="i">seqFastaItem</span> 
</code></pre></td>
</tr>
</table>
<hr>
<a name="GFF3"></a>
<h2><a name="Introducing-GFF3Parser" class="anchor" href="#Introducing-GFF3Parser">Introducing GFF3Parser</a></h2>
<br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/GFF3Parser.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/bvenn">&#128366;view author of this tutorial</a>
<br><br>
The GFF3Parser is a tool to validate, read and write **GFF3** _(Generic Feature Format Version 3)_-Files.
<p>All examples are executed on a modified version of <a href="http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff">saccharomyces_cerevisiae.gff</a>.</p>
<h2><a name="Introducing-Generic-Feature-Format-Version-3" class="anchor" href="#Introducing-Generic-Feature-Format-Version-3">Introducing Generic Feature Format Version 3</a></h2>
<p>In GFF3 files every line represents one genomic feature with nine tab-delimited fields, whereas unlimited key-value pairs can be stored in field 9.
GFFEntries are implemented as a record type containing the following fields:
<br><br></p>
<div id="responsiveTable">
<table>
<thead>
<tr class="header">
<th><p>Field name</p></th>
<th><p>type</p></th>
<th><p>Description</p></th>
<th><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>Seqid</p></td>
<td><p><code>string</code></p></td>
<td><p>ID of the landmark</p></td>
<td><p>"chrI"</p></td>
</tr>
<tr class="even">
<td><p>Source</p></td>
<td><p><code>string</code></p></td>
<td><p>The source is a free text qualifier to describe the algorithm or operating procedure that generated this feature</p></td>
<td><p>"SGD"</p></td>
</tr>
<tr class="odd">
<td><p>Feature</p></td>
<td><p><code>string</code></p></td>
<td><p>also called method or type; should be a term from the Sequence Ontology or an SO accession number</p></td>
<td><p>"gene"</p></td>
</tr>
<tr class="even">
<td><p>StartPos</p></td>
<td><p><code>int</code></p></td>
<td><p>start coordinate of the feature (1-based interger coordinate)</p></td>
<td><p>335</p></td>
</tr>
<tr class="odd">
<td><p>EndPos</p></td>
<td><p><code>int</code></p></td>
<td><p>end coordinate of the feature (1-based interger coordinate)</p></td>
<td><p>649</p></td>
</tr>
<tr class="even">
<td><p>Score</p></td>
<td><p><code>float</code></p></td>
<td><p>The score is a not well defined floating point number representing different values depending of what you are looking at (nan  if no phase is avaliable)</p></td>
<td><p>nan</p></td>
</tr>
<tr class="odd">
<td><p>Strand</p></td>
<td><p><code>char</code></p></td>
<td><p>strand of the feature ('\000'  if no phase is avaliable)</p></td>
<td><p>'+'</p></td>
</tr>
<tr class="even">
<td><p>Phase</p></td>
<td><p><code>int</code></p></td>
<td><p>for features of type CDS; indicates where the feature begins with reference to the reading frame (-1 if no phase is avaliable)</p></td>
<td><p>-1</p></td>
</tr>
<tr class="odd">
<td><p>Attributes</p></td>
<td><p><code>Map&lt;string,string list&gt;</code></p></td>
<td><p>The field Attributes can contain multiple tag-value pairs, some with predefined meanings. Please check the specification website for further information.</p></td>
<td><p>seq [[ID, [YAL069W]];[Ontology_term, [GO:0000004; GO:0005554]]]</p></td>
</tr>
</tbody>
</table>

<p>| Supplement | <code>string []</code>               | If additional columns (e.g. counts) are present, they are written in this tab-delimited string array.                                                     | [|""|]                                                           |</p>
</div>
<br>
Below you can see an example GFF3 file.
<p><img src="@Root/img/GFF3.png" alt="GFF3" style="width:150px;margin:10px" />
<br>
<div id="responsiveTable"></p>
<table>
<thead>
<tr class="header">
<th><p>Line number</p></th>
<th><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>1</p></td>
<td><p>Declaration of file version</p></td>
</tr>
<tr class="even">
<td><p>2</p></td>
<td><p>Comment</p></td>
</tr>
<tr class="odd">
<td><p>4+5</p></td>
<td><p>GFF entry lines</p></td>
</tr>
<tr class="even">
<td><p>6</p></td>
<td><p>FastA directive</p></td>
</tr>
<tr class="odd">
<td><p>7+</p></td>
<td><p>Fasta Sequence</p></td>
</tr>
</tbody>
</table>

</div>
<br>
<p>Directives (marked with "##[...]") provide additional information like the gff-version which has to be the first line of each file ("##gff-version 3[...]").
Comment lines have to start with a single "#[...]". It is possible that sequences in FastA format are at the end of the file. This has to be announced by a "##FASTA" directive line.</p>
<p>For further information visit <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3-Specifications</a>.</p>
<h2><a name="How-to-use-GFF3Parser" class="anchor" href="#How-to-use-GFF3Parser">How to use GFF3Parser</a></h2>
<p>To read in a gff you have to insert a filepath and optional a FASTA converter. For further information about FASTA check the <a href="https://csbiology.github.io/BioFSharp/Readers.html#FastA">FASTA section</a> on this site
or visit <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span>

<span class="c">///path and name of the input file</span>
<span class="k">let</span> <span class="i">filepathGFF</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gff3Example.gff&quot;</span>)

<span class="c">///To read in a gff file you have to insert a FASTA converter and a source filepath. </span>
<span class="c">///If no FASTA Sequence is included you directly can use GFF3Parser.Reader.GFF3readerWithoutFasta [filepathGFF].</span>
<span class="k">let</span> <span class="i">gffExampleRead</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">GFF3reader</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">BioArray</span><span class="o">.</span><span class="i">ofNucleotideString</span> <span class="i">filepathGFF</span> 
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3SanityCheck" class="anchor" href="#How-to-use-GFF3SanityCheck">How to use GFF3SanityCheck</a></h2>
<p>The GFF3SanityCheck prints wheter your GFF3 file is valid or not. It returns all specified errors including the lines in which the errors occured.
In contrast to GFF2 the field 3 (type, feature or method) of a GFF3 entry is restricted to terms defined by the sequence ontology (SO) so this validator checks if the entry is a valid SO term.
You can find new versions of the SO at (<a href="https://sourceforge.net/projects/song/files/SO_Feature_Annotation).">https://sourceforge.net/projects/song/files/SO_Feature_Annotation).</a></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///to validate the gff file without SOTerm verification use this function and only insert filepath</span>
<span class="k">let</span> <span class="i">gffExampleSanityCheckSimple</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">sanityCheck</span> <span class="i">filepathGFF</span>

<span class="c">///path, name and version of the &#39;Sequence Ontology terms&#39;-file</span>
<span class="k">let</span> <span class="i">filepathSO_Terms</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Sequence_Ontology_Terms_2_5_3.txt&quot;</span>)

<span class="c">///to validate the gff file insert filepath</span>
<span class="k">let</span> <span class="i">gffExampleSanityCheck</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">sanityCheckWithSOTerm</span> <span class="i">filepathSO_Terms</span> <span class="i">filepathGFF</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3RelationshipSearch" class="anchor" href="#How-to-use-GFF3RelationshipSearch">How to use GFF3RelationshipSearch</a></h2>
<p>You also can do a simple search for "<em>Parent</em> - <em>child of</em>" relationships giving back all genomic features which contain the <em>searchterm</em> in <strong>ID/Id</strong> or <strong>Parent</strong> field.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Term to search for:</span>
<span class="k">let</span> <span class="i">searchterm</span> <span class="o">=</span> <span class="s">&quot;YAL069W&quot;</span>

<span class="c">///with this function you can search for a searchterm and you will get features which are related</span>
<span class="k">let</span> <span class="i">gffExampleSearch</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">relationshipSearch</span> <span class="i">gffExampleRead</span> <span class="i">searchterm</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3Writer" class="anchor" href="#How-to-use-GFF3Writer">How to use GFF3Writer</a></h2>
<p>This writer takes GFF lines in stacks of 20,000 items and write/append them to a given filepath.
If FastA sequences are included they are appended by a FastA writer described in the <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<p><em>Note: The order of key value pairs in field 9 (attributes) may be changed.</em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FASTA converter and a destination filepath and writes it into a .gff.</span>
<span class="k">let</span> <span class="i">gffExampleWrite</span> <span class="o">=</span> <span class="i">GFF3Parser</span><span class="o">.</span><span class="i">GFF3Writer</span> <span class="i">gffExampleRead</span> <span class="i">BioItem</span><span class="o">.</span><span class="i">symbol</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gffExampleWrite.gff&quot;</span>)
</code></pre></td>
</tr>
</table>
<br>
<a name="GenBank"></a>
<h2><a name="GenBank" class="anchor" href="#GenBank">GenBank</a></h2>
<br>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp.IO/GenBank.fs">&lt;/&gt;view source code</a>
<a id="APILink" href= >&#128194;View module documentation</a>
<a id="Author" href="https://github.com/kMutagene">&#128366;view author of this tutorial</a>
<br><br>
GenBank is the NIH genetic sequence database, an annotated collection of all publicly available DNA sequences
([Nucleic Acids Research, 2013 Jan;41(D1):D36-42](https://www.ncbi.nlm.nih.gov/pubmed/23193287)). 
A GenBank file contains various information about a sequence and its features.
<p>The file structure of a GenBank (.gb) flat file can be split into 4 subsections:
<br>
<div id="responsiveTable"></p>
<table>
<thead>
<tr class="header">
<th><p>Section</p></th>
<th><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>Meta</p></td>
<td><p>Contains meta information about the annotated sequence, the file itself and the organism the sequence is found in</p></td>
</tr>
<tr class="even">
<td><p>References</p></td>
<td><p>A collection of publications about the annotated sequence(or a subsequence of it) and information associated with them</p></td>
</tr>
<tr class="odd">
<td><p>Features</p></td>
<td><p>A collection of features and their position within the sequence</p></td>
</tr>
<tr class="even">
<td><p>Origin</p></td>
<td><p>The annotated sequence itself</p></td>
</tr>
</tbody>
</table>

</div>
<br>
With the origins section being optional and the features section usually being by far the largest amongst them. 
For more information about the sections and their formatting, have a look at this 
[annotated sample record](https://www.ncbi.nlm.nih.gov/genbank/samplerecord/) over at NCBI. 
<p>This file will also be used for the purpose of this tutorial and in plain text looks like this:</p>
<br>
<button type="button" class="btn" data-toggle="collapse" data-target="#gbFileExample">Show/Hide example file</button>
<div id="gbFileExample" class="collapse fileExamples ">
    <pre>
        LOCUS       SCU49845                5028 bp    DNA     linear   PLN 14-JUL-2016 
        DEFINITION  Saccharomyces cerevisiae TCP1-beta gene, partial cds; and Axl2p     
                    (AXL2) and Rev7p (REV7) genes, complete cds.                        
        ACCESSION   U49845                                                              
        VERSION     U49845.1                                                            
        KEYWORDS    .                                                                   
        SOURCE      Saccharomyces cerevisiae (baker's yeast)                            
          ORGANISM  Saccharomyces cerevisiae                                            
                    Eukaryota; Fungi; Dikarya; Ascomycota; Saccharomycotina;            
                    Saccharomycetes; Saccharomycetales; Saccharomycetaceae;             
                    Saccharomyces.                                                      
        REFERENCE   1  (bases 1 to 5028)                                                
          AUTHORS   Roemer,T., Madden,K., Chang,J. and Snyder,M.                        
          TITLE     Selection of axial growth sites in yeast requires Axl2p, a novel    
                    plasma membrane glycoprotein                                        
          JOURNAL   Genes Dev. 10 (7), 777-793 (1996)                                   
           PUBMED   8846915                                                             
        REFERENCE   2  (bases 1 to 5028)                                                
          AUTHORS   Roemer,T.                                                           
          TITLE     Direct Submission                                                   
          JOURNAL   Submitted (22-FEB-1996) Biology, Yale University, New Haven, CT     
                    06520, USA                                                          
        FEATURES             Location/Qualifiers                                        
             source          1..5028                                                    
                             /organism="Saccharomyces cerevisiae"                       
                             /mol_type="genomic DNA"                                    
                             /db_xref="taxon:4932"                                      
                             /chromosome="IX"                                           
             mRNA            <1..>206                                                   
                             /product="TCP1-beta"                                       
             CDS             <1..206                                                    
                             /codon_start=3                                             
                             /product="TCP1-beta"                                       
                             /protein_id="AAA98665.1"                                   
                             /translation="SSIYNGISTSGLDLNNGTIADMRQLGIVESYKLKRAVVSSASEA 
                             AEVLLRVDNIIRARPRTANRQHM"                                   
             gene            <687..>3158                                                
                             /gene="AXL2"                                               
             mRNA            <687..>3158                                                
                             /gene="AXL2"                                               
                             /product="Axl2p"                                           
             CDS             687..3158                                                  
                             /gene="AXL2"                                               
                             /note="plasma membrane glycoprotein"                       
                             /codon_start=1                                             
                             /product="Axl2p"                                           
                             /protein_id="AAA98666.1"                                   
                             /translation="MTQLQISLLLTATISLLHLVVATPYEAYPIGKQYPPVARVNESF 
                             TFQISNDTYKSSVDKTAQITYNCFDLPSWLSFDSSSRTFSGEPSSDLLSDANTTLYFN 
                             VILEGTDSADSTSLNNTYQFVVTNRPSISLSSDFNLLALLKNYGYTNGKNALKLDPNE 
                             VFNVTFDRSMFTNEESIVSYYGRSQLYNAPLPNWLFFDSGELKFTGTAPVINSAIAPE 
                             TSYSFVIIATDIEGFSAVEVEFELVIGAHQLTTSIQNSLIINVTDTGNVSYDLPLNYV 
                             YLDDDPISSDKLGSINLLDAPDWVALDNATISGSVPDELLGKNSNPANFSVSIYDTYG 
                             DVIYFNFEVVSTTDLFAISSLPNINATRGEWFSYYFLPSQFTDYVNTNVSLEFTNSSQ 
                             DHDWVKFQSSNLTLAGEVPKNFDKLSLGLKANQGSQSQELYFNIIGMDSKITHSNHSA 
                             NATSTRSSHHSTSTSSYTSSTYTAKISSTSAAATSSAPAALPAANKTSSHNKKAVAIA 
                             CGVAIPLGVILVALICFLIFWRRRRENPDDENLPHAISGPDLNNPANKPNQENATPLN 
                             NPFDDDASSYDDTSIARRLAALNTLKLDNHSATESDISSVDEKRDSLSGMNTYNDQFQ 
                             SQSKEELLAKPPVQPPESPFFDPQNRSSSVYMDSEPAVNKSWRYTGNLSPVSDIVRDS 
                             YGSQKTVDTEKLFDLEAPEKEKRTSRDVTMSSLDPWNSNISPSPVRKSVTPSPYNVTK 
                             HRNRHLQNIQDSQSGKNGITPTTMSTSSSDDFVPVKDGENFCWVHSMEPDRRPSKKRL 
                             VDFSNKSNVNVGQVKDIHGRIPEML"                                 
             gene            complement(<3300..>4037)                                   
                             /gene="REV7"                                               
             mRNA            complement(<3300..>4037)                                   
                             /gene="REV7"                                               
                             /product="Rev7p"                                           
             CDS             complement(3300..4037)                                     
                             /gene="REV7"                                               
                             /codon_start=1                                             
                             /product="Rev7p"                                           
                             /protein_id="AAA98667.1"                                   
                             /translation="MNRWVEKWLRVYLKCYINLILFYRNVYPPQSFDYTTYQSFNLPQ 
                             FVPINRHPALIDYIEELILDVLSKLTHVYRFSICIINKKNDLCIEKYVLDFSELQHVD 
                             KDDQIITETEVFDEFRSSLNSLIMHLEKLPKVNDDTITFEAVINAIELELGHKLDRNR 
                             RVDSLEEKAEIERDSNWVKCQEDENLPDNNGFQPPKIKLTSLVGSDVGPLIIHQFSEK 
                             LISGDDKILNGVYSQYEEGESIFGSLF"                               
        ORIGIN                                                                          
                1 gatcctccat atacaacggt atctccacct caggtttaga tctcaacaac ggaaccattg     
               61 ccgacatgag acagttaggt atcgtcgaga gttacaagct aaaacgagca gtagtcagct     
              121 ctgcatctga agccgctgaa gttctactaa gggtggataa catcatccgt gcaagaccaa     
              181 gaaccgccaa tagacaacat atgtaacata tttaggatat acctcgaaaa taataaaccg     
              241 ccacactgtc attattataa ttagaaacag aacgcaaaaa ttatccacta tataattcaa     
              301 agacgcgaaa aaaaaagaac aacgcgtcat agaacttttg gcaattcgcg tcacaaataa     
              361 attttggcaa cttatgtttc ctcttcgagc agtactcgag ccctgtctca agaatgtaat     
              421 aatacccatc gtaggtatgg ttaaagatag catctccaca acctcaaagc tccttgccga     
              481 gagtcgccct cctttgtcga gtaattttca cttttcatat gagaacttat tttcttattc     
              541 tttactctca catcctgtag tgattgacac tgcaacagcc accatcacta gaagaacaga     
              601 acaattactt aatagaaaaa ttatatcttc ctcgaaacga tttcctgctt ccaacatcta     
              661 cgtatatcaa gaagcattca cttaccatga cacagcttca gatttcatta ttgctgacag     
              721 ctactatatc actactccat ctagtagtgg ccacgcccta tgaggcatat cctatcggaa     
              781 aacaataccc cccagtggca agagtcaatg aatcgtttac atttcaaatt tccaatgata     
              841 cctataaatc gtctgtagac aagacagctc aaataacata caattgcttc gacttaccga     
              901 gctggctttc gtttgactct agttctagaa cgttctcagg tgaaccttct tctgacttac     
              961 tatctgatgc gaacaccacg ttgtatttca atgtaatact cgagggtacg gactctgccg     
             1021 acagcacgtc tttgaacaat acataccaat ttgttgttac aaaccgtcca tccatctcgc     
             1081 tatcgtcaga tttcaatcta ttggcgttgt taaaaaacta tggttatact aacggcaaaa     
             1141 acgctctgaa actagatcct aatgaagtct tcaacgtgac ttttgaccgt tcaatgttca     
             1201 ctaacgaaga atccattgtg tcgtattacg gacgttctca gttgtataat gcgccgttac     
             1261 ccaattggct gttcttcgat tctggcgagt tgaagtttac tgggacggca ccggtgataa     
             1321 actcggcgat tgctccagaa acaagctaca gttttgtcat catcgctaca gacattgaag     
             1381 gattttctgc cgttgaggta gaattcgaat tagtcatcgg ggctcaccag ttaactacct     
             1441 ctattcaaaa tagtttgata atcaacgtta ctgacacagg taacgtttca tatgacttac     
             1501 ctctaaacta tgtttatctc gatgacgatc ctatttcttc tgataaattg ggttctataa     
             1561 acttattgga tgctccagac tgggtggcat tagataatgc taccatttcc gggtctgtcc     
             1621 cagatgaatt actcggtaag aactccaatc ctgccaattt ttctgtgtcc atttatgata     
             1681 cttatggtga tgtgatttat ttcaacttcg aagttgtctc cacaacggat ttgtttgcca     
             1741 ttagttctct tcccaatatt aacgctacaa ggggtgaatg gttctcctac tattttttgc     
             1801 cttctcagtt tacagactac gtgaatacaa acgtttcatt agagtttact aattcaagcc     
             1861 aagaccatga ctgggtgaaa ttccaatcat ctaatttaac attagctgga gaagtgccca     
             1921 agaatttcga caagctttca ttaggtttga aagcgaacca aggttcacaa tctcaagagc     
             1981 tatattttaa catcattggc atggattcaa agataactca ctcaaaccac agtgcgaatg     
             2041 caacgtccac aagaagttct caccactcca cctcaacaag ttcttacaca tcttctactt     
             2101 acactgcaaa aatttcttct acctccgctg ctgctacttc ttctgctcca gcagcgctgc     
             2161 cagcagccaa taaaacttca tctcacaata aaaaagcagt agcaattgcg tgcggtgttg     
             2221 ctatcccatt aggcgttatc ctagtagctc tcatttgctt cctaatattc tggagacgca     
             2281 gaagggaaaa tccagacgat gaaaacttac cgcatgctat tagtggacct gatttgaata     
             2341 atcctgcaaa taaaccaaat caagaaaacg ctacaccttt gaacaacccc tttgatgatg     
             2401 atgcttcctc gtacgatgat acttcaatag caagaagatt ggctgctttg aacactttga     
             2461 aattggataa ccactctgcc actgaatctg atatttccag cgtggatgaa aagagagatt     
             2521 ctctatcagg tatgaataca tacaatgatc agttccaatc ccaaagtaaa gaagaattat     
             2581 tagcaaaacc cccagtacag cctccagaga gcccgttctt tgacccacag aataggtctt     
             2641 cttctgtgta tatggatagt gaaccagcag taaataaatc ctggcgatat actggcaacc     
             2701 tgtcaccagt ctctgatatt gtcagagaca gttacggatc acaaaaaact gttgatacag     
             2761 aaaaactttt cgatttagaa gcaccagaga aggaaaaacg tacgtcaagg gatgtcacta     
             2821 tgtcttcact ggacccttgg aacagcaata ttagcccttc tcccgtaaga aaatcagtaa     
             2881 caccatcacc atataacgta acgaagcatc gtaaccgcca cttacaaaat attcaagact     
             2941 ctcaaagcgg taaaaacgga atcactccca caacaatgtc aacttcatct tctgacgatt     
             3001 ttgttccggt taaagatggt gaaaattttt gctgggtcca tagcatggaa ccagacagaa     
             3061 gaccaagtaa gaaaaggtta gtagattttt caaataagag taatgtcaat gttggtcaag     
             3121 ttaaggacat tcacggacgc atcccagaaa tgctgtgatt atacgcaacg atattttgct     
             3181 taattttatt ttcctgtttt attttttatt agtggtttac agatacccta tattttattt     
             3241 agtttttata cttagagaca tttaatttta attccattct tcaaatttca tttttgcact     
             3301 taaaacaaag atccaaaaat gctctcgccc tcttcatatt gagaatacac tccattcaaa     
             3361 attttgtcgt caccgctgat taatttttca ctaaactgat gaataatcaa aggccccacg     
             3421 tcagaaccga ctaaagaagt gagttttatt ttaggaggtt gaaaaccatt attgtctggt     
             3481 aaattttcat cttcttgaca tttaacccag tttgaatccc tttcaatttc tgctttttcc     
             3541 tccaaactat cgaccctcct gtttctgtcc aacttatgtc ctagttccaa ttcgatcgca     
             3601 ttaataactg cttcaaatgt tattgtgtca tcgttgactt taggtaattt ctccaaatgc     
             3661 ataatcaaac tatttaagga agatcggaat tcgtcgaaca cttcagtttc cgtaatgatc     
             3721 tgatcgtctt tatccacatg ttgtaattca ctaaaatcta aaacgtattt ttcaatgcat     
             3781 aaatcgttct ttttattaat aatgcagatg gaaaatctgt aaacgtgcgt taatttagaa     
             3841 agaacatcca gtataagttc ttctatatag tcaattaaag caggatgcct attaatggga     
             3901 acgaactgcg gcaagttgaa tgactggtaa gtagtgtagt cgaatgactg aggtgggtat     
             3961 acatttctat aaaataaaat caaattaatg tagcatttta agtataccct cagccacttc     
             4021 tctacccatc tattcataaa gctgacgcaa cgattactat tttttttttc ttcttggatc     
             4081 tcagtcgtcg caaaaacgta taccttcttt ttccgacctt ttttttagct ttctggaaaa     
             4141 gtttatatta gttaaacagg gtctagtctt agtgtgaaag ctagtggttt cgattgactg     
             4201 atattaagaa agtggaaatt aaattagtag tgtagacgta tatgcatatg tatttctcgc     
             4261 ctgtttatgt ttctacgtac ttttgattta tagcaagggg aaaagaaata catactattt     
             4321 tttggtaaag gtgaaagcat aatgtaaaag ctagaataaa atggacgaaa taaagagagg     
             4381 cttagttcat cttttttcca aaaagcaccc aatgataata actaaaatga aaaggatttg     
             4441 ccatctgtca gcaacatcag ttgtgtgagc aataataaaa tcatcacctc cgttgccttt     
             4501 agcgcgtttg tcgtttgtat cttccgtaat tttagtctta tcaatgggaa tcataaattt     
             4561 tccaatgaat tagcaatttc gtccaattct ttttgagctt cttcatattt gctttggaat     
             4621 tcttcgcact tcttttccca ttcatctctt tcttcttcca aagcaacgat ccttctaccc     
             4681 atttgctcag agttcaaatc ggcctctttc agtttatcca ttgcttcctt cagtttggct     
             4741 tcactgtctt ctagctgttg ttctagatcc tggtttttct tggtgtagtt ctcattatta     
             4801 gatctcaagt tattggagtc ttcagccaat tgctttgtat cagacaattg actctctaac     
             4861 ttctccactt cactgtcgag ttgctcgttt ttagcggaca aagatttaat ctcgttttct     
             4921 ttttcagtgt tagattgctc taattctttg agctgttctc tcagctcctc atatttttct     
             4981 tgccatgact cagattctaa ttttaagcta ttcaatttct ctttgatc                  
        //                                                                              
    </pre>
</div>
<br>
<h2><a name="Reading-GenBank-files" class="anchor" href="#Reading-GenBank-files">Reading GenBank files</a></h2>
<p>The type equivalent for a GenBank file in BioFSharp is a dictionary, mapping <code>string</code> keys to the <code>GenBankItem</code> type.
More information about the type modelling can be found in our <a href="">API reference</a>.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="i">BioFSharp</span><span class="o">.</span><span class="i">IO</span>

<span class="c">///Path of the example file</span>
<span class="k">let</span> <span class="i">exampleFilePath</span> <span class="o">=</span> <span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">@&quot;\data\sequence.gb&quot;</span>

<span class="c">///Parsed Example File </span>
<span class="k">let</span> <span class="i">parsedGBFile</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">Read</span><span class="o">.</span><span class="i">fromFile</span> <span class="i">exampleFilePath</span>
</code></pre></td>
</tr>
</table>
<p>The GenBank module provides various helper functions for querying items on the created Dictionary:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///All features contained in this GenBank file representation</span>
<span class="k">let</span> <span class="i">features</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">getFeatures</span> <span class="i">parsedGBFile</span> 

<span class="c">///All features with the &quot;CDS&quot; tag</span>
<span class="k">let</span> <span class="i">cdsFeatures</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">getFeaturesWithType</span> <span class="s">&quot;CDS&quot;</span> <span class="i">parsedGBFile</span>
</code></pre></td>
</tr>
</table>
<p>The <code>getfeaturesWithBaseSpan</code> function returns all Features that span the input BaseSpan of the sequence contained in the
file. However, due to the way the BaseSpans are modelled it is currently only possible to find features that have exactly the query BaseSpan,
or contain the exact BaseSpan in a join.
Getting all features that fall within the input BaseSpan will be a goal for a future update of this parser.</p>
<p>We first need to create the BaseSpan that we want to find.
For that, we create a <code>BaseSpanRange</code> union case containing the type of BaseSpan (modelled as the <code>BaseSpanInformation</code> type)
and the start and end value as an integer tuple:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///we are searching for a complete base span</span>
<span class="k">let</span> <span class="i">queryBaseSpanType</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">BaseSpanInformation</span><span class="o">.</span><span class="i">Complete</span>

<span class="c">///spanning nucleotides 687-3158</span>
<span class="k">let</span> <span class="i">queryBaseSpan</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">BaseSpanRange</span><span class="o">.</span><span class="i">Single</span> (<span class="i">queryBaseSpanType</span>,(<span class="n">687</span>,<span class="n">3158</span>))

<span class="c">///</span>
<span class="k">let</span> <span class="i">bsQueryResult</span> <span class="o">=</span> <span class="i">GenBank</span><span class="o">.</span><span class="i">getFeaturesWithBaseSpan</span> <span class="i">queryBaseSpan</span> <span class="k">true</span> <span class="i">parsedGBFile</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Writing-GenBank-files" class="anchor" href="#Writing-GenBank-files">Writing GenBank files</a></h2>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li>
                    <a href="/BioFSharp/IBioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/SeqProperties.html">Sequence Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF.html">GFF format</a></li>
                <li><a href="/BioFSharp/Genebank.html">Genebank</a></li>
                <li><a href="/BioFSharp/Clustal.html">Clustal</a></li>
                <li><a href="/BioFSharp/Newick.html">Newick</a></li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/GESA.html">Web logo</a></li>

                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
