<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Using Bio Format Readers
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>

    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="Using-Bio-Format-Readers" class="anchor" href="#Using-Bio-Format-Readers">Using Bio Format Readers</a></h1>
<p>BioFSharp contains a set of readers for different biology-associated file formats. This documentation aims to give an introduction for them.</p>
<a name="FastQ"></a>
<h2><a name="FastQ" class="anchor" href="#FastQ">FastQ</a></h2>
<p>This module allows to parse FASTQ format data with original 4-lines entries into this record type</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// FastqItem record contains header, sequence, qualityheader, qualitysequence of one entry</span>
<span class="k">type</span> <span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="t">FastqItem</span><span class="o">&lt;</span><span class="o">&#39;</span><span class="i">a</span>,<span class="o">&#39;</span><span class="i">b</span><span class="o">&gt;</span> <span class="o">=</span> {
    <span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="i">Header</span>    <span class="o">:</span> <span onmouseout="hideTip(event, 'fs6', 7)" onmouseover="showTip(event, 'fs6', 7)" class="t">string</span>;
    <span onmouseout="hideTip(event, 'fs7', 8)" onmouseover="showTip(event, 'fs7', 8)" class="i">Sequence</span>  <span class="o">:</span> <span class="o">&#39;</span><span class="i">a</span>;
    <span onmouseout="hideTip(event, 'fs8', 9)" onmouseover="showTip(event, 'fs8', 9)" class="i">QualityHeader</span> <span class="o">:</span> <span onmouseout="hideTip(event, 'fs6', 10)" onmouseover="showTip(event, 'fs6', 10)" class="t">string</span>;
    <span onmouseout="hideTip(event, 'fs9', 11)" onmouseover="showTip(event, 'fs9', 11)" class="i">QualitySequence</span> <span class="o">:</span> <span class="o">&#39;</span><span class="i">b</span>;       
}
</code></pre></td>
</tr>
</table>
<p>To be able to use this parser you need to define two converter functions,
one example for each you can also find in our module, but you also may need to write your own.</p>
<p>We can convert sequence string to predefined option type of Amino Acids, using converter function
from our library 'BioFSharp.BioItemsConverter.OptionConverter'</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// get characters as sequence units</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 12)" onmouseover="showTip(event, 'fs10', 12)" class="f">converterToAA</span> <span onmouseout="hideTip(event, 'fs11', 13)" onmouseover="showTip(event, 'fs11', 13)" class="i">string</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs11', 14)" onmouseover="showTip(event, 'fs11', 14)" class="i">string</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 15)" onmouseover="showTip(event, 'fs12', 15)" class="t">String</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs13', 16)" onmouseover="showTip(event, 'fs13', 16)" class="f">map</span> (<span onmouseout="hideTip(event, 'fs2', 17)" onmouseover="showTip(event, 'fs2', 17)" class="i">BioFSharp</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs14', 18)" onmouseover="showTip(event, 'fs14', 18)" class="t">BioItemsConverter</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs15', 19)" onmouseover="showTip(event, 'fs15', 19)" class="t">OptionConverter</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs16', 20)" onmouseover="showTip(event, 'fs16', 20)" class="f">charToOptionAminoAcid</span>)
</code></pre></td>
</tr>
</table>
<p>If you have following possible values for quality sequence:
'!""#$%&amp;'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`abcdefghijklmnopqrstuvwxyz{|}~',
with Sanger format, that can encode a Phred quality score from 0 to 93 using ASCII 33 to 126,
then you can use our converting function:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">/// get Phred quality score</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 21)" onmouseover="showTip(event, 'fs17', 21)" class="f">qualityConvertFn</span> <span onmouseout="hideTip(event, 'fs18', 22)" onmouseover="showTip(event, 'fs18', 22)" class="i">string</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs18', 23)" onmouseover="showTip(event, 'fs18', 23)" class="i">string</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs12', 24)" onmouseover="showTip(event, 'fs12', 24)" class="i">String</span><span class="o">.</span><span class="i">toCharArray</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs19', 25)" onmouseover="showTip(event, 'fs19', 25)" class="t">Array</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs20', 26)" onmouseover="showTip(event, 'fs20', 26)" class="f">map</span> (<span class="k">fun</span> <span onmouseout="hideTip(event, 'fs21', 27)" onmouseover="showTip(event, 'fs21', 27)" class="i">i</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs21', 28)" onmouseover="showTip(event, 'fs21', 28)" class="i">i</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs22', 29)" onmouseover="showTip(event, 'fs22', 29)" class="f">GetHashCode</span>()<span class="o">-</span><span class="n">33</span>)
</code></pre></td>
</tr>
</table>
<p>And then you can easily use this module to read your FastQ file</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 30)" onmouseover="showTip(event, 'fs23', 30)" class="i">yourFastqFile</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/FastQtest.fastq&quot;</span>)

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs24', 31)" onmouseover="showTip(event, 'fs24', 31)" class="i">FastQSequence</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs25', 32)" onmouseover="showTip(event, 'fs25', 32)" class="t">FastQ</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs26', 33)" onmouseover="showTip(event, 'fs26', 33)" class="f">fromFile</span> <span class="i">convertFn</span> <span onmouseout="hideTip(event, 'fs17', 34)" onmouseover="showTip(event, 'fs17', 34)" class="f">qualityConvertFn</span> <span onmouseout="hideTip(event, 'fs23', 35)" onmouseover="showTip(event, 'fs23', 35)" class="i">yourFastqFile</span>
</code></pre></td>
</tr>
</table>
<hr>
<a name="FastA"></a>
<h2><a name="FastA" class="anchor" href="#FastA">FastA</a></h2>
<p>This documentation is not yet here. Hopefully it will be soon =)</p>
<hr>
<a name="GFF3"></a>
<h2><a name="Introducing-GFF3Parser" class="anchor" href="#Introducing-GFF3Parser">Introducing GFF3Parser</a></h2>
<p>The GFF3Parser is a tool to validate, read and write <strong>GFF3</strong> <em>(Generic Feature Format Version 3)</em>-Files.</p>
<p>All examples are executed on a modified version of <a href="http://downloads.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff">saccharomyces_cerevisiae.gff</a>.</p>
<h2><a name="Introducing-Generic-Feature-Format-Version-3" class="anchor" href="#Introducing-Generic-Feature-Format-Version-3">Introducing Generic Feature Format Version 3</a></h2>
<p>In GFF3 files every line represents one genomic feature with nine tab-delimited fields, whereas unlimited key-value pairs can be stored in field 9.
The nine fields are:</p>
<table>
<thead>
<tr class="header">
<th><p>Field</p></th>
<th align="center"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>1</p></td>
<td align="center"><p>seqID</p></td>
</tr>
<tr class="even">
<td><p>2</p></td>
<td align="center"><p>source</p></td>
</tr>
<tr class="odd">
<td><p>3</p></td>
<td align="center"><p>type, method or feature</p></td>
</tr>
<tr class="even">
<td><p>4</p></td>
<td align="center"><p>startPos</p></td>
</tr>
<tr class="odd">
<td><p>5</p></td>
<td align="center"><p>endPos</p></td>
</tr>
<tr class="even">
<td><p>6</p></td>
<td align="center"><p>score</p></td>
</tr>
<tr class="odd">
<td><p>7</p></td>
<td align="center"><p>strand</p></td>
</tr>
<tr class="even">
<td><p>8</p></td>
<td align="center"><p>phase</p></td>
</tr>
<tr class="odd">
<td><p>9</p></td>
<td align="center"><p>attributes</p></td>
</tr>
</tbody>
</table>

<img src="BioFSharp/img/GFF3.png" alt="GFF3" style="max-width:100%;margin:10px" />
<p>Directives (marked with "##[...]") provide additional information like the gff-version which has to be the first line of each file ("##gff-version 3[...]").
Comment lines have to start with a single "#[...]". It is possible that sequences in FastA format are at the end of the file. This has to be announced by a "##FASTA" directive line.</p>
<p>For further information visit <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3-Specifications</a>.</p>
<h2><a name="How-to-use-GFF3Parser" class="anchor" href="#How-to-use-GFF3Parser">How to use GFF3Parser</a></h2>
<p>To read in a gff you have to insert a filepath and optional a FastA converter. For further information about FastA visit <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs2', 36)" onmouseover="showTip(event, 'fs2', 36)" class="i">BioFSharp</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs3', 37)" onmouseover="showTip(event, 'fs3', 37)" class="i">IO</span>

<span class="c">///path and name of the input file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs27', 38)" onmouseover="showTip(event, 'fs27', 38)" class="i">filepathGFF</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gff3Example.gff&quot;</span>)

<span class="c">///converter for FastA.FastaItem</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs28', 39)" onmouseover="showTip(event, 'fs28', 39)" class="f">fastaConverter</span> (<span onmouseout="hideTip(event, 'fs29', 40)" onmouseover="showTip(event, 'fs29', 40)" class="i">x</span><span class="o">:</span><span onmouseout="hideTip(event, 'fs30', 41)" onmouseover="showTip(event, 'fs30', 41)" class="t">seq</span><span class="o">&lt;</span><span onmouseout="hideTip(event, 'fs31', 42)" onmouseover="showTip(event, 'fs31', 42)" class="t">char</span><span class="o">&gt;</span>) <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs32', 43)" onmouseover="showTip(event, 'fs32', 43)" class="f">id</span> <span onmouseout="hideTip(event, 'fs29', 44)" onmouseover="showTip(event, 'fs29', 44)" class="i">x</span>

<span class="c">///To read in a gff file you have to insert a FastA converter and a source filepath. </span>
<span class="c">///If no FastA Sequence is included you directly can use GFF3Parser.Reader.GFF3readerWithoutFasta filepathGFF</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs33', 45)" onmouseover="showTip(event, 'fs33', 45)" class="i">gffExampleRead</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 46)" onmouseover="showTip(event, 'fs34', 46)" class="t">GFF3Parser</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs35', 47)" onmouseover="showTip(event, 'fs35', 47)" class="t">Reader</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs36', 48)" onmouseover="showTip(event, 'fs36', 48)" class="f">GFF3reader</span> <span onmouseout="hideTip(event, 'fs28', 49)" onmouseover="showTip(event, 'fs28', 49)" class="f">fastaConverter</span> <span onmouseout="hideTip(event, 'fs27', 50)" onmouseover="showTip(event, 'fs27', 50)" class="i">filepathGFF</span> 
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3SanityCheck" class="anchor" href="#How-to-use-GFF3SanityCheck">How to use GFF3SanityCheck</a></h2>
<p>The GFF3SanityCheck prints wheter your GFF3 file is valid or not. If not, then it also returns the specified error and the line in which the error occured.
In contrast to GFF2 the field 3 (type) of a GFF3 entry is restricted to terms defined by the sequence ontology (SO) so this validator checks if the entry is a valid SO term.
You can find new versions of the SO at (<a href="https://sourceforge.net/projects/song/files/SO_Feature_Annotation).">https://sourceforge.net/projects/song/files/SO_Feature_Annotation).</a></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///to validate the gff file without SOTerm verification use this function and only insert filepath</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs37', 51)" onmouseover="showTip(event, 'fs37', 51)" class="i">gffExampleSanityCheckSimple</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 52)" onmouseover="showTip(event, 'fs34', 52)" class="t">GFF3Parser</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs38', 53)" onmouseover="showTip(event, 'fs38', 53)" class="t">SanityCheck</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs39', 54)" onmouseover="showTip(event, 'fs39', 54)" class="f">sanityCheck</span> <span onmouseout="hideTip(event, 'fs27', 55)" onmouseover="showTip(event, 'fs27', 55)" class="i">filepathGFF</span>

<span class="c">///path, name and version of the &#39;Sequence Ontology terms&#39;-file</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs40', 56)" onmouseover="showTip(event, 'fs40', 56)" class="i">filepathSO_Terms</span> <span class="o">=</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Sequence_Ontology_Terms_2_5_3.txt&quot;</span>)

<span class="c">///to validate the gff file insert filepath</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs41', 57)" onmouseover="showTip(event, 'fs41', 57)" class="i">gffExampleSanityCheck</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 58)" onmouseover="showTip(event, 'fs34', 58)" class="t">GFF3Parser</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs38', 59)" onmouseover="showTip(event, 'fs38', 59)" class="t">SanityCheck</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs42', 60)" onmouseover="showTip(event, 'fs42', 60)" class="f">sanityCheckWithSOTerm</span> <span onmouseout="hideTip(event, 'fs40', 61)" onmouseover="showTip(event, 'fs40', 61)" class="i">filepathSO_Terms</span> <span onmouseout="hideTip(event, 'fs27', 62)" onmouseover="showTip(event, 'fs27', 62)" class="i">filepathGFF</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3RelationshipSearch" class="anchor" href="#How-to-use-GFF3RelationshipSearch">How to use GFF3RelationshipSearch</a></h2>
<p>You also can do a simple search for "<em>Parent</em> - <em>child of</em>" relationships giving back all genomic features which contain the <em>searchterm</em> in <strong>ID/Id</strong> or <strong>Parent</strong> field.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Term to search for:</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs43', 63)" onmouseover="showTip(event, 'fs43', 63)" class="i">searchterm</span> <span class="o">=</span> <span class="s">&quot;YAL069W&quot;</span>

<span class="c">///with this function you can search for a searchterm and you will get features which are related</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs44', 64)" onmouseover="showTip(event, 'fs44', 64)" class="i">gffExampleSearch</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 65)" onmouseover="showTip(event, 'fs34', 65)" class="t">GFF3Parser</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs45', 66)" onmouseover="showTip(event, 'fs45', 66)" class="t">Relationship</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs46', 67)" onmouseover="showTip(event, 'fs46', 67)" class="f">relationshipSearch</span> <span onmouseout="hideTip(event, 'fs33', 68)" onmouseover="showTip(event, 'fs33', 68)" class="i">gffExampleRead</span> <span onmouseout="hideTip(event, 'fs43', 69)" onmouseover="showTip(event, 'fs43', 69)" class="i">searchterm</span>
</code></pre></td>
</tr>
</table>
<h2><a name="How-to-use-GFF3Writer" class="anchor" href="#How-to-use-GFF3Writer">How to use GFF3Writer</a></h2>
<p>This writer takes GFF lines in stacks of 20,000 items and write/append them to a given filepath.
If FastA sequences are included they are appended by a FastA writer described in the <a href="https://csbiology.github.io/BioFSharp/reference/biofsharp-io-fasta.html">API Reference - FastA</a>.</p>
<p><em>Note: The order of key value pairs in field 9 (attributes) may be changed.</em></p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FastA converter and a destination filepath and writes it into a .gff.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs47', 70)" onmouseover="showTip(event, 'fs47', 70)" class="i">gffExampleWrite</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs34', 71)" onmouseover="showTip(event, 'fs34', 71)" class="t">GFF3Parser</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs48', 72)" onmouseover="showTip(event, 'fs48', 72)" class="t">Writer</span><span class="o">.</span><span onmouseout="hideTip(event, 'fs49', 73)" onmouseover="showTip(event, 'fs49', 73)" class="f">GFF3Writer</span> <span onmouseout="hideTip(event, 'fs33', 74)" onmouseover="showTip(event, 'fs33', 74)" class="i">gffExampleRead</span> <span onmouseout="hideTip(event, 'fs32', 75)" onmouseover="showTip(event, 'fs32', 75)" class="f">id</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/gffExampleWrite.gff&quot;</span>)
</code></pre></td>
</tr>
</table>

<div class="tip" id="fs1">namespace System</div>
<div class="tip" id="fs2">namespace BioFSharp</div>
<div class="tip" id="fs3">namespace BioFSharp.IO</div>
<div class="tip" id="fs4">type FastqItem&lt;&#39;a,&#39;b&gt; =<br />&#160;&#160;{Header: string;<br />&#160;&#160;&#160;Sequence: &#39;a;<br />&#160;&#160;&#160;QualityHeader: string;<br />&#160;&#160;&#160;QualitySequence: &#39;b;}<br /><br />Full name: Readers.FastqItem&lt;_,_&gt;<br /><em><br /><br />&#160;FastqItem record contains header, sequence, qualityheader, qualitysequence of one entry</em></div>
<div class="tip" id="fs5">FastqItem.Header: string</div>
<div class="tip" id="fs6">Multiple items<br />val string : value:&#39;T -&gt; string<br /><br />Full name: Microsoft.FSharp.Core.Operators.string<br /><br />--------------------<br />type string = String<br /><br />Full name: Microsoft.FSharp.Core.string</div>
<div class="tip" id="fs7">FastqItem.Sequence: &#39;a</div>
<div class="tip" id="fs8">FastqItem.QualityHeader: string</div>
<div class="tip" id="fs9">FastqItem.QualitySequence: &#39;b</div>
<div class="tip" id="fs10">val converterToAA : string:string -&gt; string<br /><br />Full name: Readers.converterToAA<br /><em><br /><br />&#160;get characters as sequence units</em></div>
<div class="tip" id="fs11">Multiple items<br />val string : string<br /><br />--------------------<br />type string = String<br /><br />Full name: Microsoft.FSharp.Core.string</div>
<div class="tip" id="fs12">Multiple items<br />type String =<br />&#160;&#160;new : value:char -&gt; string + 7 overloads<br />&#160;&#160;member Chars : int -&gt; char<br />&#160;&#160;member Clone : unit -&gt; obj<br />&#160;&#160;member CompareTo : value:obj -&gt; int + 1 overload<br />&#160;&#160;member Contains : value:string -&gt; bool<br />&#160;&#160;member CopyTo : sourceIndex:int * destination:char[] * destinationIndex:int * count:int -&gt; unit<br />&#160;&#160;member EndsWith : value:string -&gt; bool + 2 overloads<br />&#160;&#160;member Equals : obj:obj -&gt; bool + 2 overloads<br />&#160;&#160;member GetEnumerator : unit -&gt; CharEnumerator<br />&#160;&#160;member GetHashCode : unit -&gt; int<br />&#160;&#160;...<br /><br />Full name: System.String<br /><br />--------------------<br />String(value: nativeptr&lt;char&gt;) : unit<br />String(value: nativeptr&lt;sbyte&gt;) : unit<br />String(value: char []) : unit<br />String(c: char, count: int) : unit<br />String(value: nativeptr&lt;char&gt;, startIndex: int, length: int) : unit<br />String(value: nativeptr&lt;sbyte&gt;, startIndex: int, length: int) : unit<br />String(value: char [], startIndex: int, length: int) : unit<br />String(value: nativeptr&lt;sbyte&gt;, startIndex: int, length: int, enc: Text.Encoding) : unit</div>
<div class="tip" id="fs13">val map : mapping:(char -&gt; char) -&gt; str:string -&gt; string<br /><br />Full name: Microsoft.FSharp.Core.String.map</div>
<div class="tip" id="fs14">module BioItemsConverter<br /><br />from BioFSharp</div>
<div class="tip" id="fs15">module OptionConverter<br /><br />from BioFSharp.BioItemsConverter</div>
<div class="tip" id="fs16">val charToOptionAminoAcid : aac:char -&gt; AminoAcids.AminoAcid option<br /><br />Full name: BioFSharp.BioItemsConverter.OptionConverter.charToOptionAminoAcid</div>
<div class="tip" id="fs17">val qualityConvertFn : string:&#39;a -&gt; int []<br /><br />Full name: Readers.qualityConvertFn<br /><em><br /><br />&#160;get Phred quality score</em></div>
<div class="tip" id="fs18">Multiple items<br />val string : &#39;a<br /><br />--------------------<br />type string = String<br /><br />Full name: Microsoft.FSharp.Core.string</div>
<div class="tip" id="fs19">type Array =<br />&#160;&#160;member Clone : unit -&gt; obj<br />&#160;&#160;member CopyTo : array:Array * index:int -&gt; unit + 1 overload<br />&#160;&#160;member GetEnumerator : unit -&gt; IEnumerator<br />&#160;&#160;member GetLength : dimension:int -&gt; int<br />&#160;&#160;member GetLongLength : dimension:int -&gt; int64<br />&#160;&#160;member GetLowerBound : dimension:int -&gt; int<br />&#160;&#160;member GetUpperBound : dimension:int -&gt; int<br />&#160;&#160;member GetValue : [&lt;ParamArray&gt;] indices:int[] -&gt; obj + 7 overloads<br />&#160;&#160;member Initialize : unit -&gt; unit<br />&#160;&#160;member IsFixedSize : bool<br />&#160;&#160;...<br /><br />Full name: System.Array</div>
<div class="tip" id="fs20">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; array:&#39;T [] -&gt; &#39;U []<br /><br />Full name: Microsoft.FSharp.Collections.Array.map</div>
<div class="tip" id="fs21">val i : obj</div>
<div class="tip" id="fs22">Object.GetHashCode() : int</div>
<div class="tip" id="fs23">val yourFastqFile : string<br /><br />Full name: Readers.yourFastqFile</div>
<div class="tip" id="fs24">val FastQSequence : seq&lt;FastQ.FastqItem&lt;obj,int []&gt;&gt;<br /><br />Full name: Readers.FastQSequence</div>
<div class="tip" id="fs25">module FastQ<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs26">val fromFile : converter:(string -&gt; &#39;a) -&gt; qualityConverter:(string -&gt; &#39;b) -&gt; filePath:string -&gt; seq&lt;FastQ.FastqItem&lt;&#39;a,&#39;b&gt;&gt;<br /><br />Full name: BioFSharp.IO.FastQ.fromFile</div>
<div class="tip" id="fs27">val filepathGFF : string<br /><br />Full name: Readers.filepathGFF<br /><em><br /><br />path and name of the input file</em></div>
<div class="tip" id="fs28">val fastaConverter : x:seq&lt;char&gt; -&gt; seq&lt;char&gt;<br /><br />Full name: Readers.fastaConverter<br /><em><br /><br />converter for FastA.FastaItem</em></div>
<div class="tip" id="fs29">val x : seq&lt;char&gt;</div>
<div class="tip" id="fs30">Multiple items<br />val seq : sequence:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;T&gt;<br /><br />Full name: Microsoft.FSharp.Core.Operators.seq<br /><br />--------------------<br />type seq&lt;&#39;T&gt; = Collections.Generic.IEnumerable&lt;&#39;T&gt;<br /><br />Full name: Microsoft.FSharp.Collections.seq&lt;_&gt;</div>
<div class="tip" id="fs31">Multiple items<br />val char : value:&#39;T -&gt; char (requires member op_Explicit)<br /><br />Full name: Microsoft.FSharp.Core.Operators.char<br /><br />--------------------<br />type char = Char<br /><br />Full name: Microsoft.FSharp.Core.char</div>
<div class="tip" id="fs32">val id : x:&#39;T -&gt; &#39;T<br /><br />Full name: Microsoft.FSharp.Core.Operators.id</div>
<div class="tip" id="fs33">val gffExampleRead : seq&lt;GFF3Parser.GFF&lt;seq&lt;char&gt;&gt;&gt;<br /><br />Full name: Readers.gffExampleRead<br /><em><br /><br />To read in a gff file you have to insert a FastA converter and a source filepath. <br />If no FastA Sequence is included you directly can use GFF3Parser.Reader.GFF3readerWithoutFasta filepathGFF</em></div>
<div class="tip" id="fs34">module GFF3Parser<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs35">module Reader<br /><br />from BioFSharp.IO.GFF3Parser</div>
<div class="tip" id="fs36">val GFF3reader : fastAconverter:(seq&lt;char&gt; -&gt; &#39;a) -&gt; filepath:string -&gt; seq&lt;GFF3Parser.GFF&lt;&#39;a&gt;&gt;<br /><br />Full name: BioFSharp.IO.GFF3Parser.Reader.GFF3reader</div>
<div class="tip" id="fs37">val gffExampleSanityCheckSimple : unit<br /><br />Full name: Readers.gffExampleSanityCheckSimple<br /><em><br /><br />to validate the gff file without SOTerm verification use this function and only insert filepath</em></div>
<div class="tip" id="fs38">module SanityCheck<br /><br />from BioFSharp.IO.GFF3Parser</div>
<div class="tip" id="fs39">val sanityCheck : filepath:string -&gt; unit<br /><br />Full name: BioFSharp.IO.GFF3Parser.SanityCheck.sanityCheck</div>
<div class="tip" id="fs40">val filepathSO_Terms : string<br /><br />Full name: Readers.filepathSO_Terms<br /><em><br /><br />path, name and version of the &#39;Sequence Ontology terms&#39;-file</em></div>
<div class="tip" id="fs41">val gffExampleSanityCheck : unit<br /><br />Full name: Readers.gffExampleSanityCheck<br /><em><br /><br />to validate the gff file insert filepath</em></div>
<div class="tip" id="fs42">val sanityCheckWithSOTerm : so_TermsPath:string -&gt; filepath:string -&gt; unit<br /><br />Full name: BioFSharp.IO.GFF3Parser.SanityCheck.sanityCheckWithSOTerm</div>
<div class="tip" id="fs43">val searchterm : string<br /><br />Full name: Readers.searchterm<br /><em><br /><br />Term to search for:</em></div>
<div class="tip" id="fs44">val gffExampleSearch : seq&lt;GFF3Parser.GFFEntry&gt;<br /><br />Full name: Readers.gffExampleSearch<br /><em><br /><br />with this function you can search for a searchterm and you will get features which are related</em></div>
<div class="tip" id="fs45">module Relationship<br /><br />from BioFSharp.IO.GFF3Parser</div>
<div class="tip" id="fs46">val relationshipSearch : gffList:seq&lt;GFF3Parser.GFF&lt;&#39;a&gt;&gt; -&gt; searchterm:string -&gt; seq&lt;GFF3Parser.GFFEntry&gt;<br /><br />Full name: BioFSharp.IO.GFF3Parser.Relationship.relationshipSearch</div>
<div class="tip" id="fs47">val gffExampleWrite : unit<br /><br />Full name: Readers.gffExampleWrite<br /><em><br /><br />Takes a seq&lt;GFF&lt;&#39;a&gt;&gt;, a FastA converter and a destination filepath and writes it into a .gff.</em></div>
<div class="tip" id="fs48">module Writer<br /><br />from BioFSharp.IO.GFF3Parser</div>
<div class="tip" id="fs49">val GFF3Writer : input:seq&lt;GFF3Parser.GFF&lt;seq&lt;&#39;a&gt;&gt;&gt; -&gt; converter:(&#39;a -&gt; char) -&gt; path:string -&gt; unit<br /><br />Full name: BioFSharp.IO.GFF3Parser.Writer.GFF3Writer</div>

        </div>
        <div class="span3">
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
          <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
            <li class="nav-header">BioFSharp</li>
            <li><a href="/BioFSharp/index.html">Home page</a></li>
            <li class="divider"></li>
            <li><a href="http://nuget.org/packages/BioFSharp">Get Library via NuGet</a></li>
            <li><a href="http://github.com/CSBiology/BioFSharp">Source Code on GitHub</a></li>
            <li><a href="/BioFSharp/license.html">License</a></li>
            <li><a href="/BioFSharp/release-notes.html">Release Notes</a></li>
            
            <li class="nav-header">Getting started</li>
              
              <li><a href="/BioFSharp/BioSequence.html">BioSequence</a></li>
			  <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
			  <li><a href="/BioFSharp/BoyerMooreUsage.html">BoyerMoore</a></li>
			  <li><a href="/BioFSharp/Readers.html">I/O</a>
				  <ul class="nav nav-list">
					<li><a href="/BioFSharp/Readers.html#FastQ">FastQ</a></li>
					<li><a href="/BioFSharp/Readers.html#FastA">FastA</a></li>
					<li><a href="/BioFSharp/Readers.html#GFF3">GFF3</a></li>
				  </ul>
			  </li>
              <li class="divider"></li>
              <li><a href="/BioFSharp/CentroidSpectra.html">Spectrum centroidization</a></li>
              <li><a href="/BioFSharp/ChargeStateDetermination.html">Charge state determination</a></li>
              <li><a href="/BioFSharp/PeptideLookUp.html">Peptide look up</a></li>
              <li><a href="/BioFSharp/PeptideIdentification.html">Peptide identification</a></li>
			  <li><a href="/BioFSharp/MoleculeFinding.html">Molecule Finding</a>
              <li class="divider"></li>
              <li><a href="/BioFSharp/BlastWrapper.html">Use NCBI-Blast</a></li>
            <li class="nav-header">Documentation</li>
            <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>
          </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
