<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Parsing SOFT formatted family files
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="An open source bioinformatics toolbox written in F#. <https://csbiology.github.io/BioFSharp/>"/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <meta http-equiv="X-UA-Compatible" content="IE=11" >
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" /> <!--/BioFSharp-->
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" /> <!--/BioFSharp-->
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script> <!--/BioFSharp-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
	<body>
		<div class="container">
		  <div class="masthead">
			<ul class="nav nav-pills pull-right">
			  <li><a href="http://fsharp.org">fsharp.org</a></li>
			  <li><a href="https://github.com/CSBiology/BioFSharp">github page</a></li>
			</ul>
			<h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
		  </div>
		  <hr />
		  <div class="row">
			<div class="span9" id="main">
			  
<h1><a name="Parsing-SOFT-formatted-family-files" class="anchor" href="#Parsing-SOFT-formatted-family-files">Parsing SOFT formatted family files</a></h1>
<p>GEO (Gene Expression Omnibus) is a data repository of high-throughput gene expression and hybridization array data. All record metadata are provided in the <a href="https://www.ncbi.nlm.nih.gov/geo/info/soft.html">soft format</a> ,
which can be parsed and analyzed using the <code>BioFSharp.IO.SOFT</code> module.</p>
<p>SOFT types are very large and hard to read from standard FSI output, especially the nested top level types, so we will omitt the standard output for readability and use
the FSI printers provided in <code>BioFSharp.IO.FSIPrinters</code>.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
<span class="l">9: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 3)" onmouseover="showTip(event, 'fs1', 3)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="id">IO</span>

<span class="c">//register the desired printers</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 5)" onmouseover="showTip(event, 'fs4', 5)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs5', 6)" onmouseover="showTip(event, 'fs5', 6)" class="id">prettyPrintGPL</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 7)" onmouseover="showTip(event, 'fs4', 7)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 8)" onmouseover="showTip(event, 'fs6', 8)" class="id">prettyPrintGSE</span><span class="pn">)</span>

<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 9)" onmouseover="showTip(event, 'fs4', 9)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs7', 10)" onmouseover="showTip(event, 'fs7', 10)" class="id">prettyPrintSampleRecord</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 11)" onmouseover="showTip(event, 'fs4', 11)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 12)" onmouseover="showTip(event, 'fs8', 12)" class="id">prettyPrintSeriesRecord</span><span class="pn">)</span>
<span class="id">fsi</span><span class="pn">.</span><span class="id">AddPrinter</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs4', 13)" onmouseover="showTip(event, 'fs4', 13)" class="id">FSIPrinters</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs9', 14)" onmouseover="showTip(event, 'fs9', 14)" class="id">prettyPrintPlatformRecord</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Parsing-platform-GPL-files" class="anchor" href="#Parsing-platform-GPL-files">Parsing platform (GPL) files</a></h2>
<p>Soft formatted platform family files can be parsed using the <code>SOFT.Platform.fromFile</code> function.</p>
<p>As the format (.soft) does not specify the type of record (GSE or GPL), please make sure that you only
parse GPL*.soft files with this functions, as other files may return errors</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="id">testPlatform</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs11', 16)" onmouseover="showTip(event, 'fs11', 16)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 17)" onmouseover="showTip(event, 'fs12', 17)" class="m">Platform</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 18)" onmouseover="showTip(event, 'fs13', 18)" class="id">fromFile</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/GPL15922_family.soft&quot;</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Parsing-series-GSE-files" class="anchor" href="#Parsing-series-GSE-files">Parsing series (GSE) files</a></h2>
<p>Soft formatted series family files can be parsed using the <code>SOFT.Platform.fromFile</code> function.</p>
<p>As the format (.soft) does not specify the type of record (GSE or GPL), please make sure that you only
parse GSE*.soft files with this functions, as other files may return errors</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 19)" onmouseover="showTip(event, 'fs14', 19)" class="id">testSeries</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs11', 20)" onmouseover="showTip(event, 'fs11', 20)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs15', 21)" onmouseover="showTip(event, 'fs15', 21)" class="m">Series</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 22)" onmouseover="showTip(event, 'fs16', 22)" class="id">fromFile</span> <span class="pn">(</span><span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/GSE71469_family.soft&quot;</span><span class="pn">)</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Convenience-functions" class="anchor" href="#Convenience-functions">Convenience functions</a></h2>
<p>We implemented some convenience functions for <code>SOFT.Platform.GPL</code> and <code>SOFT.Series.GSE</code>.</p>
<p><code>Platform.getAssociatedSampleAccessions</code> for example retrieves all associated sample accessions. This could be usefull for batch downloads of these files.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 23)" onmouseover="showTip(event, 'fs17', 23)" class="id">sampleAccessions</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs10', 24)" onmouseover="showTip(event, 'fs10', 24)" class="id">testPlatform</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs11', 25)" onmouseover="showTip(event, 'fs11', 25)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 26)" onmouseover="showTip(event, 'fs12', 26)" class="m">Platform</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 27)" onmouseover="showTip(event, 'fs18', 27)" class="id">getAssociatedSampleAccessions</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>seq ["GSM1087792"; "GSM1087793"; "GSM1087794"; "GSM1087795"; ...]</code></pre></td></tr></table>
<p>This can be especially usefull to retrieve all samples that are associated with this platform (e.g. for example for meta analysis of the files).</p>
<p>The full sample records can also be retrieved, which makes it possible to access even more metadata:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs19', 28)" onmouseover="showTip(event, 'fs19', 28)" class="id">relations</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs10', 29)" onmouseover="showTip(event, 'fs10', 29)" class="id">testPlatform</span> 
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs11', 30)" onmouseover="showTip(event, 'fs11', 30)" class="m">SOFT</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs12', 31)" onmouseover="showTip(event, 'fs12', 31)" class="m">Platform</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 32)" onmouseover="showTip(event, 'fs20', 32)" class="id">getAssociatedSamples</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs21', 33)" onmouseover="showTip(event, 'fs21', 33)" class="m">List</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 34)" onmouseover="showTip(event, 'fs22', 34)" class="id">map</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs23', 35)" onmouseover="showTip(event, 'fs23', 35)" class="id">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs23', 36)" onmouseover="showTip(event, 'fs23', 36)" class="id">x</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 37)" onmouseover="showTip(event, 'fs24', 37)" class="id">Relation</span><span class="pn">)</span>
    <span class="c">//showing only the first 5 relations for ease of view</span>
    <span class="o">|&gt;</span> <span class="k">fun</span> <span onmouseout="hideTip(event, 'fs25', 38)" onmouseover="showTip(event, 'fs25', 38)" class="id">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs25', 39)" onmouseover="showTip(event, 'fs25', 39)" class="fn">x</span><span class="pn">.</span><span class="pn">[</span><span class="n">0..</span><span class="n">4</span><span class="pn">]</span>
</code></pre></td>
</tr>
</table>
<table class="pre"><tr><td><pre><code>[seq
   ["BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN01924672";
    "SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX245324"];
 seq
   ["BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN01924673";
    "SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX245325"];
 seq
   ["BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN01924674";
    "SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX245326"];
 seq
   ["BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN01924675";
    "SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX245327"];
 seq
   ["BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN01924676";
    "SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX245328"]]</code></pre></td></tr></table>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace FSharpAux</div>
<div class="tip" id="fs3">namespace BioFSharp.IO</div>
<div class="tip" id="fs4">module FSIPrinters<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs5">val prettyPrintGPL : gpl:Platform.GPL -&gt; string</div>
<div class="tip" id="fs6">val prettyPrintGSE : gse:Series.GSE -&gt; string</div>
<div class="tip" id="fs7">val prettyPrintSampleRecord : sample:SampleRecord -&gt; string</div>
<div class="tip" id="fs8">val prettyPrintSeriesRecord : series:SeriesRecord -&gt; string</div>
<div class="tip" id="fs9">val prettyPrintPlatformRecord : platform:PlatformRecord -&gt; string</div>
<div class="tip" id="fs10">val testPlatform : Platform.GPL</div>
<div class="tip" id="fs11">module SOFT<br /><br />from BioFSharp.IO</div>
<div class="tip" id="fs12">module Platform<br /><br />from BioFSharp.IO.SOFT</div>
<div class="tip" id="fs13">val fromFile : path:string -&gt; Platform.GPL</div>
<div class="tip" id="fs14">val testSeries : Series.GSE</div>
<div class="tip" id="fs15">module Series<br /><br />from BioFSharp.IO.SOFT</div>
<div class="tip" id="fs16">val fromFile : path:string -&gt; Series.GSE</div>
<div class="tip" id="fs17">val sampleAccessions : string list</div>
<div class="tip" id="fs18">val getAssociatedSampleAccessions : gpl:Platform.GPL -&gt; string list</div>
<div class="tip" id="fs19">val relations : string list list</div>
<div class="tip" id="fs20">val getAssociatedSamples : gpl:Platform.GPL -&gt; SampleRecord list</div>
<div class="tip" id="fs21">Multiple items<br />module List<br /><br />from FSharpAux<br /><br />--------------------<br />module List<br /><br />from Microsoft.FSharp.Collections<br /><br />--------------------<br />type List&lt;&#39;T&gt; =<br />&#160;&#160;| ( [] )<br />&#160;&#160;| ( :: ) of Head: &#39;T * Tail: &#39;T list<br />&#160;&#160;&#160;&#160;interface IReadOnlyList&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IReadOnlyCollection&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;interface IEnumerable<br />&#160;&#160;&#160;&#160;interface IEnumerable&lt;&#39;T&gt;<br />&#160;&#160;&#160;&#160;member GetSlice : startIndex:int option * endIndex:int option -&gt; &#39;T list<br />&#160;&#160;&#160;&#160;member Head : &#39;T<br />&#160;&#160;&#160;&#160;member IsEmpty : bool<br />&#160;&#160;&#160;&#160;member Item : index:int -&gt; &#39;T with get<br />&#160;&#160;&#160;&#160;member Length : int<br />&#160;&#160;&#160;&#160;member Tail : &#39;T list<br />&#160;&#160;&#160;&#160;...</div>
<div class="tip" id="fs22">val map : mapping:(&#39;T -&gt; &#39;U) -&gt; list:&#39;T list -&gt; &#39;U list</div>
<div class="tip" id="fs23">val x : SampleRecord</div>
<div class="tip" id="fs24">SampleRecord.Relation: string list</div>
<div class="tip" id="fs25">val x : string list list</div>

			</div>
			<div class="span3" >
			  <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
              <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                  <li class="nav-header">BioFSharp</li>
                  <li><a href="/BioFSharp/index.html">Home page</a></li>
                  <li class="divider"></li>
                  <li><a href="/BioFSharp/release-notes.html">Release Notes</a></li>
                  <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                  
                  <li class="nav-header">BioFSharp core</li>
                  <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                  <li>
                      <a href="/BioFSharp/BioItem.html">BioItems</a>
                  </li>
                  <li><a href="/BioFSharp/BioCollections.html">BioCollections</a></li>
                  <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                  <li class="nav-header">BioContainers</li>
                  <li><a href="/BioFSharp/BioContainers.html">Introduction</a></li>
                  <li><a href="/BioFSharp/BioContainersDesignGuide.html">Container API Design Guide</a></li>
                  <li class="nav-header">Algorithm</li>
                  <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                  <li><a href="/BioFSharp/ClustalOWrapper.html">Clustal Omega</a></li>
                  <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                  <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                  <li>
                      <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                  </li>
                  <li class="nav-header">File I/O</li>
                  <li>
                      <a href="/BioFSharp/Clustal.html">Clustal</a>
                  </li>
                  <li><a href="/BioFSharp/CSV.html">CSV / Tabular</a></li>
                  <li>
                      <a href="/BioFSharp/FastA.html">FastA</a>
                  </li>
                  <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                  <li>
                      <a href="/BioFSharp/GenBank.html">GenBank</a>
                  </li>
                  <li><a href="/BioFSharp/SOFT.html">GEO Metadata (SOFT)</a></li>
                  <li>
                      <a href="/BioFSharp/GFF3.html">GFF3</a>
                  </li>
                  <li>
                      <a href="/BioFSharp/Newick.html">Newick</a>
                  </li>
                  <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                  <li>
                      <a href="/BioFSharp/FSIPrinters.html">FSIPrinters</a>
                  </li>
                  <li class="nav-header">Web I/O</li>
                  <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                  <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                  <li class="nav-header">Biostatistics</li>
                  <li><a href="/BioFSharp/GSEA.html">Ontology enrichment</a></li>
                  <li class="nav-header">Biovisualization</li>
                  <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                  <li class="nav-header">GPU Computing</li>
                  <li><a href="/BioFSharp/ParallelAlignment.html">Sequence Alignment</a></li>
                  <li class="nav-header">Documentation</li>
                  <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                  
              </ul>
			</div>
		  </div>
		</div>
		<a href="https://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
	  </body>
  </html>
