<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>String matching algorithms
</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="BioFSharp aims to be a user-friendly library for Bioinformatics written in F# as the successor official of FSharpBio."/>
    <meta name="author" content="Timo Mühlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" />
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" />
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="container">
      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="http://fsharp.org">fsharp.org</a></li>
          <li><a href="http://github.com/CSBiology/BioFSharp">github page</a></li>
        </ul>
        <h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
      </div>
      <hr />
      <div class="row">
        <div class="span9" id="main">
          
<h1><a name="String-matching-algorithms" class="anchor" href="#String-matching-algorithms">String matching algorithms</a></h1>
<a id="SourceCode" href="https://github.com/CSBiology/BioFSharp/blob/master/src/BioFSharp/Algorithm/PatternQuery.fs">&lt;/&gt;view source code</a>
<a id="Author" href="https://github.com/kMutagene">&#128366;view author of this tutorial</a>
<br><br>
<p>String matching algorithms are concerned with finding a single or multiple matches of a query pattern within a source. The sub-modules of the <code>BioFSharp.Algorithm.StringMatching</code> module are organized as following:</p>
<ul>
<li><p>AlgorithmName.<strong>findAll</strong> : returns the (zero-based) starting position of all matches of the query pattern in the source as an integer list</p></li>
<li><p>AlgorithmName.<strong>find</strong> : returns the (zero-based) starting position of the first match of the query pattern in the source as an integer</p></li>
<li><p>AlgorithmName.<strong>findFrom</strong> : returns the (zero-based) starting position of the first match of the query pattern in the source as an integer, starting from a specified (zero-based) position in the source</p></li>
</ul>
<p>All algorithms contained in this library are implemented as generic as possible regarding the data type that the input source and query pattern can contain, being only restricted by implementing the <code>IEquatable</code> interface, as there must be a way to assure that two equal elements are indeed equal.
However, the nature of the specific algorithm may impose additional restrictions, for example concerning the size of a single comparable item.</p>
<p>The type of query pattern and source has been constrained to be an <code>array&lt;'a&gt;</code> because many index accessions are made during the search.</p>
<p>Runtimes are provided using <a href="https://en.wikipedia.org/wiki/Big_O_notation#Family_of_Bachmann.E2.80.93Landau_notations">Bachmann-Landau notations</a>:</p>
<ul>
<li>n : length of the source</li>
<li>m : length of the query pattern</li>
<li>k : size of the Alphabet</li>
</ul>
<p>The following query pattern and source will be used in all further examples demonstrating how to use the specific module.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///source containing the query pattern at positions 0,10 and 17 </span>
<span class="k">let</span> <span class="i">source</span> <span class="o">=</span> <span class="s">&quot;AAABBBBBBBAAABBBBAAA&quot;</span> <span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>

<span class="c">///query pattern to find in the source</span>
<span class="k">let</span> <span class="i">queryPattern</span> <span class="o">=</span> <span class="s">&quot;AAA&quot;</span><span class="o">|&gt;</span> <span class="i">BioArray</span><span class="o">.</span><span class="i">ofAminoAcidString</span>
</code></pre></td>
</tr>
</table>
<hr></hr>
<br>
<h1><a name="The-Naive-Algorithm" class="anchor" href="#The-Naive-Algorithm">The Naive Algorithm</a></h1>
<p>The naive approach to the string matching problem is walking through the source starting from the beginning and checking at each position if the resulting substring equals the query pattern.
While being inefficient, it may be beneficial to use it in cases where the speed advantage of another algorithm is neglegible or does not outhweigh the additional setup needed (for example if your source and query pattern are really short)</p>
<h2><a name="Usage" class="anchor" href="#Usage">Usage</a></h2>
<p>No additional setup needed. Just use query pattern and source as input parameters for the respective function</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//returns [0;10;17]</span>
<span class="i">Naive</span><span class="o">.</span><span class="i">findAll</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 0</span>
<span class="i">Naive</span><span class="o">.</span><span class="i">find</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 17</span>
<span class="i">Naive</span><span class="o">.</span><span class="i">findFrom</span> <span class="n">12</span> <span class="i">queryPattern</span> <span class="i">source</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Runtime" class="anchor" href="#Runtime">Runtime</a></h2>
<ul>
<li><p>Preprocessing: <em>None</em></p></li>
<li><p>Matching time: <em>O(nm)</em></p></li>
</ul>
<hr></hr>
<br>
<h1><a name="Knuth-Morris-Pratt" class="anchor" href="#Knuth-Morris-Pratt">Knuth Morris Pratt</a></h1>
<p>The KMP algorithm makes use of previous match information to determine an amount of skips that can be made until the next position in the source gets examined as a possible match.
To achieve that, a prefix table (or failure function) of the pattern needs to be computed, which determines the amount of skippable elements depending on the previous (partial) match.</p>
<h2><a name="Usage-1" class="anchor" href="#Usage-1">Usage</a></h2>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///prefix table computed from the query pattern</span>
<span class="k">let</span> <span class="i">prefixTable</span> <span class="o">=</span> <span class="i">KnuthMorrisPratt</span><span class="o">.</span><span class="i">createPrefixTable</span> <span class="i">queryPattern</span>

<span class="c">//returns [0;10;17]</span>
<span class="i">KnuthMorrisPratt</span><span class="o">.</span><span class="i">findAll</span> <span class="i">prefixTable</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 0</span>
<span class="i">KnuthMorrisPratt</span><span class="o">.</span><span class="i">find</span> <span class="i">prefixTable</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 17</span>
<span class="i">KnuthMorrisPratt</span><span class="o">.</span><span class="i">findFrom</span> <span class="i">prefixTable</span> <span class="i">queryPattern</span> <span class="n">12</span> <span class="i">source</span>
</code></pre></td>
</tr>
</table>
<h2><a name="Runtime-1" class="anchor" href="#Runtime-1">Runtime</a></h2>
<ul>
<li><p>Preprocessing: <em>Θ(m)</em></p></li>
<li><p>Matching time: average: <em>Θ(n)</em></p></li>
<li><p>Worst case(if the prefix table results in no possible skips): <em>Θ(mn)</em></p></li>
</ul>
<hr></hr>
<br>
<h1><a name="Rabin-Karp" class="anchor" href="#Rabin-Karp">Rabin Karp</a></h1>
<p>The RK Algorithm saves time by not comparing every single element of pattern and a substring of the same length from the source. Instead, it compares hashed versions of them.
The RabinKarp module contains two submodules:</p>
<ul>
<li><p>RKStandard: uses the build in .NET hash function <code>hash</code></p></li>
<li>
<p>CP: uses a cyclic polynomial hash, which is way faster than the built in hash function. Elements must be castable to <code>uint64</code> , as they get hashed as a 64 bit unsigned integer and updating the hash uses bitwise operators.
Using large query patterns is problematic, as the combined hash may exceed the size of an <code>uint64</code>.</p>
</li>
</ul>
<h2><a name="Usage-2" class="anchor" href="#Usage-2">Usage</a></h2>
<h3><a name="RKStandard" class="anchor" href="#RKStandard">RKStandard</a></h3>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
<span class="l">4: </span>
<span class="l">5: </span>
<span class="l">6: </span>
<span class="l">7: </span>
<span class="l">8: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//returns [0;10;17]</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">RKStandard</span><span class="o">.</span><span class="i">findAll</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 0</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">RKStandard</span><span class="o">.</span><span class="i">find</span> <span class="i">queryPattern</span> <span class="i">source</span>

<span class="c">//returns 17</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">RKStandard</span><span class="o">.</span><span class="i">findFrom</span> <span class="n">12</span> <span class="i">queryPattern</span> <span class="i">source</span>
</code></pre></td>
</tr>
</table>
<h3><a name="CP" class="anchor" href="#CP">CP</a></h3>
<p>As the AminoAcid type cannot be casted to an <code>uint64</code>, a conversion to their one character code is needed</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///source converted to char array </span>
<span class="k">let</span> <span class="i">source&#39;</span> <span class="o">=</span> <span class="i">source</span> <span class="o">|&gt;</span> <span class="i">Array</span><span class="o">.</span><span class="i">map</span> (<span class="k">fun</span> <span class="i">x</span> <span class="k">-&gt;</span> (<span class="i">x</span> <span class="o">:&gt;</span> <span class="i">IBioItem</span>)<span class="o">.</span><span class="i">Symbol</span>)

<span class="c">///query pattern converted to char array</span>
<span class="k">let</span> <span class="i">queryPattern&#39;</span> <span class="o">=</span> <span class="i">queryPattern</span> <span class="o">|&gt;</span> <span class="i">Array</span><span class="o">.</span><span class="i">map</span> (<span class="k">fun</span> <span class="i">x</span> <span class="k">-&gt;</span> (<span class="i">x</span> <span class="o">:&gt;</span> <span class="i">IBioItem</span>)<span class="o">.</span><span class="i">Symbol</span>)

<span class="c">//returns [0;10;17]</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">CP</span><span class="o">.</span><span class="i">findAll</span> <span class="i">queryPattern&#39;</span> <span class="i">source&#39;</span>

<span class="c">//returns 0</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">CP</span><span class="o">.</span><span class="i">find</span> <span class="i">queryPattern&#39;</span> <span class="i">source&#39;</span>

<span class="c">//returns 17</span>
<span class="i">RabinKarp</span><span class="o">.</span><span class="i">CP</span><span class="o">.</span><span class="i">findFrom</span> <span class="n">12</span> <span class="i">queryPattern&#39;</span> <span class="i">source&#39;</span>
</code></pre></td>
</tr>
</table>
<h3><a name="Using-your-own-hash-functions" class="anchor" href="#Using-your-own-hash-functions">Using your own hash functions</a></h3>
<p>The <code>RabinKarp</code> module provides generic <code>findAll</code>, <code>find</code> and <code>findFrom</code> functions that take the following additional parameters:</p>
<ul>
<li><p><code>blockHash</code>: a function that hashes an entire array (used to hash the pattern and the first substring of the source)</p></li>
<li><p><code>updateHash</code>: a function that removes an element from an existing hash and adds a new one</p></li>
</ul>
<p>Just use functional composition to build your own findX functions.
In fact the two provided algorithms are built the same way:</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
<span class="l">2: </span>
<span class="l">3: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">///RabinKarp.CP.findAll built by functional composition</span>
<span class="k">let</span> <span class="k">inline</span> <span class="i">findAll</span> (<span class="i">pattern</span> <span class="o">:</span> <span class="i">array</span><span class="o">&lt;</span><span class="o">&#39;</span><span class="i">a</span><span class="o">&gt;</span>) (<span class="i">s</span> <span class="o">:</span> <span class="i">array</span><span class="o">&lt;</span><span class="o">&#39;</span><span class="i">a</span><span class="o">&gt;</span>) <span class="o">=</span> 
    <span class="i">RabinKarp</span><span class="o">.</span><span class="i">findAllGeneric</span> (<span class="i">RabinKarp</span><span class="o">.</span><span class="i">CP</span><span class="o">.</span><span class="i">updateHash</span> <span class="i">pattern</span><span class="o">.</span><span class="i">Length</span>) (<span class="i">RabinKarp</span><span class="o">.</span><span class="i">CP</span><span class="o">.</span><span class="i">blockHash</span>) <span class="i">pattern</span> <span class="i">s</span> 
</code></pre></td>
</tr>
</table>
<h2><a name="Runtime-2" class="anchor" href="#Runtime-2">Runtime</a></h2>
<ul>
<li>Preprocessing: <em>Θ(m)</em></li>
<li>Matching time: average <em>Θ(n + m)</em></li>
<li>Worst case: Θ((n−m)m)</li>
</ul>
<hr></hr>
<br>
<h1><a name="Boyer-Moore" class="anchor" href="#Boyer-Moore">Boyer Moore</a></h1>
<a id="Author" href="https://github.com/WieczorekE">&#128366;view author of this tutorial</a>
<h2><a name="Usage-3" class="anchor" href="#Usage-3">Usage</a></h2>
<p>This part of the documentation gives a short introduction into the Boyer Moore string search algorithm which is particulary desined to work on IBioItems.This algorithm searches a query within a longer source. In using two arrays that are created using the bad character rule and the good suffix heuristics the best shifting value is determined. This accounts for higher shifts which prevents unnecessary comparisons to save time.</p>
<p>The following code will return an array created out of all the proteins that are contained in the fasta file. First, the Symbols in the file get converted
into amino acids and then they get concatenated and transformed into an array.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span class="i">chlamyproteins</span> <span class="o">=</span> 
    <span class="k">let</span> <span class="i">converter</span> <span class="i">letters</span> <span class="o">:</span> <span class="i">BioList</span><span class="o">.</span><span class="i">BioList</span><span class="o">&lt;</span><span class="i">AminoAcid</span><span class="o">&gt;</span> <span class="o">=</span> 
        <span class="i">letters</span>
            <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">map</span> ( <span class="k">fun</span> <span class="i">x</span> <span class="k">-&gt;</span> <span class="k">match</span> (<span class="i">BioFSharp</span><span class="o">.</span><span class="i">AminoAcids</span><span class="o">.</span><span class="i">charToParsedAminoAcidChar</span> <span class="i">x</span>) <span class="k">with</span>
                                  | <span class="i">ParsedAminoAcidChar</span><span class="o">.</span><span class="i">StandardCodes</span>  <span class="i">v</span> <span class="k">-&gt;</span> <span class="i">v</span>
                                  | <span class="i">ParsedAminoAcidChar</span><span class="o">.</span><span class="i">AmbiguityCodes</span> <span class="i">v</span> <span class="k">-&gt;</span> <span class="i">v</span>
                                  | <span class="i">ParsedAminoAcidChar</span><span class="o">.</span><span class="i">GapTer</span>         <span class="i">v</span> <span class="k">-&gt;</span> <span class="i">v</span>
                                  | <span class="i">ParsedAminoAcidChar</span><span class="o">.</span><span class="i">NoAAChar</span>       <span class="i">v</span> <span class="k">-&gt;</span> <span class="i">failwith</span> <span class="s">&quot;No AA Char given&quot;</span>)
            <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">toList</span>

    <span class="i">FileIO</span><span class="o">.</span><span class="i">readFile</span> (<span class="k">__SOURCE_DIRECTORY__</span> <span class="o">+</span> <span class="s">&quot;/data/Proteins.fasta&quot;</span>)
    <span class="o">|&gt;</span> <span class="i">FastA</span><span class="o">.</span><span class="i">fromFileEnumerator</span> (<span class="i">converter</span>)
    <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">map</span> (<span class="k">fun</span> <span class="i">item</span> <span class="k">-&gt;</span> <span class="i">item</span><span class="o">.</span><span class="i">Sequence</span> <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">toArray</span>)
    <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">concat</span>
    <span class="o">|&gt;</span> <span class="i">Seq</span><span class="o">.</span><span class="i">toArray</span>
</code></pre></td>
</tr>
</table>
<p>The following code will return the first occurence of the searched query within the source.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">BoyerMoore</span><span class="o">.</span><span class="i">findFirst</span> <span class="i">chlamyproteins</span> <span class="i">chlamyproteins</span><span class="o">.</span>[<span class="n">16988..</span><span class="n">17978</span>]
</code></pre></td>
</tr>
</table>
<p>The following code will return the first occurence of the query after a given staring point.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">BoyerMoore</span><span class="o">.</span><span class="i">findFrom</span> <span class="n">7</span> <span class="i">chlamyproteins</span> <span class="i">chlamyproteins</span><span class="o">.</span>[<span class="n">16988..</span><span class="n">17978</span>]
</code></pre></td>
</tr>
</table>
<p>The following code will return the all the occurence of the query within the given source.</p>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="i">BoyerMoore</span><span class="o">.</span><span class="i">findAll</span> <span class="i">chlamyproteins</span> <span class="i">chlamyproteins</span><span class="o">.</span>[<span class="n">16988..</span><span class="n">17978</span>]
</code></pre></td>
</tr>
</table>
<h2><a name="Runtime-3" class="anchor" href="#Runtime-3">Runtime</a></h2>
<ul>
<li>Preprocessing: <em>Θ(m+k)</em></li>
<li>Matching time: best <em>Ω(n/m)</em></li>
<li>Worst Case: <em>O(nm)</em></li>
</ul>
<h1><a name="Speed-Comparison" class="anchor" href="#Speed-Comparison">Speed Comparison</a></h1>
<p>Coming soon...</p>


        </div>
        <div class="span3" >
          <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
            <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                <li class="nav-header">BioFSharp</li>
                <li><a href="/BioFSharp/index.html">Home page</a></li>
                <li class="divider"></li>
                <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                
                <li class="nav-header">BioFSharp core</li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li>
                    <a href="/BioFSharp/IBioItem.html">BioItems</a>
                </li>
                <li>
                    <a href="/BioFSharp/BioCollections.html">BioCollections</a>
                </li>
                <li><a href="/BioFSharp/SeqProperties.html">Sequence Properties</a></li>
                <li class="nav-header">Algorithm</li>
                <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                <li>
                    <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                        <ul class="nav nav-list">
                            <li><a href="/BioFSharp/StringMatching.html#The-Naive-Algorithm">Naive</a></li>
                            <li><a href="/BioFSharp/StringMatching.html#Knuth-Morris-Pratt">KnuthMorrisPratt</a></li>
                            <li><a href="/BioFSharp/StringMatching.html#Rabin-Karp">RabinKarp</a></li>
                            <li><a href="/BioFSharp/StringMatching.html#Boyer-Moore">BoyerMoore</a></li>
                        </ul>
                </li>
                <li class="nav-header">File I/O</li>
                <li><a href="/BioFSharp/FastA.html">FastA</a></li>
                <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                <li><a href="/BioFSharp/GFF.html">GFF format</a></li>
                <li><a href="/BioFSharp/Genebank.html">Genebank</a></li>
                <li><a href="/BioFSharp/Clustal.html">Clustal</a></li>
                <li><a href="/BioFSharp/Newick.html">Newick</a></li>                
                <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                <li><a href="/BioFSharp/Formula.html">Chemical formula</a></li>
                <li class="nav-header">Web I/O</li>
                <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                <li class="nav-header">Biostatistics</li>
                <li><a href="/BioFSharp/GESA.html">Ontology enrichment</a></li>
                <li class="nav-header">Biovisualization</li>
                <li><a href="/BioFSharp/GESA.html">Web logo</a></li>

                
            </ul>
        </div>
      </div>
    </div>
    <a href="http://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
  </body>
  </html>
