<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>GSEA</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content="An open source bioinformatics toolbox written in F#. <https://csbiology.github.io/BioFSharp/>"/>
    <meta name="author" content="Timo Mï¿½hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <meta http-equiv="X-UA-Compatible" content="IE=11" >
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="https://CSBiology.github.io/BioFSharp//content/style.css" /> <!--https://CSBiology.github.io/BioFSharp/-->
    <link type="text/css" rel="stylesheet" href="https://CSBiology.github.io/BioFSharp//style/CSBstyle.css" /> <!--https://CSBiology.github.io/BioFSharp/-->
    <script type="text/javascript" src="https://CSBiology.github.io/BioFSharp//content/tips.js"></script> <!--https://CSBiology.github.io/BioFSharp/-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
	<body>
		<div class="container">
		  <div class="masthead">
			<ul class="nav nav-pills pull-right">
			  <li><a href="http://fsharp.org">fsharp.org</a></li>
			  <li><a href="https://github.com/CSBiology/BioFSharp">github page</a></li>
			</ul>
			<h3 class="muted"><a href="https://CSBiology.github.io/BioFSharp//index.html">BioFSharp</a></h3>
		  </div>
		  <hr />
		  <div class="row">
			<div class="span9" id="main">
			  <table class="HeadAPI">
<td class="Head"><h1>Gene set enrichment analysis</h1></td>
<td class="API">
    <a id="APILink" href="https://csbiology.github.io/BioFSharp/reference/Ontology_enrichment.html" >&#128194;View module documentation</a>
</td>
</table>
<p>When performing high-troughput experiments the question arises how to analyze the resulting huge data sets.
Of course the analysis of each single item separately is possible but unefficient and cumberersome.
If you want to compare the transcript/protein composition of a system before and after a treatment it is useful
to make use of a gene set enrichment analysis (GSEA) to identify overrepresented transcript/protein groups.
The grouping can be based on MapMan- or GO-annotations or other shared properties.
For further information visit <a href="https://academic.oup.com/bioinformatics/article/23/4/401/181853">GSEA publication</a>.</p>
<h2><a name="Method" class="anchor" href="#Method">Method</a></h2>
<p>Imagine an experiment, in which 1000 proteins were measured. Within these 1000 proteins there are 25 proteins
of a specific functional group (heat shock response).
By treating the organism(s) 200 proteins are significantly altered. 10 of the 25 heat responsive proteins
can be found within these 200.</p>
<p>Null-Hypothesis: The number of identified altered items is due to chance.</p>
<p>Question: What is the probability, that these 10 or more proteins occuring in the 200 altered proteins by chance?</p>
<p>The hypergeomentric distribution can answer this question:</p>
<ul>
<li>N is the population size (1000 proteins)</li>
<li>K is the number of success states in the population (25 heat responsive proteins)</li>
<li>n is the number of draws (200 significantly altered proteins)</li>
<li>k is the number of observed successes (10 or higher)</li>
</ul>
<p>Result: 0.0161 = 1.6%</p>
<p>If the calculated probability is small, it is very unlikely, that the observed effect is due to chance.
This implies, that under the given condition the heat response proteins are enriched.</p>
<h2><a name="Usage" class="anchor" href="#Usage">Usage</a></h2>
<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l"> 1: </span>
<span class="l"> 2: </span>
<span class="l"> 3: </span>
<span class="l"> 4: </span>
<span class="l"> 5: </span>
<span class="l"> 6: </span>
<span class="l"> 7: </span>
<span class="l"> 8: </span>
<span class="l"> 9: </span>
<span class="l">10: </span>
<span class="l">11: </span>
<span class="l">12: </span>
<span class="l">13: </span>
<span class="l">14: </span>
<span class="l">15: </span>
<span class="l">16: </span>
<span class="l">17: </span>
<span class="l">18: </span>
<span class="l">19: </span>
<span class="l">20: </span>
<span class="l">21: </span>
<span class="l">22: </span>
<span class="l">23: </span>
<span class="l">24: </span>
<span class="l">25: </span>
<span class="l">26: </span>
<span class="l">27: </span>
<span class="l">28: </span>
<span class="l">29: </span>
<span class="l">30: </span>
<span class="l">31: </span>
<span class="l">32: </span>
<span class="l">33: </span>
<span class="l">34: </span>
<span class="l">35: </span>
<span class="l">36: </span>
<span class="l">37: </span>
<span class="l">38: </span>
<span class="l">39: </span>
<span class="l">40: </span>
<span class="l">41: </span>
<span class="l">42: </span>
<span class="l">43: </span>
<span class="l">44: </span>
<span class="l">45: </span>
<span class="l">46: </span>
<span class="l">47: </span>
<span class="l">48: </span>
<span class="l">49: </span>
<span class="l">50: </span>
<span class="l">51: </span>
<span class="l">52: </span>
<span class="l">53: </span>
<span class="l">54: </span>
<span class="l">55: </span>
</pre></td>
<td class="snippet"><pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//Usage:</span>
<span class="c">//At first all items are converted into an OntologyItem type:</span>

<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 3)" onmouseover="showTip(event, 'fs1', 3)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 4)" onmouseover="showTip(event, 'fs1', 4)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs3', 5)" onmouseover="showTip(event, 'fs3', 5)" class="id">Stats</span>

<span class="c">//Example item</span>
<span class="c">//Protein:                  HSP70</span>
<span class="c">//MapManAnnotation:         GMM:29.6.2.3;GMM:20.2.1</span>
<span class="c">//Significantly altered?:   true</span>

<span class="c">//Creates an OntologyItem with an identifier, an Ontologyterm, a group index and an arbitrary and </span>
<span class="c">//optional field. The groupIndex is based on prior analyis and defines if the items belongs to a </span>
<span class="c">//specific group. If significantly altered proteins were identified, you can assign index &#39;1&#39; to </span>
<span class="c">//all altered proteins and &#39;0&#39; to all other. The choice of the index is up to you, but you have </span>
<span class="c">//to remind it for later steps.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 6)" onmouseover="showTip(event, 'fs4', 6)" class="id">item</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 7)" onmouseover="showTip(event, 'fs5', 7)" class="m">OntologyEnrichment</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 8)" onmouseover="showTip(event, 'fs6', 8)" class="id">createOntologyItem</span> <span class="s">&quot;Cre06.g250100&quot;</span> <span class="s">&quot;GMM:29.6.2.3;GMM:20.2.1&quot;</span> <span class="n">1</span> <span class="s">&quot;HSP70&quot;</span>


<span class="c">//If more than one ontology term is possible, you can use this function to split the item, so that</span>
<span class="c">//every resulting item contains just one ontology term.</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="id">splitItems</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 10)" onmouseover="showTip(event, 'fs5', 10)" class="m">OntologyEnrichment</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 11)" onmouseover="showTip(event, 'fs8', 11)" class="id">splitMultipleAnnotationsBy</span> <span class="s">&#39;;&#39;</span> <span onmouseout="hideTip(event, 'fs4', 12)" onmouseover="showTip(event, 'fs4', 12)" class="id">item</span>

<span class="c">//create a list of all items in your experiment</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 13)" onmouseover="showTip(event, 'fs9', 13)" class="id">items</span> <span class="o">=</span> <span class="pn">[</span><span onmouseout="hideTip(event, 'fs7', 14)" onmouseover="showTip(event, 'fs7', 14)" class="id">splitItems</span><span class="c">(*...*)</span><span class="pn">]</span> <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs10', 15)" onmouseover="showTip(event, 'fs10', 15)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs11', 16)" onmouseover="showTip(event, 'fs11', 16)" class="id">concat</span>

<span class="c">//When dealing with MapMan annotations all levels of the descriptions should be considered.</span>
<span class="c">//GMM:26.6.2 -&gt; [GMM:26.6.2;GMM:26.6;GMM:26] </span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 17)" onmouseover="showTip(event, 'fs12', 17)" class="id">expandedItems</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 18)" onmouseover="showTip(event, 'fs5', 18)" class="m">OntologyEnrichment</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs13', 19)" onmouseover="showTip(event, 'fs13', 19)" class="id">expandOntologyTree</span> <span onmouseout="hideTip(event, 'fs9', 20)" onmouseover="showTip(event, 'fs9', 20)" class="id">items</span>


<span class="c">//now the dataset is ready for performing the enrichment analysis</span>
<span class="c">//deGroupIndex:         which index should be considered as success? (in this case &#39;1&#39;)</span>
<span class="c">//splitPValueThreshold: if a bin size is below this threshold, the resulting pvalue </span>
<span class="c">//                      is split up (-&gt; increased to lower the impact of small bins) (default 5).</span>
<span class="c">//minNumberInTerm:      what is the minimal number of items that should be </span>
<span class="c">//                      within a bin to consider the bin at all (default 2).</span>
<span class="c">//data:                 sequence of expanded and split ontology items</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 21)" onmouseover="showTip(event, 'fs14', 21)" class="id">gseaResult</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs5', 22)" onmouseover="showTip(event, 'fs5', 22)" class="m">OntologyEnrichment</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs15', 23)" onmouseover="showTip(event, 'fs15', 23)" class="id">CalcOverEnrichment</span> <span class="n">1</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 24)" onmouseover="showTip(event, 'fs16', 24)" class="uc">Some</span> <span class="n">5</span><span class="pn">)</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs16', 25)" onmouseover="showTip(event, 'fs16', 25)" class="uc">Some</span> <span class="n">2</span><span class="pn">)</span> <span onmouseout="hideTip(event, 'fs12', 26)" onmouseover="showTip(event, 'fs12', 26)" class="id">expandedItems</span>


<span class="k">let</span> <span onmouseout="hideTip(event, 'fs17', 27)" onmouseover="showTip(event, 'fs17', 27)" class="id">filterAndSortEnrichedModules</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs14', 28)" onmouseover="showTip(event, 'fs14', 28)" class="id">gseaResult</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs10', 29)" onmouseover="showTip(event, 'fs10', 29)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs18', 30)" onmouseover="showTip(event, 'fs18', 30)" class="id">filter</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs19', 31)" onmouseover="showTip(event, 'fs19', 31)" class="id">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs19', 32)" onmouseover="showTip(event, 'fs19', 32)" class="id">x</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs20', 33)" onmouseover="showTip(event, 'fs20', 33)" class="id">PValue</span> <span class="o">&lt;</span> <span class="n">0.05</span><span class="pn">)</span>
    <span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs10', 34)" onmouseover="showTip(event, 'fs10', 34)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs21', 35)" onmouseover="showTip(event, 'fs21', 35)" class="id">sortByDescending</span> <span class="pn">(</span><span class="k">fun</span> <span onmouseout="hideTip(event, 'fs19', 36)" onmouseover="showTip(event, 'fs19', 36)" class="id">x</span> <span class="k">-&gt;</span> <span onmouseout="hideTip(event, 'fs19', 37)" onmouseover="showTip(event, 'fs19', 37)" class="id">x</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 38)" onmouseover="showTip(event, 'fs22', 38)" class="id">NumberOfDEsInBin</span><span class="pn">)</span>

<span class="c">//the result type contains the following fields:</span>
<span class="c">//GseaResult:</span>
<span class="c">//    OntologyTerm     : OntologyTerm(MapMan-Term/GO-Term/...)</span>
<span class="c">//    ItemsInBin       : Sequence of single items belonging to the ontology term </span>
<span class="c">//    NumberOfDEsInBin : number of significantly altered items in &#39;OntologyItem&#39; bin</span>
<span class="c">//    NumberInBin      : number of items in &#39;OntologyItem&#39; bin</span>
<span class="c">//    TotalNumberOfDE  : number of significantly altered items</span>
<span class="c">//    TotalUnivers     : number of all items (expanded)</span>
<span class="c">//    PValue           : p value</span>
</code></pre></td>
</tr>
</table>

<div class="tip" id="fs1">namespace BioFSharp</div>
<div class="tip" id="fs2">namespace FSharpAux</div>
<div class="tip" id="fs3">namespace BioFSharp.Stats</div>
<div class="tip" id="fs4">val item : OntologyEnrichment.OntologyItem&lt;string&gt;</div>
<div class="tip" id="fs5">module OntologyEnrichment<br /><br />from BioFSharp.Stats</div>
<div class="tip" id="fs6">val createOntologyItem : id:string -&gt; ontologyTerm:string -&gt; groupIndex:int -&gt; item:&#39;a -&gt; OntologyEnrichment.OntologyItem&lt;&#39;a&gt;</div>
<div class="tip" id="fs7">val splitItems : seq&lt;OntologyEnrichment.OntologyItem&lt;string&gt;&gt;</div>
<div class="tip" id="fs8">val splitMultipleAnnotationsBy : separator:char -&gt; item:OntologyEnrichment.OntologyItem&lt;&#39;A&gt; -&gt; seq&lt;OntologyEnrichment.OntologyItem&lt;&#39;A&gt;&gt;</div>
<div class="tip" id="fs9">val items : seq&lt;OntologyEnrichment.OntologyItem&lt;string&gt;&gt;</div>
<div class="tip" id="fs10">Multiple items<br />module Seq<br /><br />from FSharpAux<br /><br />--------------------<br />module Seq<br /><br />from Microsoft.FSharp.Collections</div>
<div class="tip" id="fs11">val concat : sources:seq&lt;#seq&lt;&#39;T&gt;&gt; -&gt; seq&lt;&#39;T&gt;</div>
<div class="tip" id="fs12">val expandedItems : seq&lt;OntologyEnrichment.OntologyItem&lt;string&gt;&gt;</div>
<div class="tip" id="fs13">val expandOntologyTree : data:seq&lt;OntologyEnrichment.OntologyItem&lt;&#39;a&gt;&gt; -&gt; seq&lt;OntologyEnrichment.OntologyItem&lt;&#39;a&gt;&gt;</div>
<div class="tip" id="fs14">val gseaResult : seq&lt;OntologyEnrichment.GseaResult&lt;string&gt;&gt;</div>
<div class="tip" id="fs15">val CalcOverEnrichment : deGroupIndex:int -&gt; splitPvalueThreshold:int option -&gt; minNumberInTerm:int option -&gt; data:seq&lt;OntologyEnrichment.OntologyItem&lt;&#39;a&gt;&gt; -&gt; seq&lt;OntologyEnrichment.GseaResult&lt;&#39;a&gt;&gt;</div>
<div class="tip" id="fs16">union case Option.Some: Value: &#39;T -&gt; Option&lt;&#39;T&gt;</div>
<div class="tip" id="fs17">val filterAndSortEnrichedModules : seq&lt;OntologyEnrichment.GseaResult&lt;string&gt;&gt;</div>
<div class="tip" id="fs18">val filter : predicate:(&#39;T -&gt; bool) -&gt; source:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;T&gt;</div>
<div class="tip" id="fs19">val x : OntologyEnrichment.GseaResult&lt;string&gt;</div>
<div class="tip" id="fs20">OntologyEnrichment.GseaResult.PValue: float</div>
<div class="tip" id="fs21">val sortByDescending : projection:(&#39;T -&gt; &#39;Key) -&gt; source:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;T&gt; (requires comparison)</div>
<div class="tip" id="fs22">OntologyEnrichment.GseaResult.NumberOfDEsInBin: int</div>

			</div>
			<div class="span3" >
			  <img src="https://CSBiology.github.io/BioFSharp//img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
              <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                  <li class="nav-header">BioFSharp</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//index.html">Home page</a></li>
                  <li class="divider"></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//release-notes.html">Release Notes</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//Introduction.html">Getting started</a></li>

                  <li class="nav-header">BioFSharp core</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//Formula.html">Formula</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//BioItem.html">BioItems</a>
                  </li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//BioCollections.html">BioCollections</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//AminoProperties.html">Peptide Properties</a></li>
                  <li class="nav-header">BioContainers</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//BioContainers.html">Introduction</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//BioContainersDesignGuide.html">Container API Design Guide</a></li>
                  <li class="nav-header">Algorithm</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//Alignment.html">Alignment</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//ClustalOWrapper.html">Clustal Omega</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//MotiveSearch.html">Motive search</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//BlastWrapper.html">NCBI-Blast</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//StringMatching.html">Pattern Queries</a>
                  </li>
                  <li class="nav-header">File I/O</li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//Clustal.html">Clustal</a>
                  </li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//CSV.html">CSV / Tabular</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//FastA.html">FastA</a>
                  </li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//FastQ.html">FastQ</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//GenBank.html">GenBank</a>
                  </li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//SOFT.html">GEO Metadata (SOFT)</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//GFF3.html">GFF3</a>
                  </li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//Newick.html">Newick</a>
                  </li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//Obo.html">Obo ontology format</a></li>
                  <li>
                      <a href="https://CSBiology.github.io/BioFSharp//FSIPrinters.html">FSIPrinters</a>
                  </li>
                  <li class="nav-header">Web I/O</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//UniProt.html">UniProt</a></li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//BioDB.html">BioDB</a></li>
                  <li class="nav-header">Biostatistics</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//GSEA.html">Ontology enrichment</a></li>
                  <li class="nav-header">Biovisualization</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//WebLogo.html">Web logo</a></li>
                  <li class="nav-header">GPU Computing</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//ParallelAlignment.html">Sequence Alignment</a></li>
                  <li class="nav-header">Documentation</li>
                  <li><a href="https://CSBiology.github.io/BioFSharp//reference/index.html">API Reference</a></li>


              </ul>
			</div>
		  </div>
		</div>
		<a href="https://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
	  </body>
  </html>
