<!DOCTYPE html>
<html lang="en" data-root="https://csbiology.github.io/BioFSharp/">
<head>
    <meta charset="UTF-8">
    <meta content="IE=edge" http-equiv="X-UA-Compatible">
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <meta content="Timo MÃ¼hlhaus, Kevin Schneider, Heinrich Lukas Weil, David Zimmer, F# open source contributors" name="author">
    <!-- Opengraph properties (https://ogp.me/) -->
    <meta property="og:site_name" content="BioFSharp">
    <meta property="og:title" content="Biocollections" />
    <meta property="og:url" content="https://csbiology.github.io/BioFSharp/script.html">
    <meta property="og:type" content="website" />
    <!-- Twitter cards (https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image) -->
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:site" content="https://csbiology.github.io/BioFSharp/">
    <meta name="twitter:title" content="Biocollections">
    
    <title>Biocollections | BioFSharp</title>
    <link href="https://fonts.googleapis.com" rel="preconnect">
    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect">
    <link href="https://fonts.googleapis.com/css2?family=Fira+Code&display=swap" rel="stylesheet">
    <script src="https://code.iconify.design/iconify-icon/1.0.7/iconify-icon.min.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/img/favicon.ico" rel="icon" sizes="32x32" type="image/png"/>
    <script type="application/javascript" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-set-dark.js"></script>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-default.css" rel="stylesheet" type="text/css"/>
    <link href="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.css" rel="stylesheet" type="text/css"/>
    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.0.1/es5/tex-mml-chtml.js"></script>
    
</head>
<body class="content">
<header>
    <div class="start">
        <div id="menu-toggle">
            <iconify-icon class="icon closed" height="22" icon="eva:menu-fill" width="22"></iconify-icon>
            <iconify-icon class="icon open" height="22" icon="mi:close" width="22"></iconify-icon>
            <input type="checkbox" name="mobile-menu" />
            <ul class="menu">
                <li class="nav-header">Links</li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
                </li>
                <li class="nav-item"><a class="nav-link"
                                        href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
                    Notes</a></li>
                <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
                    Repository</a></li>
                <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/biocollections.html">
    Biocollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
                <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
            </ul>
        </div>
        <a href="https://csbiology.github.io/BioFSharp/">
            <img alt="Header menu logo" src="https://csbiology.github.io/BioFSharp/img/logo.png">
            <strong>BioFSharp</strong>
        </a>
    </div>
    <div class="end">
        <a href="https://github.com/CSBiology/BioFSharp/" target="_blank">
            <iconify-icon icon="uil:github" width="26" height="26"></iconify-icon>
        </a>
        <iconify-icon id="search-btn" icon="carbon:search" class="search" width="24" height="24"></iconify-icon>
        <fsdocs-theme-toggle></fsdocs-theme-toggle>
    </div>
</header>
<aside id="fsdocs-main-menu">
    <ul class="menu">
        <li class="nav-header">Links</li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/blob/developer/LICENSE">License</a>
        </li>
        <li class="nav-item"><a class="nav-link"
                                href="https://github.com/CSBiology/BioFSharp/blob/developer/RELEASE_NOTES.md">Release
            Notes</a></li>
        <li class="nav-item"><a class="nav-link" href="https://github.com/CSBiology/BioFSharp/">Source
            Repository</a></li>
        <li class="nav-header active">
  BioFSharp Core
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/formula.html">
    Formulas
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/bioitems.html">
    BioItems
  </a>
</li>             
<li class="nav-item active">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/core/biocollections.html">
    Biocollections
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/01_04_sequence_properties.html">
    Sequence properties
  </a>
</li>             
<li class="nav-header ">
  Algorithms
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_01_alignment.html">
    Alignment
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_02_blast.html">
    Blast
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/02_03_pattern_query_algorithms.html">
    Pattern query algorithms
  </a>
</li>             
<li class="nav-header ">
  BioParsers
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_01_fasta_parsing.html">
    Fasta
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_02_fastq_parsing.html">
    Fastq
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_03_gff3_parsing.html">
    GFF3
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_04_genbank_parsing.html">
    GenBank
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_05_obo.html">
    Obo
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/03_06_newick.html">
    Newick
  </a>
</li>             
<li class="nav-header ">
  Other
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/rnaseq_normalization.html">
    RNA-Seq data normalization
  </a>
</li>             
<li class="nav-item ">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/img/release-notes.html">
    release-notes
  </a>
</li>
        <li class="nav-header">
  API Reference
</li>             
<li class="nav-item">
  <a class="nav-link" href="https://csbiology.github.io/BioFSharp/reference/index.html">
    All Namespaces
  </a>
</li>
    </ul>
</aside>
<main>
    <div id="content">
        
<h1><a name="BioCollections" class="anchor" href="#BioCollections">BioCollections</a></h1>
<p><em>Summary:</em> This example shows how to use collections of biological items in BioFSharp</p>
<p>Analogous to the build-in collections BioFSharp provides BioSeq, BioList and BioArray for individual collection specific optimized operations.
The easiest way to create them are the <code>ofBioItemString</code> -functions</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span class="id">BioFSharp</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">s1</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs2', 2)" onmouseover="showTip(event, 'fs2', 2)" class="id">s2</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span class="id">BioList</span><span class="pn">.</span><span class="id">ofAminoAcidString</span> 
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs3', 3)" onmouseover="showTip(event, 'fs3', 3)" class="id">s3</span> <span class="o">=</span> <span class="s">&quot;TAGCAT&quot;</span>  <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">ofNucleotideString</span> 

<span onmouseout="hideTip(event, 'fs1', 4)" onmouseover="showTip(event, 'fs1', 4)" class="id">s1</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs2', 5)" onmouseover="showTip(event, 'fs2', 5)" class="id">s2</span><span class="pn">,</span> <span onmouseout="hideTip(event, 'fs3', 6)" onmouseover="showTip(event, 'fs3', 6)" class="id">s3</span>
</code></pre>
<p><p><details open="open" class="dni-treeview"><summary><span class="dni-code-hint"><code>(Microsoft.FSharp.Collections.IEnumerator+mkSeq@177[BioFSharp.AminoAcids+AminoAcid], [Pro; Glu; Pro; ... ], BioFSharp.Nucleotides+Nucleotide[])</code></span></summary><div><table><thead><tr></tr></thead><tbody><tr><td>Item1</td><td><pre>
1  PEPTIDE
</pre></td></tr><tr><td>Item2</td><td><pre>
1  PEPTIDE
</pre></td></tr><tr><td>Item3</td><td><pre>
1  TAGCAT
</pre></td></tr></tbody></table></div></details><style>
.dni-code-hint {
font-style: italic;
overflow: hidden;
white-space: nowrap;
}
.dni-treeview {
white-space: nowrap;
}
.dni-treeview td {
vertical-align: top;
text-align: start;
}
details.dni-treeview {
padding-left: 1em;
}
table td {
text-align: start;
}
table tr {
vertical-align: top;
margin: 0em 0px;
}
table tr td pre
{
vertical-align: top !important;
margin: 0em 0px !important;
}
table th {
text-align: start;
}
</style>
</p></p>
<h2><a name="Nucleotides" class="anchor" href="#Nucleotides">Nucleotides</a></h2>
<p><img src="../img/Nucleotides.svg" alt="Nucleotides1" /></p>
<p><strong>Figure 1: Selection of covered nucleotide operations</strong> (A) Biological principle. (B) Workflow with <code>BioSeq</code>. (C) Other covered functionalities.</p>
<p>Let's imagine you have a given gene sequence and want to find out what the according protein might look like.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs4', 7)" onmouseover="showTip(event, 'fs4', 7)" class="id">myGene</span> <span class="o">=</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">ofNucleotideString</span> <span class="s">&quot;ATGGCTAGATCGATCGATCGGCTAACGTAA&quot;</span>

<span onmouseout="hideTip(event, 'fs4', 8)" onmouseover="showTip(event, 'fs4', 8)" class="id">myGene</span>
</code></pre>
<p><pre>
         1  ATGGCTAGAT CGATCGATCG GCTAACGTAA
</pre>
</p>
<p>Yikes! Unfortunately we got the 5'-3' coding strand. For proper transcription we should get the complementary strand first:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs5', 9)" onmouseover="showTip(event, 'fs5', 9)" class="id">myProperGene</span> <span class="o">=</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">complement</span> <span onmouseout="hideTip(event, 'fs4', 10)" onmouseover="showTip(event, 'fs4', 10)" class="id">myGene</span>

<span onmouseout="hideTip(event, 'fs5', 11)" onmouseover="showTip(event, 'fs5', 11)" class="id">myProperGene</span>
</code></pre>
<p><pre>
         1  TACCGATCTA GCTAGCTAGC CGATTGCATT
</pre>
</p>
<p>Now let's transcribe and translate it:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs6', 12)" onmouseover="showTip(event, 'fs6', 12)" class="id">myTranslatedGene</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs5', 13)" onmouseover="showTip(event, 'fs5', 13)" class="id">myProperGene</span>
    <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">transcribeTemplateStrand</span>
    <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">translate</span> <span class="n">0</span>

<span onmouseout="hideTip(event, 'fs6', 14)" onmouseover="showTip(event, 'fs6', 14)" class="id">myTranslatedGene</span>
</code></pre>
<p><pre>
         1  MARSIDRLT*
</pre>
</p>
<p>Of course, if your input sequence originates from the coding strand, you can directly transcribe it to mRNA since the
only difference between the coding strand and the mRNA is the replacement of 'T' by 'U' (Figure 1B)</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs7', 15)" onmouseover="showTip(event, 'fs7', 15)" class="id">myTranslatedGeneFromCodingStrand</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs4', 16)" onmouseover="showTip(event, 'fs4', 16)" class="id">myGene</span>
    <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">transcribeCodingStrand</span>
    <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">translate</span> <span class="n">0</span>

<span onmouseout="hideTip(event, 'fs7', 17)" onmouseover="showTip(event, 'fs7', 17)" class="id">myTranslatedGeneFromCodingStrand</span>
</code></pre>
<p><pre>
         1  MARSIDRLT*
</pre>
</p>
<p>Other Nucleotide conversion operations are also covered:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs8', 18)" onmouseover="showTip(event, 'fs8', 18)" class="id">mySmallGene</span> <span class="o">=</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofNucleotideString</span>  <span class="s">&quot;ATGTTCCGAT&quot;</span>

<span onmouseout="hideTip(event, 'fs8', 19)" onmouseover="showTip(event, 'fs8', 19)" class="id">mySmallGene</span>
</code></pre>
<p><pre>
         1  ATGTTCCGAT
</pre>
</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">BioSeq</span><span class="pn">.</span><span class="id">reverse</span> <span onmouseout="hideTip(event, 'fs8', 20)" onmouseover="showTip(event, 'fs8', 20)" class="id">mySmallGene</span> 
</code></pre>
<p><pre>
         1  TAGCCTTGTA
</pre>
</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">BioSeq</span><span class="pn">.</span><span class="id">complement</span> <span onmouseout="hideTip(event, 'fs8', 21)" onmouseover="showTip(event, 'fs8', 21)" class="id">mySmallGene</span>
</code></pre>
<p><pre>
         1  TACAAGGCTA
</pre>
</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">BioSeq</span><span class="pn">.</span><span class="id">reverseComplement</span> <span onmouseout="hideTip(event, 'fs8', 22)" onmouseover="showTip(event, 'fs8', 22)" class="id">mySmallGene</span>
</code></pre>
<p><pre>
         1  ATCGGAACAT
</pre>
</p>
<h2><a name="AminoAcids" class="anchor" href="#AminoAcids">AminoAcids</a></h2>
<h3><a name="Basics" class="anchor" href="#Basics">Basics</a></h3>
<p>Some functions which might be needed regularly are defined to work with nucleotides and amino acids:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 23)" onmouseover="showTip(event, 'fs9', 23)" class="id">myPeptide</span> <span class="o">=</span> <span class="s">&quot;PEPTIDE&quot;</span> <span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">ofAminoAcidString</span> 

<span onmouseout="hideTip(event, 'fs9', 24)" onmouseover="showTip(event, 'fs9', 24)" class="id">myPeptide</span>
</code></pre>
<p><pre>
         1  PEPTIDE
</pre>
</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span onmouseout="hideTip(event, 'fs9', 25)" onmouseover="showTip(event, 'fs9', 25)" class="id">myPeptide</span> 
<span class="o">|&gt;</span> <span class="id">BioSeq</span><span class="pn">.</span><span class="id">toFormula</span> 
<span class="o">|&gt;</span> <span class="id">Formula</span><span class="pn">.</span><span class="id">toString</span> 
</code></pre>
<table class="pre"><tbody><tr><td><pre><code>C34.00 H51.00 N7.00 O14.00</code></pre></td></tr></tbody></table>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="id">BioSeq</span><span class="pn">.</span><span class="id">toAverageMass</span> <span onmouseout="hideTip(event, 'fs9', 26)" onmouseover="showTip(event, 'fs9', 26)" class="id">myPeptide</span> 
</code></pre>
<p><div class="dni-plaintext"><pre>781.8103169999999</pre></div><style>
.dni-code-hint {
    font-style: italic;
    overflow: hidden;
    white-space: nowrap;
}
.dni-treeview {
    white-space: nowrap;
}
.dni-treeview td {
    vertical-align: top;
    text-align: start;
}
details.dni-treeview {
    padding-left: 1em;
}
table td {
    text-align: start;
}
table tr { 
    vertical-align: top; 
    margin: 0em 0px;
}
table tr td pre 
{ 
    vertical-align: top !important; 
    margin: 0em 0px !important;
} 
table th {
    text-align: start;
}
</style>
</p>
<h3><a name="Digestion" class="anchor" href="#Digestion">Digestion</a></h3>
<p>BioFSharp also comes equipped with a set of tools aimed at cutting apart amino acid sequences. To demonstrate the usage, we'll throw some <code>trypsin</code> at the small RuBisCO subunit of <em>Arabidopos thaliana</em>:<br />
In the first step, we define our input sequence and the protease we want to use.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs10', 27)" onmouseover="showTip(event, 'fs10', 27)" class="id">RBCS</span> <span class="o">=</span> 
    <span class="s">&quot;&quot;&quot;MASSMLSSATMVASPAQATMVAPFNGLKSSAAFPATRKANNDITSITSNGGRVNCMQVWP</span>
<span class="s">    PIGKKKFETLSYLPDLTDSELAKEVDYLIRNKWIPCVEFELEHGFVYREHGNSPGYYDGR</span>
<span class="s">    YWTMWKLPLFGCTDSAQVLKEVEECKKEYPNAFIRIIGFDNTRQVQCISFIAYKPPSFT&quot;&quot;&quot;</span> 
    <span class="o">|&gt;</span> <span class="id">BioArray</span><span class="pn">.</span><span class="id">ofAminoAcidString</span>

<span onmouseout="hideTip(event, 'fs10', 28)" onmouseover="showTip(event, 'fs10', 28)" class="id">RBCS</span>
</code></pre>
<p><pre>
         1  MASSMLSSAT MVASPAQATM VAPFNGLKSS AAFPATRKAN NDITSITSNG GRVNCMQVWP
        61  PIGKKKFETL SYLPDLTDSE LAKEVDYLIR NKWIPCVEFE LEHGFVYREH GNSPGYYDGR
       121  YWTMWKLPLF GCTDSAQVLK EVEECKKEYP NAFIRIIGFD NTRQVQCISF IAYKPPSFT
</pre>
</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs11', 29)" onmouseover="showTip(event, 'fs11', 29)" class="id">trypsin</span> <span class="o">=</span> <span class="id">Digestion</span><span class="pn">.</span><span class="id">Table</span><span class="pn">.</span><span class="id">getProteaseBy</span> <span class="s">&quot;Trypsin&quot;</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs12', 30)" onmouseover="showTip(event, 'fs12', 30)" class="id">digestedRBCS</span> <span class="o">=</span> <span class="id">Digestion</span><span class="pn">.</span><span class="id">BioArray</span><span class="pn">.</span><span class="id">digest</span> <span onmouseout="hideTip(event, 'fs11', 31)" onmouseover="showTip(event, 'fs11', 31)" class="id">trypsin</span> <span class="n">0</span> <span onmouseout="hideTip(event, 'fs10', 32)" onmouseover="showTip(event, 'fs10', 32)" class="id">RBCS</span> 

<span onmouseout="hideTip(event, 'fs12', 33)" onmouseover="showTip(event, 'fs12', 33)" class="id">digestedRBCS</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs13', 34)" onmouseover="showTip(event, 'fs13', 34)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs14', 35)" onmouseover="showTip(event, 'fs14', 35)" class="id">head</span>
</code></pre>
<p><details open="open" class="dni-treeview"><summary><span class="dni-code-hint"><code>{ ProteinID = 0\n  MissCleavages = 0\n  CleavageStart = 0\n  CleavageEnd = 27\n  PepSequence =\n   [Met; Ala; Ser; Ser; Met; Leu; Ser; Ser; Ala; Thr; Met; Val; Ala; Ser; Pro;\n    Ala; Gln; Ala; Thr; Met; Val; Ala; Pro; Phe; Asn; Gly; Leu; Lys] }</code></span></summary><div><table><thead><tr></tr></thead><tbody><tr><td>ProteinID</td><td><div class="dni-plaintext"><pre>0</pre></div></td></tr><tr><td>MissCleavages</td><td><div class="dni-plaintext"><pre>0</pre></div></td></tr><tr><td>CleavageStart</td><td><div class="dni-plaintext"><pre>0</pre></div></td></tr><tr><td>CleavageEnd</td><td><div class="dni-plaintext"><pre>27</pre></div></td></tr><tr><td>PepSequence</td><td><pre>
         1  MASSMLSSAT MVASPAQATM VAPFNGLK
</pre></td></tr></tbody></table></div></details><style>
.dni-code-hint {
    font-style: italic;
    overflow: hidden;
    white-space: nowrap;
}
.dni-treeview {
    white-space: nowrap;
}
.dni-treeview td {
    vertical-align: top;
    text-align: start;
}
details.dni-treeview {
    padding-left: 1em;
}
table td {
    text-align: start;
}
table tr { 
    vertical-align: top; 
    margin: 0em 0px;
}
table tr td pre 
{ 
    vertical-align: top !important; 
    margin: 0em 0px !important;
} 
table th {
    text-align: start;
}
</style>
</p>
<p>In reality, proteases don't always completely cut the protein down. Instead, some sites stay intact and should be considered for in silico analysis.
This can easily be done with the <code>concernMissCleavages</code> function. It takes the minimum and maximum amount of misscleavages you want to have and also the digested protein. As a result you get all possible combinations arising from this information.</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs15', 36)" onmouseover="showTip(event, 'fs15', 36)" class="id">digestedRBCS&#39;</span> <span class="o">=</span> <span class="id">Digestion</span><span class="pn">.</span><span class="id">BioArray</span><span class="pn">.</span><span class="id">concernMissCleavages</span> <span class="n">0</span> <span class="n">2</span> <span onmouseout="hideTip(event, 'fs12', 37)" onmouseover="showTip(event, 'fs12', 37)" class="id">digestedRBCS</span>

<span onmouseout="hideTip(event, 'fs12', 38)" onmouseover="showTip(event, 'fs12', 38)" class="id">digestedRBCS</span>
<span class="o">|&gt;</span> <span onmouseout="hideTip(event, 'fs13', 39)" onmouseover="showTip(event, 'fs13', 39)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 40)" onmouseover="showTip(event, 'fs16', 40)" class="id">item</span> <span class="n">1</span>
</code></pre>
<p><details open="open" class="dni-treeview"><summary><span class="dni-code-hint"><code>{ ProteinID = 0\n  MissCleavages = 0\n  CleavageStart = 28\n  CleavageEnd = 36\n  PepSequence = [Ser; Ser; Ala; Ala; Phe; Pro; Ala; Thr; Arg] }</code></span></summary><div><table><thead><tr></tr></thead><tbody><tr><td>ProteinID</td><td><div class="dni-plaintext"><pre>0</pre></div></td></tr><tr><td>MissCleavages</td><td><div class="dni-plaintext"><pre>0</pre></div></td></tr><tr><td>CleavageStart</td><td><div class="dni-plaintext"><pre>28</pre></div></td></tr><tr><td>CleavageEnd</td><td><div class="dni-plaintext"><pre>36</pre></div></td></tr><tr><td>PepSequence</td><td><pre>
         1  SSAAFPATR
</pre></td></tr></tbody></table></div></details><style>
.dni-code-hint {
    font-style: italic;
    overflow: hidden;
    white-space: nowrap;
}
.dni-treeview {
    white-space: nowrap;
}
.dni-treeview td {
    vertical-align: top;
    text-align: start;
}
details.dni-treeview {
    padding-left: 1em;
}
table td {
    text-align: start;
}
table tr { 
    vertical-align: top; 
    margin: 0em 0px;
}
table tr td pre 
{ 
    vertical-align: top !important; 
    margin: 0em 0px !important;
} 
table th {
    text-align: start;
}
</style>
</p>

        <div class="fsdocs-tip" id="fs1">val s1: obj</div>
<div class="fsdocs-tip" id="fs2">val s2: obj</div>
<div class="fsdocs-tip" id="fs3">val s3: obj</div>
<div class="fsdocs-tip" id="fs4">val myGene: obj</div>
<div class="fsdocs-tip" id="fs5">val myProperGene: obj</div>
<div class="fsdocs-tip" id="fs6">val myTranslatedGene: obj</div>
<div class="fsdocs-tip" id="fs7">val myTranslatedGeneFromCodingStrand: obj</div>
<div class="fsdocs-tip" id="fs8">val mySmallGene: obj</div>
<div class="fsdocs-tip" id="fs9">val myPeptide: obj</div>
<div class="fsdocs-tip" id="fs10">val RBCS: obj</div>
<div class="fsdocs-tip" id="fs11">val trypsin: obj</div>
<div class="fsdocs-tip" id="fs12">val digestedRBCS: obj seq</div>
<div class="fsdocs-tip" id="fs13">module Seq

from Microsoft.FSharp.Collections</div>
<div class="fsdocs-tip" id="fs14">val head: source: &#39;T seq -&gt; &#39;T</div>
<div class="fsdocs-tip" id="fs15">val digestedRBCS&#39;: obj</div>
<div class="fsdocs-tip" id="fs16">val item: index: int -&gt; source: &#39;T seq -&gt; &#39;T</div>

    </div>
</main>
<aside id="fsdocs-page-menu">
    <p id="on-this-page">On this page</p>
    <ul>
  <li class="level-1">
    <a href="#BioCollections">
      BioCollections
    </a>
  </li>
  <li class="level-2">
    <a href="#Nucleotides">
      Nucleotides
    </a>
  </li>
  <li class="level-2">
    <a href="#AminoAcids">
      AminoAcids
    </a>
  </li>
  <li class="level-3">
    <a href="#Basics">
      Basics
    </a>
  </li>
  <li class="level-3">
    <a href="#Digestion">
      Digestion
    </a>
  </li>
</ul>
</aside>
<dialog>
    <input type="search" placeholder="Search docs" />
    <div class="results">
        <ul></ul>
        <p class="empty">Type something to start searching.</p>
    </div>
</dialog>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-tips.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme-toggle.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-theme.js"></script>
<script type="module" src="https://csbiology.github.io/BioFSharp/content/fsdocs-search.js"></script>

</body>
</html>