<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>Classification - BioFSharp</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
    <meta name="description" content=""/>
    <meta name="author" content="Timo MÃ¼hlhaus"/>

    <script src="https://code.jquery.com/jquery-1.8.0.js"></script>
    <script src="https://code.jquery.com/ui/1.8.23/jquery-ui.js"></script>
    <script src="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/js/bootstrap.min.js"></script>
    <link href="https://netdna.bootstrapcdn.com/twitter-bootstrap/2.2.1/css/bootstrap-combined.min.css" rel="stylesheet"/>
    <meta http-equiv="X-UA-Compatible" content="IE=11" >
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <link type="text/css" rel="stylesheet" href="/BioFSharp/content/style.css" /> <!--/BioFSharp-->
    <link type="text/css" rel="stylesheet" href="/BioFSharp/style/CSBstyle.css" /> <!--/BioFSharp-->
    <script type="text/javascript" src="/BioFSharp/content/tips.js"></script> <!--/BioFSharp-->
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <![endif]-->
  </head>
	<body>
		<div class="container">
		  <div class="masthead">
			<ul class="nav nav-pills pull-right">
			  <li><a href="http://fsharp.org">fsharp.org</a></li>
			  <li><a href="https://github.com/CSBiology/BioFSharp">github page</a></li>
			</ul>
			<h3 class="muted"><a href="/BioFSharp/index.html">BioFSharp</a></h3>
		  </div>
		  <hr />
		  <div class="row">
			<div class="span9" id="main">
			  



<h1>Classification</h1>
<p>
  <span>Namespace: BioFSharp.ML</span><br />
  <span>Parent Module: <a href="biofsharp-ml-dppop.html">DPPOP</a></span>
  </p>
<div class="xmldoc">
<p>Contains functions to extract the features used in dppop to classify peptides for observability prediction</p>

</div>

<!-- Render nested types and modules, if there are any -->

  <h3>Functions and values</h3>
  <table class="table table-bordered member-list">
    <thead>
      <tr><td>Function or value</td><td>Description</td></tr>
    </thead>
    <tbody>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '751', 751)" onmouseover="showTip(event, '751', 751)">
            createDigestionEfficiencyMapFromFasta(...)
          </code>
          <div class="tip" id="751">
            <strong>Signature:</strong> fa:seq&lt;FastaItem&lt;BioArray&lt;AminoAcid&gt;&gt;&gt; -&gt; Map&lt;(string * AminoAcidSymbol []),(float&lt;MeasureOne&gt; * float&lt;MeasureOne&gt; * float&lt;MeasureOne&gt;)&gt;<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L539-539" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>returns a map mapping from a (proteinID<em>sequence) touple to the three digestion efficiency scores in the form of a (float</em>float*float) tuple from the input fasta item collection</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '752', 752)" onmouseover="showTip(event, '752', 752)">
            digestTryptic(aminoAcidSeq)
          </code>
          <div class="tip" id="752">
            <strong>Signature:</strong> aminoAcidSeq:BioArray&lt;AminoAcid&gt; -&gt; seq&lt;AminoAcid []&gt;<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L470-470" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>Tryptic Digestion2 of an amino acid sequence with the settings used for the dppop web API</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '753', 753)" onmouseover="showTip(event, '753', 753)">
            digestTrypticWith(...)
          </code>
          <div class="tip" id="753">
            <strong>Signature:</strong> maxMissCleavages:int -&gt; minPeptideLength:int -&gt; aminoAcidSeq:BioArray&lt;AminoAcid&gt; -&gt; seq&lt;AminoAcid []&gt;<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L463-463" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>Tryptic digestion of an amino acid sequence with the ability to control the maximal amount of misscleavages and filtering of a minimal peptide length</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '754', 754)" onmouseover="showTip(event, '754', 754)">
            getDistinctTrypticPeptidesFromFasta(fa)
          </code>
          <div class="tip" id="754">
            <strong>Signature:</strong> fa:seq&lt;FastaItem&lt;BioArray&lt;AminoAcid&gt;&gt;&gt; -&gt; Set&lt;string&gt;<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L475-475" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>Returns a distinct set of peptides that map uniquely to a single protein from the given fasta input</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '755', 755)" onmouseover="showTip(event, '755', 755)">
            getDistinctTrypticPeptidesFromFastaFile(...)
          </code>
          <div class="tip" id="755">
            <strong>Signature:</strong> filePath:string -&gt; Set&lt;string&gt;<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L487-487" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>Returns a distinct set of peptides that map uniquely to a single protein from the given fasta file input</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '756', 756)" onmouseover="showTip(event, '756', 756)">
            getPeptideFeatures(...)
          </code>
          <div class="tip" id="756">
            <strong>Signature:</strong> digestionEfficiencyMap:Map&lt;(string * BioArray&lt;AminoAcidSymbol&gt;),(float * float * float)&gt; -&gt; protId:string -&gt; peptide:AminoAcidSymbol [] -&gt; PredictionInput option<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L580-580" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>Returns a PredictionInput record type for the input peptide containing the calculated feature vector given a digestion efficiency map and the proteinId the peptide maps to</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '757', 757)" onmouseover="showTip(event, '757', 757)">
            getPhysicochemicalProperties(peptide)
          </code>
          <div class="tip" id="757">
            <strong>Signature:</strong> peptide:BioArray&lt;AminoAcidSymbol&gt; -&gt; float []<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L546-546" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>get the physicochemical properties of a peptide: length, MolecularWeight, NetCharge, PositiveCharge, NegativeCharge, piI, Relative frewuencies of polar, hydrophobic, and negatively charge amino acids</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '758', 758)" onmouseover="showTip(event, '758', 758)">
            zNormalizeNonPlantFeatureVector(...)
          </code>
          <div class="tip" id="758">
            <strong>Signature:</strong> features:PredictionInput -&gt; PredictionInput<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L683-683" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>returns the z normalized version of the input feature vector by using the normalization vector dppop uses for its non-plant model</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '759', 759)" onmouseover="showTip(event, '759', 759)">
            zNormalizePeptideFeaturesBy(...)
          </code>
          <div class="tip" id="759">
            <strong>Signature:</strong> norm:(float * float) [] -&gt; features:PredictionInput -&gt; PredictionInput<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L614-614" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>normalizes feature vector given a normalization vector containing tuples of (stDev,mean) of all features across a dataset</p>


        </td>
      </tr>
      <tr>
        <td class="member-name">

          <code onmouseout="hideTip(event, '760', 760)" onmouseover="showTip(event, '760', 760)">
            zNormalizePlantFeatureVector(features)
          </code>
          <div class="tip" id="760">
            <strong>Signature:</strong> features:PredictionInput -&gt; PredictionInput<br />
                                  </div>
        </td>
        <td class="xmldoc">
                      <a href="https://github.com/CSBiology/BioFSharp/tree/master/src/BioFSharp.ML/DPPOP.fs#L679-679" class="github-link">
              <img src="../content/img/github.png" class="normal" />
              <img src="../content/img/github-blue.png" class="hover" />
            </a>
          <p>returns the z normalized version of the input feature vector by using the normalization vector dppop uses for its plant model</p>


        </td>
      </tr>
    </tbody>
  </table>

			</div>
			<div class="span3" >
			  <img src="/BioFSharp/img/logo.png" alt="F# Project" style="width:150px;margin:10px" />  
              <ul class="nav nav-list" id="menu" style="margin-top: 20px;">
                  <li class="nav-header">BioFSharp</li>
                  <li><a href="/BioFSharp/index.html">Home page</a></li>
                  <li class="divider"></li>
                  <li><a href="/BioFSharp/release-notes.html">Release Notes</a></li>
                  <li><a href="/BioFSharp/Introduction.html">Getting started</a></li>

                  
                  <li class="nav-header">BioFSharp core</li>
                  <li><a href="/BioFSharp/Formula.html">Formula</a></li>
                  <li>
                      <a href="/BioFSharp/BioItem.html">BioItems</a>
                  </li>
                  <li><a href="/BioFSharp/BioCollections.html">BioCollections</a></li>
                  <li><a href="/BioFSharp/AminoProperties.html">Peptide Properties</a></li>
                  <li class="nav-header">BioContainers</li>
                  <li><a href="/BioFSharp/BioContainers.html">Introduction</a></li>
                  <li><a href="/BioFSharp/BioContainersDesignGuide.html">Container API Design Guide</a></li>
                  <li class="nav-header">Algorithm</li>
                  <li><a href="/BioFSharp/Alignment.html">Alignment</a></li>
                  <li><a href="/BioFSharp/ClustalOWrapper.html">Clustal Omega</a></li>
                  <li><a href="/BioFSharp/MotiveSearch.html">Motive search</a></li>
                  <li><a href="/BioFSharp/BlastWrapper.html">NCBI-Blast</a></li>
                  <li>
                      <a href="/BioFSharp/StringMatching.html">Pattern Queries</a>
                  </li>
                  <li class="nav-header">File I/O</li>
                  <li>
                      <a href="/BioFSharp/Clustal.html">Clustal</a>
                  </li>
                  <li><a href="/BioFSharp/CSV.html">CSV / Tabular</a></li>
                  <li>
                      <a href="/BioFSharp/FastA.html">FastA</a>
                  </li>
                  <li><a href="/BioFSharp/FastQ.html">FastQ</a></li>
                  <li>
                      <a href="/BioFSharp/GenBank.html">GenBank</a>
                  </li>
                  <li><a href="/BioFSharp/SOFT.html">GEO Metadata (SOFT)</a></li>
                  <li>
                      <a href="/BioFSharp/GFF3.html">GFF3</a>
                  </li>
                  <li>
                      <a href="/BioFSharp/Newick.html">Newick</a>
                  </li>
                  <li><a href="/BioFSharp/Obo.html">Obo ontology format</a></li>
                  <li>
                      <a href="/BioFSharp/FSIPrinters.html">FSIPrinters</a>
                  </li>
                  <li class="nav-header">Web I/O</li>
                  <li><a href="/BioFSharp/UniProt.html">UniProt</a></li>
                  <li><a href="/BioFSharp/BioDB.html">BioDB</a></li>
                  <li class="nav-header">Biostatistics</li>
                  <li><a href="/BioFSharp/GSEA.html">Ontology enrichment</a></li>
                  <li class="nav-header">Biovisualization</li>
                  <li><a href="/BioFSharp/WebLogo.html">Web logo</a></li>
                  <li class="nav-header">GPU Computing</li>
                  <li><a href="/BioFSharp/ParallelAlignment.html">Sequence Alignment</a></li>
                  <li class="nav-header">Documentation</li>
                  <li><a href="/BioFSharp/reference/index.html">API Reference</a></li>


                  
              </ul>
			</div>
		  </div>
		</div>
		<a href="https://github.com/CSBiology/BioFSharp"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" alt="Fork me on GitHub"/></a>
	  </body>
  </html>
